# Installation

## Install scRNASequest

First, we install the scRNASequest by downloading the scripts from GitHub:

```{bash, eval=FALSE}
git clone https://github.com/interactivereport/scRNAsequest.git
cd scRNAsequest

#Install scRNAsequest conda environment and R packages
#Please make sure you have conda installed before, and this step may take a while (30min ~1 hour)
bash install

#Activate the conda environment
conda activate scRNAsequest

#Also, .env will be created under the src directory:
ls ~/scRNAsequest/src/.env

#Check the path of current directory and add it to $PATH:
CurrentDir=`pwd`
export PATH="$CurrentDir:$PATH"

#However, the above command only adds the scRNAsequest directory to $PATH temporarily
#To add it to the environment permanently, edit ~/.bash_profile or ~/.bashrc:
vim ~/.bash_profile
#Add the full path of the scRNAsequest directory to $PATH, for example, $HOME/scRNAsequest
PATH=$PATH:$HOME/RNAsequest
#Source the file
source ~/.bash_profile
```

If the installation is successful, scAnalyzer and scRef should be able to run without passing any parameters:

```{bash, eval=FALSE}
$ scAnalyzer

***** 2022-06-23 17:40:17 *****
###########
## scRNAsequest: https://github.com/interactivereport/scRNAsequest.git
## Pipeline Path: /camhpc/ngs/tools/scRNAsequest
## Pipeline Date: 2022-06-23 15:00:23 -0400
## git HEAD: 61d9e676d31849ccc8ac8a3442704128504be248
###########

scAnalyzer /path/to/a/DNAnexus/download/folder === or === scAnalyzer /path/to/a/config/file

The config file will be generated automatically when a DNAnexus download folder is provided
Available reference data:
	human_cortex: more information @ https://azimuth.hubmapconsortium.org/references/
	human_fetal: more information @ https://azimuth.hubmapconsortium.org/references/
	human_pbmc: more information @ https://azimuth.hubmapconsortium.org/references/
	mouse_thalamus: more information @ http://omicsoft01.hpc.biogen.com/celldepot/app/core/app_project_review.php?ID=301
	mouse_cortex: more information @ https://azimuth.hubmapconsortium.org/references/
	mouse_spinalCord: more information @ http://omicsoft01.hpc.biogen.com/celldepot/app/core/app_project_review.php?ID=285
	mouse_telencephalon_diencephalon: more information @ https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-11115/
	NHP_Substantia_nigra: more information @ search NHP_Substantia_nigra in celldepot @ http://omicsoft01.hpc.biogen.com/celldepot
	NHP_Cerebellum: more information @ search NHP_Cerebellum in celldepot @ http://omicsoft01.hpc.biogen.com/celldepot
If one of the above can be used as a reference for your datasets, please update the config file with the name in 'ref_name'.

Powered by the Research Data Sciences Group [zhengyu.ouyang@biogen.com;yuhenry.sun@biogen.com]
------------
```

Typing scRef directly without any parameters will activate the user manual page:

```{bash, eval=FALSE}
$ scRef

***** 2022-06-23 17:38:28 *****
###########
## scRNAsequest: https://github.com/interactivereport/scRNAsequest.git
## Pipeline Path: /camhpc/ngs/tools/scRNAsequest
## Pipeline Date: 2022-06-23 15:00:23 -0400
## git HEAD: 61d9e676d31849ccc8ac8a3442704128504be248
###########


Loading resources

scRef /path/to/a/output/folder === or === scRef /path/to/a/Ref/config/file

The folder has to be existed.
The Ref config file will be generated automatically when a path is provided
===== CAUTION =====
	1. This process will add a seurat reference data into the scRNAsequest pipeline PERMANENTLY!
	2. Make sure the data provided for reference building is SCT transformed!

Powered by the Research Data Sciences group [zhengyu.ouyang@biogen.com;yuhenry.sun@biogen.com]
------------
```

## Configure sys.yml file

The **sys.yml** file contains critical information for the pipeline. You can use this [sys_example.yml](https://github.com/interactivereport/scRNAsequest/blob/main/src/sys_example.yml){target="_blank"} file as a template to prepare your file. The file name must be **sys.yml** and is located under the src/ directory of the pipeline.

## Install cellxgene_VIP

The cellxgene_VIP platform can be used to visualize the h5ad file generated by the scRNASequest pipeline. This visualization step can be performed before the differential expression (DE) analysis to pinpoint meaningful clusters for downstream steps.

Please follow the detailed instructions below to install cellxgene_VIP:

[https://github.com/interactivereport/cellxgene_VIP](https://github.com/interactivereport/cellxgene_VIP){target="_blank"}

However, it is up to the user how to visualize the h5ad file, and other tools with similar functions ([scanpy](https://scanpy.readthedocs.io/en/stable/){target="_blank"} or [Seurat](https://satijalab.org/seurat/){target="_blank"}) will also work.

## Install CellDepot

Please refer to the detailed instructions below to install CellDepot:

[https://celldepot.bxgenomics.com/celldepot_manual/install_environment.php](https://celldepot.bxgenomics.com/celldepot_manual/install_environment.php){target="_blank"}

The installation of CellDepot is also optional if the user only needs to analyze the data, without publishing it into the CellDepot database.