# Data preparation

scRNASequest pipeline is compatible with different single-cell experiment outputs, including the **10X mtx folder** and the **10X h5 output** from cellranger. The user may need to manually separate the annotation file, so that no cell filtering was performed. In this chapter, we will walk through how to prepare the data before running the pipeline.

## Public data in h5 format

Here, we first present an example of the data processing steps using a public EMBL EBI dataset: [E-MTAB-11115](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-11115/){target="_blank"}. The nine processed zip files were downloaded and unzipped.

There are total 6 datasets, with 6 corresponding *raw_feature_bc_matrix.h5 files:

```
#Annotation files
-rwxrwx---. 1 zouyang ngs 440K Apr 21 16:59 5705STDY8058280_annotation.csv
-rwxrwx---. 1 zouyang ngs 446K Apr 21 16:59 5705STDY8058281_annotation.csv
-rwxrwx---. 1 zouyang ngs 310K Apr 21 16:59 5705STDY8058282_annotation.csv
-rwxrwx---. 1 zouyang ngs 282K Apr 21 16:59 5705STDY8058283_annotation.csv
-rwxrwx---. 1 zouyang ngs 157K Apr 21 16:59 5705STDY8058284_annotation.csv
-rwxrwx---. 1 zouyang ngs 555K Apr 21 16:59 5705STDY8058285_annotation.csv

#h5 matrix files
-rwxrwxr--. 1 zouyang ngs 165M Oct 25  2021 5705STDY8058280_raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 165M Oct 25  2021 5705STDY8058281_raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 156M Oct 25  2021 5705STDY8058282_raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 162M Oct 25  2021 5705STDY8058283_raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 149M Oct 25  2021 5705STDY8058284_raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 177M Oct 25  2021 5705STDY8058285_raw_feature_bc_matrix.h5
```

## Public data in 10X mtx format

Another example is based on the 10X mtx folder output. **?????Waiting for an example**





## Self-prepared files

For users that have their own dataset to run EA, several files can be prepared from scratch. Here is the detailed instruction on what files are required to run the pipeline.

Please use the same file hierarchy as 4.2 after creating them.

- **comparison.csv**
    
  This file contains comparison groups, which must be defined in the **meta.csv**.
    
  Here is an example of the **comparison.csv** file:
    
  ```
  metacol,case,ctrl
  group,grp2,grp1
  group,grp3,grp2
  ```

- **counts.tsv**

  This file stores the raw counts of each sample in a tab-delimited format (tsv).

  ```
  $ head -4 counts.tsv

  ""      "SRR9139048"     "SRR9139049"        "SRR9139050"
  "ENSMUSG00000079800.2"      7       9       11
  "ENSMUSG00000095092.1"      2       1       10
  "ENSMUSG00000079192.2"      7       7       1
  ```

- **geneAnnotation.csv**

   This file contains the detailed annotation of each gene shown in the count.tsv table.
   
   Here is an example:

  ```
  $ head -4 geneAnnotation.csv
  
  "","seqnames","start","end","width","strand","source","type","bp_length","phase","UniqueID","gene_type","Gene.Name","level","mgi_id","havana_gene","tag","id"
"ENSMUSG00000079800.2","GL456210.1",9124,58882,49759,"-","ENSEMBL","gene",1271,NA,"ENSMUSG00000079800.2","protein_coding","AC125149.3","3",NA,NA,NA,1
"ENSMUSG00000095092.1","GL456210.1",108390,110303,1914,"-","ENSEMBL","gene",366,NA,"ENSMUSG00000095092.1","protein_coding","AC125149.5","3",NA,NA,NA,2
"ENSMUSG00000079192.2","GL456210.1",123792,124928,1137,"+","ENSEMBL","gene",255,NA,"ENSMUSG00000079192.2","protein_coding","AC125149.1","3",NA,NA,NA,3
   ```

- **geneLength.tsv**

  This file has the gene length information. It must match the dimension of the count.tsv file.

  ```
  $ head -4 geneLength.tsv
  
  ""      "SRR9139048"     "SRR9139049"        "SRR9139050"
  "ENSMUSG00000079800.2"        1271      1271        1271
  "ENSMUSG00000095092.1"        366       366     366
  "ENSMUSG00000079192.2"        255       255     255
  ```

- **meta.csv**

  This **meta.csv** file can contain many annotation columns for the dataset. For example, this is a meta file containing group information:
  
  ```
  $ head -4 meta.tsv
  
  "","group"
  "SRR9139048","grp1"
  "SRR9139049","grp2"
  "SRR9139050","grp3"
  ```

- **project.yml**

  The project.yml file contains a high-level summary of the project, including project name, species, etc.

  ```
  $ cat project.yml
  
  project: SRP199678
  species: mouse
  file_source: sra
  project_home: data_sources_sra
  project_type: data_sources
  number_samples: 155
  ```
