# Input configuration

First, make sure the .env has been created in the src folder under the pipeline directory:

```
ls ~/scRNAsequest/src/.env
```

## Prepare the config.yml file

To run the scRNAsequest pipeline, a config.yml file is required to be filled in. Please use the following template as an example to prepare this file:

[config.yml](https://github.com/interactivereport/scRNAsequest/blob/main/src/template.yml){target="_blank"}

```
## Project info and filtering parameters
prj_name: E-MTAB-11115            #required, project name
prj_title: "Cell2location maps fine-grained cell types in spatial transcriptomics" #required, and quotes might be needed
ref_name:                         #optional. You can choose one from scAnalyzer call without an argument
output: /home/ysun4/E-MTAB-11115/ #required, output path
sample_name: Sample_Name
sample_meta: ~/E-MTAB-11115/sampleMeta.csv   #required, sample meta information, see section 5.1
gene_group:
  MT:
    startwith: ["MT-","Mt-","mt-"]
    cutoff: 20           # percentage cutoff to filter out the cells (larger than this cutoff)
    rm: False
  RP:
    startwith: []
    cutoff: 20
    rm: False
filter_step: True        #if False, the above 'gene_group' filtering will be skipped as well
min.cells: 3             #filtering genes by minimal cell
min.features: 50         #filtering cells by minimal genes
highCount.cutoff: 10000  #any cells with higher total counts to be removed
highGene.cutoff: 3000    #any cells with a higher number of detected genes to be removed

group:                   #if provided, a 10X QC box plot will be ploted in QC plot
runAnalysis: True
newProcess: False
parallel: False          #"sge" or "slurm"
core: 2
overwrite: True
#parallelTime: 180       #for a job (not the entire pipeline) in minutes, default 3hr

## DEG analysis for an annotation (such as disease vs health) within a cell type annotation
DEG_desp:  ~/E-MTAB-11115/DEGinfo.csv         #required for DEG analysis, but you can leave this empty if not going to run DEG
# Please be causion of changing the following default filtering
min.cells.per.gene: 3
min.genes.per.cell: 250
min.perc.cells.per.gene: 0.00
perc_filter: TRUE
R6_min.cells.per.gene: 3
R6_min.perc.cells.per.gene: 0.1
R6_min.cells.per.gene.type: "or"
R6_cells.per.gene.filter: TRUE
R6_perc.cells.filter: TRUE
R6_perc.filter: FALSE
R6_perc.filter.type: "and"
R6_perc_threshold: 0.90
R6_min.ave.pseudo.bulk.cpm: 1
R6_pseudo.bulk.cpm.filter: FALSE
R6_min.cells.per.subj: 3

## publish to celldepot
publish: False
```

In the first run (see the below scAnalyzer section on how to run the pipeline), the user can set `runAnalysis: False` and use the above default parameters to perform cell filtering and QC.

Then this pipeline will only run basic QC checking and output a Bookdown report with QC figures. This step is semi-automated because different datasets may need distinct filtering criteria. The design of this pipeline is to pause here to make sure the filtering step is adequate before running through the whole analysis.

Once the filtering step is validated by checking the QC report, the user can change the following settings and run the full analysis:

```
runAnalysis: True
overwrite: True
```

## Prepare the sample meta file {#section52}

The sample meta file, usually named as **sampleMeta.csv**, is a csv file storing data information. The minimal columns for this file are **Sample_Name** (This name can be changed, but it must be consistent with the sample_name in the config.yml file) and **h5path**. If you have the cell type annotation information, you can provide is by adding a column called **metapath**. For other information, you can add as many columns in this csv file and the column names can be defined by yourself.

### h5 input

Here is an example with minimal information for the public dataset [E-MTAB-11115](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-11115/){target="_blank"} (10X h5 format):

```
Sample_Name,h5path
5705STDY8058280,~/E-MTAB-11115/data/5705STDY8058280.filtered_feature_bc_matrix.h5
5705STDY8058281,~/E-MTAB-11115/data/5705STDY8058281.filtered_feature_bc_matrix.h5
5705STDY8058282,~/E-MTAB-11115/data/5705STDY8058282.filtered_feature_bc_matrix.h5
5705STDY8058283,~/E-MTAB-11115/data/5705STDY8058283.filtered_feature_bc_matrix.h5
5705STDY8058284,~/E-MTAB-11115/data/5705STDY8058284.filtered_feature_bc_matrix.h5
5705STDY8058285,~/E-MTAB-11115/data/5705STDY8058285.filtered_feature_bc_matrix.h5
```

Another example if you would like to use the **cell type annotation** defined by their analysis, which can be included in the **metapath** column. Moreover, the user has the freedom to provide further information such as sex and age:

```
Sample_Name,h5path,metapath,sex,age
5705STDY8058280,~/E-MTAB-11115/data/5705STDY8058280_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058280_annotation.csv,Female,56d
5705STDY8058281,~/E-MTAB-11115/data/5705STDY8058281_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058281_annotation.csv,Female,56d
5705STDY8058282,~/E-MTAB-11115/data/5705STDY8058282_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058282_annotation.csv,Female,56d
5705STDY8058283,~/E-MTAB-11115/data/5705STDY8058283_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058283_annotation.csv,Male,56d
5705STDY8058284,~/E-MTAB-11115/data/5705STDY8058284_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058284_annotation.csv,Male,56d
5705STDY8058285,~/E-MTAB-11115/data/5705STDY8058285_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058285_annotation.csv,Male,56d
```

### MEX input

Since each data has three MEX files (barcodes.tsv.gz, features.tsv.gz and matrix.mtx.gz), here we provide the path to each MEX result directory. The minimal information in the sample meta file:

```
Sample_Name,h5path
miR155cKO_rep1,~/10XGenomicsData/GSM6213399_IT_YZ_0309_2904
WT_rep1,~/10XGenomicsData/GSM6213400_IT_YZ_0309_2905
WT_rep2,~/10XGenomicsData/GSM6213401_IT_YZ_0309_2908
miR155cKO_rep2,~/10XGenomicsData/GSM6213402_IT_YZ_0322_3125
```

Another example with additional information:

```
Sample_Name,h5path,Genotype,Sex
miR155cKO_rep1,~/10XGenomicsData/GSM6213399_IT_YZ_0309_2904,miR155cKO,Female
WT_rep1,~/10XGenomicsData/GSM6213400_IT_YZ_0309_2905,WT,Female
WT_rep2,~/10XGenomicsData/GSM6213401_IT_YZ_0309_2908,WT,Female
miR155cKO_rep2,~/10XGenomicsData/GSM6213402_IT_YZ_0322_3125,miR155cKO,Female
```

## File hierarchy

Before running the pipeline, please check the file hierarchy. Here are examples for h5 and MEX inputs, and we suggest creating a separate folder (such as *processing*) to store the config.yml and sampleMeta.csv files. The output files will be generated in the directory specified in config.yml.

For data downloaded in 10X h5 format:

```
E-MTAB-11115/
    ├── data
        ├── 5705STDY8058280.annotation.csv           (optional)
        ├── 5705STDY8058280.metrics_summary.csv      (optional)
        ├── 5705STDY8058280.raw_feature_bc_matrix.h5
        ├── 5705STDY8058281.annotation.csv           (optional)
        ├── 5705STDY8058281.metrics_summary.csv      (optional)
        ├── 5705STDY8058281.raw_feature_bc_matrix.h5
        ...
    ├── processing
        ├── config.yml
        └── sampleMeta.csv
```

For data downloaded in 10X MEX format:

```
GSE205483/
    ├── GSM6213399_IT_YZ_0309_2904
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ├── GSM6213400_IT_YZ_0309_2905
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ...
    ├── processing
        ├── config.yml
        └── sampleMeta.csv
```
