# Running the pipeline

After preparing the input data and configuration files (config.yml and sampleMeta.csv), we are ready to run the pipeline. Please be aware that scRNASequest is a semi-automated pipeline, and certain steps require the user to review the outputs and determine the parameters. Thus, the same step may need to be run for more than once.

## Initial scAnalyzer run

The initial scAnalyzer run intends to test the default filtering parameters on the single-cell RNA-seq data. We suggest the following settings in the config.yml file to activate the filtering step, but pause the pipeline from additional analysis (such as data integration, harmonization, etc.):

```
# Change the following settings in the config.yml file:
filter_step: True
runAnalysis: False
```

To run scAnalyzer, simply pass the path to the config.yml file to the program:

```{bash, eval=FALSE}
scAnalyzer Path/to/a/config/file

#Example:
scAnalyzer ~/E-MTAB-11115/processing/config.yml
```

Output files:

```
outputdir
    ├── BookdownReport/                #Full bookdown report folder. This is same as the BookdownReport.tar.gz below
    ├── config.yml
    ├── config.yml.20220617.log        #log file
    ├── ProjName_BookdownReport.tar.gz #Bookdown report tar ball. 
                                       #Please download this file and open the index.html file to view the report
    ├── ProjName_raw_prefilter.h5ad    #h5ad after pre-filtering
    ├── prefilter.QC.pdf               #QC plots before filtering. These plots are included in the Bookdown report
    ├── postfilter.QC.pdf              #QC plots after filtering
    └── Rmarkdown/                     #Source files for the Bookdown report
```

## Review the bookdown report

After the initial scAnalyzer run, a bookdown report, named **ProjName_BookdownReport.tar.gz** will be generated. Please open the report (index.html) to review cell numbers and figures before and after the filtering step, and adjust the following thresholds in the config.yml file if needed:

```
# Cell filtering parameters in the config.yml file:
gene_group:
  MT:
    startwith: ["MT-","Mt-","mt-"]
    cutoff: 20           #percentage cutoff to filter out the cells (larger than this cutoff)
    rm: False
  RP:                    #for ribosomal genes if you would like to filter them
    startwith: []
    cutoff: 20
    rm: False
filter_step: True        #if False, the above 'gene_group' filtering will be skipped as well
min.cells: 3             #filtering genes by minimum number of cells with non-zero count for a gene
min.features: 50         #filtering cells by minimal genes detected
highCount.cutoff: 10000  #any cells with higher total counts to be removed
highGene.cutoff: 3000    #any cells with a higher number of detected genes to be removed
```

The **min.cells** cutoff is to filter out *genes* that have low representation in cells, and the **min.features** cutoff is to filter out *cells* that have low number of genes (features) detected. The last two cutoffs, **highCount.cutoff** and **highGene.cutoff** can be used to filter cells if they have too many total counts or UMI detected, possibly due to multiplets.

The user may need to adjust the thresholds and rerun scAnalyzer to visualize the results. This step may be repeated until the threshold is adequate for the data.

Please find an example of the bookdown report here: [E-MTAB-11115 bookdown](https://interactivereport.github.io/scRNAsequest/examples/E-MTAB-11115/bookdown/index.html){target="_blank"}

## Generate a slide deck

This is an optional step if you have already reviewed the Bookdown report. This slide deck is another way to organize and present the same results.

The QC report can be turned into a online slide deck format for better sharing and demonstration. To achieve this, please use the **Slidedeck.Rmd** file under the /src folder and follow the steps below:

```
# create a new directory inside the project, and link the existed files and folder:
cd ~/E-MTAB-11115/processing/
mkdir Slidedeck && cd Slidedeck
ln -s ../BookdownReport/images
ln -s ../BookdownReport/bookdown.info.txt

# Copy the Slidedeck.Rmd file from src directory to the working directory containing config.yml file
cp ~/scRNAsequest/src/Slidedeck.Rmd ~/E-MTAB-11115/processing/Slidedeck

# Source the environment
src="~/scRNAsequest/src/.env"
source $src/.env;eval $condaEnv

# Run the Slidedeck script, and set the output file name using 'output_file' parameter
Rscript -e 'rmarkdown::render("Slidedeck.Rmd", output_file="Output.html")'
```

Please download the whole Slidedeck and open the **Output.html** file in a web browser.

Please find an example of the bookdown report here: [E-MTAB-11115 slidedeck](https://interactivereport.github.io/scRNAsequest/examples/E-MTAB-11115/slidedeck/index.html){target="_blank"}

## Run the full analysis

After the filtering step is done, we are ready to turn on the *runAnalysis* parameter and run the full pipeline:

```
# Change the following setting in the config.yml file:
runAnalysis: True
overwrite: True    #Overwrite the previous results if we run this again, for example, after adjusting some parameters.
```

Please also pay attention to the **reference** data, and if the user would like to turn on the label transfer function for cell type annotation, set the following parameter, for example, using **mouse_spinalCord** as the reference:

```
ref_name: mouse_spinalCord
```

To check all available references, run scAnalyzer without any parameter. To add new references, please see section 8: [Reference Building].

Alternatively, you can provide a path to the **ref.Rds** reference file in the [Azimuth reference format](https://github.com/satijalab/azimuth/wiki/Azimuth-Reference-Format){target="_blank"}:

```
ref_name: path/to/ref.Rds
```

Then run the following command again:

```{bash, eval=FALSE}
scAnalyzer Path/to/a/config/file

#Example:
scAnalyzer ~/E-MTAB-11115/processing/config.yml
```

This step will be time-consuming, and it took ~1 hour for the [E-MTAB-11115](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-11115/){target="_blank"} data (final cell number: 29511) to finish. For a larger dataset, such as [GSE205483](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE205483){target="_blank"} (final cell number: 111,011), the full pipeline took ~7 hours.

## Visualize the harmonized results

After running the pipeline, you will see the following files in the working directory. Each harmonization method (Harmony, Seurat, Liger) will also generate corresponding files.

```
outputdir
    # Input files
        ├── config.yml:     The analysis config file specify the scAnalyzer parameters
        ├── sampleMeta.csv: List all sample information including expression matrix (and sample/cell level meta information)
        └── DEGinfo.csv:    Define the sc DEG (between phenotypes within a cluster annotatino)
    # Log files:
        ├── config.yml.[date].log: All stdout/stderr will be recorded in this dated log file
        └── [Software].error       Standard error files of multiple software, including: 
                                        Harmony, Liger, SeuratRPCA, SeuratRef, SCT, sctHarmony, silhouette, 
    # QC files:
        ├── sequencingQC.csv:  Merged all sequence QC files from individual samples if available
        ├── sequencingQC.pdf:  Plots of sequence QC if they are available
        ├── prefilter.QC.pdf:  Initial data QC plots without filtering
        └── postfilter.QC.pdf: Data QC plots after filtering
    # Initial files:
        ├── [prj_name]_raw_prefilter.h5ad: The internal temporary raw counts created/merged directly from sample before filtering 
        └── [prj_name]_raw.h5ad:           The internal temporary raw counts after the cell/gene filtering
    # SCT:
        ├── [prj_name]_SCT.h5:     The sparse matrix contains SCT expression values
        ├── [prj_name]_SCT.h5.rds: The Seurat object contains counts (RNA assay) and 
                                      SCT expression (SCT assay) along with original meta information from SampleMeta file
        └── [prj_name]_SCT.h5ad:   The SCT expression, along with PCA, tSNE, UMAP reduction layout, will be 
                                      merged into [prj_name].h5ad
    # Seurat_integration:
        ├── [prj_name]_seurat_r2py_integrated.csv.gz:       Expression matrix from seurat integration assay of data entry 
        ├── [prj_name]_seurat_r2py_integrated.genes.csv.gz: Variable genes from integration assay
        └── [prj_name]_Seurat.h5ad:                         The seurat integration with reduction embedding and clustering, 
                                      will be merged into [prj_name].h5ad
    # Seurat_RPCA_integration:
        ├── [prj_name]_seurat_rpca.csv.gz: The reuction embedding and clustering results from seurat RPCA integration 
        └── [prj_name]_SeuratRPCA.h5ad:    The seurat RPCA integration with reduction embedding and clustering, 
                                      will be merged into [prj_name].h5ad
    # Seurat_refMapping:
        ├── [prj_name]_seuratRef.csv:  The reuction embedding and label transfer results from seurat RPCA integration 
        └── [prj_name]_SeuratRef.h5ad: The seurat reference mapping results with reduction embedding, will be merged into [prj_name].h5ad
    # Harmony:
        ├── [prj_name]_harmony_py2r_batch.Rdata: The library/batch information of each cell
        ├── [prj_name]_harmony_py2r_pca.Rdata:   The PCA of each cell
        ├── [prj_name]_harmony_r2py.Rdata:       The harmony PCA results
        └── [prj_name]_Harmony.h5ad:             The harmony integration with reduction embedding and clustering, 
                                      will be merged into [prj_name].h5ad
    # Liger:
        ├── [prj_name]_liger_hvg.csv:  High variable gene list
        ├── [prj_name]_liger.csv":     The liger integration embedding and cluster
        └── [prj_name]_Liger.h5ad":    The liger integration with reduction embedding and clustering, 
                                          will be merged into [prj_name].h5ad
    # scDEG (if you ran the DEG step):
        ├── [prj_name]_scDEG.cmd.json: The scDEG cmd jobs, if scDEG jobs defined in DEGinfo.csv are available
        ├── [prj_name]_scDEG":         The folder contains all scDEG jobs, each sub folder contains scDEG csv files from one entry in DEGinfo.csv
        └── [comparison].error         Standard error files for all comparison groups
    # Other output:
        ├── [prj_name].h5ad:           The final h5ad file which is used for cellxgene VIP load (will be copied into celldepot folder)
        ├── [prj_name].db:             The final h5ad file which is used for cellxgene VIP load (will be copied into celldepot folder)
        └── [prj_name]_raw_added.h5ad: The h5ad file contains raw counts in X along with all annotation and embeding 
                                          (will be copied into celldepot folder)
        ├── [prj_name]_kBET_umap_k0_100.pdf": The kBET evaluation plot for all integration methods
        └── [prj_name]_Silhouette_boxplot_pc50.pdf": The Silhouette evaluation plot for all integration methods
    # Rbookdown:
        ├── Rmarkdown:                 The folder contains png files for R bookdown
        ├── [prj_name]_BookdownReport.tar.gz: R bookdown document tarball for downloading and sharing
        └── BookdownReport:            The folder contains R bookdown reports
    # Parallel directory (if you ran the pipeline in parallel mode):
        └── j[numeric]:                A colder contains parallel jobs submit script and log file
```

[kBET](https://www.nature.com/articles/s41592-018-0254-1){target="_blank"} is a useful metric to evaluate and compare the harmonization results. Here is an example of the **ProjName_kBET_umap_k0_100.pdf** result:

[![](https://interactivereport.github.io/scRNAsequest/tutorial/images/kBET.png)](https://interactivereport.github.io/scRNAsequest/tutorial/images/kBET.png)

Please review the **ProjName.h5ad** file for the harmonization result. If you have [Cellxgene VIP](https://github.com/interactivereport/cellxgene_VIP){target="_blank"} installed, the visualization will be easier.

Here is an example of the ProjName.h5ad project visualized by [Cellxgene VIP](https://github.com/interactivereport/cellxgene_VIP){target="_blank"}:

[![](https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png)](https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png)
