# Differential expression (DE) analysis

Another critical function of the scRNASequest pipeline is the **Differential expression (DE)** analysis. By running the steps in Section: [Running the pipeline], we have performed quality control of the data, and generated an integrated dataset. 

Before running the DE analysis, the **DEGinfo.csv** comparison file needs to be prepared:

[![](https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_empty.png)](https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_empty.png)

The first column, **comparisonName** is designed to store the name of the comparison. These names will be used to create separate folders under the result directory. In the DEG comparison file above, **“sample”**, **“cluster”**, **“group”** and **“covars”** are the annotation headers, and **“alt”** and **“ref”** are the two entries from **“group”** column. The **“covars”** is optional and can be empty.

The DEG is performing between **“alt”** and **“ref”** cells from **“group”** within each entry of **“cluster”** considering **“sample”** variations. The **“group”** variable should contain conditions to compare, such as Mutant v.s. Control. Thus, this pipeline is designed to loop through each cluster, and perform DEG analysis between **“alt”** v.s. **“ref”**.

**Covariate adjustment:**

Adjusting covariates is optional. By default, NEBULA considers the value of **log of total UMI (library size)** as the offset covariate in our pipeline, so you don't need to adjust this again in **"covars"**.

If needed, you can put a formula in the **"covars"** column using the meta information from the sampleMeta.csv file, such as Sex, Age, etc. You can also include QC metrics such as mitochondria percentage (pct_MT), etc. Please find a full list of QC metrics below. You can also open the Cellxgene VIP and check the right side. All numeric covariates are listed there.

```{bash, eval=FALSE}
n_genes_by_counts, log1p_n_genes_by_counts, pct_counts_in_top_50_genes, pct_counts_in_top_100_genes, pct_counts_in_top_200_genes, pct_counts_in_top_500_genes, pct_MT, n_genes, predicted.CellType.score
```

You may consider **pct_MT** as a covariate to adjust, but be careful to adjust other covariates here unless you have a good reason to do so. An example of the formula to be filled out in the **"covars"** column: **Age+Sex+pct_MT**.

**DE software choices:**

Based on the previous [benckmarking analysis](https://www.mdpi.com/2075-1729/12/6/850), NEBULA outperforms other software in singl-cell RNA-seq DE analysis. Thus, our pipeline uses NEBULA by default to run DE. However, we also offer the following methods as well, and you can directly use them in the DEGinfo.csv file, see Example 3 later.

  - **u_test**: Mann–Whitney U test, also called Wilcoxon rank-sum test, is a nonparametric test. See [Wiki](https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test). In our single-cell RNA-seq analysis, this is a pseudobulk method.
  
  - **t_test**: Student's t-test, statistical test based on Student's t-distribution. See [Wiki](https://en.wikipedia.org/wiki/Student%27s_t-test). In our single-cell RNA-seq analysis, this is a pseudobulk method.
  
  - **edgeR**: Using the [edgeR package](https://bioconductor.org/packages/release/bioc/html/edgeR.html) to perform pseudobulk DE analysis. Refer to their [paper](https://academic.oup.com/bioinformatics/article/26/1/139/182458) for more details.
  
  - **limma**: Using the [limma-voom method](http://bioconductor.org/packages/release/bioc/html/limma.html) to perform pseudobulk DE analysis. Refer to their [paper](https://academic.oup.com/nar/article/43/7/e47/2414268) for more details.
  
  - **DESeq2**: Using the [DESeq2 package](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) to perform pseudobulk DE analysis. Refer to their [paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8) for more details.
  
  - **ANCOVA**: Using the [ANCOVA](https://www.r-bloggers.com/2021/10/analysis-of-covariance-ancova-using-r/) to perform pseudobulk DE analysis.

  - **MAST**: Using the [MAST package](https://www.bioconductor.org/packages/release/bioc/html/MAST.html) to perform single-cell DE analysis. Refer to their [paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5) for more details.
  
  - **NEBULA**: Using the [NEBULA software](https://github.com/lhe17/nebula) to perform single-cell DE analysis. Refer to their [paper](https://www.nature.com/articles/s42003-021-02146-6) for more details.
  
  - **glmmTMB**: Using the [glmmTMB (nbinom2 family function)](https://glmmtmb.github.io/glmmTMB/) to perform single-cell DE analysis.
  
  - **limma_cell_level**: This also uses [limma](http://bioconductor.org/packages/release/bioc/html/limma.html) but not a pseudobulk method.
  
Here, this tutorial provides several examples to illustrate the DE analysis in scRNASequest:

We use a UMAP with label transferred cell type annotation as an example dataset. Label transfer is strongly suggested if you would like to run DE analysis because the clusters are more meaningful than the original clusters assigned by the software:

[![](https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png)](https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png)

## Example 1

In the first example, we would like to run DE analysis between **'Female'** and **'Male'** for each cluster annotated by **predicted.celltype1**. In the first column, we input the header name **library_id**, which annotates the data sources. Then we add **predicted.celltype1** in the *cluster* column, which allows the pipeline to loop through each cluster in **predicted.celltype1**. The *group* column contains the header name storing the comparison groups, and here we use the **Sex** annotation. Each time, the pipeline can only compare two conditions, such as **'Female'** and **'Male'**. If the *group* column contains more groups, please list them in multiple lines in the DEGinfo.csv file.

We can also add *covars* if needed, but this is optional.

The default DE analysis is performed by [**NEBULA**](https://www.nature.com/articles/s42003-021-02146-6){target="_blank"}. As for the model, NEBULA provides two choices: **HL** and **LN**. For more details about them, please refer to the [NEBULA manual](https://cran.r-project.org/web/packages/nebula/vignettes/nebula_example.html#difference-between-nebula-ln-and-nebula-hl){target="_blank"}.

Here is the DEGinfo.csv we described above:

```{bash, eval=FALSE}
comparisonName,sample,cluster,group,alt,ref,covars[+ separated],method[default NEBULA],model[default HL]
Compare_Female_vs_Male,library_id,predicted.celltype1,Sex,Female,Male,,NEBULA,HL
```

After preparing the DEGinfo.csv file, simply add it to the **config.yml** file:

```
DEG_desp: ~/E-MTAB-11115/processing/DEGinfo.csv
```

Then rerun the pipeline. The pipeline won't run the previous steps (i.e. filtering, harmonization, etc.) again, so it will directly run the DE analysis based on the harmonized h5ad file.

```{bash, eval=FALSE}
scAnalyzer Path/to/a/config/file

#Example:
scAnalyzer ~/E-MTAB-11115/processing/config.yml
```

This step may take a few hours to run. The output files will be generated in the directory:

```
outputdir
    ... previous results         #Omitted, see section 6.4
    ├── ProjName_scDEG           #Folder containing DEG results
      ├── Compare_Female_vs_Male
        ├── Female.vs.Male_predicted.celltype1:astrocytes_NEBULA.csv    #Each comparison has three associated files
        ├── Female.vs.Male_predicted.celltype1:astrocytes_NEBULA.png
        ├── Female.vs.Male_predicted.celltype1:astrocytes_NEBULA.QC.pdf
        ├── Female.vs.Male_predicted.celltype1:discarded_NEBULA.csv
        ├── Female.vs.Male_predicted.celltype1:discarded_NEBULA.png
        ├── Female.vs.Male_predicted.celltype1:discarded_NEBULA.QC.pdf
        ...
        └── env.rds
```

An example of the DE csv table (Female.vs.Male_predicted.celltype1/astrocytes_NEBULA.csv, top 5 rows):

In the table below, the most significant gene is [Xist](https://www.ncbi.nlm.nih.gov/gene/213742){target="_blank"}, a well-known X-chromosome gene highly expressed in female.

```{r, echo=FALSE, results = "asis"}
suppressMessages(library(knitr))
suppressMessages(library(kableExtra))
Data <- data.frame(res.tab.ID=c("Xist","Pnpla7","Etnppl","Zbtb16","Sorcs2"),
                   res.tab.log2FC=c(9.676171398,-1.230034425,-1.592057996,-1.448309033,1.137956234),
                   res.tab.Pvalue=c(8.87E-114,1.88E-89,1.29E-72,2.12E-60,3.29E-51),
                   res.tab.FDR=c(4.16E-110,4.40E-86,2.01E-69,2.49E-57,3.08E-48),
                   res.tab.algorithm=rep("HL",5),
                   res.tab.convergence=rep(1,5),
                   res.tab.metric=c(6115.17801,1653.155284,525.8328005,1260.951787,1398.088133),
                   res.ls.summary.logFC_.Intercept.=c(-6.693213224,-9.32119215,-8.511808196,-9.896064065,-8.573612795),
                   res.ls.summary.logFC_.GrouPalt=c(0.127673738,-0.169455002,0.091793885,-0.091801372,-0.225284796),
                   res.ls.summary.se_.Intercept.=c(0.023462196,0.053417992,0.028283189,0.058472814,0.030386881),
                   res.ls.summary.se_.GrouPalt=c(0.046939517,0.106517692,0.05650592,0.116554902,0.060462474),
                   res.ls.summary.p_.Intercept.=rep(0,5),
                   res.ls.summary.p_.GrouPalt=c(0.006528926,0.111640522,0.10426907,0.430917004,0.000194522),
                   res.ls.summary.gene_id=1:5,
                   res.ls.summary.gene=c("Rgs20","Atp6v1h","Rb1cc1","4732440D04Rik","Pcmtd1"),
                   res.ls.overdispersion.Subject=c(0.001672827,0.006432008,1.00E-04,0.001474488,1.00E-04),
                   res.ls.overdispersion.Cell=c(0.43931521,0.332731593,0.172511335,0.728850422,0.350258671),
                   res.ls.convergence=c(1,1,1,1,-10),
                   res.ls.algorithm=rep("NBGMM (HL)",5)
                   )
kable(Data, caption = "DEG analysis of Female.vs.Male in astrocytes", align = "c", booktabs = TRUE) %>%
  kable_styling() %>%
  kableExtra::scroll_box(width = "100%")
```

An example of the DE png file (Female.vs.Male_predicted.celltype1/astrocytes_NEBULA.png):

[![](https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Female_vs_Male_Astrocytes.png)](https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Female_vs_Male_Astrocytes.png)

## Example 2

In the second example, we are going to show how to perform DE analysis **between two cell types**. To achieve this, we added one artificial column (called 'SelectAll') in the **sampleMeta.csv** file, and assigned the same values ('Everything') to all the data:

```
Sample_Name,h5path,Sex,SelectAll
5705STDY8058280,/home/ysun4/testSinglecell/ExternalData/5705STDY8058280_filtered_feature_bc_matrix.h5,Female,Everything
5705STDY8058281,/home/ysun4/testSinglecell/ExternalData/5705STDY8058281_filtered_feature_bc_matrix.h5,Female,Everything
5705STDY8058282,/home/ysun4/testSinglecell/ExternalData/5705STDY8058282_filtered_feature_bc_matrix.h5,Female,Everything
5705STDY8058283,/home/ysun4/testSinglecell/ExternalData/5705STDY8058283_filtered_feature_bc_matrix.h5,Male,Everything
5705STDY8058284,/home/ysun4/testSinglecell/ExternalData/5705STDY8058284_filtered_feature_bc_matrix.h5,Male,Everything
5705STDY8058285,/home/ysun4/testSinglecell/ExternalData/5705STDY8058285_filtered_feature_bc_matrix.h5,Male,Everything
```

Then we prepare the **DEGinfo.csv** file in this way. The 'SelectAll' was filled into the *cluster* column, and since it only has one value, it actually selects all cells when running the pipeline. Then it performs **astrocytes** to **Oligos** comparison.

```
comparisonName,sample,cluster,group,alt,ref,covars[+ separated],method[default NEBULA],model[default HL]
Comparison_Ast_vs_Oligo,library_id,SelectAll,predicted.celltype1,astrocytes,Oligos,,NEBULA,HL
```

Then we need to run the scRNASequest pipeline from the beginning, which allows the 'SelectAll' column to be embedded into the harmonized h5ad output. As long as we have DEGinfo.csv included in the config.yml, it will run through the DE analysis and produce the following results:

```
outputdir
    ... previous results         #Omitted, see section 6.4
    ├── ProjName_scDEG           #Folder containing DEG results
      ├── Comparison_Ast_vs_Oligo
        ├── astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.csv    #Each comparison has three associated files
        ├── astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.png
        ├── astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.QC.pdf
        ├── env.rds
    ├── ProjName_scDEG.cmd.json
    ├── deg6602_SelectAll_predicted.celltype1_NEBULA_HL_Everything_astrocytes_Oligos.error  #Standard error messages
    └── ProjName_scDEG.db
```

An example of the DE csv table (astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.csv, top 5 rows):

```{r, echo=FALSE, results = "asis"}
suppressMessages(library(knitr))
suppressMessages(library(kableExtra))
Data <- data.frame(res.tab.ID=c("Xkr4","Rgs20","St18","Adhfe1","Prex2"),
                   res.tab.log2FC=c(-3.327922728,6.386204605,-5.129500223,4.842191486,8.368190343),
                   res.tab.Pvalue=rep(0,5),
                   res.tab.FDR=rep(0,5),
                   res.tab.algorithm=rep("HL",5),
                   res.tab.convergence=rep(1,5),
                   res.tab.metric=c(4177.155918,5561.771074,14283.48725,12162.13605,8249.837425),
                   res.ls.summary.logFC_.Intercept.=c(-8.168075319,-9.825702638,-10.32269605,-10.16999644,-10.29222068),
                   res.ls.summary.logFC_.GrouPalt=c(-2.306740256,-2.50660294,0.217648395,0.111911715,4.426579716),
                   res.ls.summary.se_.Intercept.=c(0.036792379,0.089054801,0.031092244,0.029267929,0.051346211),
                   res.ls.summary.se_.GrouPalt=c(0.056102205,0.135391605,0.0690489,0.066976238,0.056385976),
                   res.ls.summary.p_.Intercept.=rep(0,5),
                   res.ls.summary.p_.GrouPalt=c(0,1.60E-76,0.001621059,0.09473813,0),
                   res.ls.summary.gene_id=1:5,
                   res.ls.summary.gene=c("Xkr4","Gm1992","Mrpl15","Tcea1","Rgs20"),
                   res.ls.overdispersion.Subject=c(0.006729713,0.039810264,1.00E-04,1.00E-04,0.004776775),
                   res.ls.overdispersion.Cell=c(0.607630339,1.494756075,0.418872119,0.721244774,0.456359428),
                   res.ls.convergence=c(1,1,1,1,1),
                   res.ls.algorithm=rep("NBGMM (HL)",5)
                   )
kable(Data, caption = "DEG analysis of astrocytes.vs.Oligos", align = "c", booktabs = TRUE) %>%
  kable_styling() %>%
  kableExtra::scroll_box(width = "100%")
```

An example of the DE png file (astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.png):

[![](https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Astrocytes_Oligos.png)](https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Astrocytes_Oligos.png)
