<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</title>
  <meta name="description" content="scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta name="generator" content="bookdown #bookdown:version# and GitBook 2.6.7" />

  <meta property="og:title" content="scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  
  
  



<meta name="date" content="2023-03-21" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



<!--bookdown:title:start-->
<div id="header">
<h1 class="title">scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</h1>
<p class="author multi-author"><em>Kejie Li</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:kejie.li@biogen.com" class="email">kejie.li@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Yu (Henry) Sun</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:yuhenry.sun@biogen.com" class="email">yuhenry.sun@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Zhengyu Ouyang</em></p>
<address class="author_afil">
BioInfoRx<br>
<a class="author_email" href="mailto:#"><a href="mailto:oyoung@bioinforx.com" class="email">oyoung@bioinforx.com</a></a>
</address>
<p class="author multi-author"><em>Soumya Negi</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:soumya.negi@biogen.com" class="email">soumya.negi@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Wanli Wang</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:wanli.wang@biogen.com" class="email">wanli.wang@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Yirui Chen</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:yirui.chen@biogen.com" class="email">yirui.chen@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Sarbottam Piya</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:sarbottam.piya@biogen.com" class="email">sarbottam.piya@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Wenxing Hu</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:wenxing.hu@biogen.com" class="email">wenxing.hu@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Maria Zavodszky</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:maria.zavodszky@biogen.com" class="email">maria.zavodszky@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Hima Yalamanchili</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:hima.yalamanchili@biogen.com" class="email">hima.yalamanchili@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Shaolong Cao</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:shaolong.cao@biogen.com" class="email">shaolong.cao@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Andrew Gehrke</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:andrew.gehrke@biogen.com" class="email">andrew.gehrke@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Mark Sheehan</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:make.sheehan@biogen.com" class="email">make.sheehan@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Dann Huh</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:dann.huh@biogen.com" class="email">dann.huh@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Fergal Casey</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:fergal.casey@biogen.com" class="email">fergal.casey@biogen.com</a></a>
</address>
<p class="author multi-author"><em>Xinmin Zhang</em></p>
<address class="author_afil">
BioInfoRx<br>
<a class="author_email" href="mailto:#"><a href="mailto:xinmin@bioinforx.com" class="email">xinmin@bioinforx.com</a></a>
</address>
<p class="author multi-author"><em>Baohong Zhang (Corresponding Author)</em></p>
<address class="author_afil">
Biogen<br>
<a class="author_email" href="mailto:#"><a href="mailto:baohong.zhang@biogen.com" class="email">baohong.zhang@biogen.com</a></a>
</address>
<p class="author multi-author"><em>These authors contribute equally Kejie Li, Yu (Henry) Sun, Zhengyu Ouyang</em></p>
<p class="date"><em>2023-03-21</em></p>
</div>
<!--bookdown:title:end-->

<!--bookdown:toc:start-->
  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">
<!--bookdown:toc2:start-->
<ul>
<li><a href="#preface"><span class="toc-section-number">1</span> Preface<span></span></a></li>
<li><a href="#introduction"><span class="toc-section-number">2</span> Introduction<span></span></a></li>
<li><a href="#installation"><span class="toc-section-number">3</span> Installation<span></span></a>
<ul>
<li><a href="#install-scrnasequest"><span class="toc-section-number">3.1</span> Install scRNASequest<span></span></a>
<ul>
<li><a href="#installation-using-conda"><span class="toc-section-number">3.1.1</span> Installation using Conda<span></span></a></li>
<li><a href="#installation-through-docker"><span class="toc-section-number">3.1.2</span> Installation through Docker<span></span></a></li>
</ul></li>
<li><a href="#configure-sys.yml-file"><span class="toc-section-number">3.2</span> Configure sys.yml file<span></span></a></li>
<li><a href="#install-cellxgene-vip"><span class="toc-section-number">3.3</span> Install cellxgene VIP<span></span></a></li>
<li><a href="#install-celldepot"><span class="toc-section-number">3.4</span> Install CellDepot<span></span></a></li>
</ul></li>
<li><a href="#demo-run"><span class="toc-section-number">4</span> Demo run<span></span></a>
<ul>
<li><a href="#demo-run-with-conda"><span class="toc-section-number">4.1</span> Demo run with Conda<span></span></a></li>
<li><a href="#demo-run-with-docker"><span class="toc-section-number">4.2</span> Demo run with Docker<span></span></a></li>
</ul></li>
<li><a href="#data-preparation"><span class="toc-section-number">5</span> Data preparation<span></span></a>
<ul>
<li><a href="#public-data-in-h5-format"><span class="toc-section-number">5.1</span> Public data in h5 format<span></span></a></li>
<li><a href="#public-data-in-10x-mex-format"><span class="toc-section-number">5.2</span> Public data in 10X MEX format<span></span></a></li>
<li><a href="#self-prepared-files"><span class="toc-section-number">5.3</span> Self-prepared files<span></span></a></li>
</ul></li>
<li><a href="#input-configuration"><span class="toc-section-number">6</span> Input configuration<span></span></a>
<ul>
<li><a href="#generate-template-files"><span class="toc-section-number">6.1</span> Generate template files<span></span></a></li>
<li><a href="#prepare-the-config.yml-file"><span class="toc-section-number">6.2</span> Prepare the config.yml file<span></span></a></li>
<li><a href="#section52"><span class="toc-section-number">6.3</span> Prepare the sample meta file<span></span></a>
<ul>
<li><a href="#h5-input"><span class="toc-section-number">6.3.1</span> h5 input<span></span></a></li>
<li><a href="#mex-input"><span class="toc-section-number">6.3.2</span> MEX input<span></span></a></li>
</ul></li>
<li><a href="#file-hierarchy"><span class="toc-section-number">6.4</span> File hierarchy<span></span></a></li>
</ul></li>
<li><a href="#running-the-pipeline"><span class="toc-section-number">7</span> Running the pipeline<span></span></a>
<ul>
<li><a href="#initial-scanalyzer-run"><span class="toc-section-number">7.1</span> Initial scAnalyzer run<span></span></a></li>
<li><a href="#review-the-bookdown-report"><span class="toc-section-number">7.2</span> Review the bookdown report<span></span></a></li>
<li><a href="#generate-a-slide-deck"><span class="toc-section-number">7.3</span> Generate a slide deck<span></span></a></li>
<li><a href="#run-the-full-analysis"><span class="toc-section-number">7.4</span> Run the full analysis<span></span></a></li>
<li><a href="#visualize-the-harmonized-results"><span class="toc-section-number">7.5</span> Visualize the harmonized results<span></span></a></li>
</ul></li>
<li><a href="#differential-expression-de-analysis"><span class="toc-section-number">8</span> Differential expression (DE) analysis<span></span></a>
<ul>
<li><a href="#example-1"><span class="toc-section-number">8.1</span> Example 1<span></span></a></li>
<li><a href="#example-2"><span class="toc-section-number">8.2</span> Example 2<span></span></a></li>
</ul></li>
<li><a href="#reference-building"><span class="toc-section-number">9</span> Reference building<span></span></a>
<ul>
<li><a href="#initialize-scref"><span class="toc-section-number">9.1</span> Initialize scRef<span></span></a></li>
<li><a href="#submit-scref"><span class="toc-section-number">9.2</span> Submit scRef<span></span></a></li>
<li><a href="#demo-scref"><span class="toc-section-number">9.3</span> Demo scRef<span></span></a></li>
</ul></li>
<li><a href="#celldepot-publishing"><span class="toc-section-number">10</span> CellDepot publishing<span></span></a>
<ul>
<li><a href="#initialize-sc2celldepot"><span class="toc-section-number">10.1</span> Initialize sc2celldepot<span></span></a></li>
<li><a href="#run-sc2celldepot"><span class="toc-section-number">10.2</span> Run sc2celldepot<span></span></a></li>
<li><a href="#demo-sc2celldepot"><span class="toc-section-number">10.3</span> Demo sc2celldepot<span></span></a></li>
</ul></li>
<li><a href="#additional-tools"><span class="toc-section-number">11</span> Additional tools<span></span></a>
<ul>
<li><a href="#scdeg"><span class="toc-section-number">11.1</span> scDEG<span></span></a>
<ul>
<li><a href="#initialize-scdeg"><span class="toc-section-number">11.1.1</span> Initialize scDEG<span></span></a></li>
<li><a href="#run-de-analysis"><span class="toc-section-number">11.1.2</span> Run DE analysis<span></span></a></li>
</ul></li>
<li><a href="#sctool"><span class="toc-section-number">11.2</span> scTool<span></span></a>
<ul>
<li><a href="#remove-annotations"><span class="toc-section-number">11.2.1</span> Remove annotations<span></span></a></li>
<li><a href="#add-annotations"><span class="toc-section-number">11.2.2</span> Add annotations<span></span></a></li>
<li><a href="#export-gene-expressions"><span class="toc-section-number">11.2.3</span> Export gene expressions<span></span></a></li>
</ul></li>
</ul></li>
</ul>
<!--bookdown:toc2:end-->
      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:toc:end-->
<!--bookdown:body:start-->
<div id="preface" class="section level1 hasAnchor" number="1">
<h1 class="hasAnchor"><span class="header-section-number">1</span> Preface<a href="#preface" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This is a <em>user manual</em> written using Bookdown, which provides a detailed guide for scRNASequest.</p>
<!--chapter:end:index.Rmd-->
</div>
<div id="introduction" class="section level1 hasAnchor" number="2">
<h1 class="hasAnchor"><span class="header-section-number">2</span> Introduction<a href="#introduction" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>scRNASequest is a semi-automated single-cell RNA-seq (scRNA-seq) data analysis workflow which allows the following five different functionalities:</p>
<ul>
<li><p>Pre-processing from single-cell RNA sequencing UMI count matrix data (better generated with Cell Ranger)</p></li>
<li><p>Applying multiple harmonization methods for batch correction</p></li>
<li><p>Reference-dataset-based cell type label transfer and embedding projection</p></li>
<li><p>Multi-sample multi-condition single-cell level differential gene expression analysis</p></li>
<li><p>Seamless integration with <a href="https://github.com/interactivereport/cellxgene_VIP" target="_blank">cellxgene VIP</a> for visualization and with <a href="https://github.com/interactivereport/CellDepot" target="_blank">CellDepot</a> for data hosting and sharing by generating a compatible h5ad file. Users have the option to run it on a local laptop computer or interact with sge/slurm schedulers on high performance computing (HPC) clusters.</p></li>
</ul>
<p>This pipeline contains the following programs:</p>
<ul>
<li><p><strong>scAnalyzer</strong></p>
<p>This is the main script in scRNASequest for single-cell data processing and analysis.</p>
<p>Also, scAnalyzer is an all-in-one program for downstream data analysis, including pre-processing, batch correction, label transfer, differential gene expression analysis and CellDepot integration. Moreover, scAnalyzer is embedded with a Bookdown report generator, named <strong>scReport</strong>, which can produce a user-friendly, well-structured quality control report after each run.</p></li>
<li><p><strong>scDEG</strong></p>
<p>scDEG is a standalone pipeline for differential expression (DE) analysis using harmonized data. It has been embedded into the scAnalyzer pipeline.</p></li>
<li><p><strong>scRef</strong></p>
<p>scRef generates a reference dataset for label transfer.</p>
<p>For each reference dataset, this program only needs to run once, and the reference h5ad file will be placed in a permanent path.</p></li>
<li><p><strong>sc2celldepot</strong></p>
<p>sc2celldepot offers a easy and fast way to convert your own RDS file or h5 results+cell annotation to Cellxgene VIP, without running the full pipeline. If you have run the full pipeline script (scAnalyzer), you should get the h5ad (UMI, embedding, etc.) and db (DE information) files for CellDepot publishing, and you won’t need to run sc2celldepot.</p></li>
<li><p><strong>scTool</strong></p>
<p>scTool provides flexibility for modifying the h5ad file, such as modifying annotation and extracting meta information.</p></li>
<li><p><strong>scRMambient</strong></p>
<p>scRMambient is a wrapper function to remove ambient RNAs using CellBender.</p></li>
</ul>
<p><strong>A detailed comparison of the above scripts can be found below:</strong></p>
<table>
<caption>
(#tab:unnamed-chunk-1)Pipeline scripts and their function
</caption>
<thead>
<tr>
<th style="text-align:center;">
Command
</th>
<th style="text-align:center;">
Description
</th>
<th style="text-align:center;">
Input
</th>
<th style="text-align:center;">
Output
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center;">
scAnalyzer
</td>
<td style="text-align:center;">
Main program to perform full scRNA-seq data anal-ysis with QC and data harmonization
</td>
<td style="text-align:center;">
A path to analysis config file (config.yml)
</td>
<td style="text-align:center;">
Final analysis results in h5ad and h5seurat format
</td>
</tr>
<tr>
<td style="text-align:center;">
scDEG
</td>
<td style="text-align:center;">
Program to perform DEG analysis between two phenotypes within each cluster of an annotation (such as cell types)
</td>
<td style="text-align:center;">
A path to a DEG config file (config_DEG.yml)
</td>
<td style="text-align:center;">
One DEG table for each cluster and an SQLite db file of all comparisons
</td>
</tr>
<tr>
<td style="text-align:center;">
scRef
</td>
<td style="text-align:center;">
Program to create Seurat ‘Azimuth’ references
</td>
<td style="text-align:center;">
A path to a reference con-fig file (config_ref.yml)
</td>
<td style="text-align:center;">
An RDS object with ‘Azi-muth’ reference for scAnalyzer
</td>
</tr>
<tr>
<td style="text-align:center;">
sc2celldepot
</td>
<td style="text-align:center;">
Program to transfer ana-lyzed data into h5ad for the cellxgene VIP (CellDepot) loading
</td>
<td style="text-align:center;">
A path to a data config file (config_convert.yml)
</td>
<td style="text-align:center;">
An h5ad file
</td>
</tr>
<tr>
<td style="text-align:center;">
scTool
</td>
<td style="text-align:center;">
Tool to add, remove or export express/annotation from an h5ad
</td>
<td style="text-align:center;">
A path to an h5ad file
</td>
<td style="text-align:center;">
A modified h5ad file or a csv file
</td>
</tr>
<tr>
<td style="text-align:center;">
scRMambient
</td>
<td style="text-align:center;">
Remove ambient RNA by the CellBender
</td>
<td style="text-align:center;">
A path to a sample metadata file containing paths to raw (unfiltered) UMI along with a few Cell-Bender parameters
</td>
<td style="text-align:center;">
CellBender filtered UMI counts in h5 format
</td>
</tr>
</tbody>
</table>
<!--chapter:end:02-Introduction.Rmd-->
</div>
<div id="installation" class="section level1 hasAnchor" number="3">
<h1 class="hasAnchor"><span class="header-section-number">3</span> Installation<a href="#installation" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="install-scrnasequest" class="section level2 hasAnchor" number="3.1">
<h2 class="hasAnchor"><span class="header-section-number">3.1</span> Install scRNASequest<a href="#install-scrnasequest" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We provide two methods to install scRNASequest. The first one uses <a href="https://docs.conda.io/en/latest/">Conda</a>, and the second one uses <a href="https://www.docker.com/">Docker</a>. We have tested both methods on Linux servers; however, if you are a Mac user, please use the Docker method.</p>
<div id="installation-using-conda" class="section level3 hasAnchor" number="3.1.1">
<h3 class="hasAnchor"><span class="header-section-number">3.1.1</span> Installation using Conda<a href="#installation-using-conda" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First, please make sure you have <a href="https://docs.conda.io/en/latest/">Conda</a> installed, or, <a href="https://docs.anaconda.com/anaconda/install/index.html">Anaconda</a>/<a href="https://docs.conda.io/en/latest/miniconda.html">Miniconda</a> installed:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span> conda</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Your conda path will be returned</span></span></code></pre></div>
<p>Then, we choose a directory and install scRNASequest by downloading the source code from GitHub.</p>
<p>The directory you choose here will be the future directory of this pipeline.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Go to the directory you choose. This tutorial uses $HOME (~) directory as an example:</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/interactivereport/scRNASequest.git</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> scRNASequest</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Install scRNASequest conda environment</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Before running this, please make sure you have conda installed before</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># This step will take a while, usually between 30min to 1h depending on the internet speed</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Thank you for your patience</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> install.sh</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># The .env will be created under the src directory</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ~/scRNASequest/src/.env</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Now the pipeline scripts under the scRNASequest folder can be used</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Users can add the scRNASequest directory to the environment permanently</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="co"># by editing ~/.bash_profile or ~/.bashrc</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="ex">vim</span> ~/.bash_profile</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the full path of the scRNASequest directory to $PATH.</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="co"># In our example, this will be: ~/scRNASequest</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="va">PATH</span><span class="op">=</span><span class="va">$PATH</span>:~/scRNASequest</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Close the vim text editor and source the file</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/.bash_profile</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="co">#To verify the installation, type the main program name, and the following message will show up:</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="co">#Output:</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a><span class="ex">=====</span></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a><span class="ex">Please</span> set the sys.yml in ~/scRNASequest.</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a><span class="ex">An</span> example is <span class="st">&#39;~/scRNASequest/src/sys_example.yml&#39;</span>.</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a><span class="ex">=====</span></span></code></pre></div>
<p>You got the above message because the <code>sys.yml</code> file is missing under the pipeline <code>src</code> directory (in our case, ~/scRNASequest/src/).</p>
<p>Please copy the <a href="https://github.com/interactivereport/scRNAsequest/blob/main/src/sys_example.yml">sys_example.yml</a> template there first:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ~/scRNASequest/src/sys_example.yml ~/scRNASequest/src/sys.yml</span></code></pre></div>
<p>Then, fill in the following required items. These directories will be used to host final results (.h5ad files) of the pipeline as well as the reference files for cell type label transfer.</p>
<p>Since we have <code>cellxgenedir</code> and <code>ref</code> directories created under the <code>demo</code> directory, we use them here:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">celldepotDir:</span> ~/scRNASequest/demo/cellxgenedir.  <span class="co"># the absolute path to the cellxgene VIP host folder, where the h5ad files will be copied to for cellxgene VIP</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">refDir:</span> ~/scRNASequest/demo/ref   <span class="co"># the absolute path to the Seurat reference folder if building reference is desired</span></span></code></pre></div>
<p>You may fill in the cellxgene VIP server path after installing cellxgene VIP later, but this is not required for running the pipeline. We leave this empty here.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">celldepotHttp:</span> <span class="co"># the cellxgene host (with --dataroot option) link  http://HOST:PORT/d/</span></span></code></pre></div>
<p>You may change the information in the sys.yml file later, following the <a href="https://interactivereport.github.io/scRNAsequest/tutorial/docs/installation.html#configure-sys.yml-file">full tutorial here</a>.</p>
<p>Then type the name of the main program, <code>scAnalyzer</code> again:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">*****</span> 2023-03-14 14:48:13 <span class="pp">*****</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co">## scRNAsequest: https://github.com/interactivereport/scRNAsequest.git</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Path: /mnt/depts/dept04/compbio/projects/ndru_projects/Software/scRNAsequest</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Date: 2023-03-01 10:10:55 -0500</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co">## git HEAD: d067bfd6dc056597d046a45f3b3b927dd122dd82</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span> /path/to/a/DNAnexus/download/folder === or === scAnalyzer /path/to/a/config/file</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="ex">The</span> config file will be generated automatically when a DNAnexus download folder is provided</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="ex">Available</span> reference data:</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>    <span class="ex">human_cortex:</span> more information @ https://azimuth.hubmapconsortium.org/references/</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="ex">If</span> one of the above can be used as a reference for your datasets, please update the config file with the name in <span class="st">&#39;ref_name&#39;</span>.</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="ex">Powered</span> by None</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="ex">------------</span></span></code></pre></div>
<p>The installation was successful if you see the above message.</p>
<p>Typing other scripts’ name, such as scRef, directly without any parameters will activate the user manual page:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scRef</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="ex">*****</span> 2023-01-26 15:29:54 <span class="pp">*****</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co">## scRNAsequest: https://github.com/interactivereport/scRNAsequest.git</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Path: /mnt/depts/dept04/compbio/edge_tools/scRNAsequest</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Date: 2023-01-13 17:41:49 -0500</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co">## git HEAD: 3d463e0b127af499942b7adc2fc5af6ddfc6f11e</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="ex">Loading</span> resources</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> /path/to/a/output/folder === or === scRef /path/to/a/Ref/config/file</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="ex">The</span> folder has to be existed.</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a><span class="ex">The</span> Ref config file will be generated automatically when a path is provided</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a><span class="ex">=====</span> CAUTION =====</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>    <span class="ex">1.</span> This process will add a seurat reference data into the scRNAsequest pipeline PERMANENTLY!</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>    <span class="ex">2.</span> Make sure the data provided for reference building is SCT transformed!</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a><span class="ex">Powered</span> by the Research Data Sciences group [zhengyu.ouyang@biogen.com<span class="kw">;</span><span class="ex">kejie.li@biogen.com]</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a><span class="ex">------------</span></span></code></pre></div>
<p>Similary, you will also see the manual message printed out for other scripts: scDEG, sc2celldepot, and scTool.</p>
</div>
<div id="installation-through-docker" class="section level3 hasAnchor" number="3.1.2">
<h3 class="hasAnchor"><span class="header-section-number">3.1.2</span> Installation through Docker<a href="#installation-through-docker" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We provide a <a href="https://www.docker.com/">Docker</a> image here: <a href="https://hub.docker.com/repository/docker/sunyumail93/scrnasequest/general" class="uri">https://hub.docker.com/repository/docker/sunyumail93/scrnasequest/general</a>. Users can pull this image to build a container, which has been tested on both Linux and Mac systems. This will take roughly 10 minutes to set up.</p>
<p>We also provide a <a href="https://github.com/interactivereport/scRNAsequest/blob/main/Dockerfile">Dockerfile</a> if you would like to build the image from scratch using the <code>docker build</code> command, which takes ~30 min.</p>
<p>First, please make sure Docker has been installed and can be recognized through the command line:</p>
<pre><code>which docker
# Your docker path will be returned</code></pre>
<p>Go to the directory you choose. This tutorial uses $HOME (~) directory as an example:</p>
<pre><code>cd ~
git clone https://github.com/interactivereport/scRNASequest.git
cd scRNASequest</code></pre>
<p>Then we pull the docker image. This step takes ~10 min.</p>
<pre><code>docker pull sunyumail93/scrnasequest</code></pre>
<p>Initiate the docker container. This command uses -v to map the <code>demo</code> directory under scRNASequest to <code>/demo</code> in the container:</p>
<pre><code>docker run -v `pwd`/demo:/demo -d sunyumail93/scrnasequest</code></pre>
<p>The above command prepars for the demo run in 4. You can mount any directory containing your data to the Docker container using the syntax old_dir:container_dir.</p>
<p>Verify your container:</p>
<pre><code>docker container ls

#Results:
CONTAINER ID   IMAGE                      COMMAND                  CREATED          STATUS          PORTS     NAMES
4e0f3a40ce1d   sunyumail93/scrnasequest   &quot;/bin/sh -c &#39;while t…&quot;   54 seconds ago   Up 52 seconds             interesting_lewin</code></pre>
<p>The last column is the <container_name>, and it will be used in the following steps.</p>
<p>Now we launch the main program of this pipeline. In our example, <container_name> is interesting_lewin. Please substitute to yours:</p>
<pre><code>docker exec -t -i &lt;container_name&gt; scAnalyzer

#Output:
=====
Please set the sys.yml in /home/scRNASequest/src.
An example is &#39;/home/scRNASequest/src/sys_example.yml&#39;.
=====</code></pre>
<p>This is because the sys.yml configuration file is missing under the src directory. There is a sys.yml file prepared for running the demo data (see section 2.2), and you can copy it to the pipeline src directory using the command below. It will work for future analysis also. However, you may change the information in the sys.yml file later, following the <a href="https://interactivereport.github.io/scRNAsequest/tutorial/docs/installation.html#configure-sys.yml-file">full tutorial here</a>. Please note that both ‘/demo/sys.yml’ and ‘/home/scRNASequest/src’ directories are in the container, rather than in your file system.</p>
<pre><code>docker exec -t -i &lt;container_name&gt; cp /demo/sys.yml /home/scRNASequest/src/</code></pre>
<p>Now we run this command again, and we will see the pipeline message printed out, same as the one at the end of section 3.1.1:</p>
<pre><code>docker exec -t -i &lt;container_name&gt; scAnalyzer</code></pre>
</div>
</div>
<div id="configure-sys.yml-file" class="section level2 hasAnchor" number="3.2">
<h2 class="hasAnchor"><span class="header-section-number">3.2</span> Configure sys.yml file<a href="#configure-sys.yml-file" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The <strong>sys.yml</strong> file contains critical information for the pipeline, which must be set up before running the pipeline. This file only needs to create once, under the src/ directory. You can use this <a href="https://github.com/interactivereport/scRNAsequest/blob/main/src/sys_example.yml" target="_blank">sys_example.yml</a> file as a template to prepare your file. The file name must be <strong>sys.yml</strong> and is located under the src/ directory of the pipeline.</p>
<p>The first two rows are related to cellxgene VIP and CellDepot. The <code>celldepotDir</code> is required because all .h5ad files will be copied to this directory, and when launching cellxgene VIP, it will look for files in this directory. Please look into the tutorial below to learn how to set up cellxgene VIP. After setting up cellxgene VIP, you will know the HOST and PORT of your links, which can be included in <code>celldepotHttp</code>. The scAnalyzer pipeline won’t generate error if you don’t have <code>celldepotHttp</code> set up.</p>
<p>In our demo dataset (demo/ directory under the pipeline folder), we provide the following information:</p>
<pre><code>celldepotDir: /demo/cellxgenedir # the absolute path to the cellxgene VIP host folder, where the h5ad files will be copied to for cellxgene VIP
celldepotHttp: http://HOST:PORT/d/ # the cellxgene host (with --dataroot option) link  http://HOST:PORT/d/
refDir: /demo/ref # the absolute path to the seurat refrence folder if building reference is desired
minCell: 50</code></pre>
<p>The <code>refDir</code> stores all Azimuth reference files for cell type label transfer. Please select a directory for this. The <code>scRef</code> pipeline will use this directory and copy all necessary files there. scRef will also modify this sys.yml file when new reference is added. An example of added reference is shown as:</p>
<pre><code>- human_cortex
human_cortex:
  ref_file: https://zenodo.org/record/4546932/files/ref.Rds # local file absolute path can be provided as well
  ref_link: https://azimuth.hubmapconsortium.org/references/
  ref_src: single nuclei
  ref_platform: SNARE-seq2
  ref_assay: refAssay
  ref_neighbors: refdr.annoy.neighbors
  ref_reduction: refDR
  ref_reduction.model: refUMAP
  ref_label:
  - class
  - cluster
  - subclass
  - cross_species_cluster</code></pre>
<p>The <code>powerby</code> parameter sets the ending message after running the pipeline. An example would be:</p>
<pre><code>powerby: the Research Data Sciences Group at Biogen [zhengyu.ouyang@biogen.com;yuhenry.sun@biogen.com] # the message at the end</code></pre>
<p>An example of the message showing:</p>
<pre><code>Powered by the Research Data Sciences Group at Biogen [zhengyu.ouyang@biogen.com;yuhenry.sun@biogen.com]</code></pre>
<p>If you leave <code>powerby</code> as blank, it will show as: Powered by None.</p>
</div>
<div id="install-cellxgene-vip" class="section level2 hasAnchor" number="3.3">
<h2 class="hasAnchor"><span class="header-section-number">3.3</span> Install cellxgene VIP<a href="#install-cellxgene-vip" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The cellxgene VIP (Visualization In Plugin) platform can be used to visualize the h5ad file generated by the scRNASequest pipeline. This visualization step can be performed before the differential expression (DE) analysis to pinpoint meaningful clusters for downstream steps. If you include DE analysis when running <code>scAnalyzer</code>, you will also see DE results in Cellxgene VIP.</p>
<p>Please follow the detailed instructions below to install cellxgene VIP:</p>
<p><a href="https://github.com/interactivereport/cellxgene_VIP" target="_blank">https://github.com/interactivereport/cellxgene_VIP</a></p>
<p>After setting up, when launching cellxgene VIP, you will designate HOST name and PORT number. You can add this information in the sys.yml file as <code>celldepotHttp</code> under src/ directory.</p>
<p>However, cellxgene VIP is not required for running any components of the scRNASequest pipeline (scAnalyzer, scRef, etc.). It is up to the user how to work with the h5ad files after running the <code>scAnalyzer</code> script, and (<a href="https://scanpy.readthedocs.io/en/stable/" target="_blank">scanpy</a> is an alternative option.</p>
</div>
<div id="install-celldepot" class="section level2 hasAnchor" number="3.4">
<h2 class="hasAnchor"><span class="header-section-number">3.4</span> Install CellDepot<a href="#install-celldepot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>If you have multiple projects, you will have to manage many cellxgene VIP links and related data files. CellDepot is a centralized database for storing single-cell/nucleus RNA-seq results. Please refer to the detailed instructions below to install CellDepot:</p>
<p><a href="https://celldepot.bxgenomics.com/celldepot_manual/install_environment.php" target="_blank">https://celldepot.bxgenomics.com/celldepot_manual/install_environment.php</a></p>
<p>The installation of CellDepot is optional if the user only needs to analyze the data, without publishing it into the CellDepot database.</p>
<!--chapter:end:03-Installation.Rmd-->
</div>
</div>
<div id="demo-run" class="section level1 hasAnchor" number="4">
<h1 class="hasAnchor"><span class="header-section-number">4</span> Demo run<a href="#demo-run" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>We provide a demo dataset under the <code>demo</code> directory. This demo uses two snRNA-seq data from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE185538">GSE185538</a> to run through the main steps, including <strong>QC</strong>, <strong>data integration</strong> (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1">SCTransform</a>, then <a href="https://www.nature.com/articles/s41592-019-0619-0"><strong>Harmony</strong></a>), <a href="https://satijalab.org/seurat/articles/integration_mapping.html"><strong>Seurat reference mapping</strong></a>, and <strong>evaluation of integration</strong> (<a href="https://www.nature.com/articles/s41592-018-0254-1">kBET</a> and <a href="https://ieeexplore.ieee.org/document/9260048">silhouette</a>). To save time, the differential expression analysis won’t be performed, and the DEGinfo.csv file is empty.</p>
<p>This demo run contains two downsampled snRNA-seq data from the original study, and will take <strong>~15-20 minutes to finish</strong>. Please note that to speed up the run, we used stringent QC cutoffs, which eliminated many cells. After running the scAnalyzer command, output files will be generated under the <code>demo</code> directory on your computer.</p>
<p>Also, we provided config.yml, sampleMeta.csv and DEGinfo.csv files for running this demo so you can run the <code>scAnalyzer</code> command directly. Please note that these files are only for the demo run. Later for your own dataset, please follow the full tutorial in the next section, <a href="#data-preparation">Data preparation</a>, to generate the template of these files first.</p>
<div id="demo-run-with-conda" class="section level2 hasAnchor" number="4.1">
<h2 class="hasAnchor"><span class="header-section-number">4.1</span> Demo run with Conda<a href="#demo-run-with-conda" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Continuing from section 3.1.1, we assume that the pipeline was installed in: ~/scRNASequest.</p>
<p>First, we set up the config file (~/scRNASequest/demo/config.yml) by pointing these parameters to your directory, and other lines don’t need to be changed:</p>
<pre><code>...
ref_name: ~/scRNASequest/demo/ref/rat_cortex_ref.rds   # choose one from scAnalyzer call without argument
output: ~/scRNASequest/demo                            # output path   
...
sample_meta: ~/scRNASequest/demo/sampleMeta.csv
...
DEG_desp: ~/scRNASequest/demo/DEGinfo.csv              # for DEG analysis. To save time, this demo run doesn&#39;t include DE analysis
...</code></pre>
<p>Then execute the following command:</p>
<pre><code>scAnalyzer ~/scRNASequest/demo/config.yml</code></pre>
</div>
<div id="demo-run-with-docker" class="section level2 hasAnchor" number="4.2">
<h2 class="hasAnchor"><span class="header-section-number">4.2</span> Demo run with Docker<a href="#demo-run-with-docker" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>After following section 3.1.2 to set up the pipeline, we are ready to run this demo:</p>
<pre><code>docker exec -t -i &lt;container_name&gt; scAnalyzer /demo/config.yml</code></pre>
<p>After running this demo, a list of files will be created:</p>
<pre><code>config.yml                        #Original config.yml file
config.yml.20230314.log           #Log file during pipeline run
data/                             #Original data files
  ├──RatFemaleCigarette.filtered_feature_bc_matrix.h5
  ├──RatFemaleCigarette.metrics_summary.csv
  ├──RatMaleCigarette.filtered_feature_bc_matrix.h5
  └──RatMaleCigarette.metrics_summary.csv
DEGinfo.csv                       #Comparison file. In this demo, this is empty
evaluation/                       #Evaluation of harmonization
  ├──scRNASequest_demo_kBET_umap_k0_100.pdf
  └──scRNASequest_demo_Silhouette_boxplot_pc50.pdf
log/                              #Log files for each step
  ├──kBET.log
  ├──sctHarmony.log
  ├──SCT.log
  ├──SeuratRef.log
  └──ssilhouette.log
QC/                               #QC files
  ├──postfilter.QC.pdf
  ├──prefilter.QC.pdf
  ├──sequencingQC.csv
  └──sequencingQC.pdf
raw/                              #Original UMI counts, with and withour filtering
  ├──scRNASequest_demo.h5ad
  └──scRNASequest_demo_raw_prefilter.h5ad
ref/
  └──ref.Rds                      #Azimuth reference RSD file for cell type label transfer
Rmarkdown/                        #Pre-generated QC figures, for generating Bookdown report
sampleMeta.csv                    #Input sample metainformation file
scRNASequest_demo.h5ad            #This is the final output h5ad file, and the expression values have been normalized
                                   cellxgene VIP will use this file for visualization
scRNASequest_demo.h5seurat        #Result in h5seurat format
scRNASequest_demo_raw_added.h5ad  #Since AnnData object can only have one count matrix, this one uses the raw UMI
SCT/                              #Results related to SCT transformation
  ├──scRNASequest_demo.cID
  ├──scRNASequest_demo.gID
  ├──scRNASequest_demo.h5
  ├──scRNASequest_demo.h5ad
  ├──scRNASequest_demo.h5.rds
  └──scRNASequest_demo.scaleF
sctHarmony/                       #SCT+Harmony results
  ├──scRNASequest_demo.csv
  └──scRNASequest_demo.h5ad
SeuratRef/                       #Seurat reference mapping results
  ├──scRNASequest_demo.csv
  └──scRNASequest_demo.h5ad</code></pre>
<!--chapter:end:04-DemoRun.Rmd-->
</div>
</div>
<div id="data-preparation" class="section level1 hasAnchor" number="5">
<h1 class="hasAnchor"><span class="header-section-number">5</span> Data preparation<a href="#data-preparation" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>scRNASequest pipeline is compatible with different single-cell experiment outputs, including the <a href="https://kb.10xgenomics.com/hc/en-us/articles/115000794686-How-is-the-MEX-format-used-for-the-gene-barcode-matrices" target="_blank"><strong>10X MEX format</strong></a> and the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices" target="_blank"><strong>10X h5 format</strong></a> from Cell Ranger. The user may need to manually separate the annotation file, so that no cell filtering was performed. In this chapter, we will walk through how to prepare the data before running the pipeline.</p>
<p>A bit more information about these two types of format:</p>
<p><strong>10X MEX format</strong>: This format has three associated files: an <strong>mtx file</strong> and associated <strong>barcodes</strong> file as well as a <strong>features</strong> file. Based on the information provided by the 10X official <a href="https://kb.10xgenomics.com/hc/en-us/articles/115000794686-How-is-the-MEX-format-used-for-the-gene-barcode-matrices-">website</a>, the MEX is also called the <strong>Market Exchange (MEX) format</strong>, and the .mtx is a plain text file showing <strong>MatrixMarket matrix coordinates</strong> like this:</p>
<pre><code>%%MatrixMarket matrix coordinate integer general
%
154 1 21
...</code></pre>
<p>This means the gene in line 154 of the genes file, the 1st barcode in the barcodes file, has UMI counts=21.</p>
<p><strong>10X h5 format</strong>: This format is a HDF5 format storing UMI count matrix, as described in the 10X official tutorial <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices">here</a>.</p>
<div id="public-data-in-h5-format" class="section level2 hasAnchor" number="5.1">
<h2 class="hasAnchor"><span class="header-section-number">5.1</span> Public data in h5 format<a href="#public-data-in-h5-format" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here, we first present an example of the data processing steps using a public EMBL EBI dataset: <a href="https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-11115/" target="_blank">E-MTAB-11115</a>. The nine processed zip files were downloaded and unzipped.</p>
<p>There are total 6 data, with 6 corresponding *raw_feature_bc_matrix.h5 files. These files are required by the pipeline:</p>
<pre><code>#h5 matrix files
-rwxrwxr--. 1 zouyang ngs 165M Oct 25  2021 5705STDY8058280.raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 165M Oct 25  2021 5705STDY8058281.raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 156M Oct 25  2021 5705STDY8058282.raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 162M Oct 25  2021 5705STDY8058283.raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 149M Oct 25  2021 5705STDY8058284.raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 177M Oct 25  2021 5705STDY8058285.raw_feature_bc_matrix.h5</code></pre>
<p>This dataset also has cell type annotation files associated with each data. These files are optional to the pipeline, but if you would like to use their cell type labels, it would be better to include them in the sample meta file (See section @ref(section52)).</p>
<pre><code>#Annotation files (optional to the pipeline)
-rwxrwx---. 1 zouyang ngs 440K Apr 21 16:59 5705STDY8058280.annotation.csv
-rwxrwx---. 1 zouyang ngs 446K Apr 21 16:59 5705STDY8058281.annotation.csv
-rwxrwx---. 1 zouyang ngs 310K Apr 21 16:59 5705STDY8058282.annotation.csv
-rwxrwx---. 1 zouyang ngs 282K Apr 21 16:59 5705STDY8058283.annotation.csv
-rwxrwx---. 1 zouyang ngs 157K Apr 21 16:59 5705STDY8058284.annotation.csv
-rwxrwx---. 1 zouyang ngs 555K Apr 21 16:59 5705STDY8058285.annotation.csv

#A brief look at the annotation file:
$ head -3 5705STDY8058280_annotation.csv
Cell.ID,sample,annotation_1,annotation_1_print
AAACCCAAGGAAGTAG-1,5705STDY8058280,Ext_L25,23_Ext_L25
AAACCCAAGGGCAGTT-1,5705STDY8058280,Ext_L56,24_Ext_L56</code></pre>
<p>If <strong>SampleName.metrics_summary.csv</strong> files (QC files generated by Cell Ranger) are available, please also add them in the same directory as the h5 files, and the pipeline will use them to generate QC plots, but they are not required files of the pipeline. The expected file names should be:</p>
<pre><code>5705STDY8058280.metrics_summary.csv
5705STDY8058281.metrics_summary.csv
5705STDY8058282.metrics_summary.csv
5705STDY8058283.metrics_summary.csv
5705STDY8058284.metrics_summary.csv
5705STDY8058285.metrics_summary.csv</code></pre>
<p><strong>!!! Important</strong> Since we don’t provide the path of these metrics_summary.csv files to the pipeline, their prefix <strong>SampleName</strong> must be consistent with the <strong>Sample_Name</strong> column in the sample meta file (See section 5.2), so that the pipeline can recognize them automatically. You can certainly rename the files and the corresponding <strong>SampleName</strong> column if you would like to change the data names and how they appear in the final results. Also, for the metrics_summary.csv file names, the concatenator between the <strong>Sample_Name</strong> and “metrics_summary.csv” must be “.” so that the pipeline can read them automatically. This follows the naming criteria of Cell Ranger.</p>
</div>
<div id="public-data-in-10x-mex-format" class="section level2 hasAnchor" number="5.2">
<h2 class="hasAnchor"><span class="header-section-number">5.2</span> Public data in 10X MEX format<a href="#public-data-in-10x-mex-format" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Another popular format for single-cell RNA-seq is the <a href="https://kb.10xgenomics.com/hc/en-us/articles/115000794686-How-is-the-MEX-format-used-for-the-gene-barcode-matrices" target="_blank"><strong>MEX format</strong></a>. Here, we use a public dataset from NCBI/GEO: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE185538" target="_blank">GSE185538</a> to walk through the procedures for pipeline setup.</p>
<p>There are total 4 single-nucleus RNA-seq data, and all the processed data can be downloaded from the Supplementary file section as a tarball: GSE185538_RAW.tar:</p>
<pre><code>#Untar the file
tar -xvf GSE185538_RAW.tar

$ ls -l  #Only show the MTX-related files

-rw-rw-r-- 1 ysun4 compbio       97631 Sep 29  2021 GSM5617891_snRNA_FCtr_barcodes.tsv.gz
-rw-rw-r-- 1 ysun4 compbio      206935 Sep 29  2021 GSM5617891_snRNA_FCtr_features.tsv.gz
-rw-rw-r-- 1 ysun4 compbio   116235040 Sep 29  2021 GSM5617891_snRNA_FCtr_matrix.mtx.gz
-rw-rw-r-- 1 ysun4 compbio       89425 Sep 29  2021 GSM5617892_snRNA_FEcig_barcodes.tsv.gz
-rw-rw-r-- 1 ysun4 compbio      206935 Sep 29  2021 GSM5617892_snRNA_FEcig_features.tsv.gz
-rw-rw-r-- 1 ysun4 compbio   101349173 Sep 29  2021 GSM5617892_snRNA_FEcig_matrix.mtx.gz
-rw-rw-r-- 1 ysun4 compbio      283865 Sep 29  2021 GSM5617893_snRNA_MCtr_barcodes.tsv.gz
-rw-rw-r-- 1 ysun4 compbio      206935 Sep 29  2021 GSM5617893_snRNA_MCtr_features.tsv.gz
-rw-rw-r-- 1 ysun4 compbio   440019486 Sep 29  2021 GSM5617893_snRNA_MCtr_matrix.mtx.gz
-rw-rw-r-- 1 ysun4 compbio      117119 Sep 29  2021 GSM5617894_snRNA_MEcig_barcodes.tsv.gz
-rw-rw-r-- 1 ysun4 compbio      229853 Sep 29  2021 GSM5617894_snRNA_MEcig_features.tsv.gz
-rw-rw-r-- 1 ysun4 compbio   159326641 Sep 29  2021 GSM5617894_snRNA_MEcig_matrix.mtx.gz</code></pre>
<p>Next, these files need to be organized into separate folders for the pipeline to read. In specific, we need to create separate folders for each data, and rename the three files to be: <strong>barcodes.tsv.gz</strong>, <strong>features.tsv.gz</strong>, and <strong>matrix.mtx.gz</strong> (Un-compressed files can also be used to run the pipeline, and the pipeline will compress them). The organized file hierarchy is below:</p>
<pre><code>GSE185538/
    ├── GSM5617891_snRNA_FCtr
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ├── GSM5617892_snRNA_FEcig
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ├── GSM5617893_snRNA_MCtr
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    └── GSM5617894_snRNA_MEcig
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz</code></pre>
</div>
<div id="self-prepared-files" class="section level2 hasAnchor" number="5.3">
<h2 class="hasAnchor"><span class="header-section-number">5.3</span> Self-prepared files<a href="#self-prepared-files" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>If you have raw data in <a href="https://support.illumina.com/bulletins/2016/04/fastq-files-explained.html" target="_blank">FASTQ format</a>, please process them using the Cell Ranger pipeline to generate the raw_feature_bc_matrix.h5 and metrics_summary.csv files. Please visit <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count" target="_blank">Cell Ranger website here</a> for more details about the outputs, in the “Output files” section.</p>
<p>In short, Cell Ranger outputs <strong>filtered_feature_bc_matrix.h5</strong> (cells after filtering, recommended to use), <strong>raw_feature_bc_matrix.h5</strong> (without filtering), and two folders for MEX format output: <strong>filtered_feature_bc_matrix</strong>, <strong>raw_feature_bc_matrix</strong>. In addition, Cell Ranger also has a <strong>metrics_summary.csv</strong> file generated, which can be provided to scRNASequest. Please re-organize, and if necessary, rename the Cell Ranger output files into the following file structures before running scRNASequest.</p>
<p>Below is the file hierarchy for h5 input files (metrics_summary.csv files are suggested to be included, but not required; If cell type classification annotation.csv files are available, it would be better to include them):</p>
<pre><code>Project/
    ├── Data1.filtered_feature_bc_matrix.h5
    ├── Data1.annotation.csv                 (optional)
    ├── Data1.metrics_summary.csv            (optional)
    ├── Data2.filtered_feature_bc_matrix.h5
    ├── Data2.annotation.csv                 (optional)
    ├── Data2.metrics_summary.csv            (optional)
    ...</code></pre>
<p>The file hierarchy for <a href="https://kb.10xgenomics.com/hc/en-us/articles/115000794686-How-is-the-MEX-format-used-for-the-gene-barcode-matrices" target="_blank">MEX files</a>:</p>
<pre><code>Project/
    ├── Data1
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ├── Data2
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ...</code></pre>
<!--chapter:end:05-DataPreparation.Rmd-->
</div>
</div>
<div id="input-configuration" class="section level1 hasAnchor" number="6">
<h1 class="hasAnchor"><span class="header-section-number">6</span> Input configuration<a href="#input-configuration" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>First, make sure the .env has been created in the src folder under the pipeline directory:</p>
<pre><code>ls ~/scRNAsequest/src/.env</code></pre>
<div id="generate-template-files" class="section level2 hasAnchor" number="6.1">
<h2 class="hasAnchor"><span class="header-section-number">6.1</span> Generate template files<a href="#generate-template-files" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Now we are ready to run the pipeline. Since the main pipeline script <code>scAnalyzer</code> requires a <strong>config.yml</strong> file, a sample metadata file <strong>sampleMeta.csv</strong>, and a <strong>DEGinfo.csv</strong> file for DE analysis (can be filled in later), we can first generate templates for these files.</p>
<p>We run <code>scAnalyzer</code> by providing a path to our working directory. This can be any directory on you system:</p>
<pre><code>scAnalyzer /path/to/working/dir</code></pre>
<p>Then you will see the following files generate:</p>
<pre><code>config.yml          #This is the pipeline configuration file
DEGinfo.csv         #This is for DE analysis, see the &#39;Differential expression (DE) analysis&#39; section for more details
                    #This file can be left blank, which means skip DE analysis
sampleMeta.csv      #Sample metadata file, storing sample level information
sc20230320.init.log #Log file</code></pre>
<p>In the log file, we can also see the following messages:</p>
<pre><code>*****
External data, please fill the config file and provide sample sheet with two mandatory columns:
Two columns with names &#39;Sample_Name&#39; and &#39;h5path&#39; indicating sample name and UMI path
Option columns for sample annotation
Option column &#39;metapath&#39; can be provided for cell level annotation first column cell bar code
*****</code></pre>
<p>Please following the instructions below to set up these files.</p>
</div>
<div id="prepare-the-config.yml-file" class="section level2 hasAnchor" number="6.2">
<h2 class="hasAnchor"><span class="header-section-number">6.2</span> Prepare the config.yml file<a href="#prepare-the-config.yml-file" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To run the scRNAsequest pipeline, a config.yml file is required to be filled in. Please use the following template as an example to prepare this file:</p>
<p><a href="https://github.com/interactivereport/scRNAsequest/blob/main/src/template.yml" target="_blank">config.yml</a></p>
<pre><code>## Project info and filtering parameters
prj_name: E-MTAB-11115            #required, project name
prj_title: &quot;Cell2location maps fine-grained cell types in spatial transcriptomics&quot; #required, and quotes might be needed
ref_name:                         #optional. You can choose one from scAnalyzer call without an argument
output: /home/ysun4/E-MTAB-11115/ #required, output path
sample_name: Sample_Name
sample_meta: ~/E-MTAB-11115/sampleMeta.csv   #required, sample meta information, see section 5.1
gene_group:
  MT:
    startwith: [&quot;MT-&quot;,&quot;Mt-&quot;,&quot;mt-&quot;]
    cutoff: 20           # percentage cutoff to filter out the cells (larger than this cutoff)
    rm: False            # this means the genes specified &quot;startwith&quot; will be REMOVED from the downstream analysis
  # The following is an example (Ribosomal genes) to add additional gene set/list to be quantified the percentage over total UMI per cell
  # additional gene set/list can be added as the format
  RP:
    startwith: [MRPL,MRPS,RPL,RPS]
    cutoff: 50
    rm: False
filter_step: True        #if False, the above &#39;gene_group&#39; filtering will be skipped as well
min.cells: 3             #filtering genes by minimal cell
min.features: 50         #filtering cells by minimal genes
highCount.cutoff: 10000  #any cells with higher total counts to be removed
highGene.cutoff: 3000    #any cells with a higher number of detected genes to be removed
group:                   #if provided, a 10X QC box plot will be ploted in QC plot
rasterizeFig: True       #should image in pdf be rasterized

reRunQC: True            #if satisfied with the above setting, please set this to False to save time 
runAnalysis: False       #False means only run QC steps before data normalization and harmonization
                         #If QC looks good, turn this to True to run the full pipeline
newProcess: False
methods: [SCT,Liger,SeuratRef,SeuratRPCA,sctHarmony]    #SCT method (including log1p) is required expression normalization
                         #The listed methods include all available analyses. If you want to ignore Liger, for example, you can delete it here
expScaler: 10000         #integer value, 0: SCT (along with following option); 1-100: logNormal with scale.factor to be the specified quantile
                         #&gt;100: logNormal with scale.factor to be the specified value                                         
PrepSCTFindMarkers: True # should the &quot;PrepSCTFindMarkers&quot; applied on SCT to be exported for visualization expression                                                             
parallel: False          #&quot;sge&quot; or &quot;slurm&quot;
core: 2
overwrite: False         #if you are rerunning the same project, set this to True to overwrite previous results
major_cluster_rate: 0.7  #the proportion of cells of an integration cluster to be assigned to a seurat reference label

## DEG analysis for an annotation (such as disease vs health) within a cell type annotation
DEG_desp:  ~/E-MTAB-11115/DEGinfo.csv         #required for DEG analysis, but you can leave this empty if not going to run DEG
# Please be causion of changing the following default filtering
# More details can be found: section 2.4 in https://pubmed.ncbi.nlm.nih.gov/35743881/
# Applies the 1st round of biostats filtering pipeline. Note that this filter is applied to all cells of the experiment
min.cells.per.gene: 3                         # if `perc_filter` is FALSE, then keep only genes that have expression in at least min.cells.per.gene
min.genes.per.cell: 250                       # keep cells with expression in at least min.genes.per.cell genes.
min.perc.cells.per.gene: 0.00 #if &#39;perc_filter&#39;`&#39; is TRUE, then keep only genes that have expression in at least min.per.cells.per.gene * 100 percent of cells
perc_filter: TRUE                             # if TRUE, apply the cells.per.gene filter using percentages (expressed as a decimal) rather than an absolute threshold
# Apply the 2nd round of biostats filtering.  For &quot;group&quot; mode, the filtering is applied to `ref_group` and `alt_group` for the given cell type of interest.
R6_min.cells.per.gene: 3                      # minimum cells expressed per gene.  This filter is applied if `R6_perc.cells.filter` is FALSE
R6_min.perc.cells.per.gene: 0.1               # minimum % cells expressed per gene filtering (use decimal form of percentage).  This threshold is applied if &#39;R6_perc.cells.filter&#39; is TRUE and &#39;R6_cells.per.gene.filter&#39; is TRUE
R6_min.cells.per.gene.type: &quot;or&quot;              # The type of cell per gene filtering.  If it has the value &quot;and&quot; then it requires the gene be expressed in both reference and non-reference groups. If it has the value &quot;or&quot; then it requires the gene be expressed in either group
R6_cells.per.gene.filter: TRUE                # TRUE means apply cells per gene filtering
R6_perc.cells.filter: TRUE                    # TRUE means apply cell.per.gene filtering by use of a percentage rather than absolute threshold.  If the percentage results in a number less than R6_min.cells.per.gene, the code will automatically switch to min.cells.per.gene absolute thresholding
R6_perc.filter: FALSE                         # If TRUE, then apply the 75th percentile gene filtering
R6_perc.filter.type: &quot;and&quot;                    # The type of percentile gene filtering.  If it has the value &quot;and&quot; then any gene that has 75th percentile of zero in both groups will be filtered out.  If it has the value &quot;or&quot; then any gene that has a 75th percentile of zero in either group will be filtered out.
R6_perc_threshold: 0.90                       # Percentile threshold, 75th percentile is default.  Express percentile as a decimal value.
R6_min.ave.pseudo.bulk.cpm: 1 #cpm filtering threshold
R6_pseudo.bulk.cpm.filter: FALSE              # if TRUE, then apply a cpm filter on the pseudo-bulk counts
R6_min.cells.per.subj: 3                      # Minimum cells required per subject, must be a nonzero number

## publish to celldepot
publish: False</code></pre>
<p>In the first run (see the below section 7.1 scAnalyzer section on how to run the pipeline), the user can set <code>runAnalysis: False</code> and use the above default parameters to perform cell filtering and QC.</p>
<p>Then this pipeline will only run basic QC checking and output a Bookdown report with QC figures. This step is semi-automated because different datasets may need distinct filtering criteria. The design of this pipeline is to pause here to make sure the filtering step is adequate before running through the whole analysis.</p>
<p>For HPC clusters, if you set <code>parallel</code> to be “sge” or “slurm” based on your cluster, the pipeline will automatically submit separate jobs and monitor the run. Finally, it will collect results and summarize them.</p>
<p>Once the filtering step is validated by checking the QC report, the user can change the following settings and run the full analysis:</p>
<pre><code>runAnalysis: True
overwrite: True</code></pre>
</div>
<div id="section52" class="section level2 hasAnchor" number="6.3">
<h2 class="hasAnchor"><span class="header-section-number">6.3</span> Prepare the sample meta file<a href="#section52" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The sample meta file, usually named as <strong>sampleMeta.csv</strong>, is a csv file storing data information. The minimal columns for this file are <strong>Sample_Name</strong> (This name can be changed, but it must be consistent with the sample_name in the config.yml file) and <strong>h5path</strong>. If you have the cell type annotation information, you can provide is by adding a column called <strong>metapath</strong>. For other information, you can add as many columns in this csv file and the column names can be defined by yourself.</p>
<div id="h5-input" class="section level3 hasAnchor" number="6.3.1">
<h3 class="hasAnchor"><span class="header-section-number">6.3.1</span> h5 input<a href="#h5-input" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Here is an example with minimal information for the public dataset <a href="https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-11115/" target="_blank">E-MTAB-11115</a> (10X h5 format):</p>
<pre><code>Sample_Name,h5path
5705STDY8058280,~/E-MTAB-11115/data/5705STDY8058280.filtered_feature_bc_matrix.h5
5705STDY8058281,~/E-MTAB-11115/data/5705STDY8058281.filtered_feature_bc_matrix.h5
5705STDY8058282,~/E-MTAB-11115/data/5705STDY8058282.filtered_feature_bc_matrix.h5
5705STDY8058283,~/E-MTAB-11115/data/5705STDY8058283.filtered_feature_bc_matrix.h5
5705STDY8058284,~/E-MTAB-11115/data/5705STDY8058284.filtered_feature_bc_matrix.h5
5705STDY8058285,~/E-MTAB-11115/data/5705STDY8058285.filtered_feature_bc_matrix.h5</code></pre>
<p>Another example if you would like to use the <strong>cell type annotation</strong> defined by their analysis, which can be included in the <strong>metapath</strong> column. Moreover, the user has the freedom to provide further information such as sex and age:</p>
<pre><code>Sample_Name,h5path,metapath,sex,age
5705STDY8058280,~/E-MTAB-11115/data/5705STDY8058280_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058280_annotation.csv,Female,56d
5705STDY8058281,~/E-MTAB-11115/data/5705STDY8058281_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058281_annotation.csv,Female,56d
5705STDY8058282,~/E-MTAB-11115/data/5705STDY8058282_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058282_annotation.csv,Female,56d
5705STDY8058283,~/E-MTAB-11115/data/5705STDY8058283_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058283_annotation.csv,Male,56d
5705STDY8058284,~/E-MTAB-11115/data/5705STDY8058284_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058284_annotation.csv,Male,56d
5705STDY8058285,~/E-MTAB-11115/data/5705STDY8058285_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058285_annotation.csv,Male,56d

#A brief look at the annotation file used in the metapath column:
$ head -3 5705STDY8058280_annotation.csv
Cell.ID,sample,annotation_1,annotation_1_print
AAACCCAAGGAAGTAG-1,5705STDY8058280,Ext_L25,23_Ext_L25
AAACCCAAGGGCAGTT-1,5705STDY8058280,Ext_L56,24_Ext_L56</code></pre>
</div>
<div id="mex-input" class="section level3 hasAnchor" number="6.3.2">
<h3 class="hasAnchor"><span class="header-section-number">6.3.2</span> MEX input<a href="#mex-input" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Since each data has three MEX files (barcodes.tsv.gz, features.tsv.gz and matrix.mtx.gz), here we provide the <strong>path to each MEX result directory</strong>. The minimal information in the sample meta file:</p>
<pre><code>Sample_Name,h5path
FCtr,~/10XGenomicsData/GSM5617891_snRNA_FCtr
FEcig,~/10XGenomicsData/GSM5617892_snRNA_FEcig
MCtr,~/10XGenomicsData/GSM5617893_snRNA_MCtr
MEcig,~/10XGenomicsData/GSM5617894_snRNA_MEcig</code></pre>
<p>Another example with additional information:</p>
<pre><code>Sample_Name,h5path,Treatment,Sex
FCtr,~/10XGenomicsData/GSM5617891_snRNA_FCtr,Control,Female
FEcig,~/10XGenomicsData/GSM5617892_snRNA_FEcig,EcTreated,Female
MCtr,~/10XGenomicsData/GSM5617893_snRNA_MCtr,Control,Male
MEcig,~/10XGenomicsData/GSM5617894_snRNA_MEcig,EcTreated,Male</code></pre>
</div>
</div>
<div id="file-hierarchy" class="section level2 hasAnchor" number="6.4">
<h2 class="hasAnchor"><span class="header-section-number">6.4</span> File hierarchy<a href="#file-hierarchy" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Before running the pipeline, please check the file hierarchy. Here are examples for h5 and MEX inputs, and we suggest creating a separate folder (such as <em>processing</em>) to store the config.yml and sampleMeta.csv files. The output files will be generated in the directory specified in config.yml.</p>
<p>For data downloaded in 10X h5 format:</p>
<pre><code>E-MTAB-11115/
    ├── data
        ├── 5705STDY8058280.annotation.csv           (optional)
        ├── 5705STDY8058280.metrics_summary.csv      (optional)
        ├── 5705STDY8058280.raw_feature_bc_matrix.h5
        ├── 5705STDY8058281.annotation.csv           (optional)
        ├── 5705STDY8058281.metrics_summary.csv      (optional)
        ├── 5705STDY8058281.raw_feature_bc_matrix.h5
        ...
    ├── processing
        ├── config.yml
        └── sampleMeta.csv</code></pre>
<p>For data downloaded in 10X MEX format:</p>
<pre><code>GSE185538/
    ├── GSM5617891_snRNA_FCtr
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ├── GSM5617892_snRNA_FEcig
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ...
    ├── processing
        ├── config.yml
        └── sampleMeta.csv</code></pre>
<!--chapter:end:06-InputConfiguration.Rmd-->
</div>
</div>
<div id="running-the-pipeline" class="section level1 hasAnchor" number="7">
<h1 class="hasAnchor"><span class="header-section-number">7</span> Running the pipeline<a href="#running-the-pipeline" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>After preparing the input data and configuration files (config.yml and sampleMeta.csv), we are ready to run the pipeline. Please be aware that scRNASequest is a semi-automated pipeline, and certain steps require the user to review the outputs and determine the parameters. Thus, the same step may need to be run for more than once.</p>
<div id="initial-scanalyzer-run" class="section level2 hasAnchor" number="7.1">
<h2 class="hasAnchor"><span class="header-section-number">7.1</span> Initial scAnalyzer run<a href="#initial-scanalyzer-run" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The initial scAnalyzer run intends to test the default filtering parameters on the single-cell RNA-seq data. We suggest the following settings in the config.yml file to activate the filtering step, but pause the pipeline from additional analysis (such as data integration, harmonization, etc.):</p>
<pre><code># Change the following settings in the config.yml file:
filter_step: True
runAnalysis: False</code></pre>
<p>To run scAnalyzer, simply pass the path to the config.yml file to the program:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span> Path/to/a/config/file</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span> ~/E-MTAB-11115/processing/config.yml</span></code></pre></div>
<p>Output files:</p>
<pre><code>outputdir
    ├── BookdownReport/                #Full bookdown report folder. This is same as the BookdownReport.tar.gz below
    ├── config.yml
    ├── config.yml.20220617.log        #log file
    ├── ProjName_BookdownReport.tar.gz #Bookdown report tar ball. 
                                       #Please download this file and open the index.html file to view the report
    ├── ProjName_raw_prefilter.h5ad    #h5ad after pre-filtering
    ├── prefilter.QC.pdf               #QC plots before filtering. These plots are included in the Bookdown report
    ├── postfilter.QC.pdf              #QC plots after filtering
    └── Rmarkdown/                     #Source files for the Bookdown report</code></pre>
</div>
<div id="review-the-bookdown-report" class="section level2 hasAnchor" number="7.2">
<h2 class="hasAnchor"><span class="header-section-number">7.2</span> Review the bookdown report<a href="#review-the-bookdown-report" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>After the initial <code>scAnalyzer</code> run, a bookdown report, named <strong>ProjName_BookdownReport.tar.gz</strong> will be generated. Please open the report (index.html) to review cell numbers and figures before and after the filtering step, and adjust the following thresholds in the config.yml file if needed:</p>
<pre><code># Cell filtering parameters in the config.yml file:
gene_group:
  MT:
    startwith: [&quot;MT-&quot;,&quot;Mt-&quot;,&quot;mt-&quot;]
    cutoff: 20           #percentage cutoff to filter out the cells (larger than this cutoff)
    rm: False
  RP:                    #for ribosomal genes if you would like to filter them
    startwith: []
    cutoff: 20
    rm: False
filter_step: True        #if False, the above &#39;gene_group&#39; filtering will be skipped as well
min.cells: 3             #filtering genes by minimum number of cells with non-zero count for a gene
min.features: 50         #filtering cells by minimal genes detected
highCount.cutoff: 10000  #any cells with higher total counts to be removed
highGene.cutoff: 3000    #any cells with a higher number of detected genes to be removed</code></pre>
<p>The <strong>min.cells</strong> cutoff is to filter out <em>genes</em> that have low representation in cells, and the <strong>min.features</strong> cutoff is to filter out <em>cells</em> that have low number of genes (features) detected. The last two cutoffs, <strong>highCount.cutoff</strong> and <strong>highGene.cutoff</strong> can be used to filter cells if they have too many total counts or UMI detected, possibly due to multiplets. Please evaluate these cutoffs carefully, because the default values here are arbitrary.</p>
<p>The user may need to adjust the thresholds and rerun <code>scAnalyzer</code> to visualize the results. This step may be repeated until the threshold is adequate for the data.</p>
<p>Please find an example of the bookdown report here: <a href="https://interactivereport.github.io/scRNAsequest/examples/E-MTAB-11115/bookdown/index.html" target="_blank">E-MTAB-11115 bookdown</a></p>
</div>
<div id="generate-a-slide-deck" class="section level2 hasAnchor" number="7.3">
<h2 class="hasAnchor"><span class="header-section-number">7.3</span> Generate a slide deck<a href="#generate-a-slide-deck" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This is an optional step if you have already reviewed the Bookdown report. This slide deck is another way to organize and present the same results.</p>
<p>The QC report can be turned into a online slide deck format for better sharing and demonstration. To achieve this, please use the <strong>Slidedeck.Rmd</strong> file under the /src folder and follow the steps below:</p>
<pre><code># create a new directory inside the project, and link the existed files and folder:
cd ~/E-MTAB-11115/processing/
mkdir Slidedeck &amp;&amp; cd Slidedeck
ln -s ../BookdownReport/images
ln -s ../BookdownReport/bookdown.info.txt

# Copy the Slidedeck.Rmd file from src directory to the working directory containing config.yml file.
# We assume your pipeline was installed under ~/scRNAsequest/
cp ~/scRNAsequest/src/Slidedeck.Rmd ~/E-MTAB-11115/processing/Slidedeck

# Source the environment
src=&quot;~/scRNAsequest/src&quot;
source $src/.env;eval $condaEnv

# Run the Slidedeck script, and set the output file name using &#39;output_file&#39; parameter
Rscript -e &#39;rmarkdown::render(&quot;Slidedeck.Rmd&quot;, output_file=&quot;Output.html&quot;)&#39;</code></pre>
<p>Please download <strong>the whole Slidedeck directory</strong> and open the <strong>Output.html</strong> file in a web browser.</p>
<p>Please find an example of the bookdown report here: <a href="https://interactivereport.github.io/scRNAsequest/examples/E-MTAB-11115/slidedeck/index.html" target="_blank">E-MTAB-11115 slidedeck</a></p>
</div>
<div id="run-the-full-analysis" class="section level2 hasAnchor" number="7.4">
<h2 class="hasAnchor"><span class="header-section-number">7.4</span> Run the full analysis<a href="#run-the-full-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>After the filtering step is done, we are ready to turn on the <em>runAnalysis</em> parameter and run the full pipeline:</p>
<pre><code># Change the following setting in the config.yml file:
runAnalysis: True
overwrite: True    #Overwrite the previous results if we run this again, for example, after adjusting some parameters.</code></pre>
<p>Please also pay attention to the <strong>reference</strong> data, and if the user would like to turn on the label transfer function for cell type annotation, set the following parameter, for example, using <strong>human_cortex</strong> as the reference:</p>
<pre><code>ref_name: human_cortex</code></pre>
<p>To check all available references, run <code>scAnalyzer</code> without any parameter, which will read the src/sys.yml file to check available references. To add new references, please see section 8: <a href="#reference-building">Reference Building</a>.</p>
<p>Alternatively, you can provide a path to the <strong>ref.Rds</strong> reference file in the <a href="https://github.com/satijalab/azimuth/wiki/Azimuth-Reference-Format" target="_blank">Azimuth reference format</a>:</p>
<pre><code>ref_name: path/to/ref.Rds</code></pre>
<p>An example reference file is the demo/ref/rat_cortex_ref.rds file associated with this pipeline.</p>
<p>Then run the following command again:</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span> Path/to/a/config/file</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span> ~/E-MTAB-11115/processing/config.yml</span></code></pre></div>
<p>This step will be time-consuming, and it took ~1 hour for the <a href="https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-11115/" target="_blank">E-MTAB-11115</a> data (final cell number: 29,511) to finish. For a larger dataset with 110k cells we have tested, the full pipeline took ~7 hours.</p>
</div>
<div id="visualize-the-harmonized-results" class="section level2 hasAnchor" number="7.5">
<h2 class="hasAnchor"><span class="header-section-number">7.5</span> Visualize the harmonized results<a href="#visualize-the-harmonized-results" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>After running the pipeline, you will see the following files in the working directory. Each harmonization method (Harmony, Seurat, Liger) will also generate corresponding files.</p>
<pre><code>outputdir
    # Input files
        ├── config.yml:      The analysis config file specify the scAnalyzer parameters
        ├── sampleMeta.csv:  List all sample information including expression matrix (and sample/cell level meta information)
        └── DEGinfo.csv:     Define the sc DEG (between phenotypes within a cluster annotatino)
    # Main results under the output dir:
        ├── [prj_name].h5ad:           The final h5ad file which is used for cellxgene VIP loading (will be copied into the celldepotDir folder set by sys.yml)
        ├── [prj_name].db:             The final db file which is used for cellxgene VIP loading (will be copied into the celldepotDir folder set by sys.yml)
        ├── [prj_name].h5seurat:       The h5seurat file of the final results
        └── [prj_name]_raw_added.h5ad: The h5ad file contains raw counts in X along with all annotation and embeding 
                                          (will be copied into celldepot folder)
    # Rmarkdown:
        ├── Rmarkdown:                 The folder contains png files for R bookdown
        ├── [prj_name]_BookdownReport.tar.gz: R bookdown document tarball for downloading and sharing
        └── BookdownReport:            The folder contains R bookdown reports
    # Log file:
        └── config.yml.[date].log:  All stdout/stderr will be recorded in this dated log file
        └── [Software].error:       Standard error files of multiple software, including: 
                                        Harmony, Liger, SeuratRPCA, SeuratRef, SCT, sctHarmony, silhouette, etc.
                                    These files will be under the log/ directory if you did not run in the parallel mode
                                    These files will be under the j[numeric] folder if you ran the pipeline in the parallel mode
    # QC files under the QC/ folder:
        ├── sequencingQC.csv:   Merged all sequence QC files from individual samples if available
        ├── sequencingQC.pdf:   Plots of sequence QC if they are available
        ├── prefilter.QC.pdf:   Initial data QC plots without filtering
        └── postfilter.QC.pdf:  Data QC plots after filtering
    # evaluation folder for harmonization evaluation:
        ├── [prj_name]_kBET_umap_k0_100.pdf          The kBET evaluation plot for all integration methods
        └── [prj_name]_Silhouette_boxplot_pc50.pdf   The Silhouette evaluation plot for all integration methods
    # raw/ folder for h5ad files before and after QC, but have not run through the pipeline:
        ├── [prj_name]_raw_prefilter.h5ad: The internal temporary raw counts created/merged directly from sample before filtering 
        └── [prj_name]_raw.h5ad:           The internal temporary raw counts after the cell/gene filtering
    # SCT/ folder:
        ├── [prj_name].h5:        The sparse matrix contains SCT expression values
        ├── [prj_name].h5.rds:    The Seurat object contains counts (RNA assay) and 
                                      SCT expression (SCT assay) along with original meta information from SampleMeta file
        ├── [prj_name].cID        Cell barcodes
        ├── [prj_name].gID        Gene ID
        ├── [prj_name].scalF      Scale factor
        └── [prj_name].h5ad:      The SCT expression, along with PCA, tSNE, UMAP reduction layout, will be 
                                      merged into [prj_name].h5ad
    # SeuratRPCA folder for Seurat RPCA integration results:
        ├── [prj_name].csv.gz:    The reuction embedding and clustering results from seurat RPCA integration 
        └── [prj_name].h5ad:      The seurat RPCA integration with reduction embedding and clustering, 
                                      will be merged into [prj_name].h5ad
    # SeuratRef folder for Seurat reference mapping results:
        ├── [prj_name].csv:       The reuction embedding and label transfer results from seurat RPCA integration 
        └── [prj_name].h5ad:      The seurat reference mapping results with reduction embedding, will be merged into [prj_name].h5ad
    # sctHarmony folder for Harmony based on SCT transformation:
        ├── [prj_name].csv:       The PCA iinformation of each cell
        └── [prj_name].h5ad:      The harmony integration with reduction embedding and clustering, 
                                      will be merged into [prj_name].h5ad
    # Liger folder for Liger integration results:
        ├── [prj_name]_hvg.csv:  High variable gene list
        ├── [prj_name].csv:      The liger integration embedding and cluster
        └── [prj_name].h5ad&quot;:    The liger integration with reduction embedding and clustering, 
                                          will be merged into [prj_name].h5ad
    # [prj_name]_scDEG folder (if you ran the DEG step):
        ├── Comparisone names folders
          ├── .csv files for each cell type: The DEG analysis results
          ├── .rds files for each cell type: The rds data
          └── .pnd files for each cell type: The volcano plots
    # Parallel directory (if you ran the pipeline in parallel mode):
        └── j[numeric]:                A colder contains parallel jobs submit script and log file</code></pre>
<p><a href="https://www.nature.com/articles/s41592-018-0254-1" target="_blank">kBET</a> is a useful metric to evaluate and compare the harmonization results. Here is an example of the <strong>ProjName_kBET_umap_k0_100.pdf</strong> result:</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/kBET.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/kBET.png" /></a></p>
<p>Please review the <strong>ProjName.h5ad</strong> file for the harmonization result. If you have <a href="https://github.com/interactivereport/cellxgene_VIP" target="_blank">Cellxgene VIP</a> installed, the visualization will be easier.</p>
<p>Here is an example of the ProjName.h5ad project visualized by <a href="https://github.com/interactivereport/cellxgene_VIP" target="_blank">Cellxgene VIP</a>:</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png" /></a></p>
<!--chapter:end:07-RunningThePipeline.Rmd-->
</div>
</div>
<div id="differential-expression-de-analysis" class="section level1 hasAnchor" number="8">
<h1 class="hasAnchor"><span class="header-section-number">8</span> Differential expression (DE) analysis<a href="#differential-expression-de-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Another critical function of the scRNASequest pipeline is the <strong>Differential expression (DE)</strong> analysis. By running the steps in Section: <a href="#running-the-pipeline">Running the pipeline</a>, we have performed quality control of the data, and generated an integrated dataset.</p>
<p>Before running the DE analysis, the <strong>DEGinfo.csv</strong> comparison file needs to be prepared:</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_empty.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_empty.png" /></a></p>
<p>The first column, <strong>comparisonName</strong> is designed to store the name of the comparison. These names will be used to create separate folders under the result directory. In the DEG comparison file above, <strong>“sample”</strong>, <strong>“cluster”</strong>, <strong>“group”</strong> and <strong>“covars”</strong> are the annotation headers, and <strong>“alt”</strong> and <strong>“ref”</strong> are the two entries from <strong>“group”</strong> column. The <strong>“covars”</strong> is optional and can be empty.</p>
<p>The DEG is performing between <strong>“alt”</strong> and <strong>“ref”</strong> cells from <strong>“group”</strong> within each entry of <strong>“cluster”</strong> considering <strong>“sample”</strong> variations. The <strong>“group”</strong> variable should contain conditions to compare, such as Mutant v.s. Control. Thus, this pipeline is designed to loop through each cluster, and perform DEG analysis between <strong>“alt”</strong> v.s. <strong>“ref”</strong>.</p>
<p><strong>Covariate adjustment:</strong></p>
<p>Adjusting covariates is optional. By default, NEBULA considers the value of <strong>log of total UMI (library size)</strong> as the offset covariate in our pipeline, so you don’t need to adjust this again in <strong>“covars”</strong>.</p>
<p>If needed, you can put a formula in the <strong>“covars”</strong> column using the meta information from the sampleMeta.csv file, such as Sex, Age, etc. You can also include QC metrics such as mitochondria percentage (pct_MT), etc. Please find a full list of QC metrics below. You can also open the Cellxgene VIP and check the right side. All numeric covariates are listed there.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="ex">n_genes_by_counts,</span> log1p_n_genes_by_counts, pct_counts_in_top_50_genes, pct_counts_in_top_100_genes, pct_counts_in_top_200_genes, pct_counts_in_top_500_genes, pct_MT, n_genes, predicted.CellType.score</span></code></pre></div>
<p>You may consider <strong>pct_MT</strong> as a covariate to adjust, but be careful to adjust other covariates here unless you have a good reason to do so. An example of the formula to be filled out in the <strong>“covars”</strong> column: <strong>Age+Sex+pct_MT</strong>.</p>
<p>Here, this tutorial provides several examples to illustrate the DE analysis in scRNASequest:</p>
<p>We use a UMAP with label transferred cell type annotation as an example dataset. Label transfer is strongly suggested if you would like to run DE analysis because the clusters are more meaningful than the original clusters assigned by the software:</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png" /></a></p>
<div id="example-1" class="section level2 hasAnchor" number="8.1">
<h2 class="hasAnchor"><span class="header-section-number">8.1</span> Example 1<a href="#example-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the first example, we would like to run DE analysis between <strong>‘Female’</strong> and <strong>‘Male’</strong> for each cluster annotated by <strong>predicted.celltype1</strong>. In the first column, we input the header name <strong>library_id</strong>, which annotates the data sources. Then we add <strong>predicted.celltype1</strong> in the <em>cluster</em> column, which allows the pipeline to loop through each cluster in <strong>predicted.celltype1</strong>. The <em>group</em> column contains the header name storing the comparison groups, and here we use the <strong>Sex</strong> annotation. Each time, the pipeline can only compare two conditions, such as <strong>‘Female’</strong> and <strong>‘Male’</strong>. If the <em>group</em> column contains more groups, please list them in multiple lines in the DEGinfo.csv file.</p>
<p>We can also add <em>covars</em> if needed, but this is optional.</p>
<p>The default DE analysis is performed by <a href="https://www.nature.com/articles/s42003-021-02146-6" target="_blank"><strong>NEBULA</strong></a>. As for the model, NEBULA provides two choices: <strong>HL</strong> and <strong>LN</strong>. For more details about them, please refer to the <a href="https://cran.r-project.org/web/packages/nebula/vignettes/nebula_example.html#difference-between-nebula-ln-and-nebula-hl" target="_blank">NEBULA manual</a>.</p>
<p>Here is the DEGinfo.csv we described above:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="ex">comparisonName,sample,cluster,group,alt,ref,covars[+</span> separated],method[default NEBULA],model[default HL]</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Compare_Female_vs_Male,library_id,predicted.celltype1,Sex,Female,Male,,NEBULA,HL</span></span></code></pre></div>
<p>After preparing the DEGinfo.csv file, simply add it to the <strong>config.yml</strong> file:</p>
<pre><code>DEG_desp: ~/E-MTAB-11115/processing/DEGinfo.csv</code></pre>
<p>Then rerun the pipeline. The pipeline won’t run the previous steps (i.e. filtering, harmonization, etc.) again, so it will directly run the DE analysis based on the harmonized h5ad file.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span> Path/to/a/config/file</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span> ~/E-MTAB-11115/processing/config.yml</span></code></pre></div>
<p>This step may take a few hours to run. The output files will be generated in the directory:</p>
<pre><code>outputdir
    ... previous results         #Omitted, see section 6.4
    ├── ProjName_scDEG           #Folder containing DEG results
      ├── Compare_Female_vs_Male
        ├── Female.vs.Male_predicted.celltype1:astrocytes_NEBULA.csv    #Each comparison has three associated files
        ├── Female.vs.Male_predicted.celltype1:astrocytes_NEBULA.png
        ├── Female.vs.Male_predicted.celltype1:astrocytes_NEBULA.QC.pdf
        ├── Female.vs.Male_predicted.celltype1:discarded_NEBULA.csv
        ├── Female.vs.Male_predicted.celltype1:discarded_NEBULA.png
        ├── Female.vs.Male_predicted.celltype1:discarded_NEBULA.QC.pdf
        ...
        └── env.rds</code></pre>
<p>An example of the DE csv table (Female.vs.Male_predicted.celltype1/astrocytes_NEBULA.csv, top 5 rows):</p>
<p>In the table below, the most significant gene is <a href="https://www.ncbi.nlm.nih.gov/gene/213742" target="_blank">Xist</a>, a well-known X-chromosome gene highly expressed in female.</p>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<caption>
(#tab:unnamed-chunk-14)DEG analysis of Female.vs.Male in astrocytes
</caption>
<thead>
<tr>
<th style="text-align:center;">
res.tab.ID
</th>
<th style="text-align:center;">
res.tab.log2FC
</th>
<th style="text-align:center;">
res.tab.Pvalue
</th>
<th style="text-align:center;">
res.tab.FDR
</th>
<th style="text-align:center;">
res.tab.algorithm
</th>
<th style="text-align:center;">
res.tab.convergence
</th>
<th style="text-align:center;">
res.tab.metric
</th>
<th style="text-align:center;">
res.ls.summary.logFC_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.logFC_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.se_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.se_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.p_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.p_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.gene_id
</th>
<th style="text-align:center;">
res.ls.summary.gene
</th>
<th style="text-align:center;">
res.ls.overdispersion.Subject
</th>
<th style="text-align:center;">
res.ls.overdispersion.Cell
</th>
<th style="text-align:center;">
res.ls.convergence
</th>
<th style="text-align:center;">
res.ls.algorithm
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center;">
Xist
</td>
<td style="text-align:center;">
9.676171
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
6115.1780
</td>
<td style="text-align:center;">
-6.693213
</td>
<td style="text-align:center;">
0.1276737
</td>
<td style="text-align:center;">
0.0234622
</td>
<td style="text-align:center;">
0.0469395
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0065289
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
Rgs20
</td>
<td style="text-align:center;">
0.0016728
</td>
<td style="text-align:center;">
0.4393152
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Pnpla7
</td>
<td style="text-align:center;">
-1.230034
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
1653.1553
</td>
<td style="text-align:center;">
-9.321192
</td>
<td style="text-align:center;">
-0.1694550
</td>
<td style="text-align:center;">
0.0534180
</td>
<td style="text-align:center;">
0.1065177
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.1116405
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
Atp6v1h
</td>
<td style="text-align:center;">
0.0064320
</td>
<td style="text-align:center;">
0.3327316
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Etnppl
</td>
<td style="text-align:center;">
-1.592058
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
525.8328
</td>
<td style="text-align:center;">
-8.511808
</td>
<td style="text-align:center;">
0.0917939
</td>
<td style="text-align:center;">
0.0282832
</td>
<td style="text-align:center;">
0.0565059
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.1042691
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;">
Rb1cc1
</td>
<td style="text-align:center;">
0.0001000
</td>
<td style="text-align:center;">
0.1725113
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Zbtb16
</td>
<td style="text-align:center;">
-1.448309
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
1260.9518
</td>
<td style="text-align:center;">
-9.896064
</td>
<td style="text-align:center;">
-0.0918014
</td>
<td style="text-align:center;">
0.0584728
</td>
<td style="text-align:center;">
0.1165549
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.4309170
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
4732440D04Rik
</td>
<td style="text-align:center;">
0.0014745
</td>
<td style="text-align:center;">
0.7288504
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Sorcs2
</td>
<td style="text-align:center;">
1.137956
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
1398.0881
</td>
<td style="text-align:center;">
-8.573613
</td>
<td style="text-align:center;">
-0.2252848
</td>
<td style="text-align:center;">
0.0303869
</td>
<td style="text-align:center;">
0.0604625
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0001945
</td>
<td style="text-align:center;">
5
</td>
<td style="text-align:center;">
Pcmtd1
</td>
<td style="text-align:center;">
0.0001000
</td>
<td style="text-align:center;">
0.3502587
</td>
<td style="text-align:center;">
-10
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
</tbody>
</table>
</div>
<p>An example of the DE png file (Female.vs.Male_predicted.celltype1/astrocytes_NEBULA.png):</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Female_vs_Male_Astrocytes.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Female_vs_Male_Astrocytes.png" /></a></p>
</div>
<div id="example-2" class="section level2 hasAnchor" number="8.2">
<h2 class="hasAnchor"><span class="header-section-number">8.2</span> Example 2<a href="#example-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the second example, we are going to show how to perform DE analysis <strong>between two cell types</strong>. To achieve this, we added one artificial column (called ‘SelectAll’) in the <strong>sampleMeta.csv</strong> file, and assigned the same values (‘Everything’) to all the data:</p>
<pre><code>Sample_Name,h5path,Sex,SelectAll
5705STDY8058280,/home/ysun4/testSinglecell/ExternalData/5705STDY8058280_filtered_feature_bc_matrix.h5,Female,Everything
5705STDY8058281,/home/ysun4/testSinglecell/ExternalData/5705STDY8058281_filtered_feature_bc_matrix.h5,Female,Everything
5705STDY8058282,/home/ysun4/testSinglecell/ExternalData/5705STDY8058282_filtered_feature_bc_matrix.h5,Female,Everything
5705STDY8058283,/home/ysun4/testSinglecell/ExternalData/5705STDY8058283_filtered_feature_bc_matrix.h5,Male,Everything
5705STDY8058284,/home/ysun4/testSinglecell/ExternalData/5705STDY8058284_filtered_feature_bc_matrix.h5,Male,Everything
5705STDY8058285,/home/ysun4/testSinglecell/ExternalData/5705STDY8058285_filtered_feature_bc_matrix.h5,Male,Everything</code></pre>
<p>Then we prepare the <strong>DEGinfo.csv</strong> file in this way. The ‘SelectAll’ was filled into the <em>cluster</em> column, and since it only has one value, it actually selects all cells when running the pipeline. Then it performs <strong>astrocytes</strong> to <strong>Oligos</strong> comparison.</p>
<pre><code>comparisonName,sample,cluster,group,alt,ref,covars[+ separated],method[default NEBULA],model[default HL]
Comparison_Ast_vs_Oligo,library_id,SelectAll,predicted.celltype1,astrocytes,Oligos,,NEBULA,HL</code></pre>
<p>Then we need to run the scRNASequest pipeline from the beginning, which allows the ‘SelectAll’ column to be embedded into the harmonized h5ad output. As long as we have DEGinfo.csv included in the config.yml, it will run through the DE analysis and produce the following results:</p>
<pre><code>outputdir
    ... previous results         #Omitted, see section 6.4
    ├── ProjName_scDEG           #Folder containing DEG results
      ├── Comparison_Ast_vs_Oligo
        ├── astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.csv    #Each comparison has three associated files
        ├── astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.png
        ├── astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.QC.pdf
        ├── env.rds
    ├── ProjName_scDEG.cmd.json
    ├── deg6602_SelectAll_predicted.celltype1_NEBULA_HL_Everything_astrocytes_Oligos.error  #Standard error messages
    └── ProjName_scDEG.db</code></pre>
<p>An example of the DE csv table (astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.csv, top 5 rows):</p>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<caption>
(#tab:unnamed-chunk-15)DEG analysis of astrocytes.vs.Oligos
</caption>
<thead>
<tr>
<th style="text-align:center;">
res.tab.ID
</th>
<th style="text-align:center;">
res.tab.log2FC
</th>
<th style="text-align:center;">
res.tab.Pvalue
</th>
<th style="text-align:center;">
res.tab.FDR
</th>
<th style="text-align:center;">
res.tab.algorithm
</th>
<th style="text-align:center;">
res.tab.convergence
</th>
<th style="text-align:center;">
res.tab.metric
</th>
<th style="text-align:center;">
res.ls.summary.logFC_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.logFC_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.se_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.se_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.p_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.p_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.gene_id
</th>
<th style="text-align:center;">
res.ls.summary.gene
</th>
<th style="text-align:center;">
res.ls.overdispersion.Subject
</th>
<th style="text-align:center;">
res.ls.overdispersion.Cell
</th>
<th style="text-align:center;">
res.ls.convergence
</th>
<th style="text-align:center;">
res.ls.algorithm
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center;">
Xkr4
</td>
<td style="text-align:center;">
-3.327923
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
4177.156
</td>
<td style="text-align:center;">
-8.168075
</td>
<td style="text-align:center;">
-2.3067403
</td>
<td style="text-align:center;">
0.0367924
</td>
<td style="text-align:center;">
0.0561022
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0000000
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
Xkr4
</td>
<td style="text-align:center;">
0.0067297
</td>
<td style="text-align:center;">
0.6076303
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Rgs20
</td>
<td style="text-align:center;">
6.386205
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
5561.771
</td>
<td style="text-align:center;">
-9.825703
</td>
<td style="text-align:center;">
-2.5066029
</td>
<td style="text-align:center;">
0.0890548
</td>
<td style="text-align:center;">
0.1353916
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0000000
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
Gm1992
</td>
<td style="text-align:center;">
0.0398103
</td>
<td style="text-align:center;">
1.4947561
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
St18
</td>
<td style="text-align:center;">
-5.129500
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
14283.487
</td>
<td style="text-align:center;">
-10.322696
</td>
<td style="text-align:center;">
0.2176484
</td>
<td style="text-align:center;">
0.0310922
</td>
<td style="text-align:center;">
0.0690489
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0016211
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;">
Mrpl15
</td>
<td style="text-align:center;">
0.0001000
</td>
<td style="text-align:center;">
0.4188721
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Adhfe1
</td>
<td style="text-align:center;">
4.842192
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
12162.136
</td>
<td style="text-align:center;">
-10.169996
</td>
<td style="text-align:center;">
0.1119117
</td>
<td style="text-align:center;">
0.0292679
</td>
<td style="text-align:center;">
0.0669762
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0947381
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
Tcea1
</td>
<td style="text-align:center;">
0.0001000
</td>
<td style="text-align:center;">
0.7212448
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Prex2
</td>
<td style="text-align:center;">
8.368190
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
8249.837
</td>
<td style="text-align:center;">
-10.292221
</td>
<td style="text-align:center;">
4.4265797
</td>
<td style="text-align:center;">
0.0513462
</td>
<td style="text-align:center;">
0.0563860
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0000000
</td>
<td style="text-align:center;">
5
</td>
<td style="text-align:center;">
Rgs20
</td>
<td style="text-align:center;">
0.0047768
</td>
<td style="text-align:center;">
0.4563594
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
</tbody>
</table>
</div>
<p>An example of the DE png file (astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.png):</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Astrocytes_Oligos.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Astrocytes_Oligos.png" /></a></p>
<!--chapter:end:08-DEAnalysis.Rmd-->
</div>
</div>
<div id="reference-building" class="section level1 hasAnchor" number="9">
<h1 class="hasAnchor"><span class="header-section-number">9</span> Reference building<a href="#reference-building" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Reference building is critical for label transfer. To add a reference dataset into the scRNASequest pipeline, the reference matrix needs to be SCT transformed.</p>
<p>By running the following command, we can see the manual page of the reference generator, scRef:</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scRef</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="ex">*****</span> 2023-01-26 15:29:54 <span class="pp">*****</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a><span class="co">## scRNAsequest: https://github.com/interactivereport/scRNAsequest.git</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Path: /mnt/depts/dept04/compbio/edge_tools/scRNAsequest</span></span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Date: 2023-01-13 17:41:49 -0500</span></span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a><span class="co">## git HEAD: 3d463e0b127af499942b7adc2fc5af6ddfc6f11e</span></span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a><span class="ex">Loading</span> resources</span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> /path/to/a/output/folder === or === scRef /path/to/a/Ref/config/file</span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a><span class="ex">The</span> folder has to be existed.</span>
<span id="cb62-16"><a href="#cb62-16" aria-hidden="true" tabindex="-1"></a><span class="ex">The</span> Ref config file will be generated automatically when a path is provided</span>
<span id="cb62-17"><a href="#cb62-17" aria-hidden="true" tabindex="-1"></a><span class="ex">=====</span> CAUTION =====</span>
<span id="cb62-18"><a href="#cb62-18" aria-hidden="true" tabindex="-1"></a>    <span class="ex">1.</span> This process will add a seurat reference data into the scRNAsequest pipeline PERMANENTLY!</span>
<span id="cb62-19"><a href="#cb62-19" aria-hidden="true" tabindex="-1"></a>    <span class="ex">2.</span> Make sure the data provided for reference building is SCT transformed!</span>
<span id="cb62-20"><a href="#cb62-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-21"><a href="#cb62-21" aria-hidden="true" tabindex="-1"></a><span class="ex">Powered</span> by the Research Data Sciences group [zhengyu.ouyang@biogen.com<span class="kw">;</span><span class="ex">kejie.li@biogen.com]</span></span>
<span id="cb62-22"><a href="#cb62-22" aria-hidden="true" tabindex="-1"></a><span class="ex">------------</span></span></code></pre></div>
<div id="initialize-scref" class="section level2 hasAnchor" number="9.1">
<h2 class="hasAnchor"><span class="header-section-number">9.1</span> Initialize scRef<a href="#initialize-scref" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This pipeline can be initialized using an empty directory. for example, we first create a directory called ‘Reference_data’, then initiate the pipeline pointing to this directory:</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> /path/to/the/directory</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> ~/Reference_data</span></code></pre></div>
<p>After the run, a config file, <strong>refConfig.yml</strong>, and a log file will be generated in the directory. The <strong>refConfig.yml</strong> is a template of the following scRef run, passing critical parameters to the pipeline. The <strong>refConfig.yml</strong> file will be same as this <a href="https://github.com/interactivereport/scRNAsequest/blob/main/src/ref.yml" target="_blank">template</a>, but the output directory will be yours. Here is an example after filling in the configuration file:</p>
<pre><code>output: ~/Reference_data
# the following is normally located in the same folder of celldepot hosting h5ad files
ref_h5ad_raw: /path/to/ProjectName_raw_added.h5ad # full path to the h5ad file contains raw UMI along with cell annotation and layout
ref_batch: library_id
# above two parameters are ignored, if a seurat object can be located in the project folder
ref_rds:                                          # full path to the processed seurat object including SCT assay, cell annotation and layout

# All information below are required
ref_name: Reference_data                # Please provide a unique name prefer to include species and tissue (check existed by calling scAnalyzer without argument)
ref_link:                               # The web link to the information of this reference. For scAnalyzer processed data, you could provide a Cellxgene VIP link here.
ref_src: sn                             # sc/&quot;single cell&quot; or sn/&quot;single nuclei&quot;
ref_platform: 10X                       # Single cell/neuclei technology e.g. 10X, SNARE-seq2, dropSeq, ...
#list a reduction to be used (at least 50 dimensions full name from VIP or one from seurat &#39;reductions&#39; ) 
# details: https://github.com/satijalab/azimuth/wiki/Azimuth-Reference-Format
# this reduction is NOT directly used, but used to find the neighbors which is then used for computing sPCA reduction which is used in the reference
# For instance, if harmony was prefered layout then providing either &#39;harmony&#39; (50 dimention) or &#39;harmony-PCA&#39;
ref_reduction: pca 
ref_label: [predicted.celltype1]        # List the annotations (case sensitive) to be used for transferring. Please check this in the data. For our case, the header is called &#39;predicted.celltype1&#39;
                                        # The cell type label/header name you would like to transfer
publish: False                          # Should this reference be published (added permanently) into scAnalyzer
overwrite: False                        # Overwrite the existing scAnalyzer reference </code></pre>
<p>For the input data, scRef can take either an h5ad file containing raw UMI and annotatino information, or an R data file in rds format. If you have finished running a dataset using scAnalyzer, then you can directly use the Projectname_raw_added.h5ad (not Projectname.h5ad file) as input, and attach its path to <strong>ref_h5ad_raw</strong>. Alternatively, you can provide an RDS file using <strong>ref_rds</strong>. Either providing <strong>ref_h5ad_raw</strong> or <strong>ref_rds</strong> would be sufficient to the pipeline.</p>
<p>The <strong>ref_label</strong> is critical for label transfer if you would like to use this data as a reference in the future. It tells the program to use the cell type information in these (can be one more more) headers to perform label transfer. For a h5ad data, the easiest way is to open it using Cellxgene VIP, and identify the header that contains cell type labels, e.g. predicted.celltype1 in our case. For an RDS data, you could read it in R using the <a href="https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/readRDS" target="_blank">readRDS</a> function and identify the column names containing cell type annotation.</p>
<p>For <strong>ref_rds</strong> input RDS file, please make sure it has been SCTransformed, and includes UMAP embeddings based on SCT values. Here are some codes to prepare it:</p>
<pre><code>library(Seurat)

Data &lt;- readRDS(&quot;Annotated.rds&quot;)                              #The meta.data table already has cell types annotated
Data &lt;- SCTransform(Data, return.only.var.genes = F)          #Turn return.only.var.genes off
Data &lt;- RunPCA(Data, verbose = FALSE)
Data &lt;- RunUMAP(Data, dims = 1:30, verbose = FALSE)
Data &lt;- FindNeighbors(Data, dims = 1:30, verbose = FALSE)
Data &lt;- FindClusters(Data, verbose = FALSE)

saveRDS(Data_clean, file = &quot;Annotated.ForRuningscRef.rds&quot;)</code></pre>
</div>
<div id="submit-scref" class="section level2 hasAnchor" number="9.2">
<h2 class="hasAnchor"><span class="header-section-number">9.2</span> Submit scRef<a href="#submit-scref" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>After filling in the information in the <strong>refConfig.yml</strong> file, we are ready to submit the full pipeline and build the reference data for label transfer:</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> /path/to/a/Ref/config/file</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> ~/Reference_data/refConfig.yml</span></code></pre></div>
<p>Output files:</p>
<pre><code>Reference_data
    ├── init_20220630.log
    ├── refConfig.yml
    ├── refConfig.yml.20220630.log               # Output log information
    ├── Reference_data_for_scAnalyzer.rds        # The data to use for label transfer and downstream analysis
    ├── ref_notFor_scAnalyzer.rds                # Output rds file NOT for lebel transfer
    ...</code></pre>
<p>Please also check the information in the refConfig.yml.20220630.log file, and pay attention to the last few lines:</p>
<pre><code>...  #log of the running process omitted

The private reference could be used by provide the following full path to &#39;ref_name&#39; in scAnalyzer config file:
    ~/Reference_data/Reference_data_for_scAnalyzer.rds
...</code></pre>
<p>This indicates that the reference has been successfully generated, and it can be passed to scAnalyzer through the config.yml file. If you turned on <strong>“publish: True”</strong> in the <strong>refConfig.yml</strong>, this reference will be added to scAnalyzer, and you can use its name to refer it when running scAnalyzer.</p>
</div>
<div id="demo-scref" class="section level2 hasAnchor" number="9.3">
<h2 class="hasAnchor"><span class="header-section-number">9.3</span> Demo scRef<a href="#demo-scref" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this demo, we use a previously processed RDS file using cell type annotation from a public project: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE172462">GSE172462</a>. The RDS file is: GSE172462.SCT.For_scRef.rds.</p>
<p>We first run scRef by providing our working directory to generate necessary template files:</p>
<pre><code>scRef ~/demo_scRef</code></pre>
<p>Then we fill in the <code>refConfig.yml</code> file:</p>
<pre><code># please check wiki page for the details
output: ~/demo_scRef
# the following is normally located in the same folder of celldepot hosting h5ad files
ref_h5ad_raw: # full path to the h5ad file contains raw UMI along with cell annotation and layout
ref_batch: library_id
# above two parameters are ignored, if a seurat object can be located in the project folder
ref_rds: ~/demo_scRef/GSE172462.SCT.For_scRef.rds #full path to the processed seurat object including SCT assay, cell annotation and layout

# All information below are required
ref_name: demo # please provide a unique name prefer to include species and tissue (check existed by calling scAnalyzer without argument)
ref_link: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE172462 # The web link to the information of this reference
ref_src: single nuclei # sc/&quot;single cell&quot; or sn/&quot;single nuclei&quot;
ref_platform: 10X # which single cell/neuclei technology e.g. 10X, SNARE-seq2, dropSeq, ...
#list a reduction to be used (at least 50 dimensions full name from VIP or one from seurat &#39;reductions&#39; ) 
# details: https://github.com/satijalab/azimuth/wiki/Azimuth-Reference-Format
# this reduction is NOT directly used, but used to find the neighbors which is then used for computing sPCA reduction which is used in the reference
# For instance, if harmony was prefered layout then providing either &#39;harmony&#39; (50 dimention) or &#39;harmony-PCA&#39;
ref_reduction: pca 
ref_label: [celltype, neuron_subtype, manual_anno] # list the annotations (case sensitive) to be used for transferring
publish: False #should this reference be published into scAnalyzer
overwrite: False #overwrite the existing scAnalyzer reference</code></pre>
<p>In this demo, we turned off <code>publish</code> and <code>overwrite</code> to make it just to test <code>scRef</code>. This is a reduced data (with only 1000 files) to test the demo, so it may not be ideal to use it in the pipeline.</p>
<p>Finally, we can run the <code>scRef</code>:</p>
<pre><code>scRef ~/demo_scRef/refConfig.yml</code></pre>
<p>This pipeline will finish in ~1 minute and the output file will be:</p>
<pre><code>demo_for_scAnalyzer.rds</code></pre>
<!--chapter:end:09-ReferenceBuilding.Rmd-->
</div>
</div>
<div id="celldepot-publishing" class="section level1 hasAnchor" number="10">
<h1 class="hasAnchor"><span class="header-section-number">10</span> CellDepot publishing<a href="#celldepot-publishing" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p><a href="https://www.sciencedirect.com/science/article/pii/S0022283621006665" target="_blank">CellDepot</a> is a comprehensive data management platform for single-cell RNA-seq datasets. Publishing to CellDepot allows easy navigation and visualization of the data, and facilitates big data deposition.</p>
<p>If you have run through the scRNASequest pipeline using <code>scAnalyzer</code>, you will be able to use the <strong>h5ad</strong> file directly and add it to CellDepot, following the ‘4.6 Import projects’ and ‘4.7 Create project’ in the full instruction <a href="https://interactivereport.github.io/CellDepot/bookdown/docs/">here</a>. In this case, you won’t need to run <strong>sc2celldepot</strong>. However, this may be cumbersome because you have to run the full <code>scAnalyzer</code> pipeline. Sometimes, you have a public data (such as an RDS file, or raw MEX/h5 UMI count files with UMAP coordinates) with cell annotation information, and you only need to use them for visualization, rather than running though the whole analysis.</p>
<p>Thus, the scRNASequest pipeline offers a function to publish a dataset to CellDepot using <strong>sc2celldepot</strong>. This workflow will use the existing cell type information (cell type, cluter id, etc.) and create an <strong>h5ad</strong> file. The h5ad file can be further used for uploading to the CellDepot platform. If your input is an RDS file, you can also view this pipeline as a converter from RDS to h5ad.</p>
<p>By running the following command, we can see the manual page of the script:</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> sc2celldepot</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a><span class="ex">*****</span> 2023-01-26 15:25:32 <span class="pp">*****</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a><span class="co">## ExpressionAnalysis: https://github.com/interactivereport/scRNAsequest.git</span></span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Path: /mnt/depts/dept04/compbio/edge_tools/scRNAsequest</span></span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Date: 2023-01-13 17:41:49 -0500</span></span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a><span class="co">## git HEAD: 3d463e0b127af499942b7adc2fc5af6ddfc6f11e</span></span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a><span class="ex">Loading</span> resources</span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-13"><a href="#cb73-13" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> /path/to/a/output/folder === or === sc2celldepot /path/to/a/config/file</span>
<span id="cb73-14"><a href="#cb73-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-15"><a href="#cb73-15" aria-hidden="true" tabindex="-1"></a><span class="ex">Please</span> create the folder before running sc2celldepot.</span>
<span id="cb73-16"><a href="#cb73-16" aria-hidden="true" tabindex="-1"></a><span class="ex">The</span> data config file will be generated automatically when a path is provided</span>
<span id="cb73-17"><a href="#cb73-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-18"><a href="#cb73-18" aria-hidden="true" tabindex="-1"></a><span class="ex">Powered</span> by the Research Data Sciences Group [zhengyu.ouyang@biogen.com<span class="kw">;</span><span class="ex">yuhenry.sun@biogen.com]</span></span>
<span id="cb73-19"><a href="#cb73-19" aria-hidden="true" tabindex="-1"></a><span class="ex">------------</span></span></code></pre></div>
<div id="initialize-sc2celldepot" class="section level2 hasAnchor" number="10.1">
<h2 class="hasAnchor"><span class="header-section-number">10.1</span> Initialize sc2celldepot<a href="#initialize-sc2celldepot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Running <strong>sc2celldepot</strong> by providing a working directory will initiate the project and generate a template of config file. The config file template can be found: <a href="https://github.com/interactivereport/scRNAsequest/blob/main/src/external2h5ad.yml" target="_blank">template</a>.</p>
<p>First, we create a new directory under the project. Then we run the <strong>sc2celldepot</strong> command:</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> ~/E-MTAB-11115/CellDepot_publish</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> ~/E-MTAB-11115/CellDepot_publish</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a><span class="ex">E-MTAB-11115/</span></span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> E-MTAB-11115.rds</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> CellDepot_publish</span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>        <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<p>Here is an example after filling in the configuration <strong>sc2celldepot.yml</strong> file.</p>
<p>For a simple example, we just provide a prefix and an RDS file path to this yml file. The pipeline will read the metadata and convert it to an h5ad file.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The config file to process public sc/sn RNAseq and generate h5ad for celldepot</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> ~/E-MTAB-11115/CellDepot_publish</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="ex">prefix:</span> E-MTAB-11115                                        <span class="co"># the prefix of the file name of the h5ad</span></span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat RDS is avaiable, otherwise please move to next section</span></span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratObj:</span> ~/E-MTAB-11115/E-MTAB-11115.rds                  <span class="co"># the full path to the seurat RDS file with SCT &amp; RNA assay along with meta.data and reduction</span></span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratUMI:</span> RNA                                              <span class="co"># the name of the assay stores raw UMI</span></span>
<span id="cb75-8"><a href="#cb75-8" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratSCT:</span> SCT                                              <span class="co"># the name of the assay stores SCT</span></span>
<span id="cb75-9"><a href="#cb75-9" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratMeta:</span> []                                              <span class="co"># the list of cell annotations to be stored in h5ad, empty list means all meta.data entry from seurat rds</span></span>
<span id="cb75-10"><a href="#cb75-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-11"><a href="#cb75-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Expression when the seurat RDS is not available (row gene/column cell)</span></span>
<span id="cb75-12"><a href="#cb75-12" aria-hidden="true" tabindex="-1"></a><span class="co"># if the annotation files are seperated the same as expression files, they should be the same order, other wise cell ID will be used to match</span></span>
<span id="cb75-13"><a href="#cb75-13" aria-hidden="true" tabindex="-1"></a><span class="ex">expression:</span> []                                              <span class="co"># full path the gene expression file/folder (h5/csv/txt/mtx), if multiple files, please provide the list separated by &#39;,&#39;</span></span>
<span id="cb75-14"><a href="#cb75-14" aria-hidden="true" tabindex="-1"></a><span class="ex">dataUMI:</span> True                                               <span class="co"># if the above expression is UMI, if the value in expression file should be used directly, please set &quot;False&quot;</span></span>
<span id="cb75-15"><a href="#cb75-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-16"><a href="#cb75-16" aria-hidden="true" tabindex="-1"></a><span class="co"># cell annotation (cell intersection will be used, first column is the cell ID)</span></span>
<span id="cb75-17"><a href="#cb75-17" aria-hidden="true" tabindex="-1"></a><span class="ex">annotation:</span> []                                              <span class="co"># full path to the cell annotation file, first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb75-18"><a href="#cb75-18" aria-hidden="true" tabindex="-1"></a><span class="ex">annotationUse:</span> []                                           <span class="co"># the column names in the annotation file to be extracted for h5ad, empty list means all columns</span></span>
<span id="cb75-19"><a href="#cb75-19" aria-hidden="true" tabindex="-1"></a><span class="ex">sample_column:</span>                                              <span class="co"># one column header from annotation file, if the one expression file needs to be splited into each sample</span></span>
<span id="cb75-20"><a href="#cb75-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-21"><a href="#cb75-21" aria-hidden="true" tabindex="-1"></a><span class="co"># cell layout: tSNE, UMAP, PCA, if separated the same as expression files, should be the same order</span></span>
<span id="cb75-22"><a href="#cb75-22" aria-hidden="true" tabindex="-1"></a><span class="ex">reduction:</span>                                                  <span class="co"># optional (if missing UMAP will be created), other keys can be removed or added new ones, keys will be used in h5ad</span></span>
<span id="cb75-23"><a href="#cb75-23" aria-hidden="true" tabindex="-1"></a>  <span class="ex">files:</span> []                                                 <span class="co"># full path to the cell layout file (contains all layouts of a set of cells), first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb75-24"><a href="#cb75-24" aria-hidden="true" tabindex="-1"></a>  <span class="ex">umap:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb75-25"><a href="#cb75-25" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tsne:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb75-26"><a href="#cb75-26" aria-hidden="true" tabindex="-1"></a>  <span class="ex">pca:</span> []                                                   <span class="co"># column headers from layout file to be used, please use quote for each column header (can be more than 2 dimentions though only first two will be shown in VIP)</span></span></code></pre></div>
<p>Alternatively, if you don’t have an RDS file, users can provide raw UMI files together with cell type annotation files and reduction embeddings in this way:</p>
<p>Besides the required <code>output</code> and <code>prefix</code> parameters, in this example, the <code>expression</code> points to several h5 files or MEX folders, which is required. The <code>annotation</code> points to cell type annotation (and other cell level information) csv files, which are also required.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The config file to process public sc/sn RNAseq and generate h5ad for celldepot</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> ~/E-MTAB-11115/CellDepot_publish</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="ex">prefix:</span> E-MTAB-11115                                        <span class="co"># the prefix of the file name of the h5ad</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat RDS is avaiable, otherwise please move to next section</span></span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratObj:</span>                                                  <span class="co"># the full path to the seurat RDS file with SCT &amp; RNA assay along with meta.data and reduction</span></span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratUMI:</span> RNA                                              <span class="co"># the name of the assay stores raw UMI</span></span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratSCT:</span> SCT                                              <span class="co"># the name of the assay stores SCT</span></span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratMeta:</span> []                                              <span class="co"># the list of cell annotations to be stored in h5ad, empty list means all meta.data entry from seurat rds</span></span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Expression when the seurat RDS is not available (row gene/column cell)</span></span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a><span class="co"># if the annotation files are seperated the same as expression files, they should be the same order, other wise cell ID will be used to match</span></span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a><span class="ex">expression:</span> [~/E-MTAB-11115/data/5705STDY8058280_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058281_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058282_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058283_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058284_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058285_filtered_feature_bc_matrix.h5] <span class="co"># full path the gene expression file/folder (h5/csv/txt/mtx), if multiple files, please provide the list separated by &#39;,&#39;</span></span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a><span class="ex">dataUMI:</span> True                                               <span class="co"># if the above expression is UMI, if the value in expression file should be used directly, please set &quot;False&quot;</span></span>
<span id="cb76-15"><a href="#cb76-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-16"><a href="#cb76-16" aria-hidden="true" tabindex="-1"></a><span class="co"># cell annotation (cell intersection will be used, first column is the cell ID)</span></span>
<span id="cb76-17"><a href="#cb76-17" aria-hidden="true" tabindex="-1"></a><span class="ex">annotation:</span> [~/E-MTAB-11115/data/5705STDY8058280_annotation.csv,~/E-MTAB-11115/data/5705STDY8058281_annotation.csv,~/E-MTAB-11115/data/5705STDY8058282_annotation.csv,~/E-MTAB-11115/data/5705STDY8058283_annotation.csv,~/E-MTAB-11115/data/5705STDY8058284_annotation.csv,~/E-MTAB-11115/data/5705STDY8058285_annotation.csv] <span class="co"># full path to the cell annotation file, first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb76-18"><a href="#cb76-18" aria-hidden="true" tabindex="-1"></a><span class="ex">annotationUse:</span> []                                           <span class="co"># the column names in the annotation file to be extracted for h5ad, empty list means all columns</span></span>
<span id="cb76-19"><a href="#cb76-19" aria-hidden="true" tabindex="-1"></a><span class="ex">sample_column:</span>                                              <span class="co"># one column header from annotation file, if the one expression file needs to be splited into each sample</span></span>
<span id="cb76-20"><a href="#cb76-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-21"><a href="#cb76-21" aria-hidden="true" tabindex="-1"></a><span class="co"># cell layout: tSNE, UMAP, PCA, if separated the same as expression files, should be the same order</span></span>
<span id="cb76-22"><a href="#cb76-22" aria-hidden="true" tabindex="-1"></a><span class="ex">reduction:</span> <span class="co">#optional (if missing UMAP will be created), other keys can be removed or added new ones, keys will be used in h5ad</span></span>
<span id="cb76-23"><a href="#cb76-23" aria-hidden="true" tabindex="-1"></a>  <span class="ex">files:</span> []                                                 <span class="co"># full path to the cell layout file (contains all layouts of a set of cells), first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb76-24"><a href="#cb76-24" aria-hidden="true" tabindex="-1"></a>  <span class="ex">umap:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb76-25"><a href="#cb76-25" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tsne:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb76-26"><a href="#cb76-26" aria-hidden="true" tabindex="-1"></a>  <span class="ex">pca:</span> []                                                   <span class="co"># column headers from layout file to be used, please use quote for each column header (can be more than 2 dimentions though only first two will be shown in VIP)```</span></span></code></pre></div>
</div>
<div id="run-sc2celldepot" class="section level2 hasAnchor" number="10.2">
<h2 class="hasAnchor"><span class="header-section-number">10.2</span> Run sc2celldepot<a href="#run-sc2celldepot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>After preparing the <strong>sc2celldepot.yml</strong> file, we are ready to run the pipeline by the following command:</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> ~/E-MTAB-11115/CellDepot_publish/sc2celldepot.yml</span></code></pre></div>
<p>This workflow will generate the output files in the working directory, <strong>CellDepot_publish</strong>:</p>
<p>The output prefix is determined by the <strong>prefix</strong> parameter in the sc2celldepot.yml file.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="ex">E-MTAB-11115/</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> E-MTAB-11115.rds</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> CellDepot_publish</span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>        <span class="ex">├──</span> E-MTAB-11115.h5ad</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>        <span class="ex">├──</span> E-MTAB-11115.raw_added.h5ad</span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>        <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<p>Finally, copy the <strong>E-MTAB-11115.h5ad</strong> to the <a href="https://github.com/interactivereport/CellDepot" target="_blank">CellDepot</a> folder defined by sys.yml (celldepotDir) and follow the <a href="https://interactivereport.github.io/CellDepot/bookdown/docs/" target="_blank">instructions</a> for publishing it.</p>
</div>
<div id="demo-sc2celldepot" class="section level2 hasAnchor" number="10.3">
<h2 class="hasAnchor"><span class="header-section-number">10.3</span> Demo sc2celldepot<a href="#demo-sc2celldepot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To demo sc2celldepot, we use the following RDS file (GSE172462.SCT.For_scRef.rds), which we had used it to demo scRef. This RDS file contains cell type labeling and embedding information.</p>
<p>Then we run the <strong>sc2celldepot</strong> program by providing a working directory, in our case, for example:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> ~/demo_celldepot</span></code></pre></div>
<p>File hierarchy after running the above command:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/demo_celldepot</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> GSE172462.SCT.For_scRef.rds</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<p>We set up the sc2celldepot.yml in the following way, by just adding information to <code>prefix</code> and <code>seuratObj</code>:</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The config file to process public sc/sn RNAseq and generate h5ad for celldepot</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> ~/demo_celldepot</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="ex">prefix:</span> demo_sc2celldepot                                   <span class="co"># the prefix of the file name of the h5ad</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat RDS is avaiable, otherwise please move to next section</span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratObj:</span> ~/demo_celldepot/GSE172462.SCT.For_scRef.rds     <span class="co"># the full path to the seurat RDS file with SCT &amp; RNA assay along with meta.data and reduction</span></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratUMI:</span> RNA                                              <span class="co"># the name of the assay stores raw UMI</span></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratSCT:</span> SCT                                              <span class="co"># the name of the assay stores SCT</span></span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratMeta:</span> []                                              <span class="co"># the list of cell annotations to be stored in h5ad, empty list means all meta.data entry from seurat rds</span></span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Expression when the seurat RDS is not available (row gene/column cell)</span></span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a><span class="co"># if the annotation files are seperated the same as expression files, they should be the same order, other wise cell ID will be used to match</span></span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a><span class="ex">expression:</span> []                                              <span class="co"># full path the gene expression file/folder (h5/csv/txt/mtx), if multiple files, please provide the list separated by &#39;,&#39;</span></span>
<span id="cb81-12"><a href="#cb81-12" aria-hidden="true" tabindex="-1"></a><span class="ex">dataUMI:</span> True                                               <span class="co">#if the above expression is UMI, if the value in expression file should be used directly, please set &quot;False&quot;</span></span>
<span id="cb81-13"><a href="#cb81-13" aria-hidden="true" tabindex="-1"></a><span class="co"># cell annotation (cell intersection will be used, first column is the cell ID)</span></span>
<span id="cb81-14"><a href="#cb81-14" aria-hidden="true" tabindex="-1"></a><span class="ex">annotation:</span> []                                              <span class="co"># full path to the cell annotation file, first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb81-15"><a href="#cb81-15" aria-hidden="true" tabindex="-1"></a><span class="ex">annotationUse:</span> []                                           <span class="co"># the column names in the annotation file to be extracted for h5ad, empty list means all columns</span></span>
<span id="cb81-16"><a href="#cb81-16" aria-hidden="true" tabindex="-1"></a><span class="ex">sample_column:</span>                                              <span class="co"># one column header from annotation file, if the one expression file needs to be splited into each sample</span></span>
<span id="cb81-17"><a href="#cb81-17" aria-hidden="true" tabindex="-1"></a><span class="co"># cell layout: tSNE, UMAP, PCA, if separated the same as expression files, should be the same order</span></span>
<span id="cb81-18"><a href="#cb81-18" aria-hidden="true" tabindex="-1"></a><span class="ex">reduction:</span> <span class="co">#optional (if missing UMAP will be created), other keys can be removed or added new ones, keys will be used in h5ad</span></span>
<span id="cb81-19"><a href="#cb81-19" aria-hidden="true" tabindex="-1"></a>  <span class="ex">files:</span> []                                                 <span class="co"># full path to the cell layout file (contains all layouts of a set of cells), first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb81-20"><a href="#cb81-20" aria-hidden="true" tabindex="-1"></a>  <span class="ex">umap:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb81-21"><a href="#cb81-21" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tsne:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb81-22"><a href="#cb81-22" aria-hidden="true" tabindex="-1"></a>  <span class="ex">pca:</span> []                                                   <span class="co"># column headers from layout file to be used, please use quote for each column header (can be more than 2 dimentions though only first two will be shown in VIP)</span></span></code></pre></div>
<p>Then run the pipeline:</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> ~/demo_celldepot/sc2celldepot.yml</span></code></pre></div>
<p>The results will be new h5ad files as below:</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/demo_celldepot</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> demo_sc2celldepot_rds.h5ad</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> demo_sc2celldepot_rds.raw_added.h5ad</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> GSE172462.SCT.For_scRef.rds</span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> sc2celldepot.20230321.log</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<p>We provide another demo using two h5 data files from our previous <a href="https://github.com/interactivereport/scRNAsequest/tree/main/demo">demo</a> dataset. Besides that, we also need two cell level annotation files, and I have prepared them: RatMaleCigarette.annotation.csv, RatFemaleCigarette.annotation.csv.</p>
<p>Take a quick look of the csv files (Usually, the cell barcodes in the list is only a subset of all barcodes in the h5, due to filtering in previous analysis):</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-3</span> RatMaleCigarette.annotation.csv</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a><span class="ex">,library_id,predicted.celltype</span></span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a><span class="ex">AAACCCACAACTGAAA-1,RatFemaleCigarette,OPC</span></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a><span class="ex">AAACCCACACTTTATC-1,RatFemaleCigarette,Neuron</span></span></code></pre></div>
<p>Please organize these four files (2 h5 files, 2 csv files) in the working dir (in this demo: ~/demo_celldepot_2), then run the script:</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> ~/demo_celldepot_2</span></code></pre></div>
<p>The above run generated sc2celldepot.yml, and the current file hierarchy is:</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/demo_celldepot_2</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatFemaleCigarette.annotation.csv</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatFemaleCigarette.filtered_feature_bc_matrix.h5</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatMaleCigarette.annotation.csv</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatMaleCigarette.filtered_feature_bc_matrix.h5</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<p>Then we set up the sc2celldepot.yml file in this way:</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The config file to process public sc/sn RNAseq and generate h5ad for celldepot</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> ~/demo_celldepot_2</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="ex">prefix:</span> demo_sc2celldepot_2                                 <span class="co"># the prefix of the file name of the h5ad</span></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat RDS is avaiable, otherwise please move to next section</span></span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratObj:</span>                                                  <span class="co"># the full path to the seurat RDS file with SCT &amp; RNA assay along with meta.data and reduction</span></span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratUMI:</span> RNA                                              <span class="co"># the name of the assay stores raw UMI</span></span>
<span id="cb87-7"><a href="#cb87-7" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratSCT:</span> SCT                                              <span class="co"># the name of the assay stores SCT</span></span>
<span id="cb87-8"><a href="#cb87-8" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratMeta:</span> []                                              <span class="co"># the list of cell annotations to be stored in h5ad, empty list means all meta.data entry from seurat rds</span></span>
<span id="cb87-9"><a href="#cb87-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Expression when the seurat RDS is not available (row gene/column cell)</span></span>
<span id="cb87-10"><a href="#cb87-10" aria-hidden="true" tabindex="-1"></a><span class="co"># if the annotation files are seperated the same as expression files, they should be the same order, other wise cell ID will be used to match</span></span>
<span id="cb87-11"><a href="#cb87-11" aria-hidden="true" tabindex="-1"></a><span class="ex">expression:</span> [~/demo_celldepot_2/RatMaleCigarette.filtered_feature_bc_matrix.h5,~/demo_celldepot_2/RatFemaleCigarette.filtered_feature_bc_matrix.h5]</span>
<span id="cb87-12"><a href="#cb87-12" aria-hidden="true" tabindex="-1"></a>                                                            <span class="co"># full path the gene expression file/folder (h5/csv/txt/mtx), if multiple files, please provide the list separated by &#39;,&#39;</span></span>
<span id="cb87-13"><a href="#cb87-13" aria-hidden="true" tabindex="-1"></a><span class="ex">dataUMI:</span> True                                               <span class="co">#if the above expression is UMI, if the value in expression file should be used directly, please set &quot;False&quot;</span></span>
<span id="cb87-14"><a href="#cb87-14" aria-hidden="true" tabindex="-1"></a><span class="co"># cell annotation (cell intersection will be used, first column is the cell ID)</span></span>
<span id="cb87-15"><a href="#cb87-15" aria-hidden="true" tabindex="-1"></a><span class="ex">annotation:</span> [~/demo_celldepot_2/RatMaleCigarette.annotation.csv,~/demo_celldepot_2/RatFemaleCigarette.annotation.csv]</span>
<span id="cb87-16"><a href="#cb87-16" aria-hidden="true" tabindex="-1"></a>                                                            <span class="co"># full path to the cell annotation file, first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb87-17"><a href="#cb87-17" aria-hidden="true" tabindex="-1"></a><span class="ex">annotationUse:</span> []                                           <span class="co"># the column names in the annotation file to be extracted for h5ad, empty list means all columns</span></span>
<span id="cb87-18"><a href="#cb87-18" aria-hidden="true" tabindex="-1"></a><span class="ex">sample_column:</span>                                              <span class="co"># one column header from annotation file, if the one expression file needs to be splited into each sample</span></span>
<span id="cb87-19"><a href="#cb87-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-20"><a href="#cb87-20" aria-hidden="true" tabindex="-1"></a><span class="co"># cell layout: tSNE, UMAP, PCA, if separated the same as expression files, should be the same order</span></span>
<span id="cb87-21"><a href="#cb87-21" aria-hidden="true" tabindex="-1"></a><span class="ex">reduction:</span> <span class="co">#optional (if missing UMAP will be created), other keys can be removed or added new ones, keys will be used in h5ad</span></span>
<span id="cb87-22"><a href="#cb87-22" aria-hidden="true" tabindex="-1"></a>  <span class="ex">files:</span> []                                                 <span class="co"># full path to the cell layout file (contains all layouts of a set of cells), first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb87-23"><a href="#cb87-23" aria-hidden="true" tabindex="-1"></a>  <span class="ex">umap:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb87-24"><a href="#cb87-24" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tsne:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb87-25"><a href="#cb87-25" aria-hidden="true" tabindex="-1"></a>  <span class="ex">pca:</span> []                                                   <span class="co"># column headers from layout file to be used, please use quote for each column header (can be more than 2 dimentions though only first two will be shown in VIP)</span></span></code></pre></div>
<p>This will let the pipeline to process the h5 files by running normalization, and match labeled cell types to the UMAP.</p>
<p>Files after running the pipeline:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/demo_celldepot_2</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> demo_sc2celldepot.h5ad</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> demo_sc2celldepot.raw_added.h5ad</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> demo_sc2celldepot.rds</span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatFemaleCigarette.annotation.csv</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatFemaleCigarette.filtered_feature_bc_matrix.h5</span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatMaleCigarette.annotation.csv</span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatMaleCigarette.filtered_feature_bc_matrix.h5</span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<!--chapter:end:10-PublishCellDepot.Rmd-->
</div>
</div>
<div id="additional-tools" class="section level1 hasAnchor" number="11">
<h1 class="hasAnchor"><span class="header-section-number">11</span> Additional tools<a href="#additional-tools" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="scdeg" class="section level2 hasAnchor" number="11.1">
<h2 class="hasAnchor"><span class="header-section-number">11.1</span> scDEG<a href="#scdeg" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>scDEG</strong> is a standalone pipeline to run the differential expression (DE) analysis independently if you alreadly have integrated scRNA-seq data in h5ad format.</p>
<p><strong>scDEG</strong> uses the same methods as in the scRNASequest pipeline to run DE analysis. It just provides a flexible way to apply the same analysis outside the pipeline.</p>
<div id="initialize-scdeg" class="section level3 hasAnchor" number="11.1.1">
<h3 class="hasAnchor"><span class="header-section-number">11.1.1</span> Initialize scDEG<a href="#initialize-scdeg" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>scDEG requires three files:</p>
<ul>
<li><strong>UMI</strong>: Gene counts matrix, can be in rds or h5ad format.</li>
<li><strong>Meta</strong>: cell type annotation and related meta information for DE analysis, in h5ad format. If you have this information in your h5ad matrix already, you can provide the same file twice (as UMI and meta file).</li>
<li><strong>config.yml</strong>: file storing the above UMI file path and meta data path, as well as the output file path.</li>
</ul>
<p>By running the <strong>scDEG</strong> without providing any parameters, we can see the manual page of the script:</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scDEG</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="ex">scDEG</span> /path/to/a/folder === or === scDEG /path/to/a/config/file</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a><span class="ex">An</span> empty config file will be generated automatically when a folder is provided</span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a><span class="ex">Powered</span> by the Research Data Sciences Group [zhengyu.ouyang@biogen.com<span class="kw">;</span><span class="ex">yuhenry.sun@biogen.com]</span></span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a><span class="ex">------------</span></span></code></pre></div>
<p>Then providing a directory to <strong>scDEG</strong>, and it will generate a template config file:</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scDEG /path/to/a/folder</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scDEG</span> ~/DEG_results</span></code></pre></div>
<p>Template configuration file:</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="ex">UMI:</span> /path/to/Prefix.h5ad   <span class="co">#required, can be a matrix rds or a h5ad file</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="ex">meta:</span> /path/to/Prefix.h5ad  <span class="co">#required, can be a cell annotation data.frame rds or a h5ad file</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> /path/to/output/directory</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="ex">DBname:</span> cellxgeneVIP</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a><span class="ex">parallel:</span> slurm             <span class="co"># False or &quot;sge&quot; or &quot;slurm&quot;</span></span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a><span class="co">## DEG analysis for an annotation (such as disease vs health) within a cell type annotation</span></span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a><span class="ex">DEG_desp:</span> /path/to/DEGinfo.csv      <span class="co">#required for DEG analysis</span></span>
<span id="cb91-9"><a href="#cb91-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-10"><a href="#cb91-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Please be causion of changing the following default filtering</span></span>
<span id="cb91-11"><a href="#cb91-11" aria-hidden="true" tabindex="-1"></a><span class="co"># More details can be found: section 2.4 in https://pubmed.ncbi.nlm.nih.gov/35743881/</span></span>
<span id="cb91-12"><a href="#cb91-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Applies the 1st round of biostats filtering pipeline. Note that this filter is applied to all cells of the experiment</span></span>
<span id="cb91-13"><a href="#cb91-13" aria-hidden="true" tabindex="-1"></a><span class="ex">min.cells.per.gene:</span> 3 <span class="co"># if `perc_filter` is FALSE, then keep only genes that have expression in at least min.cells.per.gene</span></span>
<span id="cb91-14"><a href="#cb91-14" aria-hidden="true" tabindex="-1"></a><span class="ex">min.genes.per.cell:</span> 250 <span class="co"># keep cells with expression in at least min.genes.per.cell genes.</span></span>
<span id="cb91-15"><a href="#cb91-15" aria-hidden="true" tabindex="-1"></a><span class="ex">min.perc.cells.per.gene:</span> 0.00 <span class="co">#if &#39;perc_filter&#39;`&#39; is TRUE, then keep only genes that have expression in at least min.per.cells.per.gene * 100 percent of cells</span></span>
<span id="cb91-16"><a href="#cb91-16" aria-hidden="true" tabindex="-1"></a><span class="ex">perc_filter:</span> TRUE <span class="co">#if TRUE, apply the cells.per.gene filter using percentages (expressed as a decimal) rather than an absolute threshold</span></span>
<span id="cb91-17"><a href="#cb91-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-18"><a href="#cb91-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply the 2nd round of biostats filtering.  For &quot;group&quot; mode, the filtering is applied to `ref_group` and `alt_group` for the given cell type of interest.</span></span>
<span id="cb91-19"><a href="#cb91-19" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_min.cells.per.gene:</span> 3 <span class="co">#minimum cells expressed per gene.  This filter is applied if `R6_perc.cells.filter` is FALSE</span></span>
<span id="cb91-20"><a href="#cb91-20" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_min.perc.cells.per.gene:</span> 0.1 <span class="co"># minimum % cells expressed per gene filtering (use decimal form of percentage).  This threshold is applied if &#39;R6_perc.cells.filter&#39; is TRUE and</span></span>
<span id="cb91-21"><a href="#cb91-21" aria-hidden="true" tabindex="-1"></a> <span class="st">&#39;R6_cells.per.gene.filter&#39;</span> is TRUE</span>
<span id="cb91-22"><a href="#cb91-22" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_min.cells.per.gene.type:</span> <span class="st">&quot;or&quot;</span> <span class="co">#The type of cell per gene filtering.  If it has the value &quot;and&quot; then it requires the gene be expressed in both reference and non-reference grou</span></span>
<span id="cb91-23"><a href="#cb91-23" aria-hidden="true" tabindex="-1"></a><span class="ex">ps.</span> If it has the value <span class="st">&quot;or&quot;</span> then it requires the gene be expressed in either group</span>
<span id="cb91-24"><a href="#cb91-24" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_cells.per.gene.filter:</span> TRUE <span class="co">#TRUE means apply cells per gene filtering</span></span>
<span id="cb91-25"><a href="#cb91-25" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_perc.cells.filter:</span> TRUE <span class="co">#TRUE means apply cell.per.gene filtering by use of a percentage rather than absolute threshold.  If the percentage results in a number less than R6_m</span></span>
<span id="cb91-26"><a href="#cb91-26" aria-hidden="true" tabindex="-1"></a><span class="ex">in.cells.per.gene,</span> the code will automatically switch to min.cells.per.gene absolute thresholding</span>
<span id="cb91-27"><a href="#cb91-27" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_perc.filter:</span> FALSE <span class="co">#If TRUE, then apply the 75th percentile gene filtering</span></span>
<span id="cb91-28"><a href="#cb91-28" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_perc.filter.type:</span> <span class="st">&quot;and&quot;</span> <span class="co">#The type of percentile gene filtering.  If it has the value &quot;and&quot; then any gene that has 75th percentile of zero in both groups will be filtered out.</span></span>
<span id="cb91-29"><a href="#cb91-29" aria-hidden="true" tabindex="-1"></a>  <span class="ex">If</span> it has the value <span class="st">&quot;or&quot;</span> then any gene that has a 75th percentile of zero in either group will be filtered out.</span>
<span id="cb91-30"><a href="#cb91-30" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_perc_threshold:</span> 0.90 <span class="co">#Percentile threshold, 75th percentile is default.  Express percentile as a decimal value.</span></span>
<span id="cb91-31"><a href="#cb91-31" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_min.ave.pseudo.bulk.cpm:</span> 1 <span class="co">#cpm filtering threshold</span></span>
<span id="cb91-32"><a href="#cb91-32" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_pseudo.bulk.cpm.filter:</span> FALSE <span class="co">#if TRUE, then apply a cpm filter on the pseudo-bulk counts</span></span>
<span id="cb91-33"><a href="#cb91-33" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_min.cells.per.subj:</span> 3 <span class="co">#Minimum cells required per subject, must be a nonzero number</span></span></code></pre></div>
<p>The UMI file can be a matrix rds or a h5ad file. The meta file can be a data.frame rds or a h5ad file, containing critical information that can be used for DE analysis.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The config file for scDEG</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">UMI:</span> ~/DEG_results/TST12055_raw_added.h5ad       <span class="co">#required, can be a matrix rds or a h5ad file</span></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">meta:</span> ~/DEG_results/TST12055_raw_added.h5ad      <span class="co">#required, can be a cell annotation data.frame rds or a h5ad file</span></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">output:</span> ~/DEG_results/</span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">DBname:</span> cellxgeneVIP</span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>    <span class="co">## DEG analysis for an annotation (such as disease vs. health) within a cell type annotation</span></span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a>    <span class="ex">DEG_desp:</span> ~/DEG_results/DEGinfo.csv              <span class="co">#required for DEG analysis</span></span>
<span id="cb92-8"><a href="#cb92-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb92-9"><a href="#cb92-9" aria-hidden="true" tabindex="-1"></a>    <span class="co">## Please be cautious of changing the following default filtering.</span></span>
<span id="cb92-10"><a href="#cb92-10" aria-hidden="true" tabindex="-1"></a>    <span class="ex">min.cells.per.gene:</span> 3</span>
<span id="cb92-11"><a href="#cb92-11" aria-hidden="true" tabindex="-1"></a>    <span class="ex">min.genes.per.cell:</span> 250</span>
<span id="cb92-12"><a href="#cb92-12" aria-hidden="true" tabindex="-1"></a>    <span class="ex">min.perc.cells.per.gene:</span> 0.00</span>
<span id="cb92-13"><a href="#cb92-13" aria-hidden="true" tabindex="-1"></a>    <span class="ex">perc_filter:</span> TRUE</span>
<span id="cb92-14"><a href="#cb92-14" aria-hidden="true" tabindex="-1"></a>    <span class="co">## ...</span></span></code></pre></div>
</div>
<div id="run-de-analysis" class="section level3 hasAnchor" number="11.1.2">
<h3 class="hasAnchor"><span class="header-section-number">11.1.2</span> Run DE analysis<a href="#run-de-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Please refer the previous section <a href="#differential-expression-de-analysis">Differential expression (DE) analysis</a> to prepare the DE comparison file.</p>
<p>After preparing the config.yml and DEGinfo.csv files, run the pipeline:</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scDEG /path/to/config.yml</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scDEG</span> ~/DEG_results/config.yml</span></code></pre></div>
</div>
</div>
<div id="sctool" class="section level2 hasAnchor" number="11.2">
<h2 class="hasAnchor"><span class="header-section-number">11.2</span> scTool<a href="#sctool" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We also provide <strong>scTool</strong>, a user-friendly toolkit for modifying the h5ad file. With <strong>scTool</strong>, the user can:</p>
<ul>
<li><p>Remove existing annotations in the data, such as the cell type annotation.</p></li>
<li><p>Add cell type annotation by providing a two-column csv file, specifying the cell type category of each single cell.</p></li>
<li><p>Export a list of genes along with all annotation.</p></li>
</ul>
<p>By running the <strong>scTool</strong> without providing any parameters, we can see the manual page of the script:</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scTool</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a><span class="ex">usage:</span> scTool.py [-h] <span class="dt">{rm</span><span class="op">,</span><span class="dt">add</span><span class="op">,</span><span class="dt">export}</span> h5ad [changes]</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a><span class="ex">Additional</span> sc tools. WARNING: The input h5ad files will be over-written!</span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a><span class="ex">positional</span> arguments:</span>
<span id="cb94-8"><a href="#cb94-8" aria-hidden="true" tabindex="-1"></a>  <span class="dt">{rm</span><span class="op">,</span><span class="dt">add</span><span class="op">,</span><span class="dt">export}</span>  Modify the h5ad by either <span class="st">&quot;rm&quot;</span>, <span class="st">&quot;add&quot;</span> or <span class="st">&quot;export&quot;</span> cell level annotations.</span>
<span id="cb94-9"><a href="#cb94-9" aria-hidden="true" tabindex="-1"></a>  <span class="ex">h5ad</span>             Path to a h5ad file to be modified</span>
<span id="cb94-10"><a href="#cb94-10" aria-hidden="true" tabindex="-1"></a>  <span class="ex">changes</span>          Options:</span>
<span id="cb94-11"><a href="#cb94-11" aria-hidden="true" tabindex="-1"></a>                   <span class="ex">1.</span> A list of annotaions to be removed <span class="er">(</span><span class="ex">separated</span> by <span class="st">&quot;,&quot;</span><span class="kw">)</span> </span>
<span id="cb94-12"><a href="#cb94-12" aria-hidden="true" tabindex="-1"></a>                   <span class="ex">2.</span> A path to a csv file contains cell level annotations <span class="er">(</span><span class="ex">first</span> column is the cell ID<span class="kw">)</span></span>
<span id="cb94-13"><a href="#cb94-13" aria-hidden="true" tabindex="-1"></a>                   <span class="ex">3.</span> A list of genes <span class="er">(</span><span class="ex">separated</span> by <span class="st">&quot;,&quot;</span>, empty or max 50<span class="kw">)</span> <span class="ex">to</span> be exported along with all annotations.</span>
<span id="cb94-14"><a href="#cb94-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-15"><a href="#cb94-15" aria-hidden="true" tabindex="-1"></a><span class="ex">optional</span> arguments:</span>
<span id="cb94-16"><a href="#cb94-16" aria-hidden="true" tabindex="-1"></a>  <span class="ex">-h,</span> <span class="at">--help</span>       show this help message and exit</span>
<span id="cb94-17"><a href="#cb94-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-18"><a href="#cb94-18" aria-hidden="true" tabindex="-1"></a><span class="ex">Powered</span> by the Research Data Sciences Group [zhengyu.ouyang@biogen.com<span class="kw">;</span><span class="ex">yuhenry.sun@biogen.com]</span></span>
<span id="cb94-19"><a href="#cb94-19" aria-hidden="true" tabindex="-1"></a><span class="ex">------------</span></span></code></pre></div>
<p>We use this pubic E-MTAB-11115 data as an example:</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png" /></a></p>
<div id="remove-annotations" class="section level3 hasAnchor" number="11.2.1">
<h3 class="hasAnchor"><span class="header-section-number">11.2.1</span> Remove annotations<a href="#remove-annotations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To remove annotation, we can simply pass the annotation header names to the program, and it will remove these headers, and re-write the original file. Please back-up the original file first if you would like to keep a copy ot it. As an example, here we remove the following headers: predicted.celltype2 and predicted.location.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scTool</span> rm ProjName_raw_added.h5ad predicted.celltype2,predicted.location</span></code></pre></div>
</div>
<div id="add-annotations" class="section level3 hasAnchor" number="11.2.2">
<h3 class="hasAnchor"><span class="header-section-number">11.2.2</span> Add annotations<a href="#add-annotations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To provide a new annotation, the user needs to prepare a csv file containing cell barcode and new annotation names. For example, we use the following csv file to include a <strong>manual.curated.labels</strong> cell type labeling to the original data:</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-2</span> manual.curated.labels.csv</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a><span class="ex">AAACCCAAGGAAGTAG-1-5705STDY8058280,neurons</span></span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a><span class="ex">AAACCCAAGGGCAGTT-1-5705STDY8058280,Oligos</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Run the scTool:</span></span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a><span class="ex">scTool</span> add ProjName_raw_added.h5ad manual.curated.labels.csv</span></code></pre></div>
</div>
<div id="export-gene-expressions" class="section level3 hasAnchor" number="11.2.3">
<h3 class="hasAnchor"><span class="header-section-number">11.2.3</span> Export gene expressions<a href="#export-gene-expressions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This function exports gene expression values of provided genes, along with all other annotations. This provides a flexible way to examine the expression values of a set of genes.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scTool</span> export ProjName_raw_added.h5ad App,Olig1,P2ry12</span></code></pre></div>
<p>The output result is a .csv file: ProjName_raw_added.h5ad.csv. Here is an example of the output file:</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/Expression.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/Expression.png" /></a></p>
<!--chapter:end:11-AdditionalTools.Rmd-->
</div>
</div>
</div>
<!--bookdown:body:end-->
            </section>

          </div>
        </div>
      </div>
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
    </div>
  </div>
<!--bookdown:config-->

</body>

</html>
