```{r, echo=FALSE}

#Retrieve information from the previously saved bookdown.info.txt file
PWD <- system("pwd", intern = TRUE)
Info <- paste0(PWD, "/bookdown.info.txt")
Data <- read.table(Info)
Source_code <- Data$V2[1]
Working_dir <- Data$V2[2]
Config_file <- Data$V2[3]
PNG_dir <- Data$V2[4]

#Extract information from the Config_file
config <- sapply(yaml::read_yaml(Config_file),unlist)
ProjectName <- config$prj_name
```

# QC plots

**Quality control (QC)** is of single-cell RNA-Seq experiment is critical to downstream analyses. In this section, we present a series of QC plots for better understanding of data quality.

Please be aware that these figures were generated based on the metrics_summary.csv files associated with each sample. For the processed data directly downloaded from NCBI/GEO or EBI, they may not have these information, and thus no QC plots will be generated here.

In total there are 19 figures showing different aspects of the data. These plots can be further grouped into the following categories:

- **General QC**

   - Estimated Number of Cells
   
   - Median UMI Counts per Cell
   
   - Median Genes per Cell
   
   - Number of Reads

   - Mean Reads per Cell
   
   - Total Genes Detected
   
   - Valid Barcodes%
   
   - Sequencing Saturation%
   
   - Fraction Reads in Cells%
   
- **Q30 Bases Percentage**
   
   - Q30 Bases in Barcode%
   
   - Q30 Bases in RNA Read%
   
   - Q30 Bases in UMI%
   
- **Reads Mapping Summary**

   - Reads Mapped to Genome%
   
   - Reads Mapped Confidently to Genome%
   
   - Reads Mapped Confidently to Intergenic Regions%
   
   - Reads Mapped Confidently to Intronic Regions%
   
   - Reads Mapped Confidently to Exonic Regions%
   
   - Reads Mapped Confidently to Transcriptome%
   
   - Reads Mapped Antisense to Gene%
   
## QC summary table

The following table shows a summary of all QC metrics if provided to the pipeline.

```{r, echo=FALSE, results = "asis"}
suppressMessages(library(knitr))
suppressMessages(library(kableExtra))

if (file.exists(paste0(Working_dir, "sequencingQC.csv"))){
  Data <- read.csv(paste0(Working_dir, "sequencingQC.csv"))
  kable(Data, caption = "Summary of sqeuencing QC metrics", align = "c", booktabs = TRUE, format.args = list(big.mark = ",")) %>% 
     kable_styling(full_width = F) %>%
     kableExtra::scroll_box(width = "100%")
}else{
  cat("No table generated because the sequencingQC.csv is not provided.") 
}
```

```{r, echo=FALSE, results = "asis"}
cat("\n\n")
```

- **Estimated.Number.of.Cells**: Total cells estimated. 10X strategy can target up to 10,000 cells per sample. For more information, see [here](https://kb.10xgenomics.com/hc/en-us/articles/115001800523-What-is-the-minimum-number-of-cells-that-can-be-profiled-){target="_blank"}.

- **Mean.Reads.per.Cell**: 10X recommends 30,000 and 70,000 reads per cell. Usually, 50,000 would be a good number.

- **Median.Genes.per.Cell**: This depends on the biology that how many genes expressed, and how many genes in the annotation. For mouse datasets, it is usually ~2,000 genes.

- **Sequencing.Saturation**: This measures how saturate is the sequencing. For more information, see [here](https://kb.10xgenomics.com/hc/en-us/articles/115005062366-What-is-sequencing-saturation-){target="_blank"}.

- **Reads.Mapped.Confidently.to.Intronic.Regions**: Percentage of reads mapped to annotated introns. This rate is expected to be higher in single-nucleus RNAseq experiments, usually >50%. For more information, see [here](https://kb.10xgenomics.com/hc/en-us/articles/360000087552-Why-do-I-have-a-high-percentage-of-reads-mapping-to-intronic-regions-){target="_blank"}.

- **Reads.Mapped.Antisense.to.Gene**: It it also common to see antisense mapping reads to genes. In mouse data, this rate can be >40%, but it also depends on your data.

- **Median.UMI.Counts.per.Cell**: The median UMI per cell is a metric independent to the annotation. It would be good to see 3,000-4,000 UMI per cell.

## General QC

The General QC plot section includes nine bar plots to visualize general metrics of the current single-cell project.

### Estimated Number of Cells

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Estimated_Number_of_Cells.png")){
  knitr::include_graphics("images/sequencingQC_Estimated_Number_of_Cells.png")
}else{
  cat("Missing figure: sequencingQC_Estimated_Number_of_Cells.png") 
}
```
   
### Median UMI Counts per Cell

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Median_UMI_Counts_per_Cell.png")){
  knitr::include_graphics("images/sequencingQC_Median_UMI_Counts_per_Cell.png")
}else{
  cat("Missing figure: sequencingQC_Median_UMI_Counts_per_Cell.png") 
}
```

### Median Genes per Cell

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Median_Genes_per_Cell.png")){
  knitr::include_graphics("images/sequencingQC_Median_Genes_per_Cell.png")
}else{
  cat("Missing figure: sequencingQC_Median_Genes_per_Cell.png") 
}
```

### Number of Reads

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Number_of_Reads.png")){
  knitr::include_graphics("images/sequencingQC_Number_of_Reads.png")
}else{
  cat("Missing figure: sequencingQC_Number_of_Reads.png") 
}
```

### Mean Reads per Cell

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Mean_Reads_per_Cell.png")){
  knitr::include_graphics("images/sequencingQC_Mean_Reads_per_Cell.png")
}else{
  cat("Missing figure: sequencingQC_Mean_Reads_per_Cell.png") 
}
```

### Total Genes Detected

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Total_Genes_Detected.png")){
  knitr::include_graphics("images/sequencingQC_Total_Genes_Detected.png")
}else{
  cat("Missing figure: sequencingQC_Total_Genes_Detected.png") 
}
```

### Valid Barcodes%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Valid_Barcodes.png")){
  knitr::include_graphics("images/sequencingQC_Valid_Barcodes.png")
}else{
  cat("Missing figure: sequencingQC_Valid_Barcodes.png") 
}
```

### Sequencing Saturation%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Sequencing_Saturation.png")){
  knitr::include_graphics("images/sequencingQC_Sequencing_Saturation.png")
}else{
  cat("Missing figure: sequencingQC_Sequencing_Saturation.png") 
}
```

### Fraction Reads in Cells%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Fraction_Reads_in_Cells.png")){
  knitr::include_graphics("images/sequencingQC_Fraction_Reads_in_Cells.png")
}else{
  cat("Missing figure: sequencingQC_Fraction_Reads_in_Cells.png") 
}
```

## Q30 Bases Percentage

The Q30 Bases Percentage includes three Q30 plots showing Barcode%, RNA Read% and UMI%.

### Q30 Bases in Barcode%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Q30_Bases_in_Barcode.png")){
  knitr::include_graphics("images/sequencingQC_Q30_Bases_in_Barcode.png")
}else{
  cat("Missing figure: sequencingQC_Q30_Bases_in_Barcode.png") 
}
```

### Q30 Bases in RNA Read%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Q30_Bases_in_RNA_Read.png")){
  knitr::include_graphics("images/sequencingQC_Q30_Bases_in_RNA_Read.png")
}else{
  cat("Missing figure: sequencingQC_Q30_Bases_in_RNA_Read.png") 
}
```

### Q30 Bases in UMI%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Q30_Bases_in_UMI.png")){
  knitr::include_graphics("images/sequencingQC_Q30_Bases_in_UMI.png")
}else{
  cat("Missing figure: sequencingQC_Q30_Bases_in_UMI.png") 
}
```

## Reads Mapping Summary

The Reads Mapping Summary section includes seven plots showing the percentage of reads mapped to different genomic features. Ideally, most of the reads should map to the Genome, as well as the Transcriptome. For single-nucleus RNA-seq data, we expect to see more reads mapped to the intronic region, compared to a single-cell RNA-seq data.

### Genome%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_to_Genome.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_to_Genome.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_to_Genome.png") 
}
```

### Confidently to Genome%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Confidently_to_Genome.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Confidently_to_Genome.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Confidently_to_Genome.png") 
}
```

### Confidently to Intergenic Regions%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Confidently_to_Intergenic_Regions.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Confidently_to_Intergenic_Regions.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Confidently_to_Intergenic_Regions.png") 
}
```

### Confidently to Intronic Regions%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Confidently_to_Intronic_Regions.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Confidently_to_Intronic_Regions.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Confidently_to_Intronic_Regions.png") 
}
```

### Confidently to Exonic Regions%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Confidently_to_Exonic_Regions.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Confidently_to_Exonic_Regions.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Confidently_to_Exonic_Regions.png") 
}
```

### Confidently to Transcriptome%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Confidently_to_Transcriptome.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Confidently_to_Transcriptome.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Confidently_to_Transcriptome.png") 
}
```

### Antisense to Gene%

```{r echo=FALSE, out.width = '80%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Antisense_to_Gene.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Antisense_to_Gene.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Antisense_to_Gene.png") 
}
```
