<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 1 QC plots | Bookdown report of E-MTAB-11115</title>
  <meta name="description" content="Chapter 1 QC plots | Bookdown report of E-MTAB-11115" />
  <meta name="generator" content="bookdown 0.26 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 1 QC plots | Bookdown report of E-MTAB-11115" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 1 QC plots | Bookdown report of E-MTAB-11115" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="pre-filtering-plots.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />




<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">scRNASequest</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Introduction</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#project-overview"><i class="fa fa-check"></i>Project Overview</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#sample-meta-information"><i class="fa fa-check"></i>Sample meta information</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="qc-plots.html"><a href="qc-plots.html"><i class="fa fa-check"></i><b>1</b> QC plots</a>
<ul>
<li class="chapter" data-level="1.1" data-path="qc-plots.html"><a href="qc-plots.html#qc-summary-table"><i class="fa fa-check"></i><b>1.1</b> QC summary table</a></li>
<li class="chapter" data-level="1.2" data-path="qc-plots.html"><a href="qc-plots.html#general-qc"><i class="fa fa-check"></i><b>1.2</b> General QC</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="qc-plots.html"><a href="qc-plots.html#estimated-number-of-cells"><i class="fa fa-check"></i><b>1.2.1</b> Estimated Number of Cells</a></li>
<li class="chapter" data-level="1.2.2" data-path="qc-plots.html"><a href="qc-plots.html#median-umi-counts-per-cell"><i class="fa fa-check"></i><b>1.2.2</b> Median UMI Counts per Cell</a></li>
<li class="chapter" data-level="1.2.3" data-path="qc-plots.html"><a href="qc-plots.html#median-genes-per-cell"><i class="fa fa-check"></i><b>1.2.3</b> Median Genes per Cell</a></li>
<li class="chapter" data-level="1.2.4" data-path="qc-plots.html"><a href="qc-plots.html#number-of-reads"><i class="fa fa-check"></i><b>1.2.4</b> Number of Reads</a></li>
<li class="chapter" data-level="1.2.5" data-path="qc-plots.html"><a href="qc-plots.html#mean-reads-per-cell"><i class="fa fa-check"></i><b>1.2.5</b> Mean Reads per Cell</a></li>
<li class="chapter" data-level="1.2.6" data-path="qc-plots.html"><a href="qc-plots.html#total-genes-detected"><i class="fa fa-check"></i><b>1.2.6</b> Total Genes Detected</a></li>
<li class="chapter" data-level="1.2.7" data-path="qc-plots.html"><a href="qc-plots.html#valid-barcodes"><i class="fa fa-check"></i><b>1.2.7</b> Valid Barcodes%</a></li>
<li class="chapter" data-level="1.2.8" data-path="qc-plots.html"><a href="qc-plots.html#sequencing-saturation"><i class="fa fa-check"></i><b>1.2.8</b> Sequencing Saturation%</a></li>
<li class="chapter" data-level="1.2.9" data-path="qc-plots.html"><a href="qc-plots.html#fraction-reads-in-cells"><i class="fa fa-check"></i><b>1.2.9</b> Fraction Reads in Cells%</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="qc-plots.html"><a href="qc-plots.html#q30-bases-percentage"><i class="fa fa-check"></i><b>1.3</b> Q30 Bases Percentage</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="qc-plots.html"><a href="qc-plots.html#q30-bases-in-barcode"><i class="fa fa-check"></i><b>1.3.1</b> Q30 Bases in Barcode%</a></li>
<li class="chapter" data-level="1.3.2" data-path="qc-plots.html"><a href="qc-plots.html#q30-bases-in-rna-read"><i class="fa fa-check"></i><b>1.3.2</b> Q30 Bases in RNA Read%</a></li>
<li class="chapter" data-level="1.3.3" data-path="qc-plots.html"><a href="qc-plots.html#q30-bases-in-umi"><i class="fa fa-check"></i><b>1.3.3</b> Q30 Bases in UMI%</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="qc-plots.html"><a href="qc-plots.html#reads-mapping-summary"><i class="fa fa-check"></i><b>1.4</b> Reads Mapping Summary</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="qc-plots.html"><a href="qc-plots.html#genome"><i class="fa fa-check"></i><b>1.4.1</b> Genome%</a></li>
<li class="chapter" data-level="1.4.2" data-path="qc-plots.html"><a href="qc-plots.html#confidently-to-genome"><i class="fa fa-check"></i><b>1.4.2</b> Confidently to Genome%</a></li>
<li class="chapter" data-level="1.4.3" data-path="qc-plots.html"><a href="qc-plots.html#confidently-to-intergenic-regions"><i class="fa fa-check"></i><b>1.4.3</b> Confidently to Intergenic Regions%</a></li>
<li class="chapter" data-level="1.4.4" data-path="qc-plots.html"><a href="qc-plots.html#confidently-to-intronic-regions"><i class="fa fa-check"></i><b>1.4.4</b> Confidently to Intronic Regions%</a></li>
<li class="chapter" data-level="1.4.5" data-path="qc-plots.html"><a href="qc-plots.html#confidently-to-exonic-regions"><i class="fa fa-check"></i><b>1.4.5</b> Confidently to Exonic Regions%</a></li>
<li class="chapter" data-level="1.4.6" data-path="qc-plots.html"><a href="qc-plots.html#confidently-to-transcriptome"><i class="fa fa-check"></i><b>1.4.6</b> Confidently to Transcriptome%</a></li>
<li class="chapter" data-level="1.4.7" data-path="qc-plots.html"><a href="qc-plots.html#antisense-to-gene"><i class="fa fa-check"></i><b>1.4.7</b> Antisense to Gene%</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="pre-filtering-plots.html"><a href="pre-filtering-plots.html"><i class="fa fa-check"></i><b>2</b> Pre-filtering Plots</a>
<ul>
<li class="chapter" data-level="2.1" data-path="pre-filtering-plots.html"><a href="pre-filtering-plots.html#counts-and-genes"><i class="fa fa-check"></i><b>2.1</b> Counts and Genes</a></li>
<li class="chapter" data-level="2.2" data-path="pre-filtering-plots.html"><a href="pre-filtering-plots.html#number-of-genes"><i class="fa fa-check"></i><b>2.2</b> Number of Genes</a></li>
<li class="chapter" data-level="2.3" data-path="pre-filtering-plots.html"><a href="pre-filtering-plots.html#top-genes"><i class="fa fa-check"></i><b>2.3</b> Top Genes</a></li>
<li class="chapter" data-level="2.4" data-path="pre-filtering-plots.html"><a href="pre-filtering-plots.html#percentage-of-top-50-genes"><i class="fa fa-check"></i><b>2.4</b> Percentage of Top 50 Genes</a></li>
<li class="chapter" data-level="2.5" data-path="pre-filtering-plots.html"><a href="pre-filtering-plots.html#percentage-of-top-100-genes"><i class="fa fa-check"></i><b>2.5</b> Percentage of Top 100 Genes</a></li>
<li class="chapter" data-level="2.6" data-path="pre-filtering-plots.html"><a href="pre-filtering-plots.html#percentage-of-top-200-genes"><i class="fa fa-check"></i><b>2.6</b> Percentage of Top 200 Genes</a></li>
<li class="chapter" data-level="2.7" data-path="pre-filtering-plots.html"><a href="pre-filtering-plots.html#percentage-of-top-500-genes"><i class="fa fa-check"></i><b>2.7</b> Percentage of Top 500 Genes</a></li>
<li class="chapter" data-level="2.8" data-path="pre-filtering-plots.html"><a href="pre-filtering-plots.html#percentage-of-mitochondria"><i class="fa fa-check"></i><b>2.8</b> Percentage of Mitochondria</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="post-filtering-plots.html"><a href="post-filtering-plots.html"><i class="fa fa-check"></i><b>3</b> Post-filtering Plots</a>
<ul>
<li class="chapter" data-level="3.1" data-path="post-filtering-plots.html"><a href="post-filtering-plots.html#counts-and-genes-1"><i class="fa fa-check"></i><b>3.1</b> Counts and Genes</a></li>
<li class="chapter" data-level="3.2" data-path="post-filtering-plots.html"><a href="post-filtering-plots.html#number-of-genes-1"><i class="fa fa-check"></i><b>3.2</b> Number of Genes</a></li>
<li class="chapter" data-level="3.3" data-path="post-filtering-plots.html"><a href="post-filtering-plots.html#top-genes-1"><i class="fa fa-check"></i><b>3.3</b> Top Genes</a></li>
<li class="chapter" data-level="3.4" data-path="post-filtering-plots.html"><a href="post-filtering-plots.html#percentage-of-top-50-genes-1"><i class="fa fa-check"></i><b>3.4</b> Percentage of Top 50 Genes</a></li>
<li class="chapter" data-level="3.5" data-path="post-filtering-plots.html"><a href="post-filtering-plots.html#percentage-of-top-100-genes-1"><i class="fa fa-check"></i><b>3.5</b> Percentage of Top 100 Genes</a></li>
<li class="chapter" data-level="3.6" data-path="post-filtering-plots.html"><a href="post-filtering-plots.html#percentage-of-top-200-genes-1"><i class="fa fa-check"></i><b>3.6</b> Percentage of Top 200 Genes</a></li>
<li class="chapter" data-level="3.7" data-path="post-filtering-plots.html"><a href="post-filtering-plots.html#percentage-of-top-500-genes-1"><i class="fa fa-check"></i><b>3.7</b> Percentage of Top 500 Genes</a></li>
<li class="chapter" data-level="3.8" data-path="post-filtering-plots.html"><a href="post-filtering-plots.html#percentage-of-mitochondria-1"><i class="fa fa-check"></i><b>3.8</b> Percentage of Mitochondria</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bookdown report of E-MTAB-11115</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="qc-plots" class="section level1 hasAnchor" number="1">
<h1><span class="header-section-number">Chapter 1</span> QC plots<a href="qc-plots.html#qc-plots" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p><strong>Quality control (QC)</strong> is of single-cell RNA-Seq experiment is critical to downstream analyses. In this section, we present a series of QC plots for better understanding of data quality.</p>
<p>Please be aware that these figures were generated based on the metrics_summary.csv files associated with each sample. For the processed data directly downloaded from NCBI/GEO or EBI, they may not have these information, and thus no QC plots will be generated here.</p>
<p>In total there are 19 figures showing different aspects of the data. These plots can be further grouped into the following categories:</p>
<ul>
<li><p><strong>General QC</strong></p>
<ul>
<li><p>Estimated Number of Cells</p></li>
<li><p>Median UMI Counts per Cell</p></li>
<li><p>Median Genes per Cell</p></li>
<li><p>Number of Reads</p></li>
<li><p>Mean Reads per Cell</p></li>
<li><p>Total Genes Detected</p></li>
<li><p>Valid Barcodes%</p></li>
<li><p>Sequencing Saturation%</p></li>
<li><p>Fraction Reads in Cells%</p></li>
</ul></li>
<li><p><strong>Q30 Bases Percentage</strong></p>
<ul>
<li><p>Q30 Bases in Barcode%</p></li>
<li><p>Q30 Bases in RNA Read%</p></li>
<li><p>Q30 Bases in UMI%</p></li>
</ul></li>
<li><p><strong>Reads Mapping Summary</strong></p>
<ul>
<li><p>Reads Mapped to Genome%</p></li>
<li><p>Reads Mapped Confidently to Genome%</p></li>
<li><p>Reads Mapped Confidently to Intergenic Regions%</p></li>
<li><p>Reads Mapped Confidently to Intronic Regions%</p></li>
<li><p>Reads Mapped Confidently to Exonic Regions%</p></li>
<li><p>Reads Mapped Confidently to Transcriptome%</p></li>
<li><p>Reads Mapped Antisense to Gene%</p></li>
</ul></li>
</ul>
<div id="qc-summary-table" class="section level2 hasAnchor" number="1.1">
<h2><span class="header-section-number">1.1</span> QC summary table<a href="qc-plots.html#qc-summary-table" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The following table shows a summary of all QC metrics if provided to the pipeline.</p>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:unnamed-chunk-4">Table 1.1: </span>Summary of sqeuencing QC metrics
</caption>
<thead>
<tr>
<th style="text-align:center;">
X
</th>
<th style="text-align:center;">
Estimated.Number.of.Cells
</th>
<th style="text-align:center;">
Mean.Reads.per.Cell
</th>
<th style="text-align:center;">
Median.Genes.per.Cell
</th>
<th style="text-align:center;">
Number.of.Reads
</th>
<th style="text-align:center;">
Valid.Barcodes.
</th>
<th style="text-align:center;">
Sequencing.Saturation.
</th>
<th style="text-align:center;">
Q30.Bases.in.Barcode.
</th>
<th style="text-align:center;">
Q30.Bases.in.RNA.Read.
</th>
<th style="text-align:center;">
Q30.Bases.in.Sample.Index.
</th>
<th style="text-align:center;">
Q30.Bases.in.UMI.
</th>
<th style="text-align:center;">
Reads.Mapped.to.Genome.
</th>
<th style="text-align:center;">
Reads.Mapped.Confidently.to.Genome.
</th>
<th style="text-align:center;">
Reads.Mapped.Confidently.to.Intergenic.Regions.
</th>
<th style="text-align:center;">
Reads.Mapped.Confidently.to.Intronic.Regions.
</th>
<th style="text-align:center;">
Reads.Mapped.Confidently.to.Exonic.Regions.
</th>
<th style="text-align:center;">
Reads.Mapped.Confidently.to.Transcriptome.
</th>
<th style="text-align:center;">
Reads.Mapped.Antisense.to.Gene.
</th>
<th style="text-align:center;">
Fraction.Reads.in.Cells.
</th>
<th style="text-align:center;">
Total.Genes.Detected
</th>
<th style="text-align:center;">
Median.UMI.Counts.per.Cell
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center;">
5705STDY8058280
</td>
<td style="text-align:center;">
8,809
</td>
<td style="text-align:center;">
33,409
</td>
<td style="text-align:center;">
2,082
</td>
<td style="text-align:center;">
294,303,040
</td>
<td style="text-align:center;">
96.4
</td>
<td style="text-align:center;">
44.0
</td>
<td style="text-align:center;">
96.3
</td>
<td style="text-align:center;">
76.2
</td>
<td style="text-align:center;">
92.7
</td>
<td style="text-align:center;">
95.7
</td>
<td style="text-align:center;">
90.8
</td>
<td style="text-align:center;">
89.0
</td>
<td style="text-align:center;">
5.4
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
83.7
</td>
<td style="text-align:center;">
33.8
</td>
<td style="text-align:center;">
46.9
</td>
<td style="text-align:center;">
77.8
</td>
<td style="text-align:center;">
24,734
</td>
<td style="text-align:center;">
4,080
</td>
</tr>
<tr>
<td style="text-align:center;">
5705STDY8058281
</td>
<td style="text-align:center;">
8,913
</td>
<td style="text-align:center;">
35,170
</td>
<td style="text-align:center;">
2,088
</td>
<td style="text-align:center;">
313,473,408
</td>
<td style="text-align:center;">
96.3
</td>
<td style="text-align:center;">
44.8
</td>
<td style="text-align:center;">
96.2
</td>
<td style="text-align:center;">
73.8
</td>
<td style="text-align:center;">
93.1
</td>
<td style="text-align:center;">
95.5
</td>
<td style="text-align:center;">
90.3
</td>
<td style="text-align:center;">
88.5
</td>
<td style="text-align:center;">
5.3
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
83.2
</td>
<td style="text-align:center;">
32.9
</td>
<td style="text-align:center;">
47.4
</td>
<td style="text-align:center;">
79.9
</td>
<td style="text-align:center;">
24,855
</td>
<td style="text-align:center;">
4,087
</td>
</tr>
<tr>
<td style="text-align:center;">
5705STDY8058282
</td>
<td style="text-align:center;">
6,233
</td>
<td style="text-align:center;">
46,470
</td>
<td style="text-align:center;">
1,900
</td>
<td style="text-align:center;">
289,649,860
</td>
<td style="text-align:center;">
96.4
</td>
<td style="text-align:center;">
55.1
</td>
<td style="text-align:center;">
96.2
</td>
<td style="text-align:center;">
75.1
</td>
<td style="text-align:center;">
92.9
</td>
<td style="text-align:center;">
95.6
</td>
<td style="text-align:center;">
90.9
</td>
<td style="text-align:center;">
89.1
</td>
<td style="text-align:center;">
5.3
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
83.8
</td>
<td style="text-align:center;">
35.8
</td>
<td style="text-align:center;">
44.9
</td>
<td style="text-align:center;">
80.1
</td>
<td style="text-align:center;">
24,647
</td>
<td style="text-align:center;">
3,746
</td>
</tr>
<tr>
<td style="text-align:center;">
5705STDY8058283
</td>
<td style="text-align:center;">
5,751
</td>
<td style="text-align:center;">
53,592
</td>
<td style="text-align:center;">
2,521
</td>
<td style="text-align:center;">
308,213,276
</td>
<td style="text-align:center;">
96.7
</td>
<td style="text-align:center;">
55.4
</td>
<td style="text-align:center;">
96.5
</td>
<td style="text-align:center;">
72.1
</td>
<td style="text-align:center;">
88.8
</td>
<td style="text-align:center;">
95.9
</td>
<td style="text-align:center;">
90.7
</td>
<td style="text-align:center;">
88.6
</td>
<td style="text-align:center;">
5.4
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
83.3
</td>
<td style="text-align:center;">
37.7
</td>
<td style="text-align:center;">
42.2
</td>
<td style="text-align:center;">
75.3
</td>
<td style="text-align:center;">
24,695
</td>
<td style="text-align:center;">
5,414
</td>
</tr>
<tr>
<td style="text-align:center;">
5705STDY8058284
</td>
<td style="text-align:center;">
3,284
</td>
<td style="text-align:center;">
97,293
</td>
<td style="text-align:center;">
3,005
</td>
<td style="text-align:center;">
319,513,432
</td>
<td style="text-align:center;">
96.8
</td>
<td style="text-align:center;">
67.8
</td>
<td style="text-align:center;">
96.6
</td>
<td style="text-align:center;">
71.7
</td>
<td style="text-align:center;">
88.6
</td>
<td style="text-align:center;">
96.1
</td>
<td style="text-align:center;">
90.6
</td>
<td style="text-align:center;">
88.6
</td>
<td style="text-align:center;">
5.4
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
83.3
</td>
<td style="text-align:center;">
37.2
</td>
<td style="text-align:center;">
42.9
</td>
<td style="text-align:center;">
77.3
</td>
<td style="text-align:center;">
24,413
</td>
<td style="text-align:center;">
7,031
</td>
</tr>
<tr>
<td style="text-align:center;">
5705STDY8058285
</td>
<td style="text-align:center;">
11,439
</td>
<td style="text-align:center;">
28,164
</td>
<td style="text-align:center;">
1,964
</td>
<td style="text-align:center;">
322,176,634
</td>
<td style="text-align:center;">
96.3
</td>
<td style="text-align:center;">
40.6
</td>
<td style="text-align:center;">
96.2
</td>
<td style="text-align:center;">
76.3
</td>
<td style="text-align:center;">
93.2
</td>
<td style="text-align:center;">
95.5
</td>
<td style="text-align:center;">
91.3
</td>
<td style="text-align:center;">
89.5
</td>
<td style="text-align:center;">
5.4
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
84.1
</td>
<td style="text-align:center;">
37.4
</td>
<td style="text-align:center;">
43.3
</td>
<td style="text-align:center;">
78.4
</td>
<td style="text-align:center;">
24,956
</td>
<td style="text-align:center;">
3,786
</td>
</tr>
</tbody>
</table>
</div>
<ul>
<li><p><strong>Estimated.Number.of.Cells</strong>: Total cells estimated. 10X strategy can target up to 10,000 cells per sample. For more information, see <a href="https://kb.10xgenomics.com/hc/en-us/articles/115001800523-What-is-the-minimum-number-of-cells-that-can-be-profiled-" target="_blank">here</a>.</p></li>
<li><p><strong>Mean.Reads.per.Cell</strong>: 10X recommends 30,000 and 70,000 reads per cell. Usually, 50,000 would be a good number.</p></li>
<li><p><strong>Median.Genes.per.Cell</strong>: This depends on the biology that how many genes expressed, and how many genes in the annotation. For mouse datasets, it is usually ~2,000 genes.</p></li>
<li><p><strong>Sequencing.Saturation</strong>: This measures how saturate is the sequencing. For more information, see <a href="https://kb.10xgenomics.com/hc/en-us/articles/115005062366-What-is-sequencing-saturation-" target="_blank">here</a>.</p></li>
<li><p><strong>Reads.Mapped.Confidently.to.Intronic.Regions</strong>: Percentage of reads mapped to annotated introns. This rate is expected to be higher in single-nucleus RNAseq experiments, usually &gt;50%. For more information, see <a href="https://kb.10xgenomics.com/hc/en-us/articles/360000087552-Why-do-I-have-a-high-percentage-of-reads-mapping-to-intronic-regions-" target="_blank">here</a>.</p></li>
<li><p><strong>Reads.Mapped.Antisense.to.Gene</strong>: It it also common to see antisense mapping reads to genes. In mouse data, this rate can be &gt;40%, but it also depends on your data.</p></li>
<li><p><strong>Median.UMI.Counts.per.Cell</strong>: The median UMI per cell is a metric independent to the annotation. It would be good to see 3,000-4,000 UMI per cell.</p></li>
</ul>
</div>
<div id="general-qc" class="section level2 hasAnchor" number="1.2">
<h2><span class="header-section-number">1.2</span> General QC<a href="qc-plots.html#general-qc" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The General QC plot section includes nine bar plots to visualize general metrics of the current single-cell project.</p>
<div id="estimated-number-of-cells" class="section level3 hasAnchor" number="1.2.1">
<h3><span class="header-section-number">1.2.1</span> Estimated Number of Cells<a href="qc-plots.html#estimated-number-of-cells" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Estimated_Number_of_Cells.png" width="80%" /></p>
</div>
<div id="median-umi-counts-per-cell" class="section level3 hasAnchor" number="1.2.2">
<h3><span class="header-section-number">1.2.2</span> Median UMI Counts per Cell<a href="qc-plots.html#median-umi-counts-per-cell" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Median_UMI_Counts_per_Cell.png" width="80%" /></p>
</div>
<div id="median-genes-per-cell" class="section level3 hasAnchor" number="1.2.3">
<h3><span class="header-section-number">1.2.3</span> Median Genes per Cell<a href="qc-plots.html#median-genes-per-cell" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Median_Genes_per_Cell.png" width="80%" /></p>
</div>
<div id="number-of-reads" class="section level3 hasAnchor" number="1.2.4">
<h3><span class="header-section-number">1.2.4</span> Number of Reads<a href="qc-plots.html#number-of-reads" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Number_of_Reads.png" width="80%" /></p>
</div>
<div id="mean-reads-per-cell" class="section level3 hasAnchor" number="1.2.5">
<h3><span class="header-section-number">1.2.5</span> Mean Reads per Cell<a href="qc-plots.html#mean-reads-per-cell" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Mean_Reads_per_Cell.png" width="80%" /></p>
</div>
<div id="total-genes-detected" class="section level3 hasAnchor" number="1.2.6">
<h3><span class="header-section-number">1.2.6</span> Total Genes Detected<a href="qc-plots.html#total-genes-detected" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Total_Genes_Detected.png" width="80%" /></p>
</div>
<div id="valid-barcodes" class="section level3 hasAnchor" number="1.2.7">
<h3><span class="header-section-number">1.2.7</span> Valid Barcodes%<a href="qc-plots.html#valid-barcodes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Valid_Barcodes.png" width="80%" /></p>
</div>
<div id="sequencing-saturation" class="section level3 hasAnchor" number="1.2.8">
<h3><span class="header-section-number">1.2.8</span> Sequencing Saturation%<a href="qc-plots.html#sequencing-saturation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Sequencing_Saturation.png" width="80%" /></p>
</div>
<div id="fraction-reads-in-cells" class="section level3 hasAnchor" number="1.2.9">
<h3><span class="header-section-number">1.2.9</span> Fraction Reads in Cells%<a href="qc-plots.html#fraction-reads-in-cells" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Fraction_Reads_in_Cells.png" width="80%" /></p>
</div>
</div>
<div id="q30-bases-percentage" class="section level2 hasAnchor" number="1.3">
<h2><span class="header-section-number">1.3</span> Q30 Bases Percentage<a href="qc-plots.html#q30-bases-percentage" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The Q30 Bases Percentage includes three Q30 plots showing Barcode%, RNA Read% and UMI%.</p>
<div id="q30-bases-in-barcode" class="section level3 hasAnchor" number="1.3.1">
<h3><span class="header-section-number">1.3.1</span> Q30 Bases in Barcode%<a href="qc-plots.html#q30-bases-in-barcode" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Q30_Bases_in_Barcode.png" width="80%" /></p>
</div>
<div id="q30-bases-in-rna-read" class="section level3 hasAnchor" number="1.3.2">
<h3><span class="header-section-number">1.3.2</span> Q30 Bases in RNA Read%<a href="qc-plots.html#q30-bases-in-rna-read" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Q30_Bases_in_RNA_Read.png" width="80%" /></p>
</div>
<div id="q30-bases-in-umi" class="section level3 hasAnchor" number="1.3.3">
<h3><span class="header-section-number">1.3.3</span> Q30 Bases in UMI%<a href="qc-plots.html#q30-bases-in-umi" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Q30_Bases_in_UMI.png" width="80%" /></p>
</div>
</div>
<div id="reads-mapping-summary" class="section level2 hasAnchor" number="1.4">
<h2><span class="header-section-number">1.4</span> Reads Mapping Summary<a href="qc-plots.html#reads-mapping-summary" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The Reads Mapping Summary section includes seven plots showing the percentage of reads mapped to different genomic features. Ideally, most of the reads should map to the Genome, as well as the Transcriptome. For single-nucleus RNA-seq data, we expect to see more reads mapped to the intronic region, compared to a single-cell RNA-seq data.</p>
<div id="genome" class="section level3 hasAnchor" number="1.4.1">
<h3><span class="header-section-number">1.4.1</span> Genome%<a href="qc-plots.html#genome" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Reads_Mapped_to_Genome.png" width="80%" /></p>
</div>
<div id="confidently-to-genome" class="section level3 hasAnchor" number="1.4.2">
<h3><span class="header-section-number">1.4.2</span> Confidently to Genome%<a href="qc-plots.html#confidently-to-genome" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Reads_Mapped_Confidently_to_Genome.png" width="80%" /></p>
</div>
<div id="confidently-to-intergenic-regions" class="section level3 hasAnchor" number="1.4.3">
<h3><span class="header-section-number">1.4.3</span> Confidently to Intergenic Regions%<a href="qc-plots.html#confidently-to-intergenic-regions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Reads_Mapped_Confidently_to_Intergenic_Regions.png" width="80%" /></p>
</div>
<div id="confidently-to-intronic-regions" class="section level3 hasAnchor" number="1.4.4">
<h3><span class="header-section-number">1.4.4</span> Confidently to Intronic Regions%<a href="qc-plots.html#confidently-to-intronic-regions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Reads_Mapped_Confidently_to_Intronic_Regions.png" width="80%" /></p>
</div>
<div id="confidently-to-exonic-regions" class="section level3 hasAnchor" number="1.4.5">
<h3><span class="header-section-number">1.4.5</span> Confidently to Exonic Regions%<a href="qc-plots.html#confidently-to-exonic-regions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Reads_Mapped_Confidently_to_Exonic_Regions.png" width="80%" /></p>
</div>
<div id="confidently-to-transcriptome" class="section level3 hasAnchor" number="1.4.6">
<h3><span class="header-section-number">1.4.6</span> Confidently to Transcriptome%<a href="qc-plots.html#confidently-to-transcriptome" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Reads_Mapped_Confidently_to_Transcriptome.png" width="80%" /></p>
</div>
<div id="antisense-to-gene" class="section level3 hasAnchor" number="1.4.7">
<h3><span class="header-section-number">1.4.7</span> Antisense to Gene%<a href="qc-plots.html#antisense-to-gene" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="images/sequencingQC_Reads_Mapped_Antisense_to_Gene.png" width="80%" /></p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="pre-filtering-plots.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/scRNAsequest/edit/gh-pages/01_QC.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/scRNAsequest/blob/gh-pages/01_QC.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
