---
title: "scRNAsequest Slide Deck"
output: 
  revealjs::revealjs_presentation:
    self_contained: false
    transition: none
    reveal_plugins: ["search", "zoom", "notes", "chalkboard", "menu"]
    reveal_options:
      chalkboard: 
        theme: chalkboard
        toggleNotesButton: true
      menu:
        numbers: true
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=6, fig.height=5, dev='svg', echo=FALSE, warning=FALSE, message=FALSE)
```

```{css css-settings, echo=FALSE}
.reveal section img {
  border: 0px;
}
```

```{r, echo=FALSE}
#Retrieve information from the previously saved bookdown.info.txt file
PWD <- system("pwd", intern = TRUE)
Info <- paste0(PWD, "/bookdown.info.txt")
Data <- read.table(Info)
Working_dir <- Data$V2[2]
PNG_dir <- Data$V2[4]
#Extract information from the Config_file
config <- sapply(yaml::read_yaml(Config_file),unlist)
ProjectName <- config$prj_name
```

## Navigation buttons
<ul>
<li> Press "Esc" or "o" key to enter the the slide overview
<li> Alt + mouse click to zoom in/out
<li> Press "s" key to speaker view to see notes
</ul>
```{r echo=FALSE}
htmltools::tags$aside('Here are some notes.',class = "notes")
```

## Introduction

This is a slide deck (part of [scRNASequest pipeline](https://interactivereport.github.io/scRNAsequest/tutorial/docs/index.html)).

This report contains basic QC plots generated by the pipeline. If you turn on the filter_step when running [scAnalyzer](https://interactivereport.github.io/scRNAsequest/tutorial/docs/running-the-pipeline.html#scanalyzer), this report will also display QC plots before and after filtering.

Please go through the figures carefully and decide the cutoffs for data filtering.

## QC plots

This section presents a series of QC plots.

These QC plots can be grouped into the following categories:

- **General QC**
   
- **Q30 Bases Percentage**
   
- **Reads Mapping Summary**
   
## General QC

The General QC plot section includes nine bar plots to visualize general metrics of the current single-cell project.

## Estimated Number of Cells

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Estimated_Number_of_Cells.png")){
  knitr::include_graphics("images/sequencingQC_Estimated_Number_of_Cells.png")
}else{
  cat("Missing figure: sequencingQC_Estimated_Number_of_Cells.png") 
}
```
   
## Median UMI Counts per Cell

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Median_UMI_Counts_per_Cell.png")){
  knitr::include_graphics("images/sequencingQC_Median_UMI_Counts_per_Cell.png")
}else{
  cat("Missing figure: sequencingQC_Median_UMI_Counts_per_Cell.png") 
}
```

## Median Genes per Cell

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Median_Genes_per_Cell.png")){
  knitr::include_graphics("images/sequencingQC_Median_Genes_per_Cell.png")
}else{
  cat("Missing figure: sequencingQC_Median_Genes_per_Cell.png") 
}
```

## Number of Reads

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Number_of_Reads.png")){
  knitr::include_graphics("images/sequencingQC_Number_of_Reads.png")
}else{
  cat("Missing figure: sequencingQC_Number_of_Reads.png") 
}
```

## Mean Reads per Cell

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Mean_Reads_per_Cell.png")){
  knitr::include_graphics("images/sequencingQC_Mean_Reads_per_Cell.png")
}else{
  cat("Missing figure: sequencingQC_Mean_Reads_per_Cell.png") 
}
```

## Total Genes Detected

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Total_Genes_Detected.png")){
  knitr::include_graphics("images/sequencingQC_Total_Genes_Detected.png")
}else{
  cat("Missing figure: sequencingQC_Total_Genes_Detected.png") 
}
```

## Valid Barcodes%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Valid_Barcodes.png")){
  knitr::include_graphics("images/sequencingQC_Valid_Barcodes.png")
}else{
  cat("Missing figure: sequencingQC_Valid_Barcodes.png") 
}
```

## Sequencing Saturation%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Sequencing_Saturation.png")){
  knitr::include_graphics("images/sequencingQC_Sequencing_Saturation.png")
}else{
  cat("Missing figure: sequencingQC_Sequencing_Saturation.png") 
}
```

## Fraction Reads in Cells%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Fraction_Reads_in_Cells.png")){
  knitr::include_graphics("images/sequencingQC_Fraction_Reads_in_Cells.png")
}else{
  cat("Missing figure: sequencingQC_Fraction_Reads_in_Cells.png") 
}
```

## Q30 Bases Percentage

The Q30 Bases Percentage includes three Q30 plots showing Barcode%, RNA Read% and UMI%.

## Q30 Bases in Barcode%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Q30_Bases_in_Barcode.png")){
  knitr::include_graphics("images/sequencingQC_Q30_Bases_in_Barcode.png")
}else{
  cat("Missing figure: sequencingQC_Q30_Bases_in_Barcode.png") 
}
```

## Q30 Bases in RNA Read%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Q30_Bases_in_RNA_Read.png")){
  knitr::include_graphics("images/sequencingQC_Q30_Bases_in_RNA_Read.png")
}else{
  cat("Missing figure: sequencingQC_Q30_Bases_in_RNA_Read.png") 
}
```

## Q30 Bases in UMI%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Q30_Bases_in_UMI.png")){
  knitr::include_graphics("images/sequencingQC_Q30_Bases_in_UMI.png")
}else{
  cat("Missing figure: sequencingQC_Q30_Bases_in_UMI.png") 
}
```

## Reads Mapping Summary

This Reads Mapping Summary section shows the percentage of reads mapped to different genomic features.

## Genome%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_to_Genome.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_to_Genome.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_to_Genome.png") 
}
```

## Confidently to Genome%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Confidently_to_Genome.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Confidently_to_Genome.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Confidently_to_Genome.png") 
}
```

## Confidently to Intergenic Regions%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Confidently_to_Intergenic_Regions.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Confidently_to_Intergenic_Regions.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Confidently_to_Intergenic_Regions.png") 
}
```

## Confidently to Intronic Regions%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Confidently_to_Intronic_Regions.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Confidently_to_Intronic_Regions.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Confidently_to_Intronic_Regions.png") 
}
```

## Confidently to Exonic Regions%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Confidently_to_Exonic_Regions.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Confidently_to_Exonic_Regions.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Confidently_to_Exonic_Regions.png") 
}
```

## Confidently to Transcriptome%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Confidently_to_Transcriptome.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Confidently_to_Transcriptome.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Confidently_to_Transcriptome.png") 
}
```

## Antisense to Gene%

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/sequencingQC_Reads_Mapped_Antisense_to_Gene.png")){
  knitr::include_graphics("images/sequencingQC_Reads_Mapped_Antisense_to_Gene.png")
}else{
  cat("Missing figure: sequencingQC_Reads_Mapped_Antisense_to_Gene.png") 
}
```














## Pre-filtering Plots

In this section, we present 8 pre-filtering figures.
## Counts and Genes

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/prefilter.QC_couts_genes.png")){
   cat("This plot shows Total Counts (X axis) v.s. Genes (Y axis) detected in each data, before filtering.\n\n")
   knitr::include_graphics("images/prefilter.QC_couts_genes.png")
}else{
  cat("Missing figure: prefilter.QC_couts_genes.png") 
}
```

## Number of Genes

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/prefilter.QC_genes.png")){
   cat("This plot shows Number of Genes (Y axis) detected in each data, before filtering.\n\n")
   knitr::include_graphics("images/prefilter.QC_genes.png")
}else{
  cat("Missing figure: prefilter.QC_genes.png")
}
```

## Top Genes

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/prefilter.QC_topGene.png")){
   cat("This plot shows Top Genes and % of Total Counts in each data, before filtering.\n\n")
   knitr::include_graphics("images/prefilter.QC_topGene.png")
}else{
  cat("Missing figure: prefilter.QC_topGene.png")
}
```

## Percentage of Top 50 Genes

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/prefilter.QC_pct_counts_in_top_50_genes.png")){
   cat("This plot shows % of Total Counts of top 50 genes, before filtering.\n\n")
   knitr::include_graphics("images/prefilter.QC_pct_counts_in_top_50_genes.png")
}else{
  cat("Missing figure: prefilter.QC_pct_counts_in_top_50_genes.png")
}
```

## Percentage of Top 100 Genes

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/prefilter.QC_pct_counts_in_top_100_genes.png")){
   cat("This plot shows % of Total Counts of top 100 genes, before filtering.\n\n")
   knitr::include_graphics("images/prefilter.QC_pct_counts_in_top_100_genes.png")
}else{
  cat("Missing figure: prefilter.QC_pct_counts_in_top_100_genes.png")
}
```

## Percentage of Top 200 Genes

```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/prefilter.QC_pct_counts_in_top_200_genes.png")){
   cat("This plot shows % of Total Counts of top 200 genes, before filtering.\n\n")
   knitr::include_graphics("images/prefilter.QC_pct_counts_in_top_200_genes.png")
}else{
  cat("Missing figure: prefilter.QC_pct_counts_in_top_200_genes.png")
}
```

## Percentage of Top 500 Genes
```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/prefilter.QC_pct_counts_in_top_500_genes.png")){
   cat("This plot shows % of Total Counts of top 500 genes, before filtering.\n\n")
   knitr::include_graphics("images/prefilter.QC_pct_counts_in_top_500_genes.png")
}else{
  cat("Missing figure: prefilter.QC_pct_counts_in_top_500_genes.png")
}
```

## Percentage of Mitochondria
```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/prefilter.QC_pct_MT.png")){
   cat("This plot shows % of Mitochondria, before filtering.\n")
   knitr::include_graphics("images/prefilter.QC_pct_MT.png")
}else{
  cat("Missing figure: prefilter.QC_pct_MT.png")
}
```









## Post-filtering Plots
Post-filtering plots can be used to examine the filtering cutoff.
```{r, echo=FALSE, results = "asis"}
suppressMessages(library(knitr))
if (file.exists(paste0(PNG_dir, "filter.csv"))){
  Data <- read.csv(paste0(PNG_dir, "filter.csv"))
  cat("\n\n This table summarizes number of cells and genes after each filtering step.\n")
  cat("The last row shows the final number after all steps.\n\n")
  kable(Data, caption = "Data filtering", align = "c", booktabs = TRUE)
}else{
  cat("Neither filtering was performed, nor filter.csv was generated.") 
}
```

## Counts and Genes
```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/postfilter.QC_couts_genes.png")){
  cat("This plot shows Total Counts (X axis) v.s. Genes (Y axis) detected in each data, after filtering.\n\n")
  knitr::include_graphics("images/postfilter.QC_couts_genes.png")
}else{
  cat("No filtering was performed.")
}
```

## Number of Genes
```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/postfilter.QC_genes.png")){
  cat("This plot shows Number of Genes (Y axis) detected in each data, after filtering.\n\n")
  knitr::include_graphics("images/postfilter.QC_genes.png")
}else{
  cat("No filtering was performed.")
}
```

## Top Genes
```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/postfilter.QC_topGene.png")){
  cat("This plot shows Top Genes and % of Total Counts in each data, after filtering.\n\n")
  knitr::include_graphics("images/postfilter.QC_topGene.png")
}else{
  cat("No filtering was performed.")
}
```

## Percentage of Top 50 Genes
```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/postfilter.QC_pct_counts_in_top_50_genes.png")){
  cat("This plot shows % of Total Counts of top 50 genes, after filtering.\n\n")
  knitr::include_graphics("images/postfilter.QC_pct_counts_in_top_50_genes.png")
}else{
  cat("No filtering was performed.")
}
```

## Percentage of Top 100 Genes
```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/postfilter.QC_pct_counts_in_top_100_genes.png")){
  cat("This plot shows % of Total Counts of top 100 genes, after filtering.\n\n")
  knitr::include_graphics("images/postfilter.QC_pct_counts_in_top_100_genes.png")
}else{
  cat("No filtering was performed.")
}
```

## Percentage of Top 200 Genes
```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/postfilter.QC_pct_counts_in_top_200_genes.png")){
  cat("This plot shows % of Total Counts of top 200 genes, after filtering.\n\n")
  knitr::include_graphics("images/postfilter.QC_pct_counts_in_top_200_genes.png")
}else{
  cat("No filtering was performed.")
}
```

## Percentage of Top 500 Genes
```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/postfilter.QC_pct_counts_in_top_500_genes.png")){
  cat("This plot shows % of Total Counts of top 500 genes, after filtering.\n\n")
  knitr::include_graphics("images/postfilter.QC_pct_counts_in_top_500_genes.png")
}else{
  cat("No filtering was performed.")
}
```

## Percentage of Mitochondria
```{r echo=FALSE, out.width = '50%', results = "asis"}
if (file.exists("images/postfilter.QC_pct_MT.png")){
  cat("This plot shows % of Mitochondria, after filtering.\n\n")
  knitr::include_graphics("images/postfilter.QC_pct_MT.png")
}else{
  cat("No filtering was performed.")
}
```