<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 11 Additional tools | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</title>
  <meta name="description" content="Chapter 11 Additional tools | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 11 Additional tools | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 11 Additional tools | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  
  
  



<meta name="date" content="2023-08-10" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="celldepot-publishing.html"/>
<link rel="next" href="miscellaneous.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">scRNASequest</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>2</b> Introduction</a></li>
<li class="chapter" data-level="3" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>3</b> Installation</a>
<ul>
<li class="chapter" data-level="3.1" data-path="installation.html"><a href="installation.html#install-scrnasequest"><i class="fa fa-check"></i><b>3.1</b> Install scRNASequest</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="installation.html"><a href="installation.html#installation-using-conda"><i class="fa fa-check"></i><b>3.1.1</b> Installation using Conda</a></li>
<li class="chapter" data-level="3.1.2" data-path="installation.html"><a href="installation.html#installation-through-docker"><i class="fa fa-check"></i><b>3.1.2</b> Installation through Docker</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="installation.html"><a href="installation.html#configure-sys.yml-file"><i class="fa fa-check"></i><b>3.2</b> Configure sys.yml file</a></li>
<li class="chapter" data-level="3.3" data-path="installation.html"><a href="installation.html#install-cellxgene-vip"><i class="fa fa-check"></i><b>3.3</b> Install cellxgene VIP</a></li>
<li class="chapter" data-level="3.4" data-path="installation.html"><a href="installation.html#install-celldepot"><i class="fa fa-check"></i><b>3.4</b> Install CellDepot</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="demo-run.html"><a href="demo-run.html"><i class="fa fa-check"></i><b>4</b> Demo run</a>
<ul>
<li class="chapter" data-level="4.1" data-path="demo-run.html"><a href="demo-run.html#demo-run-with-conda"><i class="fa fa-check"></i><b>4.1</b> Demo run with Conda</a></li>
<li class="chapter" data-level="4.2" data-path="demo-run.html"><a href="demo-run.html#demo-run-with-docker"><i class="fa fa-check"></i><b>4.2</b> Demo run with Docker</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-preparation.html"><a href="data-preparation.html"><i class="fa fa-check"></i><b>5</b> Data preparation</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-preparation.html"><a href="data-preparation.html#public-data-in-h5-format"><i class="fa fa-check"></i><b>5.1</b> Public data in h5 format</a></li>
<li class="chapter" data-level="5.2" data-path="data-preparation.html"><a href="data-preparation.html#public-data-in-10x-mex-format"><i class="fa fa-check"></i><b>5.2</b> Public data in 10X MEX format</a></li>
<li class="chapter" data-level="5.3" data-path="data-preparation.html"><a href="data-preparation.html#self-prepared-files"><i class="fa fa-check"></i><b>5.3</b> Self-prepared files</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="input-configuration.html"><a href="input-configuration.html"><i class="fa fa-check"></i><b>6</b> Input configuration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="input-configuration.html"><a href="input-configuration.html#generate-template-files"><i class="fa fa-check"></i><b>6.1</b> Generate template files</a></li>
<li class="chapter" data-level="6.2" data-path="input-configuration.html"><a href="input-configuration.html#prepare-the-config.yml-file"><i class="fa fa-check"></i><b>6.2</b> Prepare the config.yml file</a></li>
<li class="chapter" data-level="6.3" data-path="input-configuration.html"><a href="input-configuration.html#section52"><i class="fa fa-check"></i><b>6.3</b> Prepare the sample meta file</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="input-configuration.html"><a href="input-configuration.html#h5-input"><i class="fa fa-check"></i><b>6.3.1</b> h5 input</a></li>
<li class="chapter" data-level="6.3.2" data-path="input-configuration.html"><a href="input-configuration.html#mex-input"><i class="fa fa-check"></i><b>6.3.2</b> MEX input</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="input-configuration.html"><a href="input-configuration.html#file-hierarchy"><i class="fa fa-check"></i><b>6.4</b> File hierarchy</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html"><i class="fa fa-check"></i><b>7</b> Running the pipeline</a>
<ul>
<li class="chapter" data-level="7.1" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#initial-scanalyzer-run"><i class="fa fa-check"></i><b>7.1</b> Initial scAnalyzer run</a></li>
<li class="chapter" data-level="7.2" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#review-the-bookdown-report"><i class="fa fa-check"></i><b>7.2</b> Review the bookdown report</a></li>
<li class="chapter" data-level="7.3" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#generate-a-slide-deck"><i class="fa fa-check"></i><b>7.3</b> Generate a slide deck</a></li>
<li class="chapter" data-level="7.4" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#doublet-detection"><i class="fa fa-check"></i><b>7.4</b> Doublet detection</a></li>
<li class="chapter" data-level="7.5" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#cell-type-label-transfer"><i class="fa fa-check"></i><b>7.5</b> Cell type label transfer</a></li>
<li class="chapter" data-level="7.6" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#adjust-major_cluster_rate"><i class="fa fa-check"></i><b>7.6</b> Adjust major_cluster_rate</a></li>
<li class="chapter" data-level="7.7" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#run-the-pipeline-in-parallel"><i class="fa fa-check"></i><b>7.7</b> Run the pipeline in parallel</a></li>
<li class="chapter" data-level="7.8" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#run-the-full-analysis"><i class="fa fa-check"></i><b>7.8</b> Run the full analysis</a></li>
<li class="chapter" data-level="7.9" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#visualize-the-harmonized-results"><i class="fa fa-check"></i><b>7.9</b> Visualize the harmonized results</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html"><i class="fa fa-check"></i><b>8</b> Differential expression (DE) analysis</a>
<ul>
<li class="chapter" data-level="8.1" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html#example-1"><i class="fa fa-check"></i><b>8.1</b> Example 1</a></li>
<li class="chapter" data-level="8.2" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html#example-2"><i class="fa fa-check"></i><b>8.2</b> Example 2</a></li>
<li class="chapter" data-level="8.3" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html#example-3"><i class="fa fa-check"></i><b>8.3</b> Example 3</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="reference-building.html"><a href="reference-building.html"><i class="fa fa-check"></i><b>9</b> Reference building</a>
<ul>
<li class="chapter" data-level="9.1" data-path="reference-building.html"><a href="reference-building.html#initialize-scref"><i class="fa fa-check"></i><b>9.1</b> Initialize scRef</a></li>
<li class="chapter" data-level="9.2" data-path="reference-building.html"><a href="reference-building.html#submit-scref"><i class="fa fa-check"></i><b>9.2</b> Submit scRef</a></li>
<li class="chapter" data-level="9.3" data-path="reference-building.html"><a href="reference-building.html#demo-scref"><i class="fa fa-check"></i><b>9.3</b> Demo scRef</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="celldepot-publishing.html"><a href="celldepot-publishing.html"><i class="fa fa-check"></i><b>10</b> CellDepot publishing</a>
<ul>
<li class="chapter" data-level="10.1" data-path="celldepot-publishing.html"><a href="celldepot-publishing.html#initialize-sc2celldepot"><i class="fa fa-check"></i><b>10.1</b> Initialize sc2celldepot</a></li>
<li class="chapter" data-level="10.2" data-path="celldepot-publishing.html"><a href="celldepot-publishing.html#run-sc2celldepot"><i class="fa fa-check"></i><b>10.2</b> Run sc2celldepot</a></li>
<li class="chapter" data-level="10.3" data-path="celldepot-publishing.html"><a href="celldepot-publishing.html#demo-sc2celldepot"><i class="fa fa-check"></i><b>10.3</b> Demo sc2celldepot</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="additional-tools.html"><a href="additional-tools.html"><i class="fa fa-check"></i><b>11</b> Additional tools</a>
<ul>
<li class="chapter" data-level="11.1" data-path="additional-tools.html"><a href="additional-tools.html#scrmambient"><i class="fa fa-check"></i><b>11.1</b> scRMambient</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="additional-tools.html"><a href="additional-tools.html#set-up-the-inputs"><i class="fa fa-check"></i><b>11.1.1</b> Set up the inputs</a></li>
<li class="chapter" data-level="11.1.2" data-path="additional-tools.html"><a href="additional-tools.html#run-scrmambient"><i class="fa fa-check"></i><b>11.1.2</b> Run scRMambient</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="additional-tools.html"><a href="additional-tools.html#scdeg"><i class="fa fa-check"></i><b>11.2</b> scDEG</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="additional-tools.html"><a href="additional-tools.html#initialize-scdeg"><i class="fa fa-check"></i><b>11.2.1</b> Initialize scDEG</a></li>
<li class="chapter" data-level="11.2.2" data-path="additional-tools.html"><a href="additional-tools.html#run-de-analysis"><i class="fa fa-check"></i><b>11.2.2</b> Run DE analysis</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="additional-tools.html"><a href="additional-tools.html#sctool"><i class="fa fa-check"></i><b>11.3</b> scTool</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="additional-tools.html"><a href="additional-tools.html#remove-annotations"><i class="fa fa-check"></i><b>11.3.1</b> Remove annotations</a></li>
<li class="chapter" data-level="11.3.2" data-path="additional-tools.html"><a href="additional-tools.html#add-annotations"><i class="fa fa-check"></i><b>11.3.2</b> Add annotations</a></li>
<li class="chapter" data-level="11.3.3" data-path="additional-tools.html"><a href="additional-tools.html#export-gene-expressions"><i class="fa fa-check"></i><b>11.3.3</b> Export gene expressions</a></li>
<li class="chapter" data-level="11.3.4" data-path="additional-tools.html"><a href="additional-tools.html#pseudo-bulk-aggregation"><i class="fa fa-check"></i><b>11.3.4</b> Pseudo-bulk aggregation</a></li>
<li class="chapter" data-level="11.3.5" data-path="additional-tools.html"><a href="additional-tools.html#demo-sctool"><i class="fa fa-check"></i><b>11.3.5</b> Demo scTool</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="miscellaneous.html"><a href="miscellaneous.html"><i class="fa fa-check"></i><b>12</b> Miscellaneous</a>
<ul>
<li class="chapter" data-level="12.1" data-path="miscellaneous.html"><a href="miscellaneous.html#cell-subsetting"><i class="fa fa-check"></i><b>12.1</b> Cell subsetting</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="additional-tools" class="section level1 hasAnchor" number="11">
<h1><span class="header-section-number">Chapter 11</span> Additional tools<a href="additional-tools.html#additional-tools" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="scrmambient" class="section level2 hasAnchor" number="11.1">
<h2><span class="header-section-number">11.1</span> scRMambient<a href="additional-tools.html#scrmambient" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>scRMambient</strong> removes ambient RNA using the <strong>raw</strong> h5 file (raw_feature_bc_matrix.h5) as input, run <a href="https://cellbender.readthedocs.io/en/latest/">CellBender</a>, and output new h5 files for downstream analysis.</p>
<div id="set-up-the-inputs" class="section level3 hasAnchor" number="11.1.1">
<h3><span class="header-section-number">11.1.1</span> Set up the inputs<a href="additional-tools.html#set-up-the-inputs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Besides the <strong>raw</strong> h5 file (raw_feature_bc_matrix.h5), the only required file for scRMambient is a <strong>sampleMeta_raw.csv</strong> file. There are several columns that are required: <strong>h5path</strong>, <strong>expected_cells</strong>, <strong>droplets_included</strong>, <strong>low_count_threshold</strong>,<strong>learning_rate</strong>. You can also add more sample metadata information, and they will be carried to a new sampleMeta.csv file generated after running scRMambient. Users can use that new sampleMeta.csv (it also contains some information related to CellBender run) to perform downstream <code>scAnalyzer</code> run.</p>
<ul>
<li><p><strong>expected_cells</strong>: This is the number of cells output by the Cell Ranger pipeline. We directly use the number here.</p></li>
<li><p><strong>droplets_included</strong>: This is a critical parameter for CellBender. Please follow the instruction <a href="https://cellbender.readthedocs.io/en/latest/usage/index.html">here</a> to set it. We found that in many cases 15000 is a good number, but it also depends on your data.</p></li>
<li><p><strong>low_count_threshold</strong>: We set to 15 by default.</p></li>
<li><p><strong>learning_rate</strong>: We set to 0.0001 by default.</p></li>
</ul>
<p>File hierarchy of the inputs:</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb96-1"><a href="additional-tools.html#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/test_ambient_rna_removal</span></span>
<span id="cb96-2"><a href="additional-tools.html#cb96-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> Data1.raw_feature_bc_matrix.h5</span>
<span id="cb96-3"><a href="additional-tools.html#cb96-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> Data2.raw_feature_bc_matrix.h5</span>
<span id="cb96-4"><a href="additional-tools.html#cb96-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> sampleMeta_raw.csv</span></code></pre></div>
<p>An example of the <strong>sampleMeta_raw.csv</strong> file:</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb97-1"><a href="additional-tools.html#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Sample_Name,h5path,expected_cells,droplets_included,low_count_threshold,learning_rate</span></span>
<span id="cb97-2"><a href="additional-tools.html#cb97-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Data1,~/test_ambient_rna_removal/Data1.raw_feature_bc_matrix.h5,8119,15000,15,0.0001</span></span>
<span id="cb97-3"><a href="additional-tools.html#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="ex">Data2,~/test_ambient_rna_removal/Data2.raw_feature_bc_matrix.h5,7348,15000,15,0.0001</span></span></code></pre></div>
</div>
<div id="run-scrmambient" class="section level3 hasAnchor" number="11.1.2">
<h3><span class="header-section-number">11.1.2</span> Run scRMambient<a href="additional-tools.html#run-scrmambient" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Then we run the pipeline:</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb98-1"><a href="additional-tools.html#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scRMambient</span> ~/test_ambient_rna_removal/sampleMeta_raw.csv</span></code></pre></div>
<p>The output will be a new directory called <strong>cellbender</strong>, with new h5 files generated under the h5/ directory.</p>
<p>The file hierarchy is:</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb99-1"><a href="additional-tools.html#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/test_ambient_rna_removal</span></span>
<span id="cb99-2"><a href="additional-tools.html#cb99-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> cellbender</span>
<span id="cb99-3"><a href="additional-tools.html#cb99-3" aria-hidden="true" tabindex="-1"></a>        <span class="ex">├──</span> all_cellbender.pdf                     <span class="co"># Plots generated by CellBender</span></span>
<span id="cb99-4"><a href="additional-tools.html#cb99-4" aria-hidden="true" tabindex="-1"></a>        <span class="ex">├──</span> all.log</span>
<span id="cb99-5"><a href="additional-tools.html#cb99-5" aria-hidden="true" tabindex="-1"></a>        <span class="ex">├──</span> cellbender_QC.pdf                      <span class="co"># QC plots</span></span>
<span id="cb99-6"><a href="additional-tools.html#cb99-6" aria-hidden="true" tabindex="-1"></a>        <span class="ex">├──</span> cellbender_rmRate.csv                  <span class="co"># Metrics related to the final cell number, and cell removal rates</span></span>
<span id="cb99-7"><a href="additional-tools.html#cb99-7" aria-hidden="true" tabindex="-1"></a>        <span class="ex">├──</span> h5/                                    <span class="co"># Directory for all output h5 files with ambient RNA removed</span></span>
<span id="cb99-8"><a href="additional-tools.html#cb99-8" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> Data1.cellbender_cell_barcodes.csv</span>
<span id="cb99-9"><a href="additional-tools.html#cb99-9" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> Data1.cellbender_filtered.h5</span>
<span id="cb99-10"><a href="additional-tools.html#cb99-10" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> Data1.cellbender_filtered_qc.csv</span>
<span id="cb99-11"><a href="additional-tools.html#cb99-11" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> Data1.cellbender.h5</span>
<span id="cb99-12"><a href="additional-tools.html#cb99-12" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> Data1.cellbender.log</span>
<span id="cb99-13"><a href="additional-tools.html#cb99-13" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> Data1.cellbender.pdf</span>
<span id="cb99-14"><a href="additional-tools.html#cb99-14" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> ...</span>
<span id="cb99-15"><a href="additional-tools.html#cb99-15" aria-hidden="true" tabindex="-1"></a>        <span class="ex">├──</span> sc20230317_0                           <span class="co"># File directory prepared for running scAnalyzer</span></span>
<span id="cb99-16"><a href="additional-tools.html#cb99-16" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> config.yml</span>
<span id="cb99-17"><a href="additional-tools.html#cb99-17" aria-hidden="true" tabindex="-1"></a>            <span class="ex">├──</span> DEGinfo.csv</span>
<span id="cb99-18"><a href="additional-tools.html#cb99-18" aria-hidden="true" tabindex="-1"></a>            <span class="ex">└──</span> sampleMeta.csv                     <span class="co"># In this file, &#39;h5path&#39; points to all new h5 files</span></span>
<span id="cb99-19"><a href="additional-tools.html#cb99-19" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> Data1.raw_feature_bc_matrix.h5</span>
<span id="cb99-20"><a href="additional-tools.html#cb99-20" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> Data2.raw_feature_bc_matrix.h5</span>
<span id="cb99-21"><a href="additional-tools.html#cb99-21" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> sampleMeta_raw.csv</span></code></pre></div>
</div>
</div>
<div id="scdeg" class="section level2 hasAnchor" number="11.2">
<h2><span class="header-section-number">11.2</span> scDEG<a href="additional-tools.html#scdeg" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>scDEG</strong> is a standalone pipeline to run the differential expression (DE) analysis independently if you alreadly have integrated scRNA-seq data in h5ad format.</p>
<p><strong>scDEG</strong> uses the same methods as in the scRNASequest pipeline to run DE analysis. It just provides a flexible way to apply the same analysis outside the pipeline.</p>
<div id="initialize-scdeg" class="section level3 hasAnchor" number="11.2.1">
<h3><span class="header-section-number">11.2.1</span> Initialize scDEG<a href="additional-tools.html#initialize-scdeg" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>scDEG requires three files:</p>
<ul>
<li><strong>UMI</strong>: Gene counts matrix, can be in rds or h5ad format.</li>
<li><strong>Meta</strong>: cell type annotation and related meta information for DE analysis, in h5ad format. If you have this information in your h5ad matrix already, you can provide the same file twice (as UMI and meta file).</li>
<li><strong>config.yml</strong>: file storing the above UMI file path and meta data path, as well as the output file path.</li>
</ul>
<p>By running the <strong>scDEG</strong> without providing any parameters, we can see the manual page of the script:</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb100-1"><a href="additional-tools.html#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scDEG</span>
<span id="cb100-2"><a href="additional-tools.html#cb100-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-3"><a href="additional-tools.html#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="ex">scDEG</span> /path/to/a/folder === or === scDEG /path/to/a/config/file</span>
<span id="cb100-4"><a href="additional-tools.html#cb100-4" aria-hidden="true" tabindex="-1"></a><span class="ex">An</span> empty config file will be generated automatically when a folder is provided</span>
<span id="cb100-5"><a href="additional-tools.html#cb100-5" aria-hidden="true" tabindex="-1"></a><span class="ex">Powered</span> by the Research Data Sciences Group [zhengyu.ouyang@biogen.com<span class="kw">;</span><span class="ex">yuhenry.sun@biogen.com]</span></span>
<span id="cb100-6"><a href="additional-tools.html#cb100-6" aria-hidden="true" tabindex="-1"></a><span class="ex">------------</span></span></code></pre></div>
<p>Then providing a directory to <strong>scDEG</strong>, and it will generate a template config file:</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb101-1"><a href="additional-tools.html#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scDEG /path/to/a/folder</span>
<span id="cb101-2"><a href="additional-tools.html#cb101-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-3"><a href="additional-tools.html#cb101-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb101-4"><a href="additional-tools.html#cb101-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scDEG</span> ~/DEG_results</span></code></pre></div>
<p>Template configuration file:</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb102-1"><a href="additional-tools.html#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="ex">UMI:</span> /path/to/Prefix.h5ad   <span class="co">#required, can be a matrix rds or a h5ad file</span></span>
<span id="cb102-2"><a href="additional-tools.html#cb102-2" aria-hidden="true" tabindex="-1"></a><span class="ex">meta:</span> /path/to/Prefix.h5ad  <span class="co">#required, can be a cell annotation data.frame rds or a h5ad file</span></span>
<span id="cb102-3"><a href="additional-tools.html#cb102-3" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> /path/to/output/directory</span>
<span id="cb102-4"><a href="additional-tools.html#cb102-4" aria-hidden="true" tabindex="-1"></a><span class="ex">DBname:</span> cellxgeneVIP</span>
<span id="cb102-5"><a href="additional-tools.html#cb102-5" aria-hidden="true" tabindex="-1"></a><span class="ex">parallel:</span> slurm             <span class="co"># False or &quot;sge&quot; or &quot;slurm&quot;</span></span>
<span id="cb102-6"><a href="additional-tools.html#cb102-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-7"><a href="additional-tools.html#cb102-7" aria-hidden="true" tabindex="-1"></a><span class="co">## DEG analysis for an annotation (such as disease vs health) within a cell type annotation</span></span>
<span id="cb102-8"><a href="additional-tools.html#cb102-8" aria-hidden="true" tabindex="-1"></a><span class="ex">DEG_desp:</span> /path/to/DEGinfo.csv      <span class="co">#required for DEG analysis</span></span>
<span id="cb102-9"><a href="additional-tools.html#cb102-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-10"><a href="additional-tools.html#cb102-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Please be causion of changing the following default filtering</span></span>
<span id="cb102-11"><a href="additional-tools.html#cb102-11" aria-hidden="true" tabindex="-1"></a><span class="co"># More details can be found: section 2.4 in https://pubmed.ncbi.nlm.nih.gov/35743881/</span></span>
<span id="cb102-12"><a href="additional-tools.html#cb102-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Applies the 1st round of biostats filtering pipeline. Note that this filter is applied to all cells of the experiment</span></span>
<span id="cb102-13"><a href="additional-tools.html#cb102-13" aria-hidden="true" tabindex="-1"></a><span class="ex">min.cells.per.gene:</span> 3 <span class="co"># if `perc_filter` is FALSE, then keep only genes that have expression in at least min.cells.per.gene</span></span>
<span id="cb102-14"><a href="additional-tools.html#cb102-14" aria-hidden="true" tabindex="-1"></a><span class="ex">min.genes.per.cell:</span> 250 <span class="co"># keep cells with expression in at least min.genes.per.cell genes.</span></span>
<span id="cb102-15"><a href="additional-tools.html#cb102-15" aria-hidden="true" tabindex="-1"></a><span class="ex">min.perc.cells.per.gene:</span> 0.00 <span class="co">#if &#39;perc_filter&#39;`&#39; is TRUE, then keep only genes that have expression in at least min.per.cells.per.gene * 100 percent of cells</span></span>
<span id="cb102-16"><a href="additional-tools.html#cb102-16" aria-hidden="true" tabindex="-1"></a><span class="ex">perc_filter:</span> TRUE <span class="co">#if TRUE, apply the cells.per.gene filter using percentages (expressed as a decimal) rather than an absolute threshold</span></span>
<span id="cb102-17"><a href="additional-tools.html#cb102-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-18"><a href="additional-tools.html#cb102-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply the 2nd round of biostats filtering.  For &quot;group&quot; mode, the filtering is applied to `ref_group` and `alt_group` for the given cell type of interest.</span></span>
<span id="cb102-19"><a href="additional-tools.html#cb102-19" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_min.cells.per.gene:</span> 3 <span class="co">#minimum cells expressed per gene.  This filter is applied if `R6_perc.cells.filter` is FALSE</span></span>
<span id="cb102-20"><a href="additional-tools.html#cb102-20" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_min.perc.cells.per.gene:</span> 0.1 <span class="co"># minimum % cells expressed per gene filtering (use decimal form of percentage).  This threshold is applied if &#39;R6_perc.cells.filter&#39; is TRUE and</span></span>
<span id="cb102-21"><a href="additional-tools.html#cb102-21" aria-hidden="true" tabindex="-1"></a> <span class="st">&#39;R6_cells.per.gene.filter&#39;</span> is TRUE</span>
<span id="cb102-22"><a href="additional-tools.html#cb102-22" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_min.cells.per.gene.type:</span> <span class="st">&quot;or&quot;</span> <span class="co">#The type of cell per gene filtering.  If it has the value &quot;and&quot; then it requires the gene be expressed in both reference and non-reference grou</span></span>
<span id="cb102-23"><a href="additional-tools.html#cb102-23" aria-hidden="true" tabindex="-1"></a><span class="ex">ps.</span> If it has the value <span class="st">&quot;or&quot;</span> then it requires the gene be expressed in either group</span>
<span id="cb102-24"><a href="additional-tools.html#cb102-24" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_cells.per.gene.filter:</span> TRUE <span class="co">#TRUE means apply cells per gene filtering</span></span>
<span id="cb102-25"><a href="additional-tools.html#cb102-25" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_perc.cells.filter:</span> TRUE <span class="co">#TRUE means apply cell.per.gene filtering by use of a percentage rather than absolute threshold.  If the percentage results in a number less than R6_m</span></span>
<span id="cb102-26"><a href="additional-tools.html#cb102-26" aria-hidden="true" tabindex="-1"></a><span class="ex">in.cells.per.gene,</span> the code will automatically switch to min.cells.per.gene absolute thresholding</span>
<span id="cb102-27"><a href="additional-tools.html#cb102-27" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_perc.filter:</span> FALSE <span class="co">#If TRUE, then apply the 75th percentile gene filtering</span></span>
<span id="cb102-28"><a href="additional-tools.html#cb102-28" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_perc.filter.type:</span> <span class="st">&quot;and&quot;</span> <span class="co">#The type of percentile gene filtering.  If it has the value &quot;and&quot; then any gene that has 75th percentile of zero in both groups will be filtered out.</span></span>
<span id="cb102-29"><a href="additional-tools.html#cb102-29" aria-hidden="true" tabindex="-1"></a>  <span class="ex">If</span> it has the value <span class="st">&quot;or&quot;</span> then any gene that has a 75th percentile of zero in either group will be filtered out.</span>
<span id="cb102-30"><a href="additional-tools.html#cb102-30" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_perc_threshold:</span> 0.90 <span class="co">#Percentile threshold, 75th percentile is default.  Express percentile as a decimal value.</span></span>
<span id="cb102-31"><a href="additional-tools.html#cb102-31" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_min.ave.pseudo.bulk.cpm:</span> 1 <span class="co">#cpm filtering threshold</span></span>
<span id="cb102-32"><a href="additional-tools.html#cb102-32" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_pseudo.bulk.cpm.filter:</span> FALSE <span class="co">#if TRUE, then apply a cpm filter on the pseudo-bulk counts</span></span>
<span id="cb102-33"><a href="additional-tools.html#cb102-33" aria-hidden="true" tabindex="-1"></a><span class="ex">R6_min.cells.per.subj:</span> 3 <span class="co">#Minimum cells required per subject, must be a nonzero number</span></span></code></pre></div>
<p>The UMI file can be a matrix rds or a h5ad file. The meta file can be a data.frame rds or a h5ad file, containing critical information that can be used for DE analysis.</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb103-1"><a href="additional-tools.html#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The config file for scDEG</span></span>
<span id="cb103-2"><a href="additional-tools.html#cb103-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">UMI:</span> ~/DEG_results/TST12055_raw_added.h5ad       <span class="co">#required, can be a matrix rds or a h5ad file</span></span>
<span id="cb103-3"><a href="additional-tools.html#cb103-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">meta:</span> ~/DEG_results/TST12055_raw_added.h5ad      <span class="co">#required, can be a cell annotation data.frame rds or a h5ad file</span></span>
<span id="cb103-4"><a href="additional-tools.html#cb103-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">output:</span> ~/DEG_results/</span>
<span id="cb103-5"><a href="additional-tools.html#cb103-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">DBname:</span> cellxgeneVIP</span>
<span id="cb103-6"><a href="additional-tools.html#cb103-6" aria-hidden="true" tabindex="-1"></a>    <span class="co">## DEG analysis for an annotation (such as disease vs. health) within a cell type annotation</span></span>
<span id="cb103-7"><a href="additional-tools.html#cb103-7" aria-hidden="true" tabindex="-1"></a>    <span class="ex">DEG_desp:</span> ~/DEG_results/DEGinfo.csv              <span class="co">#required for DEG analysis</span></span>
<span id="cb103-8"><a href="additional-tools.html#cb103-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb103-9"><a href="additional-tools.html#cb103-9" aria-hidden="true" tabindex="-1"></a>    <span class="co">## Please be cautious of changing the following default filtering.</span></span>
<span id="cb103-10"><a href="additional-tools.html#cb103-10" aria-hidden="true" tabindex="-1"></a>    <span class="ex">min.cells.per.gene:</span> 3</span>
<span id="cb103-11"><a href="additional-tools.html#cb103-11" aria-hidden="true" tabindex="-1"></a>    <span class="ex">min.genes.per.cell:</span> 250</span>
<span id="cb103-12"><a href="additional-tools.html#cb103-12" aria-hidden="true" tabindex="-1"></a>    <span class="ex">min.perc.cells.per.gene:</span> 0.00</span>
<span id="cb103-13"><a href="additional-tools.html#cb103-13" aria-hidden="true" tabindex="-1"></a>    <span class="ex">perc_filter:</span> TRUE</span>
<span id="cb103-14"><a href="additional-tools.html#cb103-14" aria-hidden="true" tabindex="-1"></a>    <span class="co">## ...</span></span></code></pre></div>
</div>
<div id="run-de-analysis" class="section level3 hasAnchor" number="11.2.2">
<h3><span class="header-section-number">11.2.2</span> Run DE analysis<a href="additional-tools.html#run-de-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Please refer the previous section <a href="differential-expression-de-analysis.html#differential-expression-de-analysis">Differential expression (DE) analysis</a> to prepare the DE comparison file.</p>
<p>After preparing the config.yml and DEGinfo.csv files, run the pipeline:</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb104-1"><a href="additional-tools.html#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scDEG /path/to/config.yml</span>
<span id="cb104-2"><a href="additional-tools.html#cb104-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-3"><a href="additional-tools.html#cb104-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb104-4"><a href="additional-tools.html#cb104-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scDEG</span> ~/DEG_results/config.yml</span></code></pre></div>
</div>
</div>
<div id="sctool" class="section level2 hasAnchor" number="11.3">
<h2><span class="header-section-number">11.3</span> scTool<a href="additional-tools.html#sctool" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We also provide <strong>scTool</strong>, a user-friendly toolkit for modifying the h5ad file. With <strong>scTool</strong>, the user can:</p>
<ul>
<li><p>Remove existing annotations in the data, such as the cell type annotation.</p></li>
<li><p>Add cell type annotation by providing a two-column csv file, specifying the cell type category of each single cell.</p></li>
<li><p>Export a list of genes along with all annotation.</p></li>
</ul>
<p>By running the <strong>scTool</strong> without providing any parameters, we can see the manual page of the script:</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb105-1"><a href="additional-tools.html#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scTool</span>
<span id="cb105-2"><a href="additional-tools.html#cb105-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-3"><a href="additional-tools.html#cb105-3" aria-hidden="true" tabindex="-1"></a><span class="ex">usage:</span> scTool.py [-h] <span class="dt">{rm</span><span class="op">,</span><span class="dt">add</span><span class="op">,</span><span class="dt">export</span><span class="op">,</span><span class="dt">pseudo}</span> h5ad [changes]</span>
<span id="cb105-4"><a href="additional-tools.html#cb105-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-5"><a href="additional-tools.html#cb105-5" aria-hidden="true" tabindex="-1"></a><span class="ex">Additional</span> sc tools. WARNING: The input h5ad files will be over-written!</span>
<span id="cb105-6"><a href="additional-tools.html#cb105-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-7"><a href="additional-tools.html#cb105-7" aria-hidden="true" tabindex="-1"></a><span class="ex">positional</span> arguments:</span>
<span id="cb105-8"><a href="additional-tools.html#cb105-8" aria-hidden="true" tabindex="-1"></a>  <span class="dt">{rm</span><span class="op">,</span><span class="dt">add</span><span class="op">,</span><span class="dt">export</span><span class="op">,</span><span class="dt">pseudo}</span></span>
<span id="cb105-9"><a href="additional-tools.html#cb105-9" aria-hidden="true" tabindex="-1"></a>                        <span class="ex">Modify</span> the h5ad by either <span class="st">&quot;rm&quot;</span>, <span class="st">&quot;add&quot;</span> or <span class="st">&quot;export&quot;</span> cell level annotations. </span>
<span id="cb105-10"><a href="additional-tools.html#cb105-10" aria-hidden="true" tabindex="-1"></a>                        <span class="ex">Create</span> pseudo bulk as RNAsequest analysis input by <span class="st">&quot;pseudo&quot;</span></span>
<span id="cb105-11"><a href="additional-tools.html#cb105-11" aria-hidden="true" tabindex="-1"></a>  <span class="ex">h5ad</span>                  Path to a h5ad file to be modified or extract from <span class="er">(</span><span class="ex">raw</span> UMI for pseudo<span class="kw">)</span></span>
<span id="cb105-12"><a href="additional-tools.html#cb105-12" aria-hidden="true" tabindex="-1"></a>  <span class="ex">changes</span>               Options:</span>
<span id="cb105-13"><a href="additional-tools.html#cb105-13" aria-hidden="true" tabindex="-1"></a>                           <span class="ex">1.</span> <span class="er">(</span><span class="fu">rm</span><span class="kw">)</span> <span class="ex">A</span> list of annotaions to be removed <span class="er">(</span><span class="ex">separated</span> by <span class="st">&quot;,&quot;</span><span class="kw">)</span> </span>
<span id="cb105-14"><a href="additional-tools.html#cb105-14" aria-hidden="true" tabindex="-1"></a>                           <span class="ex">2.</span> <span class="er">(</span><span class="ex">add</span><span class="kw">)</span> <span class="ex">A</span> path to a csv file contains cell level annotations <span class="er">(</span><span class="ex">first</span> column is the cell ID<span class="kw">)</span></span>
<span id="cb105-15"><a href="additional-tools.html#cb105-15" aria-hidden="true" tabindex="-1"></a>                           <span class="ex">3.</span> <span class="er">(</span><span class="bu">export</span><span class="kw">)</span> <span class="ex">A</span> list of genes <span class="er">(</span><span class="ex">separated</span> by <span class="st">&quot;,&quot;</span>, empty or max 500<span class="kw">)</span> <span class="ex">to</span> be exported along with all annotations.</span>
<span id="cb105-16"><a href="additional-tools.html#cb105-16" aria-hidden="true" tabindex="-1"></a>                           <span class="ex">4.</span> <span class="er">(</span><span class="ex">pseudo</span><span class="kw">)</span> <span class="ex">A</span> list of obs to group cells into pseudo bulk, separated by <span class="st">&quot;,&quot;</span></span>
<span id="cb105-17"><a href="additional-tools.html#cb105-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-18"><a href="additional-tools.html#cb105-18" aria-hidden="true" tabindex="-1"></a><span class="ex">optional</span> arguments:</span>
<span id="cb105-19"><a href="additional-tools.html#cb105-19" aria-hidden="true" tabindex="-1"></a>  <span class="ex">-h,</span> <span class="at">--help</span>            show this help message and exit</span></code></pre></div>
<p>We use this pubic E-MTAB-11115 data as an example:</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png" /></a></p>
<div id="remove-annotations" class="section level3 hasAnchor" number="11.3.1">
<h3><span class="header-section-number">11.3.1</span> Remove annotations<a href="additional-tools.html#remove-annotations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To remove annotation, we can simply pass the annotation header names to the program, and it will remove these headers, and re-write the original file. Please back-up the original file first if you would like to keep a copy ot it. As an example, here we remove the following headers: predicted.celltype2 and predicted.location.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb106-1"><a href="additional-tools.html#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scTool</span> rm ProjName_raw_added.h5ad predicted.celltype2,predicted.location</span></code></pre></div>
</div>
<div id="add-annotations" class="section level3 hasAnchor" number="11.3.2">
<h3><span class="header-section-number">11.3.2</span> Add annotations<a href="additional-tools.html#add-annotations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To provide a new annotation, the user needs to prepare a csv file containing cell barcode and new annotation names. For example, we use the following csv file to include a <strong>manual.curated.labels</strong> cell type labeling to the original data:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb107-1"><a href="additional-tools.html#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-2</span> manual.curated.labels.csv</span>
<span id="cb107-2"><a href="additional-tools.html#cb107-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-3"><a href="additional-tools.html#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="ex">AAACCCAAGGAAGTAG-1-5705STDY8058280,neurons</span></span>
<span id="cb107-4"><a href="additional-tools.html#cb107-4" aria-hidden="true" tabindex="-1"></a><span class="ex">AAACCCAAGGGCAGTT-1-5705STDY8058280,Oligos</span></span>
<span id="cb107-5"><a href="additional-tools.html#cb107-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-6"><a href="additional-tools.html#cb107-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Run the scTool:</span></span>
<span id="cb107-7"><a href="additional-tools.html#cb107-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-8"><a href="additional-tools.html#cb107-8" aria-hidden="true" tabindex="-1"></a><span class="ex">scTool</span> add ProjName_raw_added.h5ad manual.curated.labels.csv</span></code></pre></div>
</div>
<div id="export-gene-expressions" class="section level3 hasAnchor" number="11.3.3">
<h3><span class="header-section-number">11.3.3</span> Export gene expressions<a href="additional-tools.html#export-gene-expressions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This function exports gene expression values of provided genes, along with all other annotations. This provides a flexible way to examine the expression values of a set of genes.</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb108-1"><a href="additional-tools.html#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scTool</span> export ProjName_raw_added.h5ad App,Olig1,P2ry12</span></code></pre></div>
<p>The output result is a .csv file: ProjName_raw_added.h5ad.csv. Here is an example of the output file:</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/Expression.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/Expression.png" /></a></p>
</div>
<div id="pseudo-bulk-aggregation" class="section level3 hasAnchor" number="11.3.4">
<h3><span class="header-section-number">11.3.4</span> Pseudo-bulk aggregation<a href="additional-tools.html#pseudo-bulk-aggregation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The “scTool pseudo …” will create a folder “pseudoBulk” where the h5ad is.</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb109-1"><a href="additional-tools.html#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scTool</span> pseudo ProjName_raw_added.h5ad orig.ident,cell_type</span></code></pre></div>
<p>Output: Four files will be generated in the folder:</p>
<ul>
<li><p>genes.UMI.tsv: the UMI (raw counts) for pseudo bulk samples (columns)</p></li>
<li><p>samplesheet.tsv: contains obs (if the unique value is only one for each of pseudo bulk)</p></li>
<li><p>genes.CPM.tsv: contains normalized expression (count per million)</p></li>
<li><p>genes.CPKcell.tsv: contains normalized express (count per 1k cells)</p></li>
</ul>
<p>The output can be used in our RNASequest pipeline: <a href="https://github.com/interactivereport/RNASequest" class="uri">https://github.com/interactivereport/RNASequest</a></p>
<p>In the RNASequest pipeline, EAinit will treat this as external project, and user will have to fill in:</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb110-1"><a href="additional-tools.html#cb110-1" aria-hidden="true" tabindex="-1"></a>      <span class="ex">proj_name,</span> prj_counts <span class="er">(</span><span class="ex">genes.UMI.tsv</span><span class="kw">)</span><span class="ex">,</span> prj_TPM<span class="er">(</span><span class="ex">genes.CPM.tsv</span> or genes.CPKcell.tsv<span class="kw">)</span><span class="ex">,</span> sample_meta <span class="er">(</span><span class="ex">samplesheet.tsv</span><span class="kw">)</span><span class="ex">,</span> species</span></code></pre></div>
<p>Also, in RNASequest, <code>prj_effLength</code> is not required, but <code>covariates_adjust</code> will be disabled, and <code>gene_annotation</code> is not required. The first column in count matrix will be used as gene name.</p>
</div>
<div id="demo-sctool" class="section level3 hasAnchor" number="11.3.5">
<h3><span class="header-section-number">11.3.5</span> Demo scTool<a href="additional-tools.html#demo-sctool" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We use the h5ad file generated from demo data set to test scTool: <a href="https://github.com/interactivereport/scRNAsequest/releases/download/untagged-e92025b745dbc6f9666a/scRNASequest_demo.h5ad">scRNASequest_demo.h5ad</a>.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb111-1"><a href="additional-tools.html#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scTool</span> export scRNASequest_demo.h5ad App,Olig1,P2ry12</span></code></pre></div>
<p>The output will be a scRNASequest_demo.h5ad.csv file.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="celldepot-publishing.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="miscellaneous.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/scRNAsequest/edit/gh-pages/11-AdditionalTools.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/scRNAsequest/blob/gh-pages/11-AdditionalTools.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
