<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 CellDepot publishing | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</title>
  <meta name="description" content="Chapter 10 CellDepot publishing | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 CellDepot publishing | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 CellDepot publishing | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  
  
  



<meta name="date" content="2023-04-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="reference-building.html"/>
<link rel="next" href="additional-tools.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">scRNASequest</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>2</b> Introduction</a></li>
<li class="chapter" data-level="3" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>3</b> Installation</a>
<ul>
<li class="chapter" data-level="3.1" data-path="installation.html"><a href="installation.html#install-scrnasequest"><i class="fa fa-check"></i><b>3.1</b> Install scRNASequest</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="installation.html"><a href="installation.html#installation-using-conda"><i class="fa fa-check"></i><b>3.1.1</b> Installation using Conda</a></li>
<li class="chapter" data-level="3.1.2" data-path="installation.html"><a href="installation.html#installation-through-docker"><i class="fa fa-check"></i><b>3.1.2</b> Installation through Docker</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="installation.html"><a href="installation.html#configure-sys.yml-file"><i class="fa fa-check"></i><b>3.2</b> Configure sys.yml file</a></li>
<li class="chapter" data-level="3.3" data-path="installation.html"><a href="installation.html#install-cellxgene-vip"><i class="fa fa-check"></i><b>3.3</b> Install cellxgene VIP</a></li>
<li class="chapter" data-level="3.4" data-path="installation.html"><a href="installation.html#install-celldepot"><i class="fa fa-check"></i><b>3.4</b> Install CellDepot</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="demo-run.html"><a href="demo-run.html"><i class="fa fa-check"></i><b>4</b> Demo run</a>
<ul>
<li class="chapter" data-level="4.1" data-path="demo-run.html"><a href="demo-run.html#demo-run-with-conda"><i class="fa fa-check"></i><b>4.1</b> Demo run with Conda</a></li>
<li class="chapter" data-level="4.2" data-path="demo-run.html"><a href="demo-run.html#demo-run-with-docker"><i class="fa fa-check"></i><b>4.2</b> Demo run with Docker</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-preparation.html"><a href="data-preparation.html"><i class="fa fa-check"></i><b>5</b> Data preparation</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-preparation.html"><a href="data-preparation.html#public-data-in-h5-format"><i class="fa fa-check"></i><b>5.1</b> Public data in h5 format</a></li>
<li class="chapter" data-level="5.2" data-path="data-preparation.html"><a href="data-preparation.html#public-data-in-10x-mex-format"><i class="fa fa-check"></i><b>5.2</b> Public data in 10X MEX format</a></li>
<li class="chapter" data-level="5.3" data-path="data-preparation.html"><a href="data-preparation.html#self-prepared-files"><i class="fa fa-check"></i><b>5.3</b> Self-prepared files</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="input-configuration.html"><a href="input-configuration.html"><i class="fa fa-check"></i><b>6</b> Input configuration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="input-configuration.html"><a href="input-configuration.html#generate-template-files"><i class="fa fa-check"></i><b>6.1</b> Generate template files</a></li>
<li class="chapter" data-level="6.2" data-path="input-configuration.html"><a href="input-configuration.html#prepare-the-config.yml-file"><i class="fa fa-check"></i><b>6.2</b> Prepare the config.yml file</a></li>
<li class="chapter" data-level="6.3" data-path="input-configuration.html"><a href="input-configuration.html#section52"><i class="fa fa-check"></i><b>6.3</b> Prepare the sample meta file</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="input-configuration.html"><a href="input-configuration.html#h5-input"><i class="fa fa-check"></i><b>6.3.1</b> h5 input</a></li>
<li class="chapter" data-level="6.3.2" data-path="input-configuration.html"><a href="input-configuration.html#mex-input"><i class="fa fa-check"></i><b>6.3.2</b> MEX input</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="input-configuration.html"><a href="input-configuration.html#file-hierarchy"><i class="fa fa-check"></i><b>6.4</b> File hierarchy</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html"><i class="fa fa-check"></i><b>7</b> Running the pipeline</a>
<ul>
<li class="chapter" data-level="7.1" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#initial-scanalyzer-run"><i class="fa fa-check"></i><b>7.1</b> Initial scAnalyzer run</a></li>
<li class="chapter" data-level="7.2" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#review-the-bookdown-report"><i class="fa fa-check"></i><b>7.2</b> Review the bookdown report</a></li>
<li class="chapter" data-level="7.3" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#generate-a-slide-deck"><i class="fa fa-check"></i><b>7.3</b> Generate a slide deck</a></li>
<li class="chapter" data-level="7.4" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#doublet-detection"><i class="fa fa-check"></i><b>7.4</b> Doublet detection</a></li>
<li class="chapter" data-level="7.5" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#cell-type-label-transfer"><i class="fa fa-check"></i><b>7.5</b> Cell type label transfer</a></li>
<li class="chapter" data-level="7.6" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#adjust-major_cluster_rate"><i class="fa fa-check"></i><b>7.6</b> Adjust major_cluster_rate</a></li>
<li class="chapter" data-level="7.7" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#run-the-pipeline-in-parallel"><i class="fa fa-check"></i><b>7.7</b> Run the pipeline in parallel</a></li>
<li class="chapter" data-level="7.8" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#run-the-full-analysis"><i class="fa fa-check"></i><b>7.8</b> Run the full analysis</a></li>
<li class="chapter" data-level="7.9" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#visualize-the-harmonized-results"><i class="fa fa-check"></i><b>7.9</b> Visualize the harmonized results</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html"><i class="fa fa-check"></i><b>8</b> Differential expression (DE) analysis</a>
<ul>
<li class="chapter" data-level="8.1" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html#example-1"><i class="fa fa-check"></i><b>8.1</b> Example 1</a></li>
<li class="chapter" data-level="8.2" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html#example-2"><i class="fa fa-check"></i><b>8.2</b> Example 2</a></li>
<li class="chapter" data-level="8.3" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html#example-3"><i class="fa fa-check"></i><b>8.3</b> Example 3</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="reference-building.html"><a href="reference-building.html"><i class="fa fa-check"></i><b>9</b> Reference building</a>
<ul>
<li class="chapter" data-level="9.1" data-path="reference-building.html"><a href="reference-building.html#initialize-scref"><i class="fa fa-check"></i><b>9.1</b> Initialize scRef</a></li>
<li class="chapter" data-level="9.2" data-path="reference-building.html"><a href="reference-building.html#submit-scref"><i class="fa fa-check"></i><b>9.2</b> Submit scRef</a></li>
<li class="chapter" data-level="9.3" data-path="reference-building.html"><a href="reference-building.html#demo-scref"><i class="fa fa-check"></i><b>9.3</b> Demo scRef</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="celldepot-publishing.html"><a href="celldepot-publishing.html"><i class="fa fa-check"></i><b>10</b> CellDepot publishing</a>
<ul>
<li class="chapter" data-level="10.1" data-path="celldepot-publishing.html"><a href="celldepot-publishing.html#initialize-sc2celldepot"><i class="fa fa-check"></i><b>10.1</b> Initialize sc2celldepot</a></li>
<li class="chapter" data-level="10.2" data-path="celldepot-publishing.html"><a href="celldepot-publishing.html#run-sc2celldepot"><i class="fa fa-check"></i><b>10.2</b> Run sc2celldepot</a></li>
<li class="chapter" data-level="10.3" data-path="celldepot-publishing.html"><a href="celldepot-publishing.html#demo-sc2celldepot"><i class="fa fa-check"></i><b>10.3</b> Demo sc2celldepot</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="additional-tools.html"><a href="additional-tools.html"><i class="fa fa-check"></i><b>11</b> Additional tools</a>
<ul>
<li class="chapter" data-level="11.1" data-path="additional-tools.html"><a href="additional-tools.html#scrmambient"><i class="fa fa-check"></i><b>11.1</b> scRMambient</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="additional-tools.html"><a href="additional-tools.html#set-up-the-inputs"><i class="fa fa-check"></i><b>11.1.1</b> Set up the inputs</a></li>
<li class="chapter" data-level="11.1.2" data-path="additional-tools.html"><a href="additional-tools.html#run-scrmambient"><i class="fa fa-check"></i><b>11.1.2</b> Run scRMambient</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="additional-tools.html"><a href="additional-tools.html#scdeg"><i class="fa fa-check"></i><b>11.2</b> scDEG</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="additional-tools.html"><a href="additional-tools.html#initialize-scdeg"><i class="fa fa-check"></i><b>11.2.1</b> Initialize scDEG</a></li>
<li class="chapter" data-level="11.2.2" data-path="additional-tools.html"><a href="additional-tools.html#run-de-analysis"><i class="fa fa-check"></i><b>11.2.2</b> Run DE analysis</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="additional-tools.html"><a href="additional-tools.html#sctool"><i class="fa fa-check"></i><b>11.3</b> scTool</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="additional-tools.html"><a href="additional-tools.html#remove-annotations"><i class="fa fa-check"></i><b>11.3.1</b> Remove annotations</a></li>
<li class="chapter" data-level="11.3.2" data-path="additional-tools.html"><a href="additional-tools.html#add-annotations"><i class="fa fa-check"></i><b>11.3.2</b> Add annotations</a></li>
<li class="chapter" data-level="11.3.3" data-path="additional-tools.html"><a href="additional-tools.html#export-gene-expressions"><i class="fa fa-check"></i><b>11.3.3</b> Export gene expressions</a></li>
<li class="chapter" data-level="11.3.4" data-path="additional-tools.html"><a href="additional-tools.html#demo-sctool"><i class="fa fa-check"></i><b>11.3.4</b> Demo scTool</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="miscellaneous.html"><a href="miscellaneous.html"><i class="fa fa-check"></i><b>12</b> Miscellaneous</a>
<ul>
<li class="chapter" data-level="12.1" data-path="miscellaneous.html"><a href="miscellaneous.html#cell-subsetting"><i class="fa fa-check"></i><b>12.1</b> Cell subsetting</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="celldepot-publishing" class="section level1 hasAnchor" number="10">
<h1><span class="header-section-number">Chapter 10</span> CellDepot publishing<a href="celldepot-publishing.html#celldepot-publishing" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p><a href="https://www.sciencedirect.com/science/article/pii/S0022283621006665" target="_blank">CellDepot</a> is a comprehensive data management platform for single-cell RNA-seq datasets. Publishing to CellDepot allows easy navigation and visualization of the data, and facilitates big data deposition.</p>
<p>If you have run through the scRNASequest pipeline using <code>scAnalyzer</code>, you will be able to use the <strong>h5ad</strong> file directly and add it to CellDepot, following the ‘<a href="https://interactivereport.github.io/CellDepot/bookdown/docs/SITutorial.html#import">4.6 Import projects</a>’ and ‘<a href="https://interactivereport.github.io/CellDepot/bookdown/docs/SITutorial.html#create-project">4.7 Create project</a>’ in the full instruction <a href="https://interactivereport.github.io/CellDepot/bookdown/docs/">here</a>. In this case, you won’t need to run <strong>sc2celldepot</strong>. However, this may be cumbersome because you have to run the full <code>scAnalyzer</code> pipeline to get the h5ad files. Sometimes, you have a public data (such as an RDS file, or raw MEX/h5 UMI count files with UMAP coordinates) with cell annotation information, and you only need to use them for visualization, rather than running though the whole analysis.</p>
<p>Thus, the scRNASequest pipeline offers a function to publish a dataset to CellDepot using <strong>sc2celldepot</strong>. This workflow will use the existing cell type information (cell type, cluter id, etc.) and create an <strong>h5ad</strong> file. The h5ad file can be further used for uploading to the CellDepot platform. In sum, there are two cases to use this pipeline: 1) If your input is an RDS file, you can also view this pipeline as a converter from RDS to h5ad. 2) If you have raw UMI counts, together with annotated cell barcode information (e.g. cell type for each barcode), and some UMAP embedding coordinates (the embedding information is optional), you can use this pipeline to assemble them together and create an h5ad output.</p>
<p>By running the following command, we can see the manual page of the script:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb79-1"><a href="celldepot-publishing.html#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> sc2celldepot</span>
<span id="cb79-2"><a href="celldepot-publishing.html#cb79-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-3"><a href="celldepot-publishing.html#cb79-3" aria-hidden="true" tabindex="-1"></a><span class="ex">*****</span> 2023-01-26 15:25:32 <span class="pp">*****</span></span>
<span id="cb79-4"><a href="celldepot-publishing.html#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb79-5"><a href="celldepot-publishing.html#cb79-5" aria-hidden="true" tabindex="-1"></a><span class="co">## ExpressionAnalysis: https://github.com/interactivereport/scRNAsequest.git</span></span>
<span id="cb79-6"><a href="celldepot-publishing.html#cb79-6" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Path: /mnt/depts/dept04/compbio/edge_tools/scRNAsequest</span></span>
<span id="cb79-7"><a href="celldepot-publishing.html#cb79-7" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Date: 2023-01-13 17:41:49 -0500</span></span>
<span id="cb79-8"><a href="celldepot-publishing.html#cb79-8" aria-hidden="true" tabindex="-1"></a><span class="co">## git HEAD: 3d463e0b127af499942b7adc2fc5af6ddfc6f11e</span></span>
<span id="cb79-9"><a href="celldepot-publishing.html#cb79-9" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb79-10"><a href="celldepot-publishing.html#cb79-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-11"><a href="celldepot-publishing.html#cb79-11" aria-hidden="true" tabindex="-1"></a><span class="ex">Loading</span> resources</span>
<span id="cb79-12"><a href="celldepot-publishing.html#cb79-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-13"><a href="celldepot-publishing.html#cb79-13" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> /path/to/a/output/folder === or === sc2celldepot /path/to/a/config/file</span>
<span id="cb79-14"><a href="celldepot-publishing.html#cb79-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-15"><a href="celldepot-publishing.html#cb79-15" aria-hidden="true" tabindex="-1"></a><span class="ex">Please</span> create the folder before running sc2celldepot.</span>
<span id="cb79-16"><a href="celldepot-publishing.html#cb79-16" aria-hidden="true" tabindex="-1"></a><span class="ex">The</span> data config file will be generated automatically when a path is provided</span>
<span id="cb79-17"><a href="celldepot-publishing.html#cb79-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-18"><a href="celldepot-publishing.html#cb79-18" aria-hidden="true" tabindex="-1"></a><span class="ex">Powered</span> by the Research Data Sciences Group [zhengyu.ouyang@biogen.com<span class="kw">;</span><span class="ex">yuhenry.sun@biogen.com]</span></span>
<span id="cb79-19"><a href="celldepot-publishing.html#cb79-19" aria-hidden="true" tabindex="-1"></a><span class="ex">------------</span></span></code></pre></div>
<div id="initialize-sc2celldepot" class="section level2 hasAnchor" number="10.1">
<h2><span class="header-section-number">10.1</span> Initialize sc2celldepot<a href="celldepot-publishing.html#initialize-sc2celldepot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Running <strong>sc2celldepot</strong> by providing a working directory will initiate the project and generate a template of config file. The config file template can be found: <a href="https://github.com/interactivereport/scRNAsequest/blob/main/src/external2h5ad.yml" target="_blank">template</a>.</p>
<p>First, we create a new directory under the project. Then we run the <strong>sc2celldepot</strong> command:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb80-1"><a href="celldepot-publishing.html#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> ~/E-MTAB-11115/CellDepot_publish</span>
<span id="cb80-2"><a href="celldepot-publishing.html#cb80-2" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> ~/E-MTAB-11115/CellDepot_publish</span>
<span id="cb80-3"><a href="celldepot-publishing.html#cb80-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-4"><a href="celldepot-publishing.html#cb80-4" aria-hidden="true" tabindex="-1"></a><span class="ex">E-MTAB-11115/</span></span>
<span id="cb80-5"><a href="celldepot-publishing.html#cb80-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> E-MTAB-11115.rds</span>
<span id="cb80-6"><a href="celldepot-publishing.html#cb80-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> CellDepot_publish</span>
<span id="cb80-7"><a href="celldepot-publishing.html#cb80-7" aria-hidden="true" tabindex="-1"></a>        <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<p>Here is an example after filling in the configuration <strong>sc2celldepot.yml</strong> file.</p>
<p>For a simple example, we just provide a prefix and an RDS file path to this yml file. The pipeline will read the metadata and convert it to an h5ad file.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb81-1"><a href="celldepot-publishing.html#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The config file to process public sc/sn RNAseq and generate h5ad for celldepot</span></span>
<span id="cb81-2"><a href="celldepot-publishing.html#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> ~/E-MTAB-11115/CellDepot_publish</span>
<span id="cb81-3"><a href="celldepot-publishing.html#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="ex">prefix:</span> E-MTAB-11115                                        <span class="co"># the prefix of the file name of the h5ad</span></span>
<span id="cb81-4"><a href="celldepot-publishing.html#cb81-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-5"><a href="celldepot-publishing.html#cb81-5" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat RDS is avaiable, otherwise please move to next section</span></span>
<span id="cb81-6"><a href="celldepot-publishing.html#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratObj:</span> ~/E-MTAB-11115/E-MTAB-11115.rds                  <span class="co"># the full path to the seurat RDS file with SCT &amp; RNA assay along with meta.data and reduction</span></span>
<span id="cb81-7"><a href="celldepot-publishing.html#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratUMI:</span> RNA                                              <span class="co"># the name of the assay stores raw UMI</span></span>
<span id="cb81-8"><a href="celldepot-publishing.html#cb81-8" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratSCT:</span> SCT                                              <span class="co"># the name of the assay stores SCT</span></span>
<span id="cb81-9"><a href="celldepot-publishing.html#cb81-9" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratMeta:</span> []                                              <span class="co"># the list of cell annotations to be stored in h5ad, empty list means all meta.data entry from seurat rds</span></span>
<span id="cb81-10"><a href="celldepot-publishing.html#cb81-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-11"><a href="celldepot-publishing.html#cb81-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Expression when the seurat RDS is not available (row gene/column cell)</span></span>
<span id="cb81-12"><a href="celldepot-publishing.html#cb81-12" aria-hidden="true" tabindex="-1"></a><span class="co"># if the annotation files are seperated the same as expression files, they should be the same order, other wise cell ID will be used to match</span></span>
<span id="cb81-13"><a href="celldepot-publishing.html#cb81-13" aria-hidden="true" tabindex="-1"></a><span class="ex">expression:</span> []                                              <span class="co"># full path the gene expression file/folder (h5/csv/txt/mtx), if multiple files, please provide the list separated by &#39;,&#39;</span></span>
<span id="cb81-14"><a href="celldepot-publishing.html#cb81-14" aria-hidden="true" tabindex="-1"></a><span class="ex">dataUMI:</span> True                                               <span class="co"># if the above expression is UMI, if the value in expression file should be used directly, please set &quot;False&quot;</span></span>
<span id="cb81-15"><a href="celldepot-publishing.html#cb81-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-16"><a href="celldepot-publishing.html#cb81-16" aria-hidden="true" tabindex="-1"></a><span class="co"># cell annotation (cell intersection will be used, first column is the cell ID)</span></span>
<span id="cb81-17"><a href="celldepot-publishing.html#cb81-17" aria-hidden="true" tabindex="-1"></a><span class="ex">annotation:</span> []                                              <span class="co"># full path to the cell annotation file, first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb81-18"><a href="celldepot-publishing.html#cb81-18" aria-hidden="true" tabindex="-1"></a><span class="ex">annotationUse:</span> []                                           <span class="co"># the column names in the annotation file to be extracted for h5ad, empty list means all columns</span></span>
<span id="cb81-19"><a href="celldepot-publishing.html#cb81-19" aria-hidden="true" tabindex="-1"></a><span class="ex">sample_column:</span>                                              <span class="co"># one column header from annotation file, if the one expression file needs to be splited into each sample</span></span>
<span id="cb81-20"><a href="celldepot-publishing.html#cb81-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-21"><a href="celldepot-publishing.html#cb81-21" aria-hidden="true" tabindex="-1"></a><span class="co"># cell layout: tSNE, UMAP, PCA, if separated the same as expression files, should be the same order</span></span>
<span id="cb81-22"><a href="celldepot-publishing.html#cb81-22" aria-hidden="true" tabindex="-1"></a><span class="ex">reduction:</span>                                                  <span class="co"># optional (if missing UMAP will be created), other keys can be removed or added new ones, keys will be used in h5ad</span></span>
<span id="cb81-23"><a href="celldepot-publishing.html#cb81-23" aria-hidden="true" tabindex="-1"></a>  <span class="ex">files:</span> []                                                 <span class="co"># full path to the cell layout file (contains all layouts of a set of cells), first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb81-24"><a href="celldepot-publishing.html#cb81-24" aria-hidden="true" tabindex="-1"></a>  <span class="ex">umap:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb81-25"><a href="celldepot-publishing.html#cb81-25" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tsne:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb81-26"><a href="celldepot-publishing.html#cb81-26" aria-hidden="true" tabindex="-1"></a>  <span class="ex">pca:</span> []                                                   <span class="co"># column headers from layout file to be used, please use quote for each column header (can be more than 2 dimentions though only first two will be shown in VIP)</span></span></code></pre></div>
<p>Alternatively, if you don’t have an RDS file, users can provide raw UMI files together with cell type annotation files and reduction embeddings (optional, by default it will generate UMAP) in this way:</p>
<p>Besides the required <code>output</code> and <code>prefix</code> parameters, in this example, the <code>expression</code> points to several h5 files or MEX folders, which is required. The <code>annotation</code> points to cell type annotation (and other cell level information) csv files, which are also required.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb82-1"><a href="celldepot-publishing.html#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The config file to process public sc/sn RNAseq and generate h5ad for celldepot</span></span>
<span id="cb82-2"><a href="celldepot-publishing.html#cb82-2" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> ~/E-MTAB-11115/CellDepot_publish</span>
<span id="cb82-3"><a href="celldepot-publishing.html#cb82-3" aria-hidden="true" tabindex="-1"></a><span class="ex">prefix:</span> E-MTAB-11115                                        <span class="co"># the prefix of the file name of the h5ad</span></span>
<span id="cb82-4"><a href="celldepot-publishing.html#cb82-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-5"><a href="celldepot-publishing.html#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat RDS is avaiable, otherwise please move to next section</span></span>
<span id="cb82-6"><a href="celldepot-publishing.html#cb82-6" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratObj:</span>                                                  <span class="co"># the full path to the seurat RDS file with SCT &amp; RNA assay along with meta.data and reduction</span></span>
<span id="cb82-7"><a href="celldepot-publishing.html#cb82-7" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratUMI:</span> RNA                                              <span class="co"># the name of the assay stores raw UMI</span></span>
<span id="cb82-8"><a href="celldepot-publishing.html#cb82-8" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratSCT:</span> SCT                                              <span class="co"># the name of the assay stores SCT</span></span>
<span id="cb82-9"><a href="celldepot-publishing.html#cb82-9" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratMeta:</span> []                                              <span class="co"># the list of cell annotations to be stored in h5ad, empty list means all meta.data entry from seurat rds</span></span>
<span id="cb82-10"><a href="celldepot-publishing.html#cb82-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-11"><a href="celldepot-publishing.html#cb82-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Expression when the seurat RDS is not available (row gene/column cell)</span></span>
<span id="cb82-12"><a href="celldepot-publishing.html#cb82-12" aria-hidden="true" tabindex="-1"></a><span class="co"># if the annotation files are seperated the same as expression files, they should be the same order, other wise cell ID will be used to match</span></span>
<span id="cb82-13"><a href="celldepot-publishing.html#cb82-13" aria-hidden="true" tabindex="-1"></a><span class="ex">expression:</span> [~/E-MTAB-11115/data/5705STDY8058280_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058281_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058282_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058283_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058284_filtered_feature_bc_matrix.h5,~/E-MTAB-11115/data/5705STDY8058285_filtered_feature_bc_matrix.h5] <span class="co"># full path the gene expression file/folder (h5/csv/txt/mtx), if multiple files, please provide the list separated by &#39;,&#39;</span></span>
<span id="cb82-14"><a href="celldepot-publishing.html#cb82-14" aria-hidden="true" tabindex="-1"></a><span class="ex">dataUMI:</span> True                                               <span class="co"># if the above expression is UMI, if the value in expression file should be used directly, please set &quot;False&quot;</span></span>
<span id="cb82-15"><a href="celldepot-publishing.html#cb82-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-16"><a href="celldepot-publishing.html#cb82-16" aria-hidden="true" tabindex="-1"></a><span class="co"># cell annotation (cell intersection will be used, first column is the cell ID)</span></span>
<span id="cb82-17"><a href="celldepot-publishing.html#cb82-17" aria-hidden="true" tabindex="-1"></a><span class="ex">annotation:</span> [~/E-MTAB-11115/data/5705STDY8058280_annotation.csv,~/E-MTAB-11115/data/5705STDY8058281_annotation.csv,~/E-MTAB-11115/data/5705STDY8058282_annotation.csv,~/E-MTAB-11115/data/5705STDY8058283_annotation.csv,~/E-MTAB-11115/data/5705STDY8058284_annotation.csv,~/E-MTAB-11115/data/5705STDY8058285_annotation.csv] <span class="co"># full path to the cell annotation file, first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb82-18"><a href="celldepot-publishing.html#cb82-18" aria-hidden="true" tabindex="-1"></a><span class="ex">annotationUse:</span> []                                           <span class="co"># the column names in the annotation file to be extracted for h5ad, empty list means all columns</span></span>
<span id="cb82-19"><a href="celldepot-publishing.html#cb82-19" aria-hidden="true" tabindex="-1"></a><span class="ex">sample_column:</span>                                              <span class="co"># one column header from annotation file, if the one expression file needs to be splited into each sample</span></span>
<span id="cb82-20"><a href="celldepot-publishing.html#cb82-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-21"><a href="celldepot-publishing.html#cb82-21" aria-hidden="true" tabindex="-1"></a><span class="co"># cell layout: tSNE, UMAP, PCA, if separated the same as expression files, should be the same order</span></span>
<span id="cb82-22"><a href="celldepot-publishing.html#cb82-22" aria-hidden="true" tabindex="-1"></a><span class="ex">reduction:</span> <span class="co">#optional (if missing UMAP will be created), other keys can be removed or added new ones, keys will be used in h5ad</span></span>
<span id="cb82-23"><a href="celldepot-publishing.html#cb82-23" aria-hidden="true" tabindex="-1"></a>  <span class="ex">files:</span> []                                                 <span class="co"># full path to the cell layout file (contains all layouts of a set of cells), first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb82-24"><a href="celldepot-publishing.html#cb82-24" aria-hidden="true" tabindex="-1"></a>  <span class="ex">umap:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb82-25"><a href="celldepot-publishing.html#cb82-25" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tsne:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb82-26"><a href="celldepot-publishing.html#cb82-26" aria-hidden="true" tabindex="-1"></a>  <span class="ex">pca:</span> []                                                   <span class="co"># column headers from layout file to be used, please use quote for each column header (can be more than 2 dimentions though only first two will be shown in VIP)```</span></span></code></pre></div>
</div>
<div id="run-sc2celldepot" class="section level2 hasAnchor" number="10.2">
<h2><span class="header-section-number">10.2</span> Run sc2celldepot<a href="celldepot-publishing.html#run-sc2celldepot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>After preparing the <strong>sc2celldepot.yml</strong> file, we are ready to run the pipeline by the following command:</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb83-1"><a href="celldepot-publishing.html#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> ~/E-MTAB-11115/CellDepot_publish/sc2celldepot.yml</span></code></pre></div>
<p>This workflow will generate the output files in the working directory, <strong>CellDepot_publish</strong>:</p>
<p>The output prefix is determined by the <strong>prefix</strong> parameter in the sc2celldepot.yml file.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb84-1"><a href="celldepot-publishing.html#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="ex">E-MTAB-11115/</span></span>
<span id="cb84-2"><a href="celldepot-publishing.html#cb84-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> E-MTAB-11115.rds</span>
<span id="cb84-3"><a href="celldepot-publishing.html#cb84-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> CellDepot_publish</span>
<span id="cb84-4"><a href="celldepot-publishing.html#cb84-4" aria-hidden="true" tabindex="-1"></a>        <span class="ex">├──</span> E-MTAB-11115.h5ad</span>
<span id="cb84-5"><a href="celldepot-publishing.html#cb84-5" aria-hidden="true" tabindex="-1"></a>        <span class="ex">├──</span> E-MTAB-11115.raw_added.h5ad</span>
<span id="cb84-6"><a href="celldepot-publishing.html#cb84-6" aria-hidden="true" tabindex="-1"></a>        <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<p>Finally, copy the <strong>E-MTAB-11115.h5ad</strong> to the <a href="https://github.com/interactivereport/CellDepot" target="_blank">CellDepot</a> folder defined by sys.yml (celldepotDir) and follow the <a href="https://interactivereport.github.io/CellDepot/bookdown/docs/" target="_blank">instructions</a> for publishing it.</p>
</div>
<div id="demo-sc2celldepot" class="section level2 hasAnchor" number="10.3">
<h2><span class="header-section-number">10.3</span> Demo sc2celldepot<a href="celldepot-publishing.html#demo-sc2celldepot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To demo <strong>sc2celldepot</strong>, we use the following RDS file (<a href="https://github.com/interactivereport/scRNAsequest/releases/download/untagged-e92025b745dbc6f9666a/GSE172462.SCT.For_scRef.rds">GSE172462.SCT.For_scRef.rds</a>), which we had used it to demo scRef. This RDS file contains cell type labeling and embedding information. Please download it to the working directory, for example, ~/demo_celldepot.</p>
<p>Then we run the <strong>sc2celldepot</strong> program by providing a working directory, in our case:</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb85-1"><a href="celldepot-publishing.html#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> ~/demo_celldepot</span></code></pre></div>
<p>File hierarchy after running the above command:</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb86-1"><a href="celldepot-publishing.html#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/demo_celldepot</span></span>
<span id="cb86-2"><a href="celldepot-publishing.html#cb86-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> GSE172462.SCT.For_scRef.rds</span>
<span id="cb86-3"><a href="celldepot-publishing.html#cb86-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<p>We set up the <strong>sc2celldepot.yml</strong> in the following way, by just adding information to <code>prefix</code> and <code>seuratObj</code>:</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb87-1"><a href="celldepot-publishing.html#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The config file to process public sc/sn RNAseq and generate h5ad for celldepot</span></span>
<span id="cb87-2"><a href="celldepot-publishing.html#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> ~/demo_celldepot</span>
<span id="cb87-3"><a href="celldepot-publishing.html#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="ex">prefix:</span> demo_sc2celldepot                                   <span class="co"># the prefix of the file name of the h5ad</span></span>
<span id="cb87-4"><a href="celldepot-publishing.html#cb87-4" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat RDS is avaiable, otherwise please move to next section</span></span>
<span id="cb87-5"><a href="celldepot-publishing.html#cb87-5" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratObj:</span> ~/demo_celldepot/GSE172462.SCT.For_scRef.rds     <span class="co"># the full path to the seurat RDS file with SCT &amp; RNA assay along with meta.data and reduction</span></span>
<span id="cb87-6"><a href="celldepot-publishing.html#cb87-6" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratUMI:</span> RNA                                              <span class="co"># the name of the assay stores raw UMI</span></span>
<span id="cb87-7"><a href="celldepot-publishing.html#cb87-7" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratSCT:</span> SCT                                              <span class="co"># the name of the assay stores SCT</span></span>
<span id="cb87-8"><a href="celldepot-publishing.html#cb87-8" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratMeta:</span> []                                              <span class="co"># the list of cell annotations to be stored in h5ad, empty list means all meta.data entry from seurat rds</span></span>
<span id="cb87-9"><a href="celldepot-publishing.html#cb87-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Expression when the seurat RDS is not available (row gene/column cell)</span></span>
<span id="cb87-10"><a href="celldepot-publishing.html#cb87-10" aria-hidden="true" tabindex="-1"></a><span class="co"># if the annotation files are seperated the same as expression files, they should be the same order, other wise cell ID will be used to match</span></span>
<span id="cb87-11"><a href="celldepot-publishing.html#cb87-11" aria-hidden="true" tabindex="-1"></a><span class="ex">expression:</span> []                                              <span class="co"># full path the gene expression file/folder (h5/csv/txt/mtx), if multiple files, please provide the list separated by &#39;,&#39;</span></span>
<span id="cb87-12"><a href="celldepot-publishing.html#cb87-12" aria-hidden="true" tabindex="-1"></a><span class="ex">dataUMI:</span> True                                               <span class="co">#if the above expression is UMI, if the value in expression file should be used directly, please set &quot;False&quot;</span></span>
<span id="cb87-13"><a href="celldepot-publishing.html#cb87-13" aria-hidden="true" tabindex="-1"></a><span class="co"># cell annotation (cell intersection will be used, first column is the cell ID)</span></span>
<span id="cb87-14"><a href="celldepot-publishing.html#cb87-14" aria-hidden="true" tabindex="-1"></a><span class="ex">annotation:</span> []                                              <span class="co"># full path to the cell annotation file, first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb87-15"><a href="celldepot-publishing.html#cb87-15" aria-hidden="true" tabindex="-1"></a><span class="ex">annotationUse:</span> []                                           <span class="co"># the column names in the annotation file to be extracted for h5ad, empty list means all columns</span></span>
<span id="cb87-16"><a href="celldepot-publishing.html#cb87-16" aria-hidden="true" tabindex="-1"></a><span class="ex">sample_column:</span>                                              <span class="co"># one column header from annotation file, if the one expression file needs to be splited into each sample</span></span>
<span id="cb87-17"><a href="celldepot-publishing.html#cb87-17" aria-hidden="true" tabindex="-1"></a><span class="co"># cell layout: tSNE, UMAP, PCA, if separated the same as expression files, should be the same order</span></span>
<span id="cb87-18"><a href="celldepot-publishing.html#cb87-18" aria-hidden="true" tabindex="-1"></a><span class="ex">reduction:</span> <span class="co">#optional (if missing UMAP will be created), other keys can be removed or added new ones, keys will be used in h5ad</span></span>
<span id="cb87-19"><a href="celldepot-publishing.html#cb87-19" aria-hidden="true" tabindex="-1"></a>  <span class="ex">files:</span> []                                                 <span class="co"># full path to the cell layout file (contains all layouts of a set of cells), first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb87-20"><a href="celldepot-publishing.html#cb87-20" aria-hidden="true" tabindex="-1"></a>  <span class="ex">umap:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb87-21"><a href="celldepot-publishing.html#cb87-21" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tsne:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb87-22"><a href="celldepot-publishing.html#cb87-22" aria-hidden="true" tabindex="-1"></a>  <span class="ex">pca:</span> []                                                   <span class="co"># column headers from layout file to be used, please use quote for each column header (can be more than 2 dimentions though only first two will be shown in VIP)</span></span></code></pre></div>
<p>Then run the pipeline:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb88-1"><a href="celldepot-publishing.html#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> ~/demo_celldepot/sc2celldepot.yml</span></code></pre></div>
<p>The results will be new h5ad files as below:</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb89-1"><a href="celldepot-publishing.html#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/demo_celldepot</span></span>
<span id="cb89-2"><a href="celldepot-publishing.html#cb89-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> demo_sc2celldepot_rds.h5ad</span>
<span id="cb89-3"><a href="celldepot-publishing.html#cb89-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> demo_sc2celldepot_rds.raw_added.h5ad</span>
<span id="cb89-4"><a href="celldepot-publishing.html#cb89-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> GSE172462.SCT.For_scRef.rds</span>
<span id="cb89-5"><a href="celldepot-publishing.html#cb89-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> sc2celldepot.20230321.log</span>
<span id="cb89-6"><a href="celldepot-publishing.html#cb89-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<p>We provide another demo using two h5 data files from our previous <a href="https://github.com/interactivereport/scRNAsequest/tree/main/demo">demo</a> dataset. Besides that, we also need two cell level annotation files, and I have prepared them: <a href="https://github.com/interactivereport/scRNAsequest/releases/download/untagged-e92025b745dbc6f9666a/RatMaleCigarette.annotation.csv">RatMaleCigarette.annotation.csv</a>, <a href="https://github.com/interactivereport/scRNAsequest/releases/download/untagged-e92025b745dbc6f9666a/RatFemaleCigarette.annotation.csv">RatFemaleCigarette.annotation.csv</a>.</p>
<p>Take a quick look of the csv files (Usually, the cell barcodes in the list is only a subset of all barcodes in the h5, due to filtering in previous analysis):</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb90-1"><a href="celldepot-publishing.html#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-3</span> RatMaleCigarette.annotation.csv</span>
<span id="cb90-2"><a href="celldepot-publishing.html#cb90-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-3"><a href="celldepot-publishing.html#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="ex">,library_id,predicted.celltype</span></span>
<span id="cb90-4"><a href="celldepot-publishing.html#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="ex">AAACCCACAACTGAAA-1,RatFemaleCigarette,OPC</span></span>
<span id="cb90-5"><a href="celldepot-publishing.html#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="ex">AAACCCACACTTTATC-1,RatFemaleCigarette,Neuron</span></span></code></pre></div>
<p>Please organize these four files (2 h5 files, 2 csv files) in the working dir (in this demo: ~/demo_celldepot_2), then run the script:</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb91-1"><a href="celldepot-publishing.html#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sc2celldepot</span> ~/demo_celldepot_2</span></code></pre></div>
<p>The above run generated sc2celldepot.yml, and the current file hierarchy is:</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb92-1"><a href="celldepot-publishing.html#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/demo_celldepot_2</span></span>
<span id="cb92-2"><a href="celldepot-publishing.html#cb92-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatFemaleCigarette.annotation.csv</span>
<span id="cb92-3"><a href="celldepot-publishing.html#cb92-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatFemaleCigarette.filtered_feature_bc_matrix.h5</span>
<span id="cb92-4"><a href="celldepot-publishing.html#cb92-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatMaleCigarette.annotation.csv</span>
<span id="cb92-5"><a href="celldepot-publishing.html#cb92-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatMaleCigarette.filtered_feature_bc_matrix.h5</span>
<span id="cb92-6"><a href="celldepot-publishing.html#cb92-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>
<p>Then we set up the sc2celldepot.yml file in this way:</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb93-1"><a href="celldepot-publishing.html#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The config file to process public sc/sn RNAseq and generate h5ad for celldepot</span></span>
<span id="cb93-2"><a href="celldepot-publishing.html#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> ~/demo_celldepot_2</span>
<span id="cb93-3"><a href="celldepot-publishing.html#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="ex">prefix:</span> demo_sc2celldepot_2                                 <span class="co"># the prefix of the file name of the h5ad</span></span>
<span id="cb93-4"><a href="celldepot-publishing.html#cb93-4" aria-hidden="true" tabindex="-1"></a><span class="co"># seurat RDS is avaiable, otherwise please move to next section</span></span>
<span id="cb93-5"><a href="celldepot-publishing.html#cb93-5" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratObj:</span>                                                  <span class="co"># the full path to the seurat RDS file with SCT &amp; RNA assay along with meta.data and reduction</span></span>
<span id="cb93-6"><a href="celldepot-publishing.html#cb93-6" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratUMI:</span> RNA                                              <span class="co"># the name of the assay stores raw UMI</span></span>
<span id="cb93-7"><a href="celldepot-publishing.html#cb93-7" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratSCT:</span> SCT                                              <span class="co"># the name of the assay stores SCT</span></span>
<span id="cb93-8"><a href="celldepot-publishing.html#cb93-8" aria-hidden="true" tabindex="-1"></a><span class="ex">seuratMeta:</span> []                                              <span class="co"># the list of cell annotations to be stored in h5ad, empty list means all meta.data entry from seurat rds</span></span>
<span id="cb93-9"><a href="celldepot-publishing.html#cb93-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Expression when the seurat RDS is not available (row gene/column cell)</span></span>
<span id="cb93-10"><a href="celldepot-publishing.html#cb93-10" aria-hidden="true" tabindex="-1"></a><span class="co"># if the annotation files are seperated the same as expression files, they should be the same order, other wise cell ID will be used to match</span></span>
<span id="cb93-11"><a href="celldepot-publishing.html#cb93-11" aria-hidden="true" tabindex="-1"></a><span class="ex">expression:</span> [~/demo_celldepot_2/RatMaleCigarette.filtered_feature_bc_matrix.h5,~/demo_celldepot_2/RatFemaleCigarette.filtered_feature_bc_matrix.h5]</span>
<span id="cb93-12"><a href="celldepot-publishing.html#cb93-12" aria-hidden="true" tabindex="-1"></a>                                                            <span class="co"># full path the gene expression file/folder (h5/csv/txt/mtx), if multiple files, please provide the list separated by &#39;,&#39;</span></span>
<span id="cb93-13"><a href="celldepot-publishing.html#cb93-13" aria-hidden="true" tabindex="-1"></a><span class="ex">dataUMI:</span> True                                               <span class="co">#if the above expression is UMI, if the value in expression file should be used directly, please set &quot;False&quot;</span></span>
<span id="cb93-14"><a href="celldepot-publishing.html#cb93-14" aria-hidden="true" tabindex="-1"></a><span class="co"># cell annotation (cell intersection will be used, first column is the cell ID)</span></span>
<span id="cb93-15"><a href="celldepot-publishing.html#cb93-15" aria-hidden="true" tabindex="-1"></a><span class="ex">annotation:</span> [~/demo_celldepot_2/RatMaleCigarette.annotation.csv,~/demo_celldepot_2/RatFemaleCigarette.annotation.csv]</span>
<span id="cb93-16"><a href="celldepot-publishing.html#cb93-16" aria-hidden="true" tabindex="-1"></a>                                                            <span class="co"># full path to the cell annotation file, first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb93-17"><a href="celldepot-publishing.html#cb93-17" aria-hidden="true" tabindex="-1"></a><span class="ex">annotationUse:</span> []                                           <span class="co"># the column names in the annotation file to be extracted for h5ad, empty list means all columns</span></span>
<span id="cb93-18"><a href="celldepot-publishing.html#cb93-18" aria-hidden="true" tabindex="-1"></a><span class="ex">sample_column:</span>                                              <span class="co"># one column header from annotation file, if the one expression file needs to be splited into each sample</span></span>
<span id="cb93-19"><a href="celldepot-publishing.html#cb93-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-20"><a href="celldepot-publishing.html#cb93-20" aria-hidden="true" tabindex="-1"></a><span class="co"># cell layout: tSNE, UMAP, PCA, if separated the same as expression files, should be the same order</span></span>
<span id="cb93-21"><a href="celldepot-publishing.html#cb93-21" aria-hidden="true" tabindex="-1"></a><span class="ex">reduction:</span> <span class="co">#optional (if missing UMAP will be created), other keys can be removed or added new ones, keys will be used in h5ad</span></span>
<span id="cb93-22"><a href="celldepot-publishing.html#cb93-22" aria-hidden="true" tabindex="-1"></a>  <span class="ex">files:</span> []                                                 <span class="co"># full path to the cell layout file (contains all layouts of a set of cells), first column is the cell ID which should match cell ID in expression</span></span>
<span id="cb93-23"><a href="celldepot-publishing.html#cb93-23" aria-hidden="true" tabindex="-1"></a>  <span class="ex">umap:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb93-24"><a href="celldepot-publishing.html#cb93-24" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tsne:</span> []                                                  <span class="co"># column headers from layout file to be used, please use quote for each column header</span></span>
<span id="cb93-25"><a href="celldepot-publishing.html#cb93-25" aria-hidden="true" tabindex="-1"></a>  <span class="ex">pca:</span> []                                                   <span class="co"># column headers from layout file to be used, please use quote for each column header (can be more than 2 dimentions though only first two will be shown in VIP)</span></span></code></pre></div>
<p>This will let the pipeline to process the h5 files by running normalization, and match labeled cell types to the UMAP.</p>
<p>Files after running the pipeline:</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb94-1"><a href="celldepot-publishing.html#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/demo_celldepot_2</span></span>
<span id="cb94-2"><a href="celldepot-publishing.html#cb94-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> demo_sc2celldepot.h5ad</span>
<span id="cb94-3"><a href="celldepot-publishing.html#cb94-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> demo_sc2celldepot.raw_added.h5ad</span>
<span id="cb94-4"><a href="celldepot-publishing.html#cb94-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> demo_sc2celldepot.rds</span>
<span id="cb94-5"><a href="celldepot-publishing.html#cb94-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatFemaleCigarette.annotation.csv</span>
<span id="cb94-6"><a href="celldepot-publishing.html#cb94-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatFemaleCigarette.filtered_feature_bc_matrix.h5</span>
<span id="cb94-7"><a href="celldepot-publishing.html#cb94-7" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatMaleCigarette.annotation.csv</span>
<span id="cb94-8"><a href="celldepot-publishing.html#cb94-8" aria-hidden="true" tabindex="-1"></a>    <span class="ex">├──</span> RatMaleCigarette.filtered_feature_bc_matrix.h5</span>
<span id="cb94-9"><a href="celldepot-publishing.html#cb94-9" aria-hidden="true" tabindex="-1"></a>    <span class="ex">└──</span> sc2celldepot.yml</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="reference-building.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="additional-tools.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/scRNAsequest/edit/gh-pages/10-PublishCellDepot.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/scRNAsequest/blob/gh-pages/10-PublishCellDepot.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
