<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Data preparation | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</title>
  <meta name="description" content="Chapter 4 Data preparation | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Data preparation | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Data preparation | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  
  
  



<meta name="date" content="2022-06-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="installation.html"/>
<link rel="next" href="input-configuration.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">RNASequest</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>2</b> Introduction</a></li>
<li class="chapter" data-level="3" data-path="installation.html"><a href="installation.html"><i class="fa fa-check"></i><b>3</b> Installation</a>
<ul>
<li class="chapter" data-level="3.1" data-path="installation.html"><a href="installation.html#install-scrnasequest"><i class="fa fa-check"></i><b>3.1</b> Install scRNASequest</a></li>
<li class="chapter" data-level="3.2" data-path="installation.html"><a href="installation.html#install-cellxgene_vip"><i class="fa fa-check"></i><b>3.2</b> Install cellxgene_VIP</a></li>
<li class="chapter" data-level="3.3" data-path="installation.html"><a href="installation.html#install-celldepot"><i class="fa fa-check"></i><b>3.3</b> Install CellDepot</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-preparation.html"><a href="data-preparation.html"><i class="fa fa-check"></i><b>4</b> Data preparation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="data-preparation.html"><a href="data-preparation.html#public-data-in-h5-format"><i class="fa fa-check"></i><b>4.1</b> Public data in h5 format</a></li>
<li class="chapter" data-level="4.2" data-path="data-preparation.html"><a href="data-preparation.html#public-data-in-10x-mex-format"><i class="fa fa-check"></i><b>4.2</b> Public data in 10X MEX format</a></li>
<li class="chapter" data-level="4.3" data-path="data-preparation.html"><a href="data-preparation.html#self-prepared-files"><i class="fa fa-check"></i><b>4.3</b> Self-prepared files</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="input-configuration.html"><a href="input-configuration.html"><i class="fa fa-check"></i><b>5</b> Input configuration</a>
<ul>
<li class="chapter" data-level="5.1" data-path="input-configuration.html"><a href="input-configuration.html#prepare-the-config.yml-file"><i class="fa fa-check"></i><b>5.1</b> Prepare the config.yml file</a></li>
<li class="chapter" data-level="5.2" data-path="input-configuration.html"><a href="input-configuration.html#section52"><i class="fa fa-check"></i><b>5.2</b> Prepare the sample meta file</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="input-configuration.html"><a href="input-configuration.html#h5-input"><i class="fa fa-check"></i><b>5.2.1</b> h5 input</a></li>
<li class="chapter" data-level="5.2.2" data-path="input-configuration.html"><a href="input-configuration.html#mex-input"><i class="fa fa-check"></i><b>5.2.2</b> MEX input</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="input-configuration.html"><a href="input-configuration.html#file-hierarchy"><i class="fa fa-check"></i><b>5.3</b> File hierarchy</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html"><i class="fa fa-check"></i><b>6</b> Running the pipeline</a>
<ul>
<li class="chapter" data-level="6.1" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#initial-scanalyzer-run"><i class="fa fa-check"></i><b>6.1</b> Initial scAnalyzer run</a></li>
<li class="chapter" data-level="6.2" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#review-the-bookdown-report"><i class="fa fa-check"></i><b>6.2</b> Review the bookdown report</a></li>
<li class="chapter" data-level="6.3" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#run-the-full-analysis"><i class="fa fa-check"></i><b>6.3</b> Run the full analysis</a></li>
<li class="chapter" data-level="6.4" data-path="running-the-pipeline.html"><a href="running-the-pipeline.html#visualize-the-harmonized-results"><i class="fa fa-check"></i><b>6.4</b> Visualize the harmonized results</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html"><i class="fa fa-check"></i><b>7</b> Differential expression (DE) analysis</a>
<ul>
<li class="chapter" data-level="7.1" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html#example-1"><i class="fa fa-check"></i><b>7.1</b> Example 1</a></li>
<li class="chapter" data-level="7.2" data-path="differential-expression-de-analysis.html"><a href="differential-expression-de-analysis.html#example-2"><i class="fa fa-check"></i><b>7.2</b> Example 2</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="reference-building.html"><a href="reference-building.html"><i class="fa fa-check"></i><b>8</b> Reference Building</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-preparation" class="section level1" number="4">
<h1><span class="header-section-number">Chapter 4</span> Data preparation</h1>
<p>scRNASequest pipeline is compatible with different single-cell experiment outputs, including the <a href="https://kb.10xgenomics.com/hc/en-us/articles/115000794686-How-is-the-MEX-format-used-for-the-gene-barcode-matrices" target="_blank"><strong>10X MEX format</strong></a> and the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices" target="_blank"><strong>10X h5 format</strong></a> from Cell Ranger. The user may need to manually separate the annotation file, so that no cell filtering was performed. In this chapter, we will walk through how to prepare the data before running the pipeline.</p>
<div id="public-data-in-h5-format" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Public data in h5 format</h2>
<p>Here, we first present an example of the data processing steps using a public EMBL EBI dataset: <a href="https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-11115/" target="_blank">E-MTAB-11115</a>. The nine processed zip files were downloaded and unzipped.</p>
<p>There are total 6 data, with 6 corresponding *raw_feature_bc_matrix.h5 files. These files are required by the pipeline:</p>
<pre><code>#h5 matrix files
-rwxrwxr--. 1 zouyang ngs 165M Oct 25  2021 5705STDY8058280.raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 165M Oct 25  2021 5705STDY8058281.raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 156M Oct 25  2021 5705STDY8058282.raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 162M Oct 25  2021 5705STDY8058283.raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 149M Oct 25  2021 5705STDY8058284.raw_feature_bc_matrix.h5
-rwxrwxr--. 1 zouyang ngs 177M Oct 25  2021 5705STDY8058285.raw_feature_bc_matrix.h5</code></pre>
<p>This dataset also has cell type annotation files associated with each data. These files are optional to the pipeline, but if you would like to use their cell type labels, it would be better to include them in the sample meta file (See section <a href="input-configuration.html#section52">5.2</a>).</p>
<pre><code>#Annotation files (optional to the pipeline)
-rwxrwx---. 1 zouyang ngs 440K Apr 21 16:59 5705STDY8058280.annotation.csv
-rwxrwx---. 1 zouyang ngs 446K Apr 21 16:59 5705STDY8058281.annotation.csv
-rwxrwx---. 1 zouyang ngs 310K Apr 21 16:59 5705STDY8058282.annotation.csv
-rwxrwx---. 1 zouyang ngs 282K Apr 21 16:59 5705STDY8058283.annotation.csv
-rwxrwx---. 1 zouyang ngs 157K Apr 21 16:59 5705STDY8058284.annotation.csv
-rwxrwx---. 1 zouyang ngs 555K Apr 21 16:59 5705STDY8058285.annotation.csv

#A brief look of the annotation file:
$ head -3 5705STDY8058280_annotation.csv
Cell.ID,sample,annotation_1,annotation_1_print
AAACCCAAGGAAGTAG-1,5705STDY8058280,Ext_L25,23_Ext_L25
AAACCCAAGGGCAGTT-1,5705STDY8058280,Ext_L56,24_Ext_L56</code></pre>
<p>If <strong>SampleName.metrics_summary.csv</strong> files are available, please also add them in the same directory as the h5 files, and the pipeline will use them to generate QC plots, but they are not required files of the pipeline. The expected file names should be:</p>
<pre><code>5705STDY8058280.metrics_summary.csv
5705STDY8058281.metrics_summary.csv
5705STDY8058282.metrics_summary.csv
5705STDY8058283.metrics_summary.csv
5705STDY8058284.metrics_summary.csv
5705STDY8058285.metrics_summary.csv</code></pre>
<p><strong>!!! Important</strong> Since we don’t provide the path of these metrics_summary.csv files to the pipeline, their prefix <strong>SampleName</strong> must be consistent with the <strong>Sample_Name</strong> column in the sample meta file (See section 5.2), so that the pipeline can recognize them automatically. You can certainly rename the files and the corresponding <strong>SampleName</strong> column if you would like to change the data names and how they appear in the final results. Also, for the metrics_summary.csv file names, the concatenator between the <strong>Sample_Name</strong> and “metrics_summary.csv” must be “.” so that the pipeline can read them in automatically.</p>
</div>
<div id="public-data-in-10x-mex-format" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Public data in 10X MEX format</h2>
<p>Another popular format for single-cell RNA-seq is the <a href="https://kb.10xgenomics.com/hc/en-us/articles/115000794686-How-is-the-MEX-format-used-for-the-gene-barcode-matrices" target="_blank"><strong>MEX format</strong></a>, with a <strong>mtx file</strong> and associated <strong>barcodes</strong> file as well as a <strong>features</strong> file. Here, we use a public dataset from NCBI/GEO: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE205483" target="_blank">GSE205483</a> to walk through the procedures for pipeline setup.</p>
<p>There are total 4 data, and all the processed data can be downloaded from the Supplementary file section as a tar ball: GSE205483_RAW.tar:</p>
<pre><code>#Untar the file
tar -xvf GSE205483_RAW.tar

$ ls -l

-rw-rw-r--. 1 ysun4 ysun4   4704844 Jun  2 15:13 GSM6213399_IT_YZ_0309_2904_barcodes.tsv.gz
-rw-rw-r--. 1 ysun4 ysun4    260888 Jun  2 15:13 GSM6213399_IT_YZ_0309_2904_features.tsv.gz
-rw-rw-r--. 1 ysun4 ysun4  64071180 Jun  2 15:13 GSM6213399_IT_YZ_0309_2904_matrix.mtx.gz
-rw-rw-r--. 1 ysun4 ysun4   5474445 Jun  2 15:21 GSM6213400_IT_YZ_0309_2905_barcodes.tsv.gz
-rw-rw-r--. 1 ysun4 ysun4    260888 Jun  2 15:21 GSM6213400_IT_YZ_0309_2905_features.tsv.gz
-rw-rw-r--. 1 ysun4 ysun4  73218924 Jun  2 15:21 GSM6213400_IT_YZ_0309_2905_matrix.mtx.gz
-rw-rw-r--. 1 ysun4 ysun4   4931566 Jun  2 15:30 GSM6213401_IT_YZ_0309_2908_barcodes.tsv.gz
-rw-rw-r--. 1 ysun4 ysun4    260888 Jun  2 15:30 GSM6213401_IT_YZ_0309_2908_features.tsv.gz
-rw-rw-r--. 1 ysun4 ysun4  53887060 Jun  2 15:30 GSM6213401_IT_YZ_0309_2908_matrix.mtx.gz
-rw-rw-r--. 1 ysun4 ysun4   4132528 Jun  2 15:37 GSM6213402_IT_YZ_0322_3125_barcodes.tsv.gz
-rw-rw-r--. 1 ysun4 ysun4    260888 Jun  2 15:37 GSM6213402_IT_YZ_0322_3125_features.tsv.gz
-rw-rw-r--. 1 ysun4 ysun4  53205493 Jun  2 15:37 GSM6213402_IT_YZ_0322_3125_matrix.mtx.gz</code></pre>
<p>Next, these files need to be organized into separate folders for the pipeline to read. In specific, we need to create separate folders for each data, and rename the three files to be: <strong>barcodes.tsv.gz</strong>, <strong>features.tsv.gz</strong>, and <strong>matrix.mtx.gz</strong>. The organized file hierarchy is below:</p>
<pre><code>GSE205483/
    ├── GSM6213399_IT_YZ_0309_2904
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ├── GSM6213400_IT_YZ_0309_2905
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ├── GSM6213401_IT_YZ_0309_2908
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    └── GSM6213402_IT_YZ_0322_3125
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz</code></pre>
</div>
<div id="self-prepared-files" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Self-prepared files</h2>
<p>If you have raw data in <a href="https://support.illumina.com/bulletins/2016/04/fastq-files-explained.html" target="_blank">FASTQ format</a>, please process them using the Cell Ranger pipeline to generate the raw_feature_bc_matrix.h5 and metrics_summary.csv files. Please visit <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count" target="_blank">Cell Ranger website here</a> for more details about the outputs, in the “Output files” section.</p>
<p>In short, Cell Ranger outputs <strong>filtered_feature_bc_matrix.h5</strong> (cells after filtering, recommended to use), <strong>raw_feature_bc_matrix.h5</strong> (without filtering), and two folders for MEX format output: <strong>filtered_feature_bc_matrix</strong>, <strong>raw_feature_bc_matrix</strong>. In addition, Cell Ranger also has <strong>metrics_summary.csv</strong> generated which can be provided to scRNASequest. Please organize, and if necessary, rename the Cell Ranger output files into the following file structures before running scRNASequest.</p>
<p>Below is the file hierarchy for h5 input files (metrics_summary.csv files are suggested to be included, but not required; If cell type classification annotation.csv files are available, it would be better to include them):</p>
<pre><code>Project/
    ├── Data1.filtered_feature_bc_matrix.h5
    ├── Data1.annotation.csv                 (optional)
    ├── Data1.metrics_summary.csv            (optional)
    ├── Data2.filtered_feature_bc_matrix.h5
    ├── Data2.annotation.csv                 (optional)
    ├── Data2.metrics_summary.csv            (optional)
    ...</code></pre>
<p>The file hierarchy for <a href="https://kb.10xgenomics.com/hc/en-us/articles/115000794686-How-is-the-MEX-format-used-for-the-gene-barcode-matrices" target="_blank">MEX files</a>:</p>
<pre><code>Project/
    ├── Data1
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ├── Data2
        ├── barcodes.tsv.gz
        ├── features.tsv.gz
        └── matrix.mtx.gz
    ...</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="installation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="input-configuration.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/scRNAsequest/edit/gh-pages/04-DataPreparation.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/scRNAsequest/blob/gh-pages/04-DataPreparation.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
