<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Demo run | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</title>
  <meta name="description" content="Chapter 4 Demo run | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Demo run | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Demo run | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  
  
  



<meta name="date" content="2024-03-28" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="installation.html"/>
<link rel="next" href="data-preparation.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">scRNASequest</a></li>

<li class="divider"></li>
<li><a href="index.html#preface" id="toc-preface"><span class="toc-section-number">1</span> Preface</a></li>
<li><a href="introduction.html#introduction" id="toc-introduction"><span class="toc-section-number">2</span> Introduction</a></li>
<li><a href="installation.html#installation" id="toc-installation"><span class="toc-section-number">3</span> Installation</a>
<ul>
<li><a href="installation.html#install-scrnasequest" id="toc-install-scrnasequest"><span class="toc-section-number">3.1</span> Install scRNASequest</a>
<ul>
<li><a href="installation.html#installation-using-conda" id="toc-installation-using-conda"><span class="toc-section-number">3.1.1</span> Installation using Conda</a></li>
<li><a href="installation.html#installation-through-docker" id="toc-installation-through-docker"><span class="toc-section-number">3.1.2</span> Installation through Docker</a></li>
</ul></li>
<li><a href="installation.html#configure-sys.yml-file" id="toc-configure-sys.yml-file"><span class="toc-section-number">3.2</span> Configure sys.yml file</a></li>
<li><a href="installation.html#install-cellxgene-vip" id="toc-install-cellxgene-vip"><span class="toc-section-number">3.3</span> Install cellxgene VIP</a></li>
<li><a href="installation.html#install-celldepot" id="toc-install-celldepot"><span class="toc-section-number">3.4</span> Install CellDepot</a></li>
</ul></li>
<li><a href="demo-run.html#demo-run" id="toc-demo-run"><span class="toc-section-number">4</span> Demo run</a>
<ul>
<li><a href="demo-run.html#demo-run-with-conda" id="toc-demo-run-with-conda"><span class="toc-section-number">4.1</span> Demo run with Conda</a></li>
<li><a href="demo-run.html#demo-run-with-docker" id="toc-demo-run-with-docker"><span class="toc-section-number">4.2</span> Demo run with Docker</a></li>
</ul></li>
<li><a href="data-preparation.html#data-preparation" id="toc-data-preparation"><span class="toc-section-number">5</span> Data preparation</a>
<ul>
<li><a href="data-preparation.html#public-data-in-h5-format" id="toc-public-data-in-h5-format"><span class="toc-section-number">5.1</span> Public data in h5 format</a></li>
<li><a href="data-preparation.html#public-data-in-10x-mex-format" id="toc-public-data-in-10x-mex-format"><span class="toc-section-number">5.2</span> Public data in 10X MEX format</a></li>
<li><a href="data-preparation.html#self-prepared-files" id="toc-self-prepared-files"><span class="toc-section-number">5.3</span> Self-prepared files</a></li>
</ul></li>
<li><a href="input-configuration.html#input-configuration" id="toc-input-configuration"><span class="toc-section-number">6</span> Input configuration</a>
<ul>
<li><a href="input-configuration.html#generate-template-files" id="toc-generate-template-files"><span class="toc-section-number">6.1</span> Generate template files</a></li>
<li><a href="input-configuration.html#prepare-the-config.yml-file" id="toc-prepare-the-config.yml-file"><span class="toc-section-number">6.2</span> Prepare the config.yml file</a></li>
<li><a href="input-configuration.html#section52" id="toc-section52"><span class="toc-section-number">6.3</span> Prepare the sample meta file</a>
<ul>
<li><a href="input-configuration.html#h5-input" id="toc-h5-input"><span class="toc-section-number">6.3.1</span> h5 input</a></li>
<li><a href="input-configuration.html#mex-input" id="toc-mex-input"><span class="toc-section-number">6.3.2</span> MEX input</a></li>
</ul></li>
<li><a href="input-configuration.html#file-hierarchy" id="toc-file-hierarchy"><span class="toc-section-number">6.4</span> File hierarchy</a></li>
</ul></li>
<li><a href="running-the-pipeline.html#running-the-pipeline" id="toc-running-the-pipeline"><span class="toc-section-number">7</span> Running the pipeline</a>
<ul>
<li><a href="running-the-pipeline.html#initial-scanalyzer-run" id="toc-initial-scanalyzer-run"><span class="toc-section-number">7.1</span> Initial scAnalyzer run</a></li>
<li><a href="running-the-pipeline.html#review-the-bookdown-report" id="toc-review-the-bookdown-report"><span class="toc-section-number">7.2</span> Review the bookdown report</a></li>
<li><a href="running-the-pipeline.html#generate-a-slide-deck" id="toc-generate-a-slide-deck"><span class="toc-section-number">7.3</span> Generate a slide deck</a></li>
<li><a href="running-the-pipeline.html#doublet-detection" id="toc-doublet-detection"><span class="toc-section-number">7.4</span> Doublet detection</a></li>
<li><a href="running-the-pipeline.html#cell-type-label-transfer" id="toc-cell-type-label-transfer"><span class="toc-section-number">7.5</span> Cell type label transfer</a></li>
<li><a href="running-the-pipeline.html#adjust-major_cluster_rate" id="toc-adjust-major_cluster_rate"><span class="toc-section-number">7.6</span> Adjust major_cluster_rate</a></li>
<li><a href="running-the-pipeline.html#run-the-pipeline-in-parallel" id="toc-run-the-pipeline-in-parallel"><span class="toc-section-number">7.7</span> Run the pipeline in parallel</a></li>
<li><a href="running-the-pipeline.html#run-the-full-analysis" id="toc-run-the-full-analysis"><span class="toc-section-number">7.8</span> Run the full analysis</a></li>
<li><a href="running-the-pipeline.html#visualize-the-harmonized-results" id="toc-visualize-the-harmonized-results"><span class="toc-section-number">7.9</span> Visualize the harmonized results</a></li>
</ul></li>
<li><a href="differential-expression-de-analysis.html#differential-expression-de-analysis" id="toc-differential-expression-de-analysis"><span class="toc-section-number">8</span> Differential expression (DE) analysis</a>
<ul>
<li><a href="differential-expression-de-analysis.html#example-1" id="toc-example-1"><span class="toc-section-number">8.1</span> Example 1</a></li>
<li><a href="differential-expression-de-analysis.html#example-2" id="toc-example-2"><span class="toc-section-number">8.2</span> Example 2</a></li>
<li><a href="differential-expression-de-analysis.html#example-3" id="toc-example-3"><span class="toc-section-number">8.3</span> Example 3</a></li>
</ul></li>
<li><a href="reference-building.html#reference-building" id="toc-reference-building"><span class="toc-section-number">9</span> Reference building</a>
<ul>
<li><a href="reference-building.html#initialize-scref" id="toc-initialize-scref"><span class="toc-section-number">9.1</span> Initialize scRef</a></li>
<li><a href="reference-building.html#submit-scref" id="toc-submit-scref"><span class="toc-section-number">9.2</span> Submit scRef</a></li>
<li><a href="reference-building.html#demo-scref" id="toc-demo-scref"><span class="toc-section-number">9.3</span> Demo scRef</a></li>
</ul></li>
<li><a href="celldepot-publishing.html#celldepot-publishing" id="toc-celldepot-publishing"><span class="toc-section-number">10</span> CellDepot publishing</a>
<ul>
<li><a href="celldepot-publishing.html#initialize-sc2celldepot" id="toc-initialize-sc2celldepot"><span class="toc-section-number">10.1</span> Initialize sc2celldepot</a></li>
<li><a href="celldepot-publishing.html#run-sc2celldepot" id="toc-run-sc2celldepot"><span class="toc-section-number">10.2</span> Run sc2celldepot</a></li>
<li><a href="celldepot-publishing.html#demo-sc2celldepot" id="toc-demo-sc2celldepot"><span class="toc-section-number">10.3</span> Demo sc2celldepot</a></li>
</ul></li>
<li><a href="additional-tools.html#additional-tools" id="toc-additional-tools"><span class="toc-section-number">11</span> Additional tools</a>
<ul>
<li><a href="additional-tools.html#scrmambient" id="toc-scrmambient"><span class="toc-section-number">11.1</span> scRMambient</a>
<ul>
<li><a href="additional-tools.html#set-up-the-inputs" id="toc-set-up-the-inputs"><span class="toc-section-number">11.1.1</span> Set up the inputs</a></li>
<li><a href="additional-tools.html#run-scrmambient" id="toc-run-scrmambient"><span class="toc-section-number">11.1.2</span> Run scRMambient</a></li>
</ul></li>
<li><a href="additional-tools.html#scdeg" id="toc-scdeg"><span class="toc-section-number">11.2</span> scDEG</a>
<ul>
<li><a href="additional-tools.html#initialize-scdeg" id="toc-initialize-scdeg"><span class="toc-section-number">11.2.1</span> Initialize scDEG</a></li>
<li><a href="additional-tools.html#run-de-analysis" id="toc-run-de-analysis"><span class="toc-section-number">11.2.2</span> Run DE analysis</a></li>
</ul></li>
<li><a href="additional-tools.html#sctool" id="toc-sctool"><span class="toc-section-number">11.3</span> scTool</a>
<ul>
<li><a href="additional-tools.html#remove-annotations" id="toc-remove-annotations"><span class="toc-section-number">11.3.1</span> Remove annotations</a></li>
<li><a href="additional-tools.html#add-annotations" id="toc-add-annotations"><span class="toc-section-number">11.3.2</span> Add annotations</a></li>
<li><a href="additional-tools.html#export-gene-expressions" id="toc-export-gene-expressions"><span class="toc-section-number">11.3.3</span> Export gene expressions</a></li>
<li><a href="additional-tools.html#pseudo-bulk-aggregation" id="toc-pseudo-bulk-aggregation"><span class="toc-section-number">11.3.4</span> Pseudo-bulk aggregation</a></li>
<li><a href="additional-tools.html#demo-sctool" id="toc-demo-sctool"><span class="toc-section-number">11.3.5</span> Demo scTool</a></li>
</ul></li>
</ul></li>
<li><a href="miscellaneous.html#miscellaneous" id="toc-miscellaneous"><span class="toc-section-number">12</span> Miscellaneous</a>
<ul>
<li><a href="miscellaneous.html#cell-subsetting" id="toc-cell-subsetting"><span class="toc-section-number">12.1</span> Cell subsetting</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="demo-run" class="section level1" number="4">
<h1><span class="header-section-number">Chapter 4</span> Demo run</h1>
<p>We provide a demo dataset under the <code>demo</code> directory. This demo uses two snRNA-seq data from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE185538">GSE185538</a> to run through the main steps, including <strong>QC</strong>, <strong>data integration</strong> (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1">SCTransform</a>, then <a href="https://www.nature.com/articles/s41592-019-0619-0"><strong>Harmony</strong></a>), <a href="https://satijalab.org/seurat/articles/integration_mapping.html"><strong>Seurat reference mapping</strong></a>, and <strong>evaluation of integration</strong> (<a href="https://www.nature.com/articles/s41592-018-0254-1">kBET</a> and <a href="https://ieeexplore.ieee.org/document/9260048">silhouette</a>). To save time, the differential expression analysis won’t be performed, and the DEGinfo.csv file is empty.</p>
<p>This demo run contains two downsampled snRNA-seq data from the original study, and will take <strong>~15-20 minutes to finish</strong>. Please note that to speed up the run, we used stringent QC cutoffs, which eliminated many cells. After running the scAnalyzer command, output files will be generated under the <code>demo</code> directory on your computer.</p>
<p>Also, we provided config.yml, sampleMeta.csv and DEGinfo.csv files for running this demo so you can run the <code>scAnalyzer</code> command directly. Please note that these files are only for the demo run. Later for your own dataset, please follow the full tutorial in the next section, <a href="data-preparation.html#data-preparation">Data preparation</a>, to generate the template of these files first.</p>
<div id="demo-run-with-conda" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Demo run with Conda</h2>
<p>Continuing from section 3.1.1, we assume that the pipeline was installed in: ~/scRNASequest.</p>
<p>First, we set up the config file (~/scRNASequest/demo/config.yml) by pointing these parameters to your directory, and other lines don’t need to be changed:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="demo-run.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span>
<span id="cb20-2"><a href="demo-run.html#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="ex">ref_name:</span> ~/scRNASequest/demo/ref/rat_cortex_ref.rds   <span class="co"># choose one from scAnalyzer call without argument</span></span>
<span id="cb20-3"><a href="demo-run.html#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="ex">output:</span> ~/scRNASequest/demo                            <span class="co"># output path   </span></span>
<span id="cb20-4"><a href="demo-run.html#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span>
<span id="cb20-5"><a href="demo-run.html#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="ex">sample_meta:</span> ~/scRNASequest/demo/sampleMeta.csv</span>
<span id="cb20-6"><a href="demo-run.html#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span>
<span id="cb20-7"><a href="demo-run.html#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="ex">DEG_desp:</span> ~/scRNASequest/demo/DEGinfo.csv              <span class="co"># for DEG analysis. To save time, this demo run doesn&#39;t include DE analysis</span></span>
<span id="cb20-8"><a href="demo-run.html#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span></code></pre></div>
<p>Accordingly, we modify the sampleMeta.csv file by providing the full path to .h5 files by adding ‘~/scRNASequest’:</p>
<pre><code>Sample_Name,h5path,Sex
RatFemaleCigarette,~/scRNASequest/demo/data/RatFemaleCigarette.filtered_feature_bc_matrix.h5,Female
RatMaleCigarette,~/scRNASequest/demo/data/RatMaleCigarette.filtered_feature_bc_matrix.h5,Male</code></pre>
<p>Then execute the following command:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb22-1"><a href="demo-run.html#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span> ~/scRNASequest/demo/config.yml</span></code></pre></div>
</div>
<div id="demo-run-with-docker" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Demo run with Docker</h2>
<p>After following section 3.1.2 to set up the pipeline, we are ready to run this demo:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb23-1"><a href="demo-run.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> exec <span class="at">-t</span> <span class="at">-i</span> <span class="op">&lt;</span>container_name<span class="op">&gt;</span> scAnalyzer /demo/config.yml</span></code></pre></div>
<p>After running this demo, a list of files will be created:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb24-1"><a href="demo-run.html#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">config.yml</span>                        <span class="co">#Original config.yml file</span></span>
<span id="cb24-2"><a href="demo-run.html#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="ex">config.yml.20230314.log</span>           <span class="co">#Log file during pipeline run</span></span>
<span id="cb24-3"><a href="demo-run.html#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="ex">data/</span>                             <span class="co">#Original data files</span></span>
<span id="cb24-4"><a href="demo-run.html#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──RatFemaleCigarette.filtered_feature_bc_matrix.h5</span></span>
<span id="cb24-5"><a href="demo-run.html#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──RatFemaleCigarette.metrics_summary.csv</span></span>
<span id="cb24-6"><a href="demo-run.html#cb24-6" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──RatMaleCigarette.filtered_feature_bc_matrix.h5</span></span>
<span id="cb24-7"><a href="demo-run.html#cb24-7" aria-hidden="true" tabindex="-1"></a>  <span class="ex">└──RatMaleCigarette.metrics_summary.csv</span></span>
<span id="cb24-8"><a href="demo-run.html#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="ex">DEGinfo.csv</span>                       <span class="co">#Comparison file. In this demo, this is empty</span></span>
<span id="cb24-9"><a href="demo-run.html#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="ex">evaluation/</span>                       <span class="co">#Evaluation of harmonization</span></span>
<span id="cb24-10"><a href="demo-run.html#cb24-10" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──scRNASequest_demo_kBET_umap_k0_100.pdf</span></span>
<span id="cb24-11"><a href="demo-run.html#cb24-11" aria-hidden="true" tabindex="-1"></a>  <span class="ex">└──scRNASequest_demo_Silhouette_boxplot_pc50.pdf</span></span>
<span id="cb24-12"><a href="demo-run.html#cb24-12" aria-hidden="true" tabindex="-1"></a><span class="ex">log/</span>                              <span class="co">#Log files for each step</span></span>
<span id="cb24-13"><a href="demo-run.html#cb24-13" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──kBET.log</span></span>
<span id="cb24-14"><a href="demo-run.html#cb24-14" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──sctHarmony.log</span></span>
<span id="cb24-15"><a href="demo-run.html#cb24-15" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──SCT.log</span></span>
<span id="cb24-16"><a href="demo-run.html#cb24-16" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──SeuratRef.log</span></span>
<span id="cb24-17"><a href="demo-run.html#cb24-17" aria-hidden="true" tabindex="-1"></a>  <span class="ex">└──ssilhouette.log</span></span>
<span id="cb24-18"><a href="demo-run.html#cb24-18" aria-hidden="true" tabindex="-1"></a><span class="ex">QC/</span>                               <span class="co">#QC files</span></span>
<span id="cb24-19"><a href="demo-run.html#cb24-19" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──postfilter.QC.pdf</span></span>
<span id="cb24-20"><a href="demo-run.html#cb24-20" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──prefilter.QC.pdf</span></span>
<span id="cb24-21"><a href="demo-run.html#cb24-21" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──sequencingQC.csv</span></span>
<span id="cb24-22"><a href="demo-run.html#cb24-22" aria-hidden="true" tabindex="-1"></a>  <span class="ex">└──sequencingQC.pdf</span></span>
<span id="cb24-23"><a href="demo-run.html#cb24-23" aria-hidden="true" tabindex="-1"></a><span class="ex">raw/</span>                              <span class="co">#Original UMI counts, with and withour filtering</span></span>
<span id="cb24-24"><a href="demo-run.html#cb24-24" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──scRNASequest_demo.h5ad</span></span>
<span id="cb24-25"><a href="demo-run.html#cb24-25" aria-hidden="true" tabindex="-1"></a>  <span class="ex">└──scRNASequest_demo_raw_prefilter.h5ad</span></span>
<span id="cb24-26"><a href="demo-run.html#cb24-26" aria-hidden="true" tabindex="-1"></a><span class="ex">ref/</span></span>
<span id="cb24-27"><a href="demo-run.html#cb24-27" aria-hidden="true" tabindex="-1"></a>  <span class="ex">└──ref.Rds</span>                      <span class="co">#Azimuth reference RSD file for cell type label transfer</span></span>
<span id="cb24-28"><a href="demo-run.html#cb24-28" aria-hidden="true" tabindex="-1"></a><span class="ex">Rmarkdown/</span>                        <span class="co">#Pre-generated QC figures, for generating Bookdown report</span></span>
<span id="cb24-29"><a href="demo-run.html#cb24-29" aria-hidden="true" tabindex="-1"></a><span class="ex">sampleMeta.csv</span>                    <span class="co">#Input sample metainformation file</span></span>
<span id="cb24-30"><a href="demo-run.html#cb24-30" aria-hidden="true" tabindex="-1"></a><span class="ex">scRNASequest_demo.h5ad</span>            <span class="co">#This is the final output h5ad file, and the expression values have been normalized</span></span>
<span id="cb24-31"><a href="demo-run.html#cb24-31" aria-hidden="true" tabindex="-1"></a>                                   <span class="ex">cellxgene</span> VIP will use this file for visualization</span>
<span id="cb24-32"><a href="demo-run.html#cb24-32" aria-hidden="true" tabindex="-1"></a><span class="ex">scRNASequest_demo.h5seurat</span>        <span class="co">#Result in h5seurat format</span></span>
<span id="cb24-33"><a href="demo-run.html#cb24-33" aria-hidden="true" tabindex="-1"></a><span class="ex">scRNASequest_demo_raw_added.h5ad</span>  <span class="co">#Since AnnData object can only have one count matrix, this one uses the raw UMI</span></span>
<span id="cb24-34"><a href="demo-run.html#cb24-34" aria-hidden="true" tabindex="-1"></a><span class="ex">SCT/</span>                              <span class="co">#Results related to SCT transformation</span></span>
<span id="cb24-35"><a href="demo-run.html#cb24-35" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──scRNASequest_demo.cID</span></span>
<span id="cb24-36"><a href="demo-run.html#cb24-36" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──scRNASequest_demo.gID</span></span>
<span id="cb24-37"><a href="demo-run.html#cb24-37" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──scRNASequest_demo.h5</span></span>
<span id="cb24-38"><a href="demo-run.html#cb24-38" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──scRNASequest_demo.h5ad</span></span>
<span id="cb24-39"><a href="demo-run.html#cb24-39" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──scRNASequest_demo.h5.rds</span></span>
<span id="cb24-40"><a href="demo-run.html#cb24-40" aria-hidden="true" tabindex="-1"></a>  <span class="ex">└──scRNASequest_demo.scaleF</span></span>
<span id="cb24-41"><a href="demo-run.html#cb24-41" aria-hidden="true" tabindex="-1"></a><span class="ex">sctHarmony/</span>                       <span class="co">#SCT+Harmony results</span></span>
<span id="cb24-42"><a href="demo-run.html#cb24-42" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──scRNASequest_demo.csv</span></span>
<span id="cb24-43"><a href="demo-run.html#cb24-43" aria-hidden="true" tabindex="-1"></a>  <span class="ex">└──scRNASequest_demo.h5ad</span></span>
<span id="cb24-44"><a href="demo-run.html#cb24-44" aria-hidden="true" tabindex="-1"></a><span class="ex">SeuratRef/</span>                       <span class="co">#Seurat reference mapping results</span></span>
<span id="cb24-45"><a href="demo-run.html#cb24-45" aria-hidden="true" tabindex="-1"></a>  <span class="ex">├──scRNASequest_demo.csv</span></span>
<span id="cb24-46"><a href="demo-run.html#cb24-46" aria-hidden="true" tabindex="-1"></a>  <span class="ex">└──scRNASequest_demo.h5ad</span></span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="installation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-preparation.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/scRNAsequest/edit/gh-pages/04-DemoRun.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/scRNAsequest/blob/gh-pages/04-DemoRun.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
