<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 8 Differential expression (DE) analysis | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</title>
  <meta name="description" content="Chapter 8 Differential expression (DE) analysis | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 8 Differential expression (DE) analysis | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 Differential expression (DE) analysis | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  
  
  



<meta name="date" content="2024-03-28" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="running-the-pipeline.html"/>
<link rel="next" href="reference-building.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">scRNASequest</a></li>

<li class="divider"></li>
<li><a href="index.html#preface" id="toc-preface"><span class="toc-section-number">1</span> Preface</a></li>
<li><a href="introduction.html#introduction" id="toc-introduction"><span class="toc-section-number">2</span> Introduction</a></li>
<li><a href="installation.html#installation" id="toc-installation"><span class="toc-section-number">3</span> Installation</a>
<ul>
<li><a href="installation.html#install-scrnasequest" id="toc-install-scrnasequest"><span class="toc-section-number">3.1</span> Install scRNASequest</a>
<ul>
<li><a href="installation.html#installation-using-conda" id="toc-installation-using-conda"><span class="toc-section-number">3.1.1</span> Installation using Conda</a></li>
<li><a href="installation.html#installation-through-docker" id="toc-installation-through-docker"><span class="toc-section-number">3.1.2</span> Installation through Docker</a></li>
</ul></li>
<li><a href="installation.html#configure-sys.yml-file" id="toc-configure-sys.yml-file"><span class="toc-section-number">3.2</span> Configure sys.yml file</a></li>
<li><a href="installation.html#install-cellxgene-vip" id="toc-install-cellxgene-vip"><span class="toc-section-number">3.3</span> Install cellxgene VIP</a></li>
<li><a href="installation.html#install-celldepot" id="toc-install-celldepot"><span class="toc-section-number">3.4</span> Install CellDepot</a></li>
</ul></li>
<li><a href="demo-run.html#demo-run" id="toc-demo-run"><span class="toc-section-number">4</span> Demo run</a>
<ul>
<li><a href="demo-run.html#demo-run-with-conda" id="toc-demo-run-with-conda"><span class="toc-section-number">4.1</span> Demo run with Conda</a></li>
<li><a href="demo-run.html#demo-run-with-docker" id="toc-demo-run-with-docker"><span class="toc-section-number">4.2</span> Demo run with Docker</a></li>
</ul></li>
<li><a href="data-preparation.html#data-preparation" id="toc-data-preparation"><span class="toc-section-number">5</span> Data preparation</a>
<ul>
<li><a href="data-preparation.html#public-data-in-h5-format" id="toc-public-data-in-h5-format"><span class="toc-section-number">5.1</span> Public data in h5 format</a></li>
<li><a href="data-preparation.html#public-data-in-10x-mex-format" id="toc-public-data-in-10x-mex-format"><span class="toc-section-number">5.2</span> Public data in 10X MEX format</a></li>
<li><a href="data-preparation.html#self-prepared-files" id="toc-self-prepared-files"><span class="toc-section-number">5.3</span> Self-prepared files</a></li>
</ul></li>
<li><a href="input-configuration.html#input-configuration" id="toc-input-configuration"><span class="toc-section-number">6</span> Input configuration</a>
<ul>
<li><a href="input-configuration.html#generate-template-files" id="toc-generate-template-files"><span class="toc-section-number">6.1</span> Generate template files</a></li>
<li><a href="input-configuration.html#prepare-the-config.yml-file" id="toc-prepare-the-config.yml-file"><span class="toc-section-number">6.2</span> Prepare the config.yml file</a></li>
<li><a href="input-configuration.html#section52" id="toc-section52"><span class="toc-section-number">6.3</span> Prepare the sample meta file</a>
<ul>
<li><a href="input-configuration.html#h5-input" id="toc-h5-input"><span class="toc-section-number">6.3.1</span> h5 input</a></li>
<li><a href="input-configuration.html#mex-input" id="toc-mex-input"><span class="toc-section-number">6.3.2</span> MEX input</a></li>
</ul></li>
<li><a href="input-configuration.html#file-hierarchy" id="toc-file-hierarchy"><span class="toc-section-number">6.4</span> File hierarchy</a></li>
</ul></li>
<li><a href="running-the-pipeline.html#running-the-pipeline" id="toc-running-the-pipeline"><span class="toc-section-number">7</span> Running the pipeline</a>
<ul>
<li><a href="running-the-pipeline.html#initial-scanalyzer-run" id="toc-initial-scanalyzer-run"><span class="toc-section-number">7.1</span> Initial scAnalyzer run</a></li>
<li><a href="running-the-pipeline.html#review-the-bookdown-report" id="toc-review-the-bookdown-report"><span class="toc-section-number">7.2</span> Review the bookdown report</a></li>
<li><a href="running-the-pipeline.html#generate-a-slide-deck" id="toc-generate-a-slide-deck"><span class="toc-section-number">7.3</span> Generate a slide deck</a></li>
<li><a href="running-the-pipeline.html#doublet-detection" id="toc-doublet-detection"><span class="toc-section-number">7.4</span> Doublet detection</a></li>
<li><a href="running-the-pipeline.html#cell-type-label-transfer" id="toc-cell-type-label-transfer"><span class="toc-section-number">7.5</span> Cell type label transfer</a></li>
<li><a href="running-the-pipeline.html#adjust-major_cluster_rate" id="toc-adjust-major_cluster_rate"><span class="toc-section-number">7.6</span> Adjust major_cluster_rate</a></li>
<li><a href="running-the-pipeline.html#run-the-pipeline-in-parallel" id="toc-run-the-pipeline-in-parallel"><span class="toc-section-number">7.7</span> Run the pipeline in parallel</a></li>
<li><a href="running-the-pipeline.html#run-the-full-analysis" id="toc-run-the-full-analysis"><span class="toc-section-number">7.8</span> Run the full analysis</a></li>
<li><a href="running-the-pipeline.html#visualize-the-harmonized-results" id="toc-visualize-the-harmonized-results"><span class="toc-section-number">7.9</span> Visualize the harmonized results</a></li>
</ul></li>
<li><a href="differential-expression-de-analysis.html#differential-expression-de-analysis" id="toc-differential-expression-de-analysis"><span class="toc-section-number">8</span> Differential expression (DE) analysis</a>
<ul>
<li><a href="differential-expression-de-analysis.html#example-1" id="toc-example-1"><span class="toc-section-number">8.1</span> Example 1</a></li>
<li><a href="differential-expression-de-analysis.html#example-2" id="toc-example-2"><span class="toc-section-number">8.2</span> Example 2</a></li>
<li><a href="differential-expression-de-analysis.html#example-3" id="toc-example-3"><span class="toc-section-number">8.3</span> Example 3</a></li>
</ul></li>
<li><a href="reference-building.html#reference-building" id="toc-reference-building"><span class="toc-section-number">9</span> Reference building</a>
<ul>
<li><a href="reference-building.html#initialize-scref" id="toc-initialize-scref"><span class="toc-section-number">9.1</span> Initialize scRef</a></li>
<li><a href="reference-building.html#submit-scref" id="toc-submit-scref"><span class="toc-section-number">9.2</span> Submit scRef</a></li>
<li><a href="reference-building.html#demo-scref" id="toc-demo-scref"><span class="toc-section-number">9.3</span> Demo scRef</a></li>
</ul></li>
<li><a href="celldepot-publishing.html#celldepot-publishing" id="toc-celldepot-publishing"><span class="toc-section-number">10</span> CellDepot publishing</a>
<ul>
<li><a href="celldepot-publishing.html#initialize-sc2celldepot" id="toc-initialize-sc2celldepot"><span class="toc-section-number">10.1</span> Initialize sc2celldepot</a></li>
<li><a href="celldepot-publishing.html#run-sc2celldepot" id="toc-run-sc2celldepot"><span class="toc-section-number">10.2</span> Run sc2celldepot</a></li>
<li><a href="celldepot-publishing.html#demo-sc2celldepot" id="toc-demo-sc2celldepot"><span class="toc-section-number">10.3</span> Demo sc2celldepot</a></li>
</ul></li>
<li><a href="additional-tools.html#additional-tools" id="toc-additional-tools"><span class="toc-section-number">11</span> Additional tools</a>
<ul>
<li><a href="additional-tools.html#scrmambient" id="toc-scrmambient"><span class="toc-section-number">11.1</span> scRMambient</a>
<ul>
<li><a href="additional-tools.html#set-up-the-inputs" id="toc-set-up-the-inputs"><span class="toc-section-number">11.1.1</span> Set up the inputs</a></li>
<li><a href="additional-tools.html#run-scrmambient" id="toc-run-scrmambient"><span class="toc-section-number">11.1.2</span> Run scRMambient</a></li>
</ul></li>
<li><a href="additional-tools.html#scdeg" id="toc-scdeg"><span class="toc-section-number">11.2</span> scDEG</a>
<ul>
<li><a href="additional-tools.html#initialize-scdeg" id="toc-initialize-scdeg"><span class="toc-section-number">11.2.1</span> Initialize scDEG</a></li>
<li><a href="additional-tools.html#run-de-analysis" id="toc-run-de-analysis"><span class="toc-section-number">11.2.2</span> Run DE analysis</a></li>
</ul></li>
<li><a href="additional-tools.html#sctool" id="toc-sctool"><span class="toc-section-number">11.3</span> scTool</a>
<ul>
<li><a href="additional-tools.html#remove-annotations" id="toc-remove-annotations"><span class="toc-section-number">11.3.1</span> Remove annotations</a></li>
<li><a href="additional-tools.html#add-annotations" id="toc-add-annotations"><span class="toc-section-number">11.3.2</span> Add annotations</a></li>
<li><a href="additional-tools.html#export-gene-expressions" id="toc-export-gene-expressions"><span class="toc-section-number">11.3.3</span> Export gene expressions</a></li>
<li><a href="additional-tools.html#pseudo-bulk-aggregation" id="toc-pseudo-bulk-aggregation"><span class="toc-section-number">11.3.4</span> Pseudo-bulk aggregation</a></li>
<li><a href="additional-tools.html#demo-sctool" id="toc-demo-sctool"><span class="toc-section-number">11.3.5</span> Demo scTool</a></li>
</ul></li>
</ul></li>
<li><a href="miscellaneous.html#miscellaneous" id="toc-miscellaneous"><span class="toc-section-number">12</span> Miscellaneous</a>
<ul>
<li><a href="miscellaneous.html#cell-subsetting" id="toc-cell-subsetting"><span class="toc-section-number">12.1</span> Cell subsetting</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="differential-expression-de-analysis" class="section level1" number="8">
<h1><span class="header-section-number">Chapter 8</span> Differential expression (DE) analysis</h1>
<p>Another critical function of the scRNASequest pipeline is the <strong>Differential expression (DE)</strong> analysis. By running the steps in Section: <a href="running-the-pipeline.html#running-the-pipeline">Running the pipeline</a>, we have performed quality control of the data, and generated an integrated dataset.</p>
<p>Before running the DE analysis, the <strong>DEGinfo.csv</strong> comparison file needs to be prepared:</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_empty.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_empty.png" /></a></p>
<p>The first column, <strong>comparisonName</strong> is designed to store the name of the comparison. These names will be used to create separate folders under the result directory. In the DEG comparison file above, <strong>“sample”</strong>, <strong>“cluster”</strong>, <strong>“group”</strong> and <strong>“covars”</strong> are the annotation headers, and <strong>“alt”</strong> and <strong>“ref”</strong> are the two entries from <strong>“group”</strong> column. The <strong>“covars”</strong> is optional and can be empty.</p>
<p>The DEG is performing between <strong>“alt”</strong> and <strong>“ref”</strong> cells from <strong>“group”</strong> within each entry of <strong>“cluster”</strong> considering <strong>“sample”</strong> variations. The <strong>“group”</strong> variable should contain conditions to compare, such as Mutant v.s. Control. Thus, this pipeline is designed to loop through each cluster, and perform DEG analysis between <strong>“alt”</strong> v.s. <strong>“ref”</strong>.</p>
<p><strong>Covariate adjustment:</strong></p>
<p>Adjusting covariates is optional. By default, NEBULA considers the value of <strong>log of total UMI (library size)</strong> as the offset covariate in our pipeline, so you don’t need to adjust this again in <strong>“covars”</strong>.</p>
<p>If needed, you can put a formula in the <strong>“covars”</strong> column using the meta information from the sampleMeta.csv file, such as Sex, Age, etc. You can also include QC metrics such as mitochondria percentage (pct_MT), etc. Please find a full list of QC metrics below. You can also open the Cellxgene VIP and check the right side. All numeric covariates are listed there.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb59-1"><a href="differential-expression-de-analysis.html#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="ex">n_genes_by_counts,</span> log1p_n_genes_by_counts, pct_counts_in_top_50_genes, pct_counts_in_top_100_genes, pct_counts_in_top_200_genes, pct_counts_in_top_500_genes, pct_MT, n_genes, predicted.CellType.score</span></code></pre></div>
<p>You may consider <strong>pct_MT</strong> as a covariate to adjust, but be careful to adjust other covariates here unless you have a good reason to do so. An example of the formula to be filled out in the <strong>“covars”</strong> column: <strong>Age+Sex+pct_MT</strong>.</p>
<p><strong>Continuous/Numeric variable as the comparison group:</strong></p>
<p>If the user is interested in using a numeric variable (such as Age) as the comparison group in the comparison, please provide the column header of the numeric column in the cell/sample meta table in the <strong>“group”</strong> column of the above DEGinfo.csv table, and leave <strong>“alt”</strong> and <strong>“ref”</strong> columns of that row empty. Currently this function is only supported if DE analysis is run using the NEBULA method.</p>
<p><strong>Missing values:</strong></p>
<p>It’s important to know that if any cells don’t have such DE condition information. For example, some samples don’t have information such as ‘treated’ or ‘control’ labeled, and instead, the values were ‘NA’, ‘NaN’, etc. In such case, we can easily include such information in the main config.yml file by defining the NA values to be ignored. The pipeline will remove all these cells from the DE analysis.</p>
<pre><code>#An example of the config.yml file:
NAstring: [NA, NaN, NAN] #provide a list of strings which should be considered NA and associated cells to be removed</code></pre>
<p><strong>DE software choices:</strong></p>
<p>Based on the previous <a href="https://www.mdpi.com/2075-1729/12/6/850" target="_blank">benchmarking analysis</a>, NEBULA outperforms other software in single-cell RNA-seq DE analysis. Thus, our pipeline uses NEBULA by default to run DE. However, we also offer the following methods as well, and you can directly use them in the DEGinfo.csv file, see Example 3 later.</p>
<ul>
<li><p><strong>u_test</strong>: Mann–Whitney U test, also called Wilcoxon rank-sum test, is a nonparametric test. See <a href="https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test" target="_blank">Wiki</a>. In our single-cell RNA-seq analysis, this is a pseudobulk method.</p></li>
<li><p><strong>t_test</strong>: Student’s t-test, statistical test based on Student’s t-distribution. See <a href="https://en.wikipedia.org/wiki/Student%27s_t-test" target="_blank">Wiki</a>. In our single-cell RNA-seq analysis, this is a pseudobulk method.</p></li>
<li><p><strong>edgeR</strong>: Using the <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html" target="_blank">edgeR package</a> to perform pseudobulk DE analysis. Refer to their <a href="https://academic.oup.com/bioinformatics/article/26/1/139/182458" target="_blank">paper</a> for more details.</p></li>
<li><p><strong>limma</strong>: Using the <a href="http://bioconductor.org/packages/release/bioc/html/limma.html" target="_blank">limma-voom method</a> to perform pseudobulk DE analysis. Refer to their <a href="https://academic.oup.com/nar/article/43/7/e47/2414268" target="_blank">paper</a> for more details.</p></li>
<li><p><strong>DESeq2</strong>: Using the <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html" target="_blank">DESeq2 package</a> to perform pseudobulk DE analysis. Refer to their <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8" target="_blank">paper</a> for more details.</p></li>
<li><p><strong>ANCOVA</strong>: Using the <a href="https://www.r-bloggers.com/2021/10/analysis-of-covariance-ancova-using-r/" target="_blank">ANCOVA</a> to perform pseudobulk DE analysis.</p></li>
<li><p><strong>MAST</strong>: Using the <a href="https://www.bioconductor.org/packages/release/bioc/html/MAST.html" target="_blank">MAST package</a> to perform single-cell DE analysis. Refer to their <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5" target="_blank">paper</a> for more details.</p></li>
<li><p><strong>NEBULA</strong>: Using the <a href="https://github.com/lhe17/nebula" target="_blank">NEBULA software</a> to perform single-cell DE analysis. Refer to their <a href="https://www.nature.com/articles/s42003-021-02146-6" target="_blank">paper</a> for more details.</p></li>
<li><p><strong>glmmTMB</strong>: Using the <a href="https://glmmtmb.github.io/glmmTMB/" target="_blank">glmmTMB (nbinom2 family function)</a> to perform single-cell DE analysis.</p></li>
<li><p><strong>limma_cell_level</strong>: This also uses <a href="http://bioconductor.org/packages/release/bioc/html/limma.html" target="_blank">limma</a> but not a pseudobulk method.</p></li>
</ul>
<p>Here, this tutorial provides several examples to illustrate the DE analysis in scRNASequest:</p>
<p>We use a UMAP with label transferred cell type annotation as an example dataset. Label transfer is strongly suggested if you would like to run DE analysis because the clusters are more meaningful than the original clusters assigned by the software:</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/UMAP.png" /></a></p>
<div id="example-1" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> Example 1</h2>
<p>In the first example, we would like to run DE analysis between <strong>‘Female’</strong> and <strong>‘Male’</strong> for each cluster annotated by <strong>predicted.celltype1</strong>. In the first column, we input the header name <strong>library_id</strong>, which annotates the data sources. Then we add <strong>predicted.celltype1</strong> in the <em>cluster</em> column, which allows the pipeline to loop through each cluster in <strong>predicted.celltype1</strong>. The <em>group</em> column contains the header name storing the comparison groups, and here we use the <strong>Sex</strong> annotation. Each time, the pipeline can only compare two conditions, such as <strong>‘Female’</strong> and <strong>‘Male’</strong>. If the <em>group</em> column contains more groups, please list them in multiple lines in the DEGinfo.csv file.</p>
<p>We can also add <em>covars</em> if needed, but this is optional.</p>
<p>The default DE analysis is performed by <a href="https://www.nature.com/articles/s42003-021-02146-6" target="_blank"><strong>NEBULA</strong></a>. As for the model, NEBULA provides two choices: <strong>HL</strong> and <strong>LN</strong>. For more details about them, please refer to the <a href="https://cran.r-project.org/web/packages/nebula/vignettes/nebula_example.html#difference-between-nebula-ln-and-nebula-hl" target="_blank">NEBULA manual</a>.</p>
<p>Here is the DEGinfo.csv we described above:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb61-1"><a href="differential-expression-de-analysis.html#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="ex">comparisonName,sample,cluster,group,alt,ref,covars[+</span> separated],method[default NEBULA],model[default HL]</span>
<span id="cb61-2"><a href="differential-expression-de-analysis.html#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Compare_Female_vs_Male,library_id,predicted.celltype1,Sex,Female,Male,,NEBULA,HL</span></span></code></pre></div>
<p>After preparing the DEGinfo.csv file, simply add it to the <strong>config.yml</strong> file:</p>
<pre><code>DEG_desp: ~/E-MTAB-11115/processing/DEGinfo.csv</code></pre>
<p>Then rerun the pipeline. The pipeline won’t run the previous steps (i.e. filtering, harmonization, etc.) again, so it will directly run the DE analysis based on the harmonized h5ad file.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb63-1"><a href="differential-expression-de-analysis.html#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span> Path/to/a/config/file</span>
<span id="cb63-2"><a href="differential-expression-de-analysis.html#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="differential-expression-de-analysis.html#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb63-4"><a href="differential-expression-de-analysis.html#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scAnalyzer</span> ~/E-MTAB-11115/processing/config.yml</span></code></pre></div>
<p>This step may take a few hours to run. The output files will be generated in the directory:</p>
<pre><code>outputdir
    ... previous results         #Omitted, see section 6.4
    ├── ProjName_scDEG           #Folder containing DEG results
      ├── Compare_Female_vs_Male
        ├── Female.vs.Male_predicted.celltype1:astrocytes_NEBULA.csv    #Each comparison has three associated files
        ├── Female.vs.Male_predicted.celltype1:astrocytes_NEBULA.png
        ├── Female.vs.Male_predicted.celltype1:astrocytes_NEBULA.QC.pdf
        ├── Female.vs.Male_predicted.celltype1:discarded_NEBULA.csv
        ├── Female.vs.Male_predicted.celltype1:discarded_NEBULA.png
        ├── Female.vs.Male_predicted.celltype1:discarded_NEBULA.QC.pdf
        ...
        └── env.rds</code></pre>
<p>An example of the DE csv table (Female.vs.Male_predicted.celltype1/astrocytes_NEBULA.csv, top 5 rows):</p>
<p>In the table below, the most significant gene is <a href="https://www.ncbi.nlm.nih.gov/gene/213742" target="_blank">Xist</a>, a well-known X-chromosome gene highly expressed in female.</p>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:unnamed-chunk-40">Table 8.1: </span>DEG analysis of Female.vs.Male in astrocytes
</caption>
<thead>
<tr>
<th style="text-align:center;">
res.tab.ID
</th>
<th style="text-align:center;">
res.tab.log2FC
</th>
<th style="text-align:center;">
res.tab.Pvalue
</th>
<th style="text-align:center;">
res.tab.FDR
</th>
<th style="text-align:center;">
res.tab.algorithm
</th>
<th style="text-align:center;">
res.tab.convergence
</th>
<th style="text-align:center;">
res.tab.metric
</th>
<th style="text-align:center;">
res.ls.summary.logFC_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.logFC_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.se_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.se_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.p_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.p_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.gene_id
</th>
<th style="text-align:center;">
res.ls.summary.gene
</th>
<th style="text-align:center;">
res.ls.overdispersion.Subject
</th>
<th style="text-align:center;">
res.ls.overdispersion.Cell
</th>
<th style="text-align:center;">
res.ls.convergence
</th>
<th style="text-align:center;">
res.ls.algorithm
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center;">
Xist
</td>
<td style="text-align:center;">
9.676171
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
6115.1780
</td>
<td style="text-align:center;">
-6.693213
</td>
<td style="text-align:center;">
0.1276737
</td>
<td style="text-align:center;">
0.0234622
</td>
<td style="text-align:center;">
0.0469395
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0065289
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
Rgs20
</td>
<td style="text-align:center;">
0.0016728
</td>
<td style="text-align:center;">
0.4393152
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Pnpla7
</td>
<td style="text-align:center;">
-1.230034
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
1653.1553
</td>
<td style="text-align:center;">
-9.321192
</td>
<td style="text-align:center;">
-0.1694550
</td>
<td style="text-align:center;">
0.0534180
</td>
<td style="text-align:center;">
0.1065177
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.1116405
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
Atp6v1h
</td>
<td style="text-align:center;">
0.0064320
</td>
<td style="text-align:center;">
0.3327316
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Etnppl
</td>
<td style="text-align:center;">
-1.592058
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
525.8328
</td>
<td style="text-align:center;">
-8.511808
</td>
<td style="text-align:center;">
0.0917939
</td>
<td style="text-align:center;">
0.0282832
</td>
<td style="text-align:center;">
0.0565059
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.1042691
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;">
Rb1cc1
</td>
<td style="text-align:center;">
0.0001000
</td>
<td style="text-align:center;">
0.1725113
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Zbtb16
</td>
<td style="text-align:center;">
-1.448309
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
1260.9518
</td>
<td style="text-align:center;">
-9.896064
</td>
<td style="text-align:center;">
-0.0918014
</td>
<td style="text-align:center;">
0.0584728
</td>
<td style="text-align:center;">
0.1165549
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.4309170
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
4732440D04Rik
</td>
<td style="text-align:center;">
0.0014745
</td>
<td style="text-align:center;">
0.7288504
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Sorcs2
</td>
<td style="text-align:center;">
1.137956
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
1398.0881
</td>
<td style="text-align:center;">
-8.573613
</td>
<td style="text-align:center;">
-0.2252848
</td>
<td style="text-align:center;">
0.0303869
</td>
<td style="text-align:center;">
0.0604625
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0001945
</td>
<td style="text-align:center;">
5
</td>
<td style="text-align:center;">
Pcmtd1
</td>
<td style="text-align:center;">
0.0001000
</td>
<td style="text-align:center;">
0.3502587
</td>
<td style="text-align:center;">
-10
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
</tbody>
</table>
</div>
<p>An example of the DE png file (Female.vs.Male_predicted.celltype1/astrocytes_NEBULA.png):</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Female_vs_Male_Astrocytes.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Female_vs_Male_Astrocytes.png" /></a></p>
</div>
<div id="example-2" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> Example 2</h2>
<p>In the second example, we are going to show how to perform DE analysis <strong>between two cell types</strong>. To achieve this, we added one artificial column (called ‘SelectAll’) in the <strong>sampleMeta.csv</strong> file, and assigned the same values (‘Everything’) to all the data:</p>
<pre><code>Sample_Name,h5path,Sex,SelectAll
5705STDY8058280,/home/ysun4/testSinglecell/ExternalData/5705STDY8058280_filtered_feature_bc_matrix.h5,Female,Everything
5705STDY8058281,/home/ysun4/testSinglecell/ExternalData/5705STDY8058281_filtered_feature_bc_matrix.h5,Female,Everything
5705STDY8058282,/home/ysun4/testSinglecell/ExternalData/5705STDY8058282_filtered_feature_bc_matrix.h5,Female,Everything
5705STDY8058283,/home/ysun4/testSinglecell/ExternalData/5705STDY8058283_filtered_feature_bc_matrix.h5,Male,Everything
5705STDY8058284,/home/ysun4/testSinglecell/ExternalData/5705STDY8058284_filtered_feature_bc_matrix.h5,Male,Everything
5705STDY8058285,/home/ysun4/testSinglecell/ExternalData/5705STDY8058285_filtered_feature_bc_matrix.h5,Male,Everything</code></pre>
<p>Then we prepare the <strong>DEGinfo.csv</strong> file in this way. The ‘SelectAll’ was filled into the <em>cluster</em> column, and since it only has one value, it actually selects all cells when running the pipeline. Then it performs <strong>astrocytes</strong> to <strong>Oligos</strong> comparison.</p>
<pre><code>comparisonName,sample,cluster,group,alt,ref,covars[+ separated],method[default NEBULA],model[default HL]
Comparison_Ast_vs_Oligo,library_id,SelectAll,predicted.celltype1,astrocytes,Oligos,,NEBULA,HL</code></pre>
<p>Then we need to run the scRNASequest pipeline from the beginning, which allows the ‘SelectAll’ column to be embedded into the harmonized h5ad output. As long as we have DEGinfo.csv included in the config.yml, it will run through the DE analysis and produce the following results:</p>
<pre><code>outputdir
    ... previous results         #Omitted, see section 6.4
    ├── ProjName_scDEG           #Folder containing DEG results
      ├── Comparison_Ast_vs_Oligo
        ├── astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.csv    #Each comparison has three associated files
        ├── astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.png
        ├── astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.QC.pdf
        ├── env.rds
    ├── ProjName_scDEG.cmd.json
    ├── deg6602_SelectAll_predicted.celltype1_NEBULA_HL_Everything_astrocytes_Oligos.error  #Standard error messages
    └── ProjName_scDEG.db</code></pre>
<p>An example of the DE csv table (astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.csv, top 5 rows):</p>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:unnamed-chunk-41">Table 8.2: </span>DEG analysis of astrocytes.vs.Oligos
</caption>
<thead>
<tr>
<th style="text-align:center;">
res.tab.ID
</th>
<th style="text-align:center;">
res.tab.log2FC
</th>
<th style="text-align:center;">
res.tab.Pvalue
</th>
<th style="text-align:center;">
res.tab.FDR
</th>
<th style="text-align:center;">
res.tab.algorithm
</th>
<th style="text-align:center;">
res.tab.convergence
</th>
<th style="text-align:center;">
res.tab.metric
</th>
<th style="text-align:center;">
res.ls.summary.logFC_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.logFC_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.se_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.se_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.p_.Intercept.
</th>
<th style="text-align:center;">
res.ls.summary.p_.GrouPalt
</th>
<th style="text-align:center;">
res.ls.summary.gene_id
</th>
<th style="text-align:center;">
res.ls.summary.gene
</th>
<th style="text-align:center;">
res.ls.overdispersion.Subject
</th>
<th style="text-align:center;">
res.ls.overdispersion.Cell
</th>
<th style="text-align:center;">
res.ls.convergence
</th>
<th style="text-align:center;">
res.ls.algorithm
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center;">
Xkr4
</td>
<td style="text-align:center;">
-3.327923
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
4177.156
</td>
<td style="text-align:center;">
-8.168075
</td>
<td style="text-align:center;">
-2.3067403
</td>
<td style="text-align:center;">
0.0367924
</td>
<td style="text-align:center;">
0.0561022
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0000000
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
Xkr4
</td>
<td style="text-align:center;">
0.0067297
</td>
<td style="text-align:center;">
0.6076303
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Rgs20
</td>
<td style="text-align:center;">
6.386205
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
5561.771
</td>
<td style="text-align:center;">
-9.825703
</td>
<td style="text-align:center;">
-2.5066029
</td>
<td style="text-align:center;">
0.0890548
</td>
<td style="text-align:center;">
0.1353916
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0000000
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
Gm1992
</td>
<td style="text-align:center;">
0.0398103
</td>
<td style="text-align:center;">
1.4947561
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
St18
</td>
<td style="text-align:center;">
-5.129500
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
14283.487
</td>
<td style="text-align:center;">
-10.322696
</td>
<td style="text-align:center;">
0.2176484
</td>
<td style="text-align:center;">
0.0310922
</td>
<td style="text-align:center;">
0.0690489
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0016211
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;">
Mrpl15
</td>
<td style="text-align:center;">
0.0001000
</td>
<td style="text-align:center;">
0.4188721
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Adhfe1
</td>
<td style="text-align:center;">
4.842192
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
12162.136
</td>
<td style="text-align:center;">
-10.169996
</td>
<td style="text-align:center;">
0.1119117
</td>
<td style="text-align:center;">
0.0292679
</td>
<td style="text-align:center;">
0.0669762
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0947381
</td>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
Tcea1
</td>
<td style="text-align:center;">
0.0001000
</td>
<td style="text-align:center;">
0.7212448
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
<tr>
<td style="text-align:center;">
Prex2
</td>
<td style="text-align:center;">
8.368190
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
HL
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
8249.837
</td>
<td style="text-align:center;">
-10.292221
</td>
<td style="text-align:center;">
4.4265797
</td>
<td style="text-align:center;">
0.0513462
</td>
<td style="text-align:center;">
0.0563860
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0.0000000
</td>
<td style="text-align:center;">
5
</td>
<td style="text-align:center;">
Rgs20
</td>
<td style="text-align:center;">
0.0047768
</td>
<td style="text-align:center;">
0.4563594
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
NBGMM (HL)
</td>
</tr>
</tbody>
</table>
</div>
<p>An example of the DE png file (astrocytes.vs.Oligos_SelectAll:Everything_NEBULA.png):</p>
<p><a href="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Astrocytes_Oligos.png"><img src="https://interactivereport.github.io/scRNAsequest/tutorial/images/DEG_Astrocytes_Oligos.png" /></a></p>
</div>
<div id="example-3" class="section level2" number="8.3">
<h2><span class="header-section-number">8.3</span> Example 3</h2>
<p>In this example, we show how to perform DE analysis using multiple software. In the config file, each line is a comparison, and for the same comparison (treat v.s. control), we can set up multiple runs using different DE tools as below:</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb68-1"><a href="differential-expression-de-analysis.html#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="ex">comparisonName,sample,cluster,group,alt,ref,covars[+</span> separated],method[default NEBULA],model[default HL]</span>
<span id="cb68-2"><a href="differential-expression-de-analysis.html#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Treated_vs_Control_nebula,library_id,sctHarmony_cluster_predicted.celltype,Treatment,Treated,Vehicle,,NEBULA,HL</span></span>
<span id="cb68-3"><a href="differential-expression-de-analysis.html#cb68-3" aria-hidden="true" tabindex="-1"></a><span class="ex">Treated_vs_Control_t_test,library_id,sctHarmony_cluster_predicted.celltype,Treatment,Treated,Vehicle,,t_test,</span></span>
<span id="cb68-4"><a href="differential-expression-de-analysis.html#cb68-4" aria-hidden="true" tabindex="-1"></a><span class="ex">Treated_vs_Control_u_test,library_id,sctHarmony_cluster_predicted.celltype,Treatment,Treated,Vehicle,,u_test,</span></span>
<span id="cb68-5"><a href="differential-expression-de-analysis.html#cb68-5" aria-hidden="true" tabindex="-1"></a><span class="ex">Treated_vs_Control_edgeR,library_id,sctHarmony_cluster_predicted.celltype,Treatment,Treated,Vehicle,,edgeR,</span></span>
<span id="cb68-6"><a href="differential-expression-de-analysis.html#cb68-6" aria-hidden="true" tabindex="-1"></a><span class="ex">Treated_vs_Control_limma,library_id,sctHarmony_cluster_predicted.celltype,Treatment,Treated,Vehicle,,limma,</span></span>
<span id="cb68-7"><a href="differential-expression-de-analysis.html#cb68-7" aria-hidden="true" tabindex="-1"></a><span class="ex">Treated_vs_Control_DESeq2,library_id,sctHarmony_cluster_predicted.celltype,Treatment,Treated,Vehicle,,DESeq2,</span></span>
<span id="cb68-8"><a href="differential-expression-de-analysis.html#cb68-8" aria-hidden="true" tabindex="-1"></a><span class="ex">Treated_vs_Control_MAST,library_id,sctHarmony_cluster_predicted.celltype,Treatment,Treated,Vehicle,,MAST,</span></span>
<span id="cb68-9"><a href="differential-expression-de-analysis.html#cb68-9" aria-hidden="true" tabindex="-1"></a><span class="ex">Treated_vs_Control_limma_cell_level,library_id,sctHarmony_cluster_predicted.celltype,Treatment,Treated,Vehicle,,limma_cell_level,</span></span>
<span id="cb68-10"><a href="differential-expression-de-analysis.html#cb68-10" aria-hidden="true" tabindex="-1"></a><span class="ex">Treated_vs_Control_glmmTMB,library_id,sctHarmony_cluster_predicted.celltype,Treatment,Treated,Vehicle,,glmmTMB,poisson</span></span>
<span id="cb68-11"><a href="differential-expression-de-analysis.html#cb68-11" aria-hidden="true" tabindex="-1"></a><span class="ex">Treated_vs_Control_ancova,library_id,sctHarmony_cluster_predicted.celltype,Treatment,Treated,Vehicle,,ancova,</span></span></code></pre></div>
<p>Please note that NEBULA and glmmTMB require <strong>model</strong> information.</p>
<p>For NEBULA, users can choose from <strong>HL</strong> and <strong>LN</strong>, and the pipeline uses ‘HL’ by default. The differences between them can be found in the <a href="https://github.com/lhe17/nebula" target="_blank">NEBULA GitHub page</a>:</p>
<p>“NEBULA-LN is faster and performs particularly well when the number of cells per subject (CPS) is large. In addition, NEBULA-LN is much more accurate in estimating a very large subject-level overdispersion. In contrast, NEBULA-HL is slower but more accurate in estimating the cell-level overdispersion.”</p>
<p>For glmmTMB, it is required to choose a model from the following: “nbinom2”, “nbinom1”, “poisson”, “nbinom2zi”, “nbinom1zi”.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="running-the-pipeline.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="reference-building.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/scRNAsequest/edit/gh-pages/08-DEAnalysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/scRNAsequest/blob/gh-pages/08-DEAnalysis.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
