<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 9 Reference building | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</title>
  <meta name="description" content="Chapter 9 Reference building | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 9 Reference building | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 9 Reference building | scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing" />
  
  
  



<meta name="date" content="2024-03-28" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="differential-expression-de-analysis.html"/>
<link rel="next" href="celldepot-publishing.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">scRNASequest</a></li>

<li class="divider"></li>
<li><a href="index.html#preface" id="toc-preface"><span class="toc-section-number">1</span> Preface</a></li>
<li><a href="introduction.html#introduction" id="toc-introduction"><span class="toc-section-number">2</span> Introduction</a></li>
<li><a href="installation.html#installation" id="toc-installation"><span class="toc-section-number">3</span> Installation</a>
<ul>
<li><a href="installation.html#install-scrnasequest" id="toc-install-scrnasequest"><span class="toc-section-number">3.1</span> Install scRNASequest</a>
<ul>
<li><a href="installation.html#installation-using-conda" id="toc-installation-using-conda"><span class="toc-section-number">3.1.1</span> Installation using Conda</a></li>
<li><a href="installation.html#installation-through-docker" id="toc-installation-through-docker"><span class="toc-section-number">3.1.2</span> Installation through Docker</a></li>
</ul></li>
<li><a href="installation.html#configure-sys.yml-file" id="toc-configure-sys.yml-file"><span class="toc-section-number">3.2</span> Configure sys.yml file</a></li>
<li><a href="installation.html#install-cellxgene-vip" id="toc-install-cellxgene-vip"><span class="toc-section-number">3.3</span> Install cellxgene VIP</a></li>
<li><a href="installation.html#install-celldepot" id="toc-install-celldepot"><span class="toc-section-number">3.4</span> Install CellDepot</a></li>
</ul></li>
<li><a href="demo-run.html#demo-run" id="toc-demo-run"><span class="toc-section-number">4</span> Demo run</a>
<ul>
<li><a href="demo-run.html#demo-run-with-conda" id="toc-demo-run-with-conda"><span class="toc-section-number">4.1</span> Demo run with Conda</a></li>
<li><a href="demo-run.html#demo-run-with-docker" id="toc-demo-run-with-docker"><span class="toc-section-number">4.2</span> Demo run with Docker</a></li>
</ul></li>
<li><a href="data-preparation.html#data-preparation" id="toc-data-preparation"><span class="toc-section-number">5</span> Data preparation</a>
<ul>
<li><a href="data-preparation.html#public-data-in-h5-format" id="toc-public-data-in-h5-format"><span class="toc-section-number">5.1</span> Public data in h5 format</a></li>
<li><a href="data-preparation.html#public-data-in-10x-mex-format" id="toc-public-data-in-10x-mex-format"><span class="toc-section-number">5.2</span> Public data in 10X MEX format</a></li>
<li><a href="data-preparation.html#self-prepared-files" id="toc-self-prepared-files"><span class="toc-section-number">5.3</span> Self-prepared files</a></li>
</ul></li>
<li><a href="input-configuration.html#input-configuration" id="toc-input-configuration"><span class="toc-section-number">6</span> Input configuration</a>
<ul>
<li><a href="input-configuration.html#generate-template-files" id="toc-generate-template-files"><span class="toc-section-number">6.1</span> Generate template files</a></li>
<li><a href="input-configuration.html#prepare-the-config.yml-file" id="toc-prepare-the-config.yml-file"><span class="toc-section-number">6.2</span> Prepare the config.yml file</a></li>
<li><a href="input-configuration.html#section52" id="toc-section52"><span class="toc-section-number">6.3</span> Prepare the sample meta file</a>
<ul>
<li><a href="input-configuration.html#h5-input" id="toc-h5-input"><span class="toc-section-number">6.3.1</span> h5 input</a></li>
<li><a href="input-configuration.html#mex-input" id="toc-mex-input"><span class="toc-section-number">6.3.2</span> MEX input</a></li>
</ul></li>
<li><a href="input-configuration.html#file-hierarchy" id="toc-file-hierarchy"><span class="toc-section-number">6.4</span> File hierarchy</a></li>
</ul></li>
<li><a href="running-the-pipeline.html#running-the-pipeline" id="toc-running-the-pipeline"><span class="toc-section-number">7</span> Running the pipeline</a>
<ul>
<li><a href="running-the-pipeline.html#initial-scanalyzer-run" id="toc-initial-scanalyzer-run"><span class="toc-section-number">7.1</span> Initial scAnalyzer run</a></li>
<li><a href="running-the-pipeline.html#review-the-bookdown-report" id="toc-review-the-bookdown-report"><span class="toc-section-number">7.2</span> Review the bookdown report</a></li>
<li><a href="running-the-pipeline.html#generate-a-slide-deck" id="toc-generate-a-slide-deck"><span class="toc-section-number">7.3</span> Generate a slide deck</a></li>
<li><a href="running-the-pipeline.html#doublet-detection" id="toc-doublet-detection"><span class="toc-section-number">7.4</span> Doublet detection</a></li>
<li><a href="running-the-pipeline.html#cell-type-label-transfer" id="toc-cell-type-label-transfer"><span class="toc-section-number">7.5</span> Cell type label transfer</a></li>
<li><a href="running-the-pipeline.html#adjust-major_cluster_rate" id="toc-adjust-major_cluster_rate"><span class="toc-section-number">7.6</span> Adjust major_cluster_rate</a></li>
<li><a href="running-the-pipeline.html#run-the-pipeline-in-parallel" id="toc-run-the-pipeline-in-parallel"><span class="toc-section-number">7.7</span> Run the pipeline in parallel</a></li>
<li><a href="running-the-pipeline.html#run-the-full-analysis" id="toc-run-the-full-analysis"><span class="toc-section-number">7.8</span> Run the full analysis</a></li>
<li><a href="running-the-pipeline.html#visualize-the-harmonized-results" id="toc-visualize-the-harmonized-results"><span class="toc-section-number">7.9</span> Visualize the harmonized results</a></li>
</ul></li>
<li><a href="differential-expression-de-analysis.html#differential-expression-de-analysis" id="toc-differential-expression-de-analysis"><span class="toc-section-number">8</span> Differential expression (DE) analysis</a>
<ul>
<li><a href="differential-expression-de-analysis.html#example-1" id="toc-example-1"><span class="toc-section-number">8.1</span> Example 1</a></li>
<li><a href="differential-expression-de-analysis.html#example-2" id="toc-example-2"><span class="toc-section-number">8.2</span> Example 2</a></li>
<li><a href="differential-expression-de-analysis.html#example-3" id="toc-example-3"><span class="toc-section-number">8.3</span> Example 3</a></li>
</ul></li>
<li><a href="reference-building.html#reference-building" id="toc-reference-building"><span class="toc-section-number">9</span> Reference building</a>
<ul>
<li><a href="reference-building.html#initialize-scref" id="toc-initialize-scref"><span class="toc-section-number">9.1</span> Initialize scRef</a></li>
<li><a href="reference-building.html#submit-scref" id="toc-submit-scref"><span class="toc-section-number">9.2</span> Submit scRef</a></li>
<li><a href="reference-building.html#demo-scref" id="toc-demo-scref"><span class="toc-section-number">9.3</span> Demo scRef</a></li>
</ul></li>
<li><a href="celldepot-publishing.html#celldepot-publishing" id="toc-celldepot-publishing"><span class="toc-section-number">10</span> CellDepot publishing</a>
<ul>
<li><a href="celldepot-publishing.html#initialize-sc2celldepot" id="toc-initialize-sc2celldepot"><span class="toc-section-number">10.1</span> Initialize sc2celldepot</a></li>
<li><a href="celldepot-publishing.html#run-sc2celldepot" id="toc-run-sc2celldepot"><span class="toc-section-number">10.2</span> Run sc2celldepot</a></li>
<li><a href="celldepot-publishing.html#demo-sc2celldepot" id="toc-demo-sc2celldepot"><span class="toc-section-number">10.3</span> Demo sc2celldepot</a></li>
</ul></li>
<li><a href="additional-tools.html#additional-tools" id="toc-additional-tools"><span class="toc-section-number">11</span> Additional tools</a>
<ul>
<li><a href="additional-tools.html#scrmambient" id="toc-scrmambient"><span class="toc-section-number">11.1</span> scRMambient</a>
<ul>
<li><a href="additional-tools.html#set-up-the-inputs" id="toc-set-up-the-inputs"><span class="toc-section-number">11.1.1</span> Set up the inputs</a></li>
<li><a href="additional-tools.html#run-scrmambient" id="toc-run-scrmambient"><span class="toc-section-number">11.1.2</span> Run scRMambient</a></li>
</ul></li>
<li><a href="additional-tools.html#scdeg" id="toc-scdeg"><span class="toc-section-number">11.2</span> scDEG</a>
<ul>
<li><a href="additional-tools.html#initialize-scdeg" id="toc-initialize-scdeg"><span class="toc-section-number">11.2.1</span> Initialize scDEG</a></li>
<li><a href="additional-tools.html#run-de-analysis" id="toc-run-de-analysis"><span class="toc-section-number">11.2.2</span> Run DE analysis</a></li>
</ul></li>
<li><a href="additional-tools.html#sctool" id="toc-sctool"><span class="toc-section-number">11.3</span> scTool</a>
<ul>
<li><a href="additional-tools.html#remove-annotations" id="toc-remove-annotations"><span class="toc-section-number">11.3.1</span> Remove annotations</a></li>
<li><a href="additional-tools.html#add-annotations" id="toc-add-annotations"><span class="toc-section-number">11.3.2</span> Add annotations</a></li>
<li><a href="additional-tools.html#export-gene-expressions" id="toc-export-gene-expressions"><span class="toc-section-number">11.3.3</span> Export gene expressions</a></li>
<li><a href="additional-tools.html#pseudo-bulk-aggregation" id="toc-pseudo-bulk-aggregation"><span class="toc-section-number">11.3.4</span> Pseudo-bulk aggregation</a></li>
<li><a href="additional-tools.html#demo-sctool" id="toc-demo-sctool"><span class="toc-section-number">11.3.5</span> Demo scTool</a></li>
</ul></li>
</ul></li>
<li><a href="miscellaneous.html#miscellaneous" id="toc-miscellaneous"><span class="toc-section-number">12</span> Miscellaneous</a>
<ul>
<li><a href="miscellaneous.html#cell-subsetting" id="toc-cell-subsetting"><span class="toc-section-number">12.1</span> Cell subsetting</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scRNASequest: an ecosystem of scRNA-seq analysis, visualization and publishing</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="reference-building" class="section level1" number="9">
<h1><span class="header-section-number">Chapter 9</span> Reference building</h1>
<p>Reference building is critical for label transfer. To add a reference dataset into the scRNASequest pipeline, the reference matrix needs to be SCT transformed.</p>
<p>By running the following command, we can see the manual page of the reference generator, scRef:</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb69-1"><a href="reference-building.html#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> scRef</span>
<span id="cb69-2"><a href="reference-building.html#cb69-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-3"><a href="reference-building.html#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="ex">*****</span> 2023-01-26 15:29:54 <span class="pp">*****</span></span>
<span id="cb69-4"><a href="reference-building.html#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb69-5"><a href="reference-building.html#cb69-5" aria-hidden="true" tabindex="-1"></a><span class="co">## scRNAsequest: https://github.com/interactivereport/scRNAsequest.git</span></span>
<span id="cb69-6"><a href="reference-building.html#cb69-6" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Path: /mnt/depts/dept04/compbio/edge_tools/scRNAsequest</span></span>
<span id="cb69-7"><a href="reference-building.html#cb69-7" aria-hidden="true" tabindex="-1"></a><span class="co">## Pipeline Date: 2023-01-13 17:41:49 -0500</span></span>
<span id="cb69-8"><a href="reference-building.html#cb69-8" aria-hidden="true" tabindex="-1"></a><span class="co">## git HEAD: 3d463e0b127af499942b7adc2fc5af6ddfc6f11e</span></span>
<span id="cb69-9"><a href="reference-building.html#cb69-9" aria-hidden="true" tabindex="-1"></a><span class="co">###########</span></span>
<span id="cb69-10"><a href="reference-building.html#cb69-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-11"><a href="reference-building.html#cb69-11" aria-hidden="true" tabindex="-1"></a><span class="ex">Loading</span> resources</span>
<span id="cb69-12"><a href="reference-building.html#cb69-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-13"><a href="reference-building.html#cb69-13" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> /path/to/a/output/folder === or === scRef /path/to/a/Ref/config/file</span>
<span id="cb69-14"><a href="reference-building.html#cb69-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-15"><a href="reference-building.html#cb69-15" aria-hidden="true" tabindex="-1"></a><span class="ex">The</span> folder has to be existed.</span>
<span id="cb69-16"><a href="reference-building.html#cb69-16" aria-hidden="true" tabindex="-1"></a><span class="ex">The</span> Ref config file will be generated automatically when a path is provided</span>
<span id="cb69-17"><a href="reference-building.html#cb69-17" aria-hidden="true" tabindex="-1"></a><span class="ex">=====</span> CAUTION =====</span>
<span id="cb69-18"><a href="reference-building.html#cb69-18" aria-hidden="true" tabindex="-1"></a>    <span class="ex">1.</span> This process will add a seurat reference data into the scRNAsequest pipeline PERMANENTLY!</span>
<span id="cb69-19"><a href="reference-building.html#cb69-19" aria-hidden="true" tabindex="-1"></a>    <span class="ex">2.</span> Make sure the data provided for reference building is SCT transformed!</span>
<span id="cb69-20"><a href="reference-building.html#cb69-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-21"><a href="reference-building.html#cb69-21" aria-hidden="true" tabindex="-1"></a><span class="ex">Powered</span> by the Research Data Sciences group [zhengyu.ouyang@biogen.com<span class="kw">;</span><span class="ex">kejie.li@biogen.com]</span></span>
<span id="cb69-22"><a href="reference-building.html#cb69-22" aria-hidden="true" tabindex="-1"></a><span class="ex">------------</span></span></code></pre></div>
<div id="initialize-scref" class="section level2" number="9.1">
<h2><span class="header-section-number">9.1</span> Initialize scRef</h2>
<p>This pipeline can be initialized using an empty directory. for example, we first create a directory called ‘Reference_data’, then initiate the pipeline pointing to this directory:</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb70-1"><a href="reference-building.html#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> /path/to/the/directory</span>
<span id="cb70-2"><a href="reference-building.html#cb70-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-3"><a href="reference-building.html#cb70-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb70-4"><a href="reference-building.html#cb70-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> ~/Reference_data</span></code></pre></div>
<p>After the run, a config file, <strong>refConfig.yml</strong>, and a log file will be generated in the directory. The <strong>refConfig.yml</strong> is a template of the following scRef run, passing critical parameters to the pipeline. The <strong>refConfig.yml</strong> file will be same as this <a href="https://github.com/interactivereport/scRNAsequest/blob/main/src/ref.yml" target="_blank">template</a>, but the output directory will be yours. Here is an example after filling in the configuration file:</p>
<pre><code>output: ~/Reference_data
# the following is normally located in the same folder of celldepot hosting h5ad files
ref_h5ad_raw: /path/to/ProjectName_raw_added.h5ad # full path to the h5ad file contains raw UMI along with cell annotation and layout
ref_batch: library_id
# above two parameters are ignored, if a seurat object can be located in the project folder
ref_rds:                                          # full path to the processed seurat object including SCT assay, cell annotation and layout

# All information below are required
ref_name: Reference_data                # Please provide a unique name prefer to include species and tissue (check existed by calling scAnalyzer without argument)
ref_link:                               # The web link to the information of this reference. For scAnalyzer processed data, you could provide a Cellxgene VIP link here.
ref_src: sn                             # sc/&quot;single cell&quot; or sn/&quot;single nuclei&quot;
ref_platform: 10X                       # Single cell/neuclei technology e.g. 10X, SNARE-seq2, dropSeq, ...
#list a reduction to be used (at least 50 dimensions full name from VIP or one from seurat &#39;reductions&#39; ) 
# details: https://github.com/satijalab/azimuth/wiki/Azimuth-Reference-Format
# this reduction is NOT directly used, but used to find the neighbors which is then used for computing sPCA reduction which is used in the reference
# For instance, if harmony was prefered layout then providing either &#39;harmony&#39; (50 dimention) or &#39;harmony-PCA&#39;
ref_reduction: pca 
ref_label: [predicted.celltype1]        # List the annotations (case sensitive) to be used for transferring. Please check this in the data. For our case, the header is called &#39;predicted.celltype1&#39;
                                        # The cell type label/header name you would like to transfer
publish: False                          # Should this reference be published (added permanently) into scAnalyzer
overwrite: False                        # Overwrite the existing scAnalyzer reference </code></pre>
<p>For the input data, scRef can take either an h5ad file containing raw UMI and annotatino information, or an R data file in rds format. If you have finished running a dataset using scAnalyzer, then you can directly use the ProjectName_raw_added.h5ad (not ProjectName.h5ad file) as input, and attach its path to <strong>ref_h5ad_raw</strong>. Alternatively, you can provide an RDS file using <strong>ref_rds</strong>. Either providing <strong>ref_h5ad_raw</strong> or <strong>ref_rds</strong> would be sufficient to the pipeline.</p>
<p>The <strong>ref_label</strong> is critical for label transfer if you would like to use this data as a reference in the future. It tells the program to use the cell type information in these (can be one more more) headers to perform label transfer. For a h5ad data, the easiest way is to open it using Cellxgene VIP, and identify the header that contains cell type labels, e.g. predicted.celltype1 in our case. For an RDS data, you could read it in R using the <a href="https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/readRDS" target="_blank">readRDS</a> function and identify the column names containing cell type annotation.</p>
<p>For <strong>ref_rds</strong> input RDS file, please make sure it has been SCTransformed, and includes UMAP embeddings based on SCT values. Here are some codes to prepare it:</p>
<pre><code>library(Seurat)

Data &lt;- readRDS(&quot;Annotated.rds&quot;)                              #The meta.data table already has cell types annotated
Data &lt;- SCTransform(Data, return.only.var.genes = F)          #Turn return.only.var.genes off
Data &lt;- RunPCA(Data, verbose = FALSE)
Data &lt;- RunUMAP(Data, dims = 1:30, verbose = FALSE)
Data &lt;- FindNeighbors(Data, dims = 1:30, verbose = FALSE)
Data &lt;- FindClusters(Data, verbose = FALSE)

saveRDS(Data, file = &quot;Annotated.ForRuningscRef.rds&quot;)</code></pre>
</div>
<div id="submit-scref" class="section level2" number="9.2">
<h2><span class="header-section-number">9.2</span> Submit scRef</h2>
<p>After filling in the information in the <strong>refConfig.yml</strong> file, we are ready to submit the full pipeline and build the reference data for label transfer:</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb73-1"><a href="reference-building.html#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> /path/to/a/Ref/config/file</span>
<span id="cb73-2"><a href="reference-building.html#cb73-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-3"><a href="reference-building.html#cb73-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Example:</span></span>
<span id="cb73-4"><a href="reference-building.html#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="ex">scRef</span> ~/Reference_data/refConfig.yml</span></code></pre></div>
<p>Output files:</p>
<pre><code>Reference_data
    ├── init_20220630.log
    ├── refConfig.yml
    ├── refConfig.yml.20220630.log               # Output log information
    ├── Reference_data_for_scAnalyzer.rds        # The data to use for label transfer and downstream analysis
    ├── ref_notFor_scAnalyzer.rds                # Output rds file NOT for lebel transfer
    ...</code></pre>
<p>Please also check the information in the refConfig.yml.20220630.log file, and pay attention to the last few lines:</p>
<pre><code>...  #log of the running process omitted

The private reference could be used by provide the following full path to &#39;ref_name&#39; in scAnalyzer config file:
    ~/Reference_data/Reference_data_for_scAnalyzer.rds
...</code></pre>
<p>This indicates that the reference has been successfully generated, and it can be passed to scAnalyzer through the config.yml file. If you turned on <strong>“publish: True”</strong> in the <strong>refConfig.yml</strong>, this reference will be added to scAnalyzer, and you can use its name to refer it when running scAnalyzer.</p>
</div>
<div id="demo-scref" class="section level2" number="9.3">
<h2><span class="header-section-number">9.3</span> Demo scRef</h2>
<p>In this demo, we use a previously processed RDS file using cell type annotation from a public project: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE172462">GSE172462</a>. The RDS file is: <a href="https://github.com/interactivereport/scRNAsequest/releases/download/untagged-e92025b745dbc6f9666a/GSE172462.SCT.For_scRef.rds">GSE172462.SCT.For_scRef.rds</a>.</p>
<p>We first run scRef by providing our working directory to generate necessary template files:</p>
<pre><code>scRef ~/demo_scRef</code></pre>
<p>Then we fill in the <code>refConfig.yml</code> file:</p>
<pre><code># please check wiki page for the details
output: ~/demo_scRef
# the following is normally located in the same folder of celldepot hosting h5ad files
ref_h5ad_raw:                                       # full path to the h5ad file contains raw UMI along with cell annotation and layout
ref_batch: library_id
# above two parameters are ignored, if a seurat object can be located in the project folder
ref_rds: ~/demo_scRef/GSE172462.SCT.For_scRef.rds   #full path to the processed seurat object including SCT assay, cell annotation and layout

# All information below are required
ref_name: demo                                      # please provide a unique name prefer to include species and tissue (check existed by calling scAnalyzer without argument)
ref_link: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE172462 # The web link to the information of this reference
ref_src: single nuclei # sc/&quot;single cell&quot; or sn/&quot;single nuclei&quot;
ref_platform: 10X # which single cell/neuclei technology e.g. 10X, SNARE-seq2, dropSeq, ...
#list a reduction to be used (at least 50 dimensions full name from VIP or one from seurat &#39;reductions&#39; ) 
# details: https://github.com/satijalab/azimuth/wiki/Azimuth-Reference-Format
# this reduction is NOT directly used, but used to find the neighbors which is then used for computing sPCA reduction which is used in the reference
# For instance, if harmony was prefered layout then providing either &#39;harmony&#39; (50 dimention) or &#39;harmony-PCA&#39;
ref_reduction: pca 
ref_label: [celltype, neuron_subtype, manual_anno]  # list the annotations (case sensitive) to be used for transferring
publish: False                                      # should this reference be published into scAnalyzer
overwrite: False                                    # overwrite the existing scAnalyzer reference</code></pre>
<p>In this demo, we turned off <code>publish</code> and <code>overwrite</code> to make it just to test <code>scRef</code>. This is a reduced data (with only 1000 files) to test the demo, so it may not be ideal to use it as a real reference in the pipeline.</p>
<p>Finally, we can run the <code>scRef</code>:</p>
<pre><code>scRef ~/demo_scRef/refConfig.yml</code></pre>
<p>This pipeline will finish in ~1 minute and the output file will be:</p>
<pre><code>demo_for_scAnalyzer.rds</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="differential-expression-de-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="celldepot-publishing.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/interactivereport/scRNAsequest/edit/gh-pages/09-ReferenceBuilding.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/interactivereport/scRNAsequest/blob/gh-pages/09-ReferenceBuilding.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
